<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Vignette 0: Different Types of Fully Efficient Models</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Vignette 0: Different Types of Fully
Efficient Models</h1>



<p>Libraries for open source:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(matrixcalc) <span class="co">#for direct sum operator</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(sommer) <span class="co">#for modelling</span></span></code></pre></div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>In this vignette we will show how different types of fully-efficient,
two stage genomic selection models can be fitted to a dataset. We will
also compare them to their equivalent unweighted and single stage
models. To that end, we will use a simulated dataset:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;GmatOat.RData&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;SimulData.RData&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#True variance values from the simulation</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;True&quot;</span>, </span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>              SimulResults<span class="sc">$</span>SampleLocVar, </span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>              SimulResults<span class="sc">$</span>SampleAdditiveVar,</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>              SimulResults<span class="sc">$</span>SampleGresVar)</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#True breeding values from the simulation</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>TBV <span class="ot">&lt;-</span> SimulResults<span class="sc">$</span>AdditiveValues</span></code></pre></div>
<p>This data contains a simulated augmented randomized complete block
design within five locations. This design is sparse for each location,
i.e. only some checks are replicated across blocks within each field,
but most genotypes are unreplicated. Conversely, all genotypes are fully
replicated across locations. In the simulation, we have simulated the
phenotype as the sum of additive and non-additive genetic effects, a
location effect, spatial variation within each field and some additional
random noise.</p>
<p>We will fit models that account for all of these terms except the
non-additive one. The reason for this, is that we want to compare the
performance of the different models when they are a simplification of
reality, which is usually the case when working with empirical data. We
will start with the most standard methodology (single stage, unweighted
two stage) and later we will move into the fully-efficient two stage
models. For all models we will store the estimates of the variance
components and the genomic estimated breeding values (GEBVs), which can
be compared to the true ones (known because the data was simulated).
This will allow us to evaluate the quality of each model.</p>
<p>Finally, it is important to keep in mind that this is just a single
iteration of the simulation, making the results obtained here just
tentative. In the main paper we have performed 500 repetitions, making
the results much more reliable.</p>
</div>
<div id="single-stage-model" class="section level1">
<h1>Single Stage Model</h1>
<p>Modelling all effects in a single stage is considered to be the best
approach, but it has a larger computational cost than the two stage
alternatives, which can make it unfeasible for industrial-scale
datasets. The model used in this work is the following:</p>
<p><span class="math display">\[
\boldsymbol{y} = \mu + Z_e \boldsymbol{u_e} + Z_a \boldsymbol{u_a} + Z_p
\boldsymbol{u_p} + \boldsymbol{\epsilon}
\]</span></p>
<p>Where <span class="math inline">\(\boldsymbol{y}\)</span> is a vector
of simulated phenotypes. <span class="math inline">\(\mu\)</span> is the
fixed intercept. <span class="math inline">\(\boldsymbol{u_e} \sim
N(\boldsymbol{0}, I\sigma^2_e)\)</span> is a vector of environmental
random effects. <span class="math inline">\(\boldsymbol{u_a} \sim
N(\boldsymbol{0}, G\sigma^2_a)\)</span> is a vector of additive
genotypic random effects. <span class="math inline">\(\boldsymbol{u_p}\)</span> is a vector of random
spatial effects following the 2D p-splines framework and nested within
the environment. <span class="math inline">\(\boldsymbol{\epsilon} \sim
N(\boldsymbol{0}, I\sigma^2_\epsilon)\)</span> is a vector of residuals.
<span class="math inline">\(Z_e\)</span>, <span class="math inline">\(Z_a\)</span>, and <span class="math inline">\(Z_\epsilon\)</span> are the design matrices of
their corresponding effects and <span class="math inline">\(G\)</span>
is the VanRaden additive genomic relationship matrix. One important
consideration about the residuals in this model is that they only
contain the nugget (true error) from the simulation, which was specified
to have the same variance in all environments. Therefore, we can use
homogeneous residuals across environments for the model, but
heterogeneous residuals are common when working with empirical
datasets.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>iter <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>SingleStageDF <span class="ot">&lt;-</span> SimulResults[[<span class="st">&quot;AugmentedRCBD_Book&quot;</span>]]</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>SingleStageDF<span class="sc">$</span>Rowf <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(SingleStageDF<span class="sc">$</span>Row)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>SingleStageDF<span class="sc">$</span>Colf <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(SingleStageDF<span class="sc">$</span>Col)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>SingleStageDF<span class="sc">$</span>GIDres <span class="ot">&lt;-</span> SingleStageDF<span class="sc">$</span>GID</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="fu">head</span>(SingleStageDF)</span></code></pre></div>
<pre><code>##     GID Env Block Row Col Check     Pheno Rowf Colf GIDres
## 1 FM552  E1     1   1   1 FALSE 1006.3320    1    1  FM552
## 2  FM67  E1     1   2   1 FALSE 1000.2891    2    1   FM67
## 3 FM396  E1     1   3   1 FALSE  998.6174    3    1  FM396
## 4 FM686  E1     1   4   1  TRUE  999.5628    4    1  FM686
## 5 FM220  E1     1   5   1 FALSE  997.7516    5    1  FM220
## 6 FM251  E1     1   6   1 FALSE  992.7325    6    1  FM251</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>Nrows <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(SingleStageDF<span class="sc">$</span>Row))</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>Ncols <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(SingleStageDF<span class="sc">$</span>Col))</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#Fit model</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#This can be a bit slow</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>SingleStageModel <span class="ot">&lt;-</span> <span class="fu">mmer</span>(Pheno <span class="sc">~</span> <span class="dv">1</span>, <span class="co">#Intercept is only fixed effect</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>       <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env) <span class="sc">+</span> <span class="co">#environment</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>         <span class="fu">vsr</span>(GID, <span class="at">Gu =</span> Ga) <span class="sc">+</span> <span class="co">#additive</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>         <span class="co">#Spatial effects are split into two main effects and an interaction term:</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>         <span class="fu">vsr</span>(<span class="fu">dsr</span>(Env),Rowf)<span class="sc">+</span> <span class="co">#main effect for row within each environment</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>         <span class="fu">vsr</span>(<span class="fu">dsr</span>(Env),Colf)<span class="sc">+</span> <span class="co">#main effect for column within each environment</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>         <span class="co">#Interaction effect between rows and columns within each environment:</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>         <span class="co">#Use splines, which allow to fit more coefficients than datapoints</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>         <span class="co">#available thanks to the use of a penalization:</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>         <span class="fu">spl2Da</span>(Col,Row, <span class="at">at.var=</span>Env, </span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>           <span class="co">#Number of Segments for splines around half of the number of rows or columns</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>           <span class="at">nsegments =</span> <span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">16</span>), </span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>           <span class="co">#half the number of segments for the smooth by smooth interaction (faster computations)</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>           <span class="at">nestorder =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)), </span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>       <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units), <span class="co">#homogeneous residuals across locations</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>       <span class="at">data=</span> SingleStageDF,</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>       <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  )</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;GID&#39;</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co">#Variance values from single stage model</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Single_Stage&quot;</span>, </span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>              SingleStageModel<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>              SingleStageModel<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:GID</span><span class="st">`</span>,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#Genomic estimated breeding values from single stage model</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>GEBVs_SingleStage <span class="ot">&lt;-</span> SingleStageModel<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:GID</span><span class="st">`</span><span class="sc">$</span>Pheno</span></code></pre></div>
</div>
<div id="two-stage-models" class="section level1">
<h1>Two Stage Models</h1>
<p>All two stage models have a common first stage, but they differ on
how the second stage is implemented. Thus, we will first run the first
stage models:</p>
<div id="first-stage-model" class="section level2">
<h2>First Stage Model</h2>
<p>A different stage 1 model is fit to each environment
independently:</p>
<p><span class="math display">\[
\boldsymbol{y_j} = X_j \boldsymbol{\beta_j} + Z_j
\boldsymbol{u_{p_j}}  + \boldsymbol{\epsilon_j}
\]</span></p>
<p>Where <span class="math inline">\(\boldsymbol{y_j}\)</span> contains
the phenotypic observations in environment <span class="math inline">\(j\)</span>; <span class="math inline">\(\boldsymbol{\beta_j}\)</span> is a vector of fixed
genotypic effects and <span class="math inline">\(X_j\)</span> is its
corresponding design matrix; <span class="math inline">\(\boldsymbol{u_{p_j}}\)</span> is a vector of
random spatial effects following the 2D P-splines framework and <span class="math inline">\(Z_{p_j}\)</span> is its corresponding design
matrix; <span class="math inline">\(\boldsymbol{\epsilon_j} \sim
N(\boldsymbol{0}, I\sigma^2_{\epsilon_j})\)</span> is a vector of
residuals. <span class="math inline">\(I\)</span> is the identity matrix
of appropriate dimensions and <span class="math inline">\(\sigma^2_{\epsilon_j}\)</span> is the residual
variance component.</p>
<p>From this model, we can obtain BLUEs for the fixed genotypic effects
(<span class="math inline">\(\boldsymbol{\hat{\beta}_j}\)</span>), which
can be regarded as adjusted phenotypes without the spatial effects of
the field. We can also calculate their associated error
variance-covariance, which can be used as a matrix of weights. These
will be carried on to the second stage. For the sake of clarity, we will
call the variance-covariance associated to stage 1 genotypic BLUEs the
estimation error variance-covariance matrix (<span class="math inline">\(EEV\)</span>). The genotypic BLUEs of the models
from the different environments will be concatenated together to be used
as input for second stage models and the <span class="math inline">\(EEV\)</span> matrices will be merged though their
direct sum.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#Stage 1</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#one model per environment</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>EEVs <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>Envs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>H2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>residuals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&quot;First Step&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;First Step&quot;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co">#One model per environment</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="cf">for</span> (env <span class="cf">in</span> <span class="fu">rownames</span>(SimulResults<span class="sc">$</span>Evcov)) {</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  phenoDF <span class="ot">&lt;-</span> SimulResults[[<span class="st">&quot;AugmentedRCBD_Book&quot;</span>]]</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> phenoDF[phenoDF<span class="sc">$</span>Env <span class="sc">==</span> env,]</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  </span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  <span class="co">#Keep track of the environments for which we are making the models</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  Envs <span class="ot">&lt;-</span> <span class="fu">c</span>(Envs, <span class="fu">rep</span>(env, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>GID))))</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  </span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  <span class="co">#Prepare data</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  phenoenv<span class="sc">$</span>Block <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>Block) <span class="co">#cofactor</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  phenoenv<span class="sc">$</span>Rowf <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(phenoenv<span class="sc">$</span>Row)</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  phenoenv<span class="sc">$</span>Colf <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(phenoenv<span class="sc">$</span>Col)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  <span class="fu">rownames</span>(phenoDF) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#Fit stage 1 model</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>  models[[env]] <span class="ot">&lt;-</span> <span class="fu">mmer</span>(Pheno<span class="sc">~</span>GID <span class="sc">-</span><span class="dv">1</span>, <span class="co">#fixed genotypic effects, no intercept</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>         <span class="co">#Spatial effects are split into two main effects and an interaction term:</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>         <span class="at">random =</span> <span class="sc">~</span>Rowf<span class="sc">+</span>Colf<span class="sc">+</span> <span class="co">#main effects for row and column</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>         <span class="co">#Interaction effect between rows and columns:</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>         <span class="co">#Use splines, which allow to fit more coefficients than datapoints</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>         <span class="co">#available thanks to the use of a penalization:</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>         <span class="fu">spl2Da</span>(Col,Row, </span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>           <span class="co">#Number of Segments for splines around half of the number of rows or columns</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>           <span class="at">nsegments =</span> <span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">16</span>), </span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>           <span class="co">#half the number of segments for the smooth by smooth interaction (faster computations)</span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>           <span class="at">nestorder =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)), </span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a>         <span class="at">data=</span>phenoenv, </span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>         <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a>          )</span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a> </span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a>  EEV <span class="ot">&lt;-</span> models[[env]]<span class="sc">$</span>VarBeta</span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(EEV) <span class="ot">&lt;-</span> </span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a>    <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">&quot;^GID&quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="at">x =</span> models[[env]]<span class="sc">$</span>Beta<span class="sc">$</span>Effect)</span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a>  </span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a>  <span class="co">#keep only the BLUEs for the id fixed effect</span></span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a>  <span class="co">#Concatenate the BLUEs from the models for each environment</span></span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a>  placeholder <span class="ot">&lt;-</span> <span class="fu">names</span>(BLUEs)</span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a>  BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>(BLUEs, models[[env]]<span class="sc">$</span>Beta<span class="sc">$</span>Estimate)</span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a>  <span class="fu">names</span>(BLUEs) <span class="ot">&lt;-</span> <span class="fu">c</span>(placeholder,</span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a>                    <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">&quot;^GID&quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="at">x =</span> models[[env]]<span class="sc">$</span>Beta<span class="sc">$</span>Effect))</span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a>  </span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a>  <span class="co">#Store EEV matrix for each environment</span></span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a>  EEVs[[env]] <span class="ot">&lt;-</span> EEV</span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co">#merge EEV matrices from each environmet for their use in the second stage</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#It is done using the direct sum operator from matrixcalc library</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>()</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoDF<span class="sc">$</span>Env))) {</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  env1 <span class="ot">&lt;-</span> <span class="fu">unique</span>(phenoDF<span class="sc">$</span>Env)[i<span class="dv">-1</span>]</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  env2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(phenoDF<span class="sc">$</span>Env)[i]</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  EEV1 <span class="ot">&lt;-</span> EEVs[[env1]]</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV1) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(EEV1), <span class="st">&quot;:&quot;</span>, env1)</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>  EEV2 <span class="ot">&lt;-</span> EEVs[[env2]]</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV2) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rownames</span>(EEV2), <span class="st">&quot;:&quot;</span>, env2)</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(<span class="fu">is.na</span>(placeholder_matrix)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEV1,EEV2)</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(placeholder_matrix,EEV2)</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>  }</span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>}</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a>EEV_full <span class="ot">&lt;-</span> placeholder_matrix</span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co">#remove &quot;GID&quot; at the beginning of the rownames (added by sommer)</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="fu">rownames</span>(EEV_full) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;^GID&quot;</span>, <span class="st">&quot;&quot;</span>,<span class="fu">rownames</span>(EEV_full))</span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a><span class="fu">colnames</span>(EEV_full) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(EEV_full)</span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a><span class="fu">dim</span>(EEV_full)</span></code></pre></div>
<pre><code>## [1] 1300 1300</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">length</span>(BLUEs)</span></code></pre></div>
<pre><code>## [1] 1300</code></pre>
</div>
<div id="unweighted-second-stage-model" class="section level2">
<h2>Unweighted Second Stage Model</h2>
<p>This model the first stage BLUEs as a response variable but it does
not consider <span class="math inline">\(EEV\)</span> at all, therefore
giving the same weight to all observations regardless of how confident
we are on the quality of their estimation:</p>
<p><span class="math display">\[
\boldsymbol{\hat{\beta}_{all}} = \mu + Z_e \boldsymbol{u_e} + Z_a
\boldsymbol{u_a} + \boldsymbol{\epsilon}
\]</span></p>
<p>Where the <span class="math inline">\(\boldsymbol{\hat{\beta}_{all}}\)</span> genotypic
BLUEs from the first stage models are the response variable and all
other terms are the same as the effects with the same name in the single
stage model.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="co">#1) Prepare data</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>sommerdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">BLUEs =</span> BLUEs,</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>                               <span class="at">id =</span> <span class="fu">names</span>(BLUEs),</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>                               <span class="at">Env =</span> Envs)</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#fit model</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>Stage2ModelUnweighted <span class="ot">&lt;-</span> <span class="fu">mmer</span>(BLUEs <span class="sc">~</span> <span class="dv">1</span>, <span class="co">#only fixed effect is intercept</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>                                    <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env)<span class="sc">+</span> <span class="co">#random environmental effect</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>                                      <span class="fu">vsr</span>(id, <span class="at">Gu =</span> Ga), <span class="co">#additive effect</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>                                    <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units), <span class="co">#iid residuals</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>                                    <span class="at">data=</span> sommerdf,</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a>                                    <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;id&#39;</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co">#Variance values from unweighted two stage model</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Unweighted&quot;</span>, </span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>              Stage2ModelUnweighted<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>              Stage2ModelUnweighted<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span>,</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co">#genomic estimated breeding values from unweighted two stage model</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>GEBVs_Unweighted <span class="ot">&lt;-</span> Stage2ModelUnweighted<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span><span class="sc">$</span>BLUEs</span></code></pre></div>
</div>
<div id="fully-efficient-second-stage-models" class="section level2">
<h2>Fully Efficient Second Stage Models</h2>
<p>These models are similar to the previous one but they also consider
the EEV matrix. However, there are several ways to take EEV into
account. It can be used as the covariance matrix for the residuals or
for a random effect. Furthermore, it is possible to use the entire EEV
matrix or the off-diagonals can be discarded to accelerate computations.
In the latter case, it is important to note that, in the mixed models,
the EEV matrix is not used directly. Instead, its inverse is of
interest. Therefore, to minimize the loss of information, the
off-diagonals of <span class="math inline">\(EEV\)</span> only have to
be discarded after inverting it.</p>
<div id="fully-efficient-second-stage-full-eev-in-residuals" class="section level3">
<h3>Fully Efficient Second Stage, Full EEV in residuals</h3>
<p>This model has the following equation:</p>
<p><span class="math display">\[
\boldsymbol{\hat{\beta}_{all}} = \mu + Z_e \boldsymbol{u_e} + Z_a
\boldsymbol{u_a} + \boldsymbol{\epsilon}
\]</span> Where all terms are the same as for the unweighted second
stage model with the exception of the residuals. <span class="math inline">\(EEV\)</span> is included as the
variance-covariance matrix of the residuals: <span class="math inline">\(\boldsymbol{\epsilon} \sim N(\boldsymbol{0},
EEV)\)</span>. One important thing to note is that the variance
component of the residuals has to be fixed, as it is already known from
the first stage models that its variance-covariance matrix is <span class="math inline">\(EEV\)</span>. We don’t want the model to estimate
a residual variance component that scales <span class="math inline">\(EEV\)</span> to a different value. The main
problem of this type of model is its inability to account for any terms
not explicitly modeled in fixed or random effects. As the residuals are
constrained to follow <span class="math inline">\(EEV\)</span>, they can
only fit the true error generated in the first stage models. For
instance, we have not explicitly modeled non-additive genetic effects in
the model, which is a problem since they cannot be absorbed by the
residuals (the covariance structure of the non-additive genetic effects
does not match <span class="math inline">\(EEV\)</span> and therefore
the model cannot fit these effects into the residuals). Thus, we only
recommend including <span class="math inline">\(EEV\)</span> into the
residuals if you are completely certain that all meaningful effects have
been explicitly model in fixed or random effects.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">#Prepare data</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="co">#we need to add to our data a column for stage 1 estimation error </span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>Stage1Error <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sommerdf<span class="sc">$</span>id,<span class="st">&quot;:&quot;</span>,sommerdf<span class="sc">$</span>Env)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="fu">identical</span>(<span class="fu">rownames</span>(EEV_full), Stage1Error)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>sommerdf <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf,</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                      Stage1Error)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="co">#sommer convention</span></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co">#all variances are stored relative to the variance of the response variable</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf<span class="sc">$</span>BLUEs))</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="co">#fit model</span></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>Stage2ModelFullResid <span class="ot">&lt;-</span> <span class="fu">mmer</span>(BLUEs <span class="sc">~</span> <span class="dv">1</span>, </span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env)<span class="sc">+</span> <span class="co">#random environmental effect</span></span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>       <span class="fu">vsr</span>(id, <span class="at">Gu =</span> Ga), <span class="co">#additive effect</span></span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a>     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units, <span class="co">#residuals.</span></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a>               <span class="at">Gti =</span> Variance_scaled, <span class="co">#provide variance component to be equal to 1</span></span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a>               <span class="co">#i.e., don&#39;t let the model alter EEV by scaling it with a variance</span></span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a>               <span class="co">#component different from one.</span></span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a>               <span class="at">Gtc =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#Disable estimation of residual component, i.e.</span></span>
<span id="cb21-17"><a href="#cb21-17" tabindex="-1"></a>               <span class="co">#use the value provided in Gti = Variance_scaled</span></span>
<span id="cb21-18"><a href="#cb21-18" tabindex="-1"></a>     <span class="at">data=</span> sommerdf,</span>
<span id="cb21-19"><a href="#cb21-19" tabindex="-1"></a>     <span class="at">W=</span><span class="fu">solve</span>(EEV_full), <span class="co">#Here we set the inverse of EEV to be the matrix of weights.</span></span>
<span id="cb21-20"><a href="#cb21-20" tabindex="-1"></a>     <span class="co">#This means that we are setting EEV to be the covariance matrix of the residuals.</span></span>
<span id="cb21-21"><a href="#cb21-21" tabindex="-1"></a>     <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb21-22"><a href="#cb21-22" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;id&#39;</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co">#Variance values from fully efficient, two stage model, full EEV in residuals</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;FullResid&quot;</span>, </span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>              Stage2ModelFullResid<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>              Stage2ModelFullResid<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span>,</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="co">#genomic estimated breeding values from two stage model with full EEV in residuals</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>GEBVs_FullResid <span class="ot">&lt;-</span> Stage2ModelFullResid<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span><span class="sc">$</span>BLUEs</span></code></pre></div>
</div>
<div id="fully-efficient-second-stage-diagonal-eev-in-residuals" class="section level3">
<h3>Fully-Efficient Second Stage, Diagonal EEV in residuals</h3>
<p>This model is exaclty the same as the previous one but we remove the
off-diagonals of the inverse of EEV to accelerate computations:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co">#Prepare data</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>EEV_inv_diag <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">solve</span>(EEV_full))</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="fu">names</span>(EEV_inv_diag) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(EEV_full)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>sommerdf<span class="sc">$</span>Weights <span class="ot">=</span> EEV_inv_diag</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a><span class="co">#sommer convention</span></span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a><span class="co">#all variances are stored relative to the variance of the response variable</span></span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf<span class="sc">$</span>BLUEs))</span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a><span class="co">#fit model</span></span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a>Stage2ModelDiagResid <span class="ot">&lt;-</span> <span class="fu">mmer</span>(BLUEs <span class="sc">~</span> <span class="dv">1</span>, </span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a>     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env)<span class="sc">+</span> <span class="co">#random environmental effect</span></span>
<span id="cb24-13"><a href="#cb24-13" tabindex="-1"></a>       <span class="fu">vsr</span>(id, <span class="at">Gu =</span> Ga), <span class="co">#additive effect</span></span>
<span id="cb24-14"><a href="#cb24-14" tabindex="-1"></a>     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units, <span class="co">#residuals.</span></span>
<span id="cb24-15"><a href="#cb24-15" tabindex="-1"></a>               <span class="at">Gti =</span> Variance_scaled, <span class="co">#provide variance component to be equal to 1</span></span>
<span id="cb24-16"><a href="#cb24-16" tabindex="-1"></a>               <span class="co">#i.e., don&#39;t let the model alter EEV by scaling it with a variance</span></span>
<span id="cb24-17"><a href="#cb24-17" tabindex="-1"></a>               <span class="co">#component different from one.</span></span>
<span id="cb24-18"><a href="#cb24-18" tabindex="-1"></a>               <span class="at">Gtc =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#Disable estimation of residual component, i.e.</span></span>
<span id="cb24-19"><a href="#cb24-19" tabindex="-1"></a>               <span class="co">#use the value provided in Gti = Variance_scaled</span></span>
<span id="cb24-20"><a href="#cb24-20" tabindex="-1"></a>     <span class="at">data=</span> sommerdf,</span>
<span id="cb24-21"><a href="#cb24-21" tabindex="-1"></a>     <span class="at">weights=</span>Weights, <span class="co">#Here we set the diagonal of the inverse of EEV to be the vector of weights.</span></span>
<span id="cb24-22"><a href="#cb24-22" tabindex="-1"></a>     <span class="co">#This means that we are setting the inverse of the residual covariance matrix to be equal to</span></span>
<span id="cb24-23"><a href="#cb24-23" tabindex="-1"></a>     <span class="co">#the diagonal of the inverse of EEV</span></span>
<span id="cb24-24"><a href="#cb24-24" tabindex="-1"></a>     <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb24-25"><a href="#cb24-25" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;id&#39;</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="co">#Variance values from fully efficient, two stage model, diagonal EEV in residuals</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;DiagResid&quot;</span>, </span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>              Stage2ModelDiagResid<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>              Stage2ModelDiagResid<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span>,</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a><span class="co">#genomic estimated breeding values from two stage model with diagonal EEV in residuals</span></span>
<span id="cb26-9"><a href="#cb26-9" tabindex="-1"></a>GEBVs_DiagResid <span class="ot">&lt;-</span> Stage2ModelDiagResid<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span><span class="sc">$</span>BLUEs</span></code></pre></div>
</div>
<div id="fully-efficient-second-stage-full-eev-in-random-effect" class="section level3">
<h3>Fully-Efficient Second Stage, Full EEV in random effect</h3>
<p>This model has the following equation:</p>
<p><span class="math display">\[
\boldsymbol{\hat{\beta}_{all}} = \mu + Z_e \boldsymbol{u_e} + Z_a
\boldsymbol{u_a} + Z_s \boldsymbol{u_s} + \boldsymbol{\epsilon}
\]</span></p>
<p>The same nomenclature as in previous models has been used where
applicable. <span class="math inline">\(\boldsymbol{u_s} \sim
N(\boldsymbol{0}, EEV)\)</span> is the random effect for the stage 1
error and it contains the true error. As happened in the previous
models, no variance component for the stage 1 error term is estimated as
it is already implicitly included within <span class="math inline">\(EEV\)</span>. In this model, <span class="math inline">\(\boldsymbol{\epsilon} \sim N(\boldsymbol{0},
I\sigma^2_\epsilon)\)</span>, i.e. there are i.i.d. residuals that give
more flexibility to the model. For instance, we have not explicitly
modeled non-additive genetic effects in the model and they can be
absorbed by the residuals.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co">#fit model</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>Stage2ModelFullRaneff <span class="ot">&lt;-</span> <span class="fu">mmer</span>(BLUEs <span class="sc">~</span> <span class="dv">1</span>, </span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env)<span class="sc">+</span> <span class="co">#random environmental effect</span></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>       <span class="fu">vsr</span>(id, <span class="at">Gu =</span> Ga)<span class="sc">+</span> <span class="co">#additive effect</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>       <span class="fu">vsr</span>(Stage1Error, <span class="co">#first stage error term</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>           <span class="at">Gti =</span> Variance_scaled, <span class="co">#unit variance value</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>           <span class="at">Gtc =</span> <span class="fu">matrix</span>(<span class="dv">3</span>), <span class="co">#disable variance estimation (the model will be forced</span></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a>           <span class="co">#to use the value provided in Gti argument)</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>           <span class="at">Gu =</span> EEV_full), <span class="co">#EEV variance-covariance matrix</span></span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units), <span class="co">#iid residuals</span></span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a>     <span class="at">data=</span> sommerdf, </span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a>     <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;id&#39;</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co">#Variance values from fully efficient, two stage model, full EEV in random effect</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;FullRaneff&quot;</span>, </span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>              Stage2ModelFullRaneff<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>              Stage2ModelFullRaneff<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span>,</span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a><span class="co">#genomic estimated breeding values from two stage model with full EEV in random effect</span></span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a>GEBVs_FullRaneff <span class="ot">&lt;-</span> Stage2ModelFullRaneff<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span><span class="sc">$</span>BLUEs</span></code></pre></div>
</div>
<div id="fully-efficient-second-stage-diagonal-eev-in-random-effect" class="section level3">
<h3>Fully-Efficient Second Stage, Diagonal EEV in random effect</h3>
<p>The same model as before has been used, but the off-diagonal elements
of the inverse of EEV have been discarded to allow for faster
computations.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a>EEV_diag <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(EEV_full), <span class="at">ncol =</span> <span class="fu">ncol</span>(EEV_full))</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="fu">rownames</span>(EEV_diag) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(EEV_diag) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(EEV_full)</span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="co">#1) Invert EEV (solve() function)</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a><span class="co">#2) Discard off-diagonals of the inverse (diag() function extracts only the diagonal)</span></span>
<span id="cb30-5"><a href="#cb30-5" tabindex="-1"></a><span class="co">#3) Invert again to undo the first inversion (1/diagonal elements, as the matrix</span></span>
<span id="cb30-6"><a href="#cb30-6" tabindex="-1"></a><span class="co">#is diagonal its inverse can be done this way)</span></span>
<span id="cb30-7"><a href="#cb30-7" tabindex="-1"></a><span class="co">#This allows us to get a diagonal EEV matrix that, when inverted, is equal to </span></span>
<span id="cb30-8"><a href="#cb30-8" tabindex="-1"></a><span class="co">#the diagonal of the inverse of the full EEV matrix, i.e. we have been able to </span></span>
<span id="cb30-9"><a href="#cb30-9" tabindex="-1"></a><span class="co">#discard the off-diagonals of EEV after the inversion</span></span>
<span id="cb30-10"><a href="#cb30-10" tabindex="-1"></a><span class="co">#We want to discard them after the inversion because what the models actually</span></span>
<span id="cb30-11"><a href="#cb30-11" tabindex="-1"></a><span class="co">#use is the inverse of the covariance matrices. Therefore, discarding data before</span></span>
<span id="cb30-12"><a href="#cb30-12" tabindex="-1"></a><span class="co">#inverting will result in increased information loss</span></span>
<span id="cb30-13"><a href="#cb30-13" tabindex="-1"></a><span class="fu">diag</span>(EEV_diag) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="fu">diag</span>(<span class="fu">solve</span>(EEV_full)))</span>
<span id="cb30-14"><a href="#cb30-14" tabindex="-1"></a></span>
<span id="cb30-15"><a href="#cb30-15" tabindex="-1"></a></span>
<span id="cb30-16"><a href="#cb30-16" tabindex="-1"></a><span class="co">#fit model</span></span>
<span id="cb30-17"><a href="#cb30-17" tabindex="-1"></a>Stage2ModelDiagRaneff <span class="ot">&lt;-</span> <span class="fu">mmer</span>(BLUEs <span class="sc">~</span> <span class="dv">1</span>, </span>
<span id="cb30-18"><a href="#cb30-18" tabindex="-1"></a>     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsr</span>(Env)<span class="sc">+</span> <span class="co">#random environmental effect</span></span>
<span id="cb30-19"><a href="#cb30-19" tabindex="-1"></a>       <span class="fu">vsr</span>(id, <span class="at">Gu =</span> Ga)<span class="sc">+</span> <span class="co">#additive effect</span></span>
<span id="cb30-20"><a href="#cb30-20" tabindex="-1"></a>       <span class="fu">vsr</span>(Stage1Error, <span class="co">#first stage error term</span></span>
<span id="cb30-21"><a href="#cb30-21" tabindex="-1"></a>           <span class="at">Gti =</span> Variance_scaled, <span class="co">#unit variance value</span></span>
<span id="cb30-22"><a href="#cb30-22" tabindex="-1"></a>           <span class="at">Gtc =</span> <span class="fu">matrix</span>(<span class="dv">3</span>), <span class="co">#disable variance estimation (the model will be forced</span></span>
<span id="cb30-23"><a href="#cb30-23" tabindex="-1"></a>           <span class="co">#to use the value provided in Gti argument)</span></span>
<span id="cb30-24"><a href="#cb30-24" tabindex="-1"></a>           <span class="at">Gu =</span> EEV_diag), <span class="co">#diagonal of EEV as variance-covariance matrix</span></span>
<span id="cb30-25"><a href="#cb30-25" tabindex="-1"></a>     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units), <span class="co">#iid residuals</span></span>
<span id="cb30-26"><a href="#cb30-26" tabindex="-1"></a>     <span class="at">data=</span> sommerdf, </span>
<span id="cb30-27"><a href="#cb30-27" tabindex="-1"></a>     <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb30-28"><a href="#cb30-28" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;id&#39;</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co">#Variance values from fully efficient, two stage model, diagonal EEV in random effect</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;DiagRaneff&quot;</span>, </span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a>              Stage2ModelDiagRaneff<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:Env</span><span class="st">`</span>, </span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>              Stage2ModelDiagRaneff<span class="sc">$</span>sigma<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span>,</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a>              <span class="cn">NA</span>)</span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a>variance_components <span class="ot">&lt;-</span> <span class="fu">rbind</span>(variance_components, tmp)</span>
<span id="cb32-7"><a href="#cb32-7" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" tabindex="-1"></a><span class="co">#genomic estimated breeding values from two stage model with diagonal EEV in random effect</span></span>
<span id="cb32-9"><a href="#cb32-9" tabindex="-1"></a>GEBVs_DiagRaneff <span class="ot">&lt;-</span> Stage2ModelDiagRaneff<span class="sc">$</span>U<span class="sc">$</span><span class="st">`</span><span class="at">u:id</span><span class="st">`</span><span class="sc">$</span>BLUEs</span></code></pre></div>
</div>
</div>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<p>Now, we will compare the results from the different models. Please,
keep in mind that these results are based on a single repetition of a
simulation and therefore they are just orientative. In the main paper we
have the results for 500 iterations.</p>
<div id="variance-components" class="section level2">
<h2>Variance Components</h2>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co">#Some formatting:</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a><span class="fu">rownames</span>(variance_components) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a><span class="fu">colnames</span>(variance_components) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Model&quot;</span>,</span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a>                                   <span class="st">&quot;Environment&quot;</span>,</span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a>                                   <span class="st">&quot;Additive&quot;</span>,</span>
<span id="cb33-6"><a href="#cb33-6" tabindex="-1"></a>                                   <span class="st">&quot;Non_Additive&quot;</span>)</span>
<span id="cb33-7"><a href="#cb33-7" tabindex="-1"></a>variance_components<span class="sc">$</span>Environment <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">as.numeric</span>(variance_components<span class="sc">$</span>Environment),<span class="dv">2</span>)</span>
<span id="cb33-8"><a href="#cb33-8" tabindex="-1"></a>variance_components<span class="sc">$</span>Additive <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">as.numeric</span>(variance_components<span class="sc">$</span>Additive),<span class="dv">2</span>)</span>
<span id="cb33-9"><a href="#cb33-9" tabindex="-1"></a>variance_components<span class="sc">$</span>Non_Additive <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">as.numeric</span>(variance_components<span class="sc">$</span>Non_Additive),<span class="dv">2</span>)</span>
<span id="cb33-10"><a href="#cb33-10" tabindex="-1"></a><span class="fu">print</span>(variance_components)</span></code></pre></div>
<pre><code>##          Model Environment Additive Non_Additive
## 1         True        2.15     0.75         0.71
## 2 Single_Stage        3.32     2.29           NA
## 3   Unweighted        3.54     1.62           NA
## 4    FullResid        3.42    14.52           NA
## 5    DiagResid        4.62    29.82           NA
## 6   FullRaneff        3.23     1.37           NA
## 7   DiagRaneff        3.44     1.24           NA</code></pre>
<p>All models overestimate additive variance. This is expected, as in
the simulation there was a very substantial non-additive variance that
has not been included in any of the models. Therefore, some of the
non_additive variance was absorbed by the additive term. Furthermore,
noise can also be absorbed by the additive term, further increasing its
variance. Much better estimates of the additive variance could be
obtained with models that included a term for non-additive variance.</p>
<p>The main highlight from this results is that the models with EEV in
the residuals (“FullResid”, “DiagResid”) had a much more over-inflated
additive variance than the rest. The reason for this is that all other
models had i.i.d. residuals able to capture part of the non-additive
variance, precluding it from being wrongly included into the additive
term. However, in “FullResid” and “DiagResid” the EEV in the residuals
constrained them, and it did not allow them to capture the non-additive
effects. As a result, the non-additive effects did not fit in any of the
terms of the “FullResid” and “DiagResid”, causing them to break down and
having a very poor fit.</p>
<p>From this, we can conclude that we should only include EEV in the
residuals if we are confident that all terms influencing the phenotype
have been explicitly modeled in fixed or random effects. In general,
including EEV in a random effect is much safer.</p>
</div>
<div id="accuracy" class="section level2">
<h2>Accuracy</h2>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_SingleStage[<span class="fu">names</span>(TBV)]) <span class="co">#Single Stage</span></span></code></pre></div>
<pre><code>## [1] 0.772422</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_Unweighted[<span class="fu">names</span>(TBV)]) <span class="co">#Unweighted two stage</span></span></code></pre></div>
<pre><code>## [1] 0.7735591</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_FullResid[<span class="fu">names</span>(TBV)]) <span class="co">#Two stage, full EEV in residuals</span></span></code></pre></div>
<pre><code>## [1] 0.5351415</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_DiagResid[<span class="fu">names</span>(TBV)]) <span class="co">#Two stage, diagonal EEV in residuals</span></span></code></pre></div>
<pre><code>## [1] 0.4088922</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_FullRaneff[<span class="fu">names</span>(TBV)]) <span class="co">#Two stage, full EEV in random effects</span></span></code></pre></div>
<pre><code>## [1] 0.7838959</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">cor</span>(TBV, GEBVs_DiagRaneff[<span class="fu">names</span>(TBV)]) <span class="co">#Two stage, diagonal EEV in random effects</span></span></code></pre></div>
<pre><code>## [1] 0.7791556</code></pre>
<p>As expected, the models with EEV in residuals performed poorly due to
the reasons outlined in the previous section. Conversely, the models
with EEV in a random effect achieved the best accuracy. Single stage and
unweighted models a slightly worse performance than the models with EEV
in a random effect. This was expected for the unweighted model, but it
is unexpected for the single stage model, as it should be in theory the
best performing one. However, the differences between single stage,
unweighted and fully efficient with EEV in the residuals are very small
and we cannot conclude if their differences are significant with a
single repetition of the simulation.</p>
<p>In the main paper we performed 500 repetitions, which confirmed the
poor performance of the models with EEV in the residuals. Furthermore,
single stage models and models with EEV in a random effect were not
significantly different, showcasing the robustness of fully efficient
models. In contrast, the unweighted model was significantly worse than
the single stage, which is in line with our initial expectations.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
