<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Vignette 1: Single trait analysis with homogeneous GxE</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Vignette 1: Single trait analysis with
homogeneous GxE</h1>



<p>Libraries for open source:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(Matrix) <span class="co">#all StageWise models in the vignettes work for Matrix package </span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#version 1.5-1. Later versions may cause errors in some of the StageWise models</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(lme4)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(StageWise)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(matrixcalc)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(sommer)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(pracma)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;Vignette1Models.RData&quot;</span>)</span></code></pre></div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>StageWise package allows you to build fully-efficient two-stage
models on the multi-trait, multi-environment data. A fully efficient
two-stage model is equivalent to its corresponding single-stage model,
but it is much faster to fit.</p>
<p>The main advantages of StageWise include its ease of use and its
ability to handle different degrees of ploidy. It also allows for
modeling genotype-by-environment interactions in multi-location,
multi-year trials and supports multi-trait analyses. However,
<strong>its primary disadvantage is its reliance on ASReml for variance
estimations, which requires a license</strong>. Furthermore, while
StageWise is versatile and applicable in most situations, understanding
how to perform all modeling steps manually opens up additional
possibilities. As a result, the aim of this work is twofold:</p>
<ul>
<li><p>Replicate StageWise results using open-source software,
specifically the lme4 and Sommer R packages.</p></li>
<li><p>Provide ample theoretical background and comprehensive,
step-by-step explanations to enable readers to gain a deep understanding
of the process, empowering them to employ modeling strategies beyond
those included in the StageWise package, if the situation demand
it.</p></li>
</ul>
<p>In this document, we will explain the bulk of the theoretical
foundation and reproduce the results found in StageWise Vignette1.
Subsequent documents will focus on genotype-by-location interactions and
multitrait modeling.}</p>
<p>If you choose to cite this work, <strong>it is important to also
include a citation for the StageWise package (Endelman, J. B.,
2023)</strong> since we expand upon the vignettes he‚Äôs published,
offering open-source alternatives. For a more in-depth exploration of
the fully-efficient two-stage modeling strategy, we recommend referring
to Damesa et al.¬†(2017) and also Piepho et al (2012)} .</p>
</div>
<div id="theoretical-background" class="section level1">
<h1>Theoretical background</h1>
<p>Here, we present the theory related to the scenario covered in
StageWise vignette1: single location, single trait, several years. The
modelling is divided in two stages:</p>
<div id="stage-1-model" class="section level3">
<h3>Stage 1 model:</h3>
<p><span class="math display">\[
\boldsymbol{y_j} = X_{g_j}\boldsymbol{g_j} +
X_{Co_j}\boldsymbol{CoFix_j} + Z_{Co_j}\boldsymbol{CoRand_j} +
\boldsymbol{\epsilon_j}
\]</span></p>
<p>Where <span class="math inline">\(\boldsymbol{y_j}\)</span> is a
vector of the phenotypes in environment <span class="math inline">\(j\)</span>, <span class="math inline">\(\boldsymbol{g_j}\)</span> is a vector of fixed
genotypic effects, <span class="math inline">\(\boldsymbol{CoFix_j}\)</span> and <span class="math inline">\(\boldsymbol{CoRand_j}\)</span> refer to any
additional fixed or random covarieties or cofactors that may be included
in the model (e.g.¬†block effect) and <span class="math inline">\(\boldsymbol{\epsilon_j} \sim N(\boldsymbol{0},
I\sigma^2_{\epsilon_j})\)</span> is a vector of residuals. <span class="math inline">\(X_{g_j}\)</span>, <span class="math inline">\(X_{Co_j}\)</span> and <span class="math inline">\(Z_{Co_j}\)</span> are design matrices.</p>
<p><strong>An independent stage 1 model is performed for each
environment. The BLUEs from each environment (contained in <span class="math inline">\(\boldsymbol{g_j}\)</span>) are concatenated into a
single vector that will be used as input for the next step</strong>.
These BLUEs have a estimation error variance-covariance structure (<span class="math inline">\(EEV_j\)</span>) associated with them (equivalent
to the prediction error variance-covariance of random effects) that can
be extracted and passed to the next stage, with large values of the
error variance reflecting a low reliability of the estimations. The
calculation of <span class="math inline">\(EEV_j\)</span> is usually
done automatically by the modelling packages, but in a few situations,
it may be necessary to calculate it manually. Furthermore, knowing how
it is calculated is very helpful for understanding its meaning.</p>
<p>From Robinson, et al., 1987, we have the formula for fixed and random
effects estimation/prediction error variance in a mixed model following
Henderson equations:</p>
<p><span class="math display">\[
Var \left( \begin{matrix} \boldsymbol{\beta_j} -
\boldsymbol{\hat{\beta_j}} \\ \boldsymbol{u_j} - \boldsymbol{\hat{u_j}}
\end{matrix} \right) =   \left( \begin{matrix} X_j&#39;R_j^{-1}X_j &amp;
X_j&#39;R_j^{-1}Z_j \\ Z_j&#39;R_j^{-1}X_j &amp; Z_j&#39;R_j^{-1}Z_j +
\lambda_j K_j^{-1} \end{matrix} \right)^{-1}*\sigma^2_{\epsilon_j} =
\left( \begin{matrix} C_{11} &amp; C_{12} \\ C_{21} &amp; C_{22}
\end{matrix} \right)*\sigma^2_{\epsilon_j}
\]</span> <span class="math display">\[
EEV_j = C_{11}*\sigma^2_{\epsilon_j}
\]</span> Where <span class="math inline">\(\boldsymbol{\beta_j}\)</span> is a vector of fixed
effects in environment <span class="math inline">\(j\)</span>, <span class="math inline">\(\boldsymbol{\epsilon_j} \sim
N(\boldsymbol{0},\sigma^2_{\epsilon_j}*R_j)\)</span> which follows a
normal distribution with mean <span class="math inline">\(\boldsymbol{0}\)</span> and variance <span class="math inline">\(\sigma^2_{\epsilon_j}*R_j\)</span>, represents the
residuals, <span class="math inline">\(\boldsymbol{u} \sim
N(\boldsymbol{0},\sigma^2_{u_j}*K_j)\)</span> are the random effects,
<span class="math inline">\(R_j\)</span> and <span class="math inline">\(K_j\)</span> are the variance-covariance
structures for residuals and random effects respectively, <span class="math inline">\(\sigma^2_{\epsilon_j}\)</span> and <span class="math inline">\(\sigma^2_{u_j}\)</span> are the variance
components for residuals and random effects, <span class="math inline">\(X_j\)</span> and <span class="math inline">\(Z_j\)</span> are design matrices for fixed and
random effects respectively, <span class="math inline">\(\lambda_j =
\frac{\sigma^2_{\epsilon_j}}{\sigma^2_{u_j}}\)</span> is the residual
variance divided by the random effect variance. <span class="math inline">\(EEV_j\)</span> is the estimation error
variance-covariance <strong>(EEV)</strong> matrix of the fixed effects
in environment <span class="math inline">\(j\)</span>. In most
situations, <span class="math inline">\(R_j\)</span> is the identity
matrix and it can therefore be ignored</p>
<p><strong>In the first stage modelling, the effect that must always be
present is the fixed genotypic effect</strong>, but there may also be
several fixed and random effects.</p>
<p>If several fixed or random effects are included, the equation above
still holds. To accommodate multiple effects, we can concatenate the
design matrices and calculate the direct sum of the variance-covariance
matrices for the random effects. Imagine that there are two different
fixed effects <span class="math inline">\(\boldsymbol{\beta_{1_j}}\)</span> and <span class="math inline">\(\boldsymbol{\beta_{2_j}}\)</span> with design
matrices <span class="math inline">\(X_{1_j}\)</span> and <span class="math inline">\(X_{2_j}\)</span>. There are also two different
random effects <span class="math inline">\(\boldsymbol{u_{1_j}} \sim
N(\boldsymbol{0},\sigma^2_{1_j}*K_{1_j})\)</span> and <span class="math inline">\(\boldsymbol{u_{2_j}} \sim
N(\boldsymbol{0},\sigma^2_{2_j}*K_{2_j})\)</span> with design matrices
<span class="math inline">\(Z_{1_j}\)</span> and <span class="math inline">\(Z_{2_j}\)</span>. We can merge the fixed and
random effects into a single one as follows:</p>
<ul>
<li>Vector Creation: <strong>Create combined vectors for each set of
effects:</strong>
<ul>
<li><span class="math inline">\(\beta_j=
c(\beta_{1j},\beta_{2j})\)</span> for beta coefficients.</li>
<li><span class="math inline">\(u_j=c(u_{1j},u_{2j})\)</span> for ‚Äòu‚Äô
values.</li>
</ul></li>
<li>Matrix Binding <strong>Bind matrices for X and Z values</strong>
<ul>
<li><span class="math inline">\(Xj=cbind(X_{1j},X_{2j})\)</span> for X
matrices.</li>
<li><span class="math inline">\(Zj=cbind(Z_{1j}, Z_{2j})\)</span> for Z
matrices.</li>
</ul></li>
</ul>
<p>The merged random effects would have the following distribution,
which means <span class="math inline">\(\boldsymbol{u_j} \sim(0,\left(
\begin{matrix} \sigma^2_{1_j}*K_{1_j} &amp; 0 \\ 0 &amp;
\sigma^2_{2_j}*K_{2_j} \end{matrix} \right))\)</span>. That way, in the
equation for the calculation of <span class="math inline">\(EEV_j\)</span>, <span class="math inline">\(\lambda_j K_j^{-1}\)</span> can be replaced by
<span class="math inline">\(\left( \begin{matrix}
\frac{\sigma^2_{\epsilon_j}}{\sigma^2_{1_j}}*K_{1_j}^{-1} &amp; 0 \\ 0
&amp; \frac{\sigma^2_{\epsilon_j}}{\sigma^2_{2_j}}*K_{2_j}^{-1}
\end{matrix} \right)\)</span></p>
<p>In contrast, if only the fixed genotypic effects are included in the
model, the equation becomes very simple:</p>
<p><span class="math display">\[
EEV_j = Var (\boldsymbol{\beta_j} - \boldsymbol{\hat{\beta_j}}) =
(X_j&#39;R_j^{-1}X_j)^{-1}\sigma^2_{\epsilon_j}
\]</span> If <span class="math inline">\(R_j\)</span> is assumed to be
the identity, the equation above coincides with the variance for <span class="math inline">\(\boldsymbol{\beta_j}\)</span> when solved by
ordinary least squares (<span class="math inline">\((X_j&#39;X_j)^{-1}\sigma^2_{\epsilon_j}\)</span>).
<strong>Please note that <span class="math inline">\(X&#39;X\)</span> is
a diagonal matrix counting the number of times each genotype is observed
in environment <span class="math inline">\(j\)</span></strong>.
Therefore, <span class="math inline">\((X_j&#39;X_j)^{-1}\)</span> is a
diagonal matrix containing the inverse of the number of times each
genotype is present. The reliability of the estimations is thus higher
for genotypes frequently observed in our data.</p>
</div>
<div id="stage-2-model-without-marker-data" class="section level3">
<h3>Stage 2 model without marker data:</h3>
<p><span class="math display">\[
BLUEs = \boldsymbol{y} = X\boldsymbol{E} + Z\boldsymbol{g} +
\boldsymbol{s} + \boldsymbol{gE}
\]</span> Where <span class="math inline">\(BLUEs =
\boldsymbol{y}\)</span> is a vector of stage 1 genotypic BLUEs (BLUEs of
different environments are concatenated), <span class="math inline">\(\boldsymbol{E}\)</span> is a vector of fixed
environmental effects, <span class="math inline">\(\boldsymbol{g} \sim
N(\boldsymbol{0},I\sigma^2_g)\)</span> is a vector of random genotypic
effects, <span class="math inline">\(\boldsymbol{s} \sim
N(\boldsymbol{0},EEV)\)</span> is a vector of random Stage 1 estimation
errors and <span class="math inline">\(\boldsymbol{gE} \sim
N(\boldsymbol{0},I\sigma^2_{ge})\)</span> is a vector of residuals that
correspond to genotype by environment interaction unexplained by the
other terms. <span class="math inline">\(X\)</span> and <span class="math inline">\(Z\)</span> are design matrices for their
corresponding effects. <span class="math inline">\(\sigma^2_g\)</span>
and <span class="math inline">\(\sigma^2_{ge}\)</span> are the genotypic
and residual variances respectively and <span class="math inline">\(I\)</span> is an identity matrix with the
appropriate dimensions.</p>
<p>The <span class="math inline">\(EEV\)</span> matrix can be obtained
as the <em>direct sum</em> of the <span class="math inline">\(EEV_j\)</span> matrices from each environment:</p>
<p><span class="math display">\[
EEV = \left( \begin{matrix} EEV_1 &amp; 0 &amp; \dots &amp; 0 \\ 0 &amp;
EEV_2 &amp; &amp;  \\ \vdots &amp;  &amp; \ddots &amp;  \\ 0
&amp;  &amp; &amp; EEV_j \end{matrix} \right)
\]</span></p>
<p>Where <span class="math inline">\(EEV_1\)</span> to <span class="math inline">\(EEV_j\)</span> are the EEV matrices for the fixed
genotypic effects in the environments 1 to <span class="math inline">\(j\)</span>. The environments have to be in the
same order during the concatenation of their BLUEs vectors and during
the construction of <span class="math inline">\(EEV\)</span>.</p>
<p><strong>It is important to note that <span class="math inline">\(s\)</span> has no free variance
parameters</strong>. Until now, we have expressed the
variance-covariance of all random effects as the product of its variance
component and its variance-covariance structure. For example, in <span class="math inline">\(\boldsymbol{g} \sim
N(\boldsymbol{0},I\sigma^2_g),\)</span> <span class="math inline">\(I\)</span> is the variance-covariance structure,
and <span class="math inline">\(\sigma^2_g\)</span> is the variance
component. However, the <span class="math inline">\(EEV\)</span> matrix
we have obtained from the Stage1Errors is the result of the product
between its variance-covariance structure and its variance component. If
we want to know the variance component associated with the effect <span class="math inline">\(s\)</span>, we need a way to undo this product. To
that end, we need a way to summarize, in a scalar number (the variance
component), the overall amount of variation present in a
variance-covariance matrix. A common way to do this is by using the
determinant of the matrix, which returns a large scalar for high values
on the diagonal (variances) and low values in the off-diagonals
(covariances). In other words, high overall variance is reached when the
Stage1Errors have a high variance and are uncorrelated with each other.
Instead of using the determinant, StageWise calculates the overall
variance as the mean of the diagonal elements of the matrix minus the
mean of the entire matrix. This is an easier-to-compute approach to the
determinant while being conceptually similar. Therefore, we can easily
split <span class="math inline">\(EEV\)</span> into its variance
component <span class="math inline">\(\sigma^2_s\)</span> and its
variance-covariance structure <span class="math inline">\(EEV_{str},\)</span> where the variance-covariance
structure is the full variance-covariance matrix scaled to have a total
amount of variation equal to one, and the variance component is the
scaling factor</p>
<p><span class="math display">\[
\sigma^2_s = mean(diag(EEV)) - mean(EEV)
\]</span></p>
<p><span class="math display">\[
EEV_{str} = EEV/\sigma^2_s
\]</span></p>
<p><span class="math display">\[
s \sim N(\boldsymbol{0},EEV) = N(\boldsymbol{0},\sigma^2_s*EEV_{str})
\]</span></p>
<p>When we say that <span class="math inline">\(\boldsymbol{s}\)</span>
has no free variance parameters, we mean that <span class="math inline">\(\sigma^2_s\)</span> is known and provided as an
input to the model, it is not estimated.</p>
</div>
<div id="stage-2-model-with-marker-data" class="section level3">
<h3>Stage 2 model with marker data:</h3>
<p>It is similar to the previous model but the genotypic effects are
split into additive and dominance effects, with variance-covariance
matrices calculated from marker data:</p>
<p><span class="math display">\[
BLUEs = \boldsymbol{y} = X\boldsymbol{E} + Z\boldsymbol{a} +
Z\boldsymbol{d} -b\boldsymbol{F} + \boldsymbol{s} + \boldsymbol{gE}
\]</span> Where <span class="math inline">\(\boldsymbol{a} \sim
N(\boldsymbol{0},G\sigma^2_a)\)</span> is a vector of random additive
effects, <span class="math inline">\(\boldsymbol{d} \sim
N(\boldsymbol{0},D\sigma^2_d)\)</span> is a vector of random dominance
effects in the absence of heterosis, and <span class="math inline">\(b\boldsymbol{F}\)</span> is a vector of fixed
heterosis effects. <span class="math inline">\(G\)</span> is the
additive relationship matrix corrected for ploidy (the classical
VanRaden, 2008 relationship matrix multiplied by <span class="math inline">\(2\)</span> and divided by the degree of ploidy
<span class="math inline">\(\phi\)</span>). <span class="math inline">\(D\)</span> is the dominance relationship matrix
corrected for ploidy. To compute it, a <span class="math inline">\(Q\)</span> matrix of dominance coefficients is
calculated by substituting each element in the marker matrix <span class="math inline">\(M\)</span> as follows:</p>
<p><span class="math display">\[
Q[i,j]=‚àí2(\binom{\phi}{2})ùëù^2+2ùëù(ùúô‚àí1)M[i,j]‚àíM[i,j](M[i,j]‚àí1)
\]</span> <span class="math display">\[
D= \frac{QQ^T}{ \binom{\phi}{2} \sum_k 4 p^2_k q^2_k}
\]</span></p>
<p>Where <span class="math inline">\(D\)</span> is the dominance
relationship matrix, <span class="math inline">\(\phi\)</span> is the
ploidy number, <span class="math inline">\(p\)</span> is the minor
allele frequency, <span class="math inline">\(q = 1-p\)</span> is the
major allele frequency, and <span class="math inline">\(i\)</span> and
<span class="math inline">\(j\)</span> indicate the positions in the
<span class="math inline">\(Q\)</span> and <span class="math inline">\(M\)</span> matrices. <span class="math inline">\(M\)</span> is a numeric marker matrix containing
the number of copies of the minor allele at each locus for each
genotype. This is a generalized equation for any ploidy number, but if
we assume a diploid genotype (<span class="math inline">\(\phi =
2\)</span>, with <span class="math inline">\(M\)</span> coded as 0, 1,
2; with 2 indicating homozygosity for the minor allele), and we
elaborate from the <span class="math inline">\(Q[i,j]\)</span> equation,
we arrive at the following substitutions:</p>
<p><span class="math display">\[
M[i,j] = 0 \rightarrow Q[i,j] = -2p^2
\]</span> <span class="math display">\[
M[i,j] = 1 \rightarrow Q[i,j] = 2pq
\]</span> <span class="math display">\[
M[i,j] = 2 \rightarrow Q[i,j] = -2q^2
\]</span> Which is equivalent to the classical Vitezica et al., 2013
calculation of the dominance relationship matrix.</p>
<p>The heterosis effects <span class="math inline">\(-b\boldsymbol{F}\)</span> can be calculated in
several ways, although here we will just focus on one. <span class="math inline">\(\boldsymbol{F}\)</span> is a vector of inbreeding
coefficients that can be extracted from the main diagonal of <span class="math inline">\(G\)</span> as:</p>
<p><span class="math display">\[
\boldsymbol{F} = \frac{diag(G)-1}{\phi - 1}
\]</span> <span class="math inline">\(b\)</span> is a regression
coefficient that is estimated by the model. if <span class="math inline">\(b &gt; 0\)</span>, there is heterosis.</p>
</div>
</div>
<div id="replicate-stagewise-vignette1" class="section level1">
<h1>Replicate StageWise Vignette1:</h1>
<p>Here, we include StageWise Vignette1 and we will explain how to
replicate it with open-source software as we go through it.</p>
<div id="preface" class="section level3">
<h3>Preface</h3>
<p>This vignette illustrates basic features of the package using a
potato breeding dataset of 943 genotypes (i.e., clones), with phenotype
data from six years at one location. It is an updated version of the
dataset published by <a href="https://doi.org/10.1534/genetics.118.300685">Endelman et
al.¬†(2018)</a>. This vignette covers single trait analysis, under the
assumption that all environments have the same genetic correlation
(i.e., compound symmetry). The analysis of multiple locations with
different correlations are covered in <a href="https://jendelman.github.io/StageWise/Vignette2.html">Vignette
2</a>, and the analysis of correlated traits is covered in <a href="https://jendelman.github.io/StageWise/Vignette3.html">Vignette
3</a>.</p>
<p>There are five main functions in the package:</p>
<ul>
<li><p><code>read_geno</code></p></li>
<li><p><code>Stage1</code></p></li>
<li><p><code>Stage2</code></p></li>
<li><p><code>blup_prep</code></p></li>
<li><p><code>blup</code></p></li>
</ul>
<p>The package depends on ASReml-R (version 4.1.0.148 or later), which
requires a license from <a href="https://www.vsni.co.uk/software/asreml-r">VSN
International</a>.</p>
</div>
<div id="stage-1" class="section level3">
<h3>Stage 1</h3>
<p>In Stage 1, the data for each environment are analyzed independently,
which allows for the selection of different models tailored to different
experimental designs and patterns of spatial variation. The
<code>Stage1</code> function in the package offers a variety of commonly
used analysis methods. However, it is also possible to use other
software for Stage 1. For a linear model with only fixed and i.i.d.
random effects, for example, the argument <code>solver=&quot;asreml&quot;</code>
triggers the use of ASReml-R for variance component estimation. Another
option is <code>solver=&quot;spats&quot;</code>, which triggers the use of R
package <a href="https://cran.r-project.org/web/packages/SpATS/index.html">SpATS</a>
to fit a 2D spline (in addition to fixed or i.i.d. effects). Regardless
of which solver is used, at least some of the individuals should be
replicated (which includes augmented designs with repeated checks). If
you have no replication within environment, skip Stage 1 and go to Stage
2.</p>
<p>The CSV file of phenotype data used in <code>Stage1</code> should
have two required columns: ‚Äòid,‚Äô which contains the individual
identifier and is matched against the information from the genotype
input file; and ‚Äòenv,‚Äô which represents the name of the environment,
typically in the format of a location x year combination (This vignette
illustrates the analysis of multiple years from one location, or when
multiple locations are similar enough that the genotype x location
effect can be neglected. For the analysis of correlated locations,
consult <a href="https://jendelman.github.io/StageWise/Vignette2.html">Vignette
2</a> after you complete this vignette.) The other columns in the input
file contain traits, cofactors, or covariates.</p>
<p>The phenotypes for this tutorial are based on six years (2015-2020)
of variety trials at the Hancock Research Station of the University of
Wisconsin. Data for the first five years are in the file ‚Äúpheno1a‚Äù,
which includes a column with an incomplete blocking factor. The 2020
data are provided in the file ‚Äúpheno1b‚Äù and consists of two partially
replicated trials (preliminary and advanced), with row and range
information to illustrate spatial analysis.</p>
<ul>
<li><p>total yield (Mg per ha)</p></li>
<li><p>vine maturity (1=early to 9=late)</p></li>
<li><p>fry color, measured in units of Hunter Lightness (L) after 6
months of storage</p></li>
</ul>
<p>The data files also contain the stand count for each plot (out of 15
total plants), which is included as a covariate in the Stage 1
model.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>pheno1a.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;pheno1a.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>pheno1a <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(pheno1a.file)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">head</span>(pheno1a)</span></code></pre></div>
<pre><code>##         env         id block total.yield vine.maturity fry.color stand.count
## 1 Hancock15  A00188-3C     2        68.1             8      47.8          15
## 2 Hancock15  A00188-3C     3        71.2             8      47.3          15
## 3 Hancock15  A01143-3C     2       109.2             8      48.4          15
## 4 Hancock15  A01143-3C     3       105.3             9      51.9          15
## 5 Hancock15  A09037-6C     1        70.3             5      53.9          15
## 6 Hancock15 AAF07847-2     1        77.6             8      49.0          15</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>pheno1b.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;pheno1b.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>pheno1b <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(pheno1b.file)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">head</span>(pheno1b)</span></code></pre></div>
<pre><code>##         env   expt row range        id total.yield vine.maturity fry.color
## 1 Hancock20 prelim   1     1 W17037-24        44.7             3      58.0
## 2 Hancock20 prelim   1     2 W17062-11        38.8             4      61.5
## 3 Hancock20 prelim   1     3  W17064-7        36.8             2      61.2
## 4 Hancock20 prelim   1     4 W17039-15        45.4             3      59.6
## 5 Hancock20 prelim   1     5 W17039-30        47.1             5      61.9
## 6 Hancock20 prelim   1     6  W17039-7        64.8             3      61.4
##   stand.count
## 1          15
## 2          13
## 3          15
## 4          15
## 5          15
## 6          15</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">merge</span>(pheno1a[,<span class="fu">c</span>(<span class="st">&quot;env&quot;</span>,<span class="st">&quot;id&quot;</span>)],pheno1b[,<span class="fu">c</span>(<span class="st">&quot;env&quot;</span>,<span class="st">&quot;id&quot;</span>)],<span class="at">all=</span>T)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>tmp,<span class="fu">aes</span>(<span class="at">x=</span>env)) <span class="sc">+</span> <span class="fu">geom_bar</span>() <span class="sc">+</span> </span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Number of plots&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Environment&quot;</span>) <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">vjust=</span><span class="fl">0.5</span>,<span class="at">size=</span><span class="dv">10</span>)) </span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAGACAMAAACTGUWNAAAAxlBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6AGY6kNtNTU1NTW5NTY5NbqtNjshZWVlmAABmADpmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2Ojm6OyP+QOgCQkDqQtpCQ2/+rbk2rbm6rq+SryKur5OSr5P+2ZgC225C22/+2///Ijk3I/+TI///bkDrb/7bb/9vb///kq27k5Kvk///r6+v/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///8CvCCuAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALPUlEQVR4nO2dC1cbxwFGhU2xWeOUQB4Qt6IBN4UGgtKotkTFY///n+q+AKHsSPPcT2Lvd46tg8RlRnM1j9W+BjmRZqCuQN+DAHEQIA4CxEGAOAgQx0/AX0hwwgQsf/mrx5/0YTorKAGDADFjJeDhbJjnd8fZxy9PDy9pv7KjMevSmD6MlYBxNqwkjL95fFig/cqOxqxLY/owNgJmP/xtmN99us5n3183Dwu0X9nRmHVpTB/GQsDD538VH/vZj1/yu5/Om4e8WQF9JYGxEDA+Ksed6ceq5ZuHBX1+8qMx6/Jp9mFWCyg+8w9tPWCe9is7GrMujenDrBYwzsocMQekYayXoQ9nR/Uq6IhVUEyG7QAxw5ZwLOavbnnEEBCLQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYkQrwLNz1PUaFEOD+HmOWgwCP94iAxSAguG4IcAkC4pWDAI/3iIDFICC4bghwyfoJaM4ycCzc+2yGrsrpsG70AJesXw8IK9z1PSJgMQgIrhsCXIKAeOUgwOM9ImAxmyAgYt0Q4FNOxLohwKeciHVDgE85EeuGAJ9yItYNAT7lRKwbAnzKiVg3BPiUE7FuCPApJ2LdEOBTTsS6IcCnnIh1Q4BPORHrhgCfciLWDQE+5awzgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxYydgmmV718sv3NpVhV8bYyWgvFL0850b2m/g0FWFXxtj1wNqCUsv3t1VhV8bYy2g+MwvvYGDY+EwbucHzA4/nC+/gYNj4TCuPeD5o99+A4euKvzaGGsB+dWQOSABYyWgGXSW3sChqwq/NsauB4yzrJgD2A5IwNgJWBIEhDEIEDMIEDMIEDMIEDMIEDMIEDMIEDMIEDMIEDMIEDMmAbf7B7f7gzeXCEjMmARcbOejN5ejbQQkZgwCig5wf7KdT1Z3AQSEMWYBt/u7CEjPGATcn+xOtk7LgQgBaRmDgPxmZ7CdX7z9AwGJGZMA6yAgjDEIKOaA8oE5IDmDADHTKmA0eAyTcGqmVcBTD7AIAsIYgwD7ICCMMQqohqFdBKRmTAJG5fRbbgwjIC1jEMAqqCsGAWLGIMBhCOIEjSDGeIIGk3A3jKkHWAcBYQwCxEybgNv9p68imIRTM20CnIKAMAYBYsYooFoFWXwjh4AwxiSAryI6YgwC2BLuikGAmDEIyMuDUhiCOmAMAuw3BRAQxhgE2AcBYQwCxAwCxAwCxAwCxEybgOrYdI4L6oZBgJhpEzB3aCJbwqmZVgEcmigXYB8EhDFGARwV0Q1jEsD+gI4YgwC+ju6KQYCYMQhgCOqKMQlgEu6IMQp4kdlhlg25cGsKxkpAeanu2Xfn3MAhAWMlYFo2NxfvTsJYCWh6ATdwSMBY3sChumQ6N3BIwBh6wP3Jyy/j7o6P8j/fwgQB4YxBwMK3obPDYWmBOSA+YxCQj+avVFO3f84NHBIwxh4wv0NmnJUZsh2QgDH1AOsgIIxBgJgxChgNBgcjLlmWnDEJuHj7e33lSgSkZQwCmiNT2B+QnEGAmDEIKLYDfq+v3YqAtIxJQD5hh0wnjFGAbRAQxiBAzBgFTDhRuxPGJICjIjpiDAI4LqgrBgFixiCgbnqbg9QREMa0CeCCTWIBTkFAGIMAMWMScLPDENQJYxBgsycAATEYgwBO0uuKMQhYPDALAakYgwCbTTAExGBMApiEO2IMAu5PLPbFICACYxDAJNwVYxDAJNwVYxCQ37y3nYQ5QSOIMZyg8fLgXHpAOsbUA6yDgDAGAWLGIIAhqCvGIKDR8O3pqvZHQCCzVEA+WX18OgLCmBUCGIJSM8sFWNxUHgFhjEFAMwlvMQekZgwC7IOAMAYBYqZNAMcFiQU85cLiDA0EhDFLBNzuW8zBCAhkzAImA4uzhBEQyhgFXAzsjgxCQBhjEHB/YnN2DALCmXYBNzu2R6UgIJBpFTCyHH4QEM60CWA7QCzAKQgIYxAgZhAgZhAgZhAgZhAgZhAgZhAgZmwFVNcq5sKt8RlLAdNs7zrnBg4JGDsBVx9+KXoAF+9OwFj2gKrFuYFDAsb2Bg6lAG7gkIAJ6gEICGdcBDAHJGBcBHADhwSMiwC2AxIwtgKMQUAYgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxgwAxEy6AEzSCGNsTNMwCmkfHwmEYgtaDQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCYQYCY8RHAhVsjMh4CuIFDTMZDABfvjsl4CGi/gQPxjquA9hs4GPLV8o+GMp0VlIDx7wHztF/Z0Zh1aUwfJs4c4Fd2NGZdGtOHcV8Ftd3Awa/saMy6NKYPE2c7wK/saMy6NKYPE2dL2K/saMy6NKYPgwAxgwAxgwAxgwAxgwAxgwAxEyiABCdIwCo/HTGdFZSQQYCYQYCYSSKA2AcB4iBAHASIE11Atd9ymmXZkQt0fJTPDrPsm9W/Oo845+442/vtzKmc+s1k2d716t98zlXF2LRBEgHT8g2O7Ruo3N1W7e8c27fM3U9/f95BZJurYdGcQ6dyqt+d/fjl7njoUE5TtavV5aQQUB7Elc/tQbZh6jZxY2aHTp/kupyqhzrWrTJXHpbgxORW5ayFgIfPTa9xbZhp9uF89a++YMqjCzwEHLkzuUjA8d519Wl26ebTD5W1ciiyLqd+kw9nLqOzzxBUtH3dRR1GvMeh52o1k2AVVMx0xefSouyXTBGXj/PzQUpuxbhPwldVxcZOs/BYNwkTtyAgRYqBsfr0K+YAktdzQDkGaybhrIn9oPnqmGp+mu5da3qAxdbHK2eaBcL0438lAlwWk6+UaZyNLXoNc0CKPJw1BhCw9kkh4O6f1f+fXLZcesskEXB8VK6Eneau3jJJhqByO8T1u+K+MonmgPVdIq4bw4aYmGEVJA4CxEkiYPad6/fn/WVSCKh3SbrsQeoxk2Q7oPouymEfap+ZJENQtQBz2yfZWybZhpjrlktfGVZB4iBAHIYgMZNmGeo+afWWYRkqZugBYoY5QMywChIHAeLwbaiY4dtQMcMyVMzwbaiYYRkqZlgFiYMAcZII8Dm3vK9MmmNDHU4q7zuTbhkKY5UkQ5DLedB9Z9IMQe5jZm8ZVkHiIECcNHOAx7Ktr0yiZej045crp5M7e8skWobefbI5Sxwm2f6Ah8/nbhXuLZNqj9jU7ap9/WVYBYmDAHE4S1LMJNoj5v71YV8ZBIgZBIgZBIgZBIgZVkFihu0AcRAgDgLEQYA4CBAHAeJsmID7k0GVN5cLL9y8O/X4c//7NUalgrJxAnYj/jU/a3GDAHE2VsDNu593BoOD+onRm3+/O715/49iaCoHqe2nl5vH3Zvqh2oEK0av5sXyyZg+vbK5Anbe/lE0/OWoeCieLT7MNzvb5S9sV/8eX64fB/UP5Qt5ATy9SA9wzeMkvFu08EE1hpSNWD9Uz0zK6XlSNnzzcvXY/FC9eLt/MM+qs3ECnoeg0+q/8pnyQ90056T4bD+17dxj9d+osffiRXE2XkDR5P8pnrQTUL44xyLAOS0Cbr/9+f3ls4Ct03oIahNQvZgjICAtAvKLetVTN+fTJNwm4P6k6AKFhcefi+lA+m7KbJyAehjfmmvZyWB+Sn1ahrYIqF7cOn3+uXKnzYYJeH1BgDgIEAcB4iBAHASIgwBxECAOAsRBgDj/B4201YD87wgbAAAAAElFTkSuQmCC" /><!-- --></p>
<p>As is typical of breeding trials, the majority of clones were tested
in one year and then dropped, but there is sufficient replication across
years to estimate genotype x env interactions.</p>
<p>A data frame with variables ‚Äúname‚Äù, ‚Äúfixed‚Äù, and ‚Äúfactor‚Äù is used to
specify which columns of the input file should be included as covariates
or cofactors, and whether the effects are fixed or random. We begin with
analysis of the 2015-2019 data, using ‚Äúblock‚Äù and ‚Äústand.count‚Äù as
cofactor and covariate, respectively:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>effects <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">name=</span><span class="fu">c</span>(<span class="st">&quot;block&quot;</span>,<span class="st">&quot;stand.count&quot;</span>),</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                      <span class="at">fixed=</span><span class="fu">c</span>(<span class="cn">FALSE</span>,<span class="cn">TRUE</span>),</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>                      <span class="at">factor=</span><span class="fu">c</span>(<span class="cn">TRUE</span>,<span class="cn">FALSE</span>))</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>effects</span></code></pre></div>
<pre><code>##          name fixed factor
## 1       block FALSE   TRUE
## 2 stand.count  TRUE  FALSE</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>ans1a <span class="ot">&lt;-</span> <span class="fu">Stage1</span>(<span class="at">filename=</span>pheno1a.file,<span class="at">traits=</span><span class="st">&quot;total.yield&quot;</span>,</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>                <span class="at">effects=</span>effects,<span class="at">solver=</span><span class="st">&quot;asreml&quot;</span>)</span></code></pre></div>
<pre><code>## Online License checked out Thu Jun  6 13:07:33 2024</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#one model per environment</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>EEVs <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>Envs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>H2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>residuals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)) {</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> pheno1a[pheno1a<span class="sc">$</span>env <span class="sc">==</span> enviro,]</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  <span class="co">#1) remove entries with missing values for yield</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>total.yield),]</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  </span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>  Envs <span class="ot">&lt;-</span> <span class="fu">c</span>(Envs, <span class="fu">rep</span>(enviro, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))))</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>  </span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>  phenoenv<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>id)</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  phenoenv<span class="sc">$</span>block <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>block) <span class="co">#cofactor</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  phenoenv<span class="sc">$</span>stand.count <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenoenv<span class="sc">$</span>stand.count) <span class="co">#covariate</span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  phenoenv<span class="sc">$</span>stand.count <span class="ot">&lt;-</span> <span class="fu">scale</span>(phenoenv<span class="sc">$</span>stand.count, <span class="at">scale =</span> <span class="cn">FALSE</span>, <span class="at">center =</span> <span class="cn">TRUE</span>) </span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="co">#Important to center covaraites!</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  <span class="co">#If you don&#39;t center the data, EEV calculation will be massively off!!!</span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  </span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>  <span class="co">#remove the intercept in the model!</span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>  <span class="co">#It&#39;s important for correct EEV calculation</span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>  </span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>  <span class="co">#Make sure the first fixed effect in the formula is id!</span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>  models[[enviro]] <span class="ot">&lt;-</span> <span class="fu">lmer</span>(total.yield <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span> id <span class="sc">+</span> stand.count <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>block) ,</span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>                        <span class="at">data=</span>phenoenv)</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a>  <span class="co">#If you get this message:</span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a>  <span class="co">#boundary (singular) fit: see help(&#39;isSingular&#39;)</span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a>  <span class="co">#It means that the estimated variance for the random effect was 0. It&#39;s fine</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>  <span class="co">#because we are interested in the fixed effects of id.</span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a>  </span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>  <span class="co">#extract H2</span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>  <span class="co">#random effect variances</span></span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>  variance_block <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>varcor<span class="sc">$</span>block)</span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>  variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models[[enviro]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>  </span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a>  <span class="co">#fixed effect variances</span></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a>  <span class="co">#1) calculate design matrix</span></span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a>  X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(total.yield<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoenv) <span class="co">#remove intercept.</span></span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a>  X.stand.count <span class="ot">&lt;-</span> phenoenv<span class="sc">$</span>stand.count <span class="co">#fixed effect for a covariate is a</span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a>  <span class="co">#regression. design matrix of a regression contains the values for the variable</span></span>
<span id="cb11-45"><a href="#cb11-45" tabindex="-1"></a>  <span class="fu">colnames</span>(X.stand.count) <span class="ot">&lt;-</span> <span class="st">&quot;stand.count&quot;</span></span>
<span id="cb11-46"><a href="#cb11-46" tabindex="-1"></a>  </span>
<span id="cb11-47"><a href="#cb11-47" tabindex="-1"></a>  <span class="co">#calculate fixed effects for each plot</span></span>
<span id="cb11-48"><a href="#cb11-48" tabindex="-1"></a>  <span class="co">#id effects:</span></span>
<span id="cb11-49"><a href="#cb11-49" tabindex="-1"></a>  <span class="co"># y = Xb + ... + epsilon</span></span>
<span id="cb11-50"><a href="#cb11-50" tabindex="-1"></a>  <span class="co"># b (vector of the genotypic fixed effects) can be retrieved from:</span></span>
<span id="cb11-51"><a href="#cb11-51" tabindex="-1"></a>  <span class="co">#&quot;models[[enviro]]@beta[1:length(unique(phenoenv$id))]&quot;</span></span>
<span id="cb11-52"><a href="#cb11-52" tabindex="-1"></a>  <span class="co"># X is the design matrix linking each obervation (each plot in the field) with</span></span>
<span id="cb11-53"><a href="#cb11-53" tabindex="-1"></a>  <span class="co"># its corresponding genotypic effect.</span></span>
<span id="cb11-54"><a href="#cb11-54" tabindex="-1"></a>  id_effects <span class="ot">&lt;-</span> X.id<span class="sc">%*%</span>models[[enviro]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))]</span>
<span id="cb11-55"><a href="#cb11-55" tabindex="-1"></a>  <span class="co">#stand.count value times its regression slope</span></span>
<span id="cb11-56"><a href="#cb11-56" tabindex="-1"></a>  <span class="co">#the slope is the fixed effect found after the id effects</span></span>
<span id="cb11-57"><a href="#cb11-57" tabindex="-1"></a>  <span class="co">#it&#39;s after the id effects because that&#39;s the order in which we wrote it in </span></span>
<span id="cb11-58"><a href="#cb11-58" tabindex="-1"></a>  <span class="co">#the lmer formula</span></span>
<span id="cb11-59"><a href="#cb11-59" tabindex="-1"></a>  stand.count_effects <span class="ot">&lt;-</span> X.stand.count<span class="sc">%*%</span>models[[enviro]]<span class="sc">@</span>beta[<span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))<span class="sc">+</span><span class="dv">1</span>]</span>
<span id="cb11-60"><a href="#cb11-60" tabindex="-1"></a>  </span>
<span id="cb11-61"><a href="#cb11-61" tabindex="-1"></a>  <span class="co">#calculate variance for the fixed effects</span></span>
<span id="cb11-62"><a href="#cb11-62" tabindex="-1"></a>  id_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(id_effects)</span>
<span id="cb11-63"><a href="#cb11-63" tabindex="-1"></a>  stand.count_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(stand.count_effects)</span>
<span id="cb11-64"><a href="#cb11-64" tabindex="-1"></a>  </span>
<span id="cb11-65"><a href="#cb11-65" tabindex="-1"></a>  <span class="co">#Plot level heritability</span></span>
<span id="cb11-66"><a href="#cb11-66" tabindex="-1"></a>  H2 <span class="ot">&lt;-</span> <span class="fu">c</span>(H2, id_variance<span class="sc">/</span>(id_variance <span class="sc">+</span></span>
<span id="cb11-67"><a href="#cb11-67" tabindex="-1"></a>                       variance_block <span class="sc">+</span> </span>
<span id="cb11-68"><a href="#cb11-68" tabindex="-1"></a>                       variance_residual <span class="sc">+</span> </span>
<span id="cb11-69"><a href="#cb11-69" tabindex="-1"></a>                       stand.count_variance) )</span>
<span id="cb11-70"><a href="#cb11-70" tabindex="-1"></a>  </span>
<span id="cb11-71"><a href="#cb11-71" tabindex="-1"></a>  </span>
<span id="cb11-72"><a href="#cb11-72" tabindex="-1"></a>  <span class="co">#extract residuals</span></span>
<span id="cb11-73"><a href="#cb11-73" tabindex="-1"></a>  resid <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>residuals)</span>
<span id="cb11-74"><a href="#cb11-74" tabindex="-1"></a>  expt <span class="ot">&lt;-</span> <span class="fu">rep</span>(enviro, <span class="fu">length</span>(resid))</span>
<span id="cb11-75"><a href="#cb11-75" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">cbind</span>(expt, resid))</span>
<span id="cb11-76"><a href="#cb11-76" tabindex="-1"></a>  residuals <span class="ot">&lt;-</span> <span class="fu">rbind</span>(residuals, tmp)</span>
<span id="cb11-77"><a href="#cb11-77" tabindex="-1"></a>  </span>
<span id="cb11-78"><a href="#cb11-78" tabindex="-1"></a>  </span>
<span id="cb11-79"><a href="#cb11-79" tabindex="-1"></a>  <span class="co">#calculate EEV (estimation error variance-covariance matrix for id effects)</span></span>
<span id="cb11-80"><a href="#cb11-80" tabindex="-1"></a>  EEV_lmer <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>x, </span>
<span id="cb11-81"><a href="#cb11-81" tabindex="-1"></a>                     <span class="at">nrow =</span> <span class="fu">sqrt</span>(<span class="fu">length</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>x))) </span>
<span id="cb11-82"><a href="#cb11-82" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">1</span>]]</span>
<span id="cb11-83"><a href="#cb11-83" tabindex="-1"></a>  <span class="fu">colnames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">2</span>]]</span>
<span id="cb11-84"><a href="#cb11-84" tabindex="-1"></a>  <span class="co">#remove EEV entries for all fixed effects other than id</span></span>
<span id="cb11-85"><a href="#cb11-85" tabindex="-1"></a>  <span class="co">#in this case, the only fixed effect to remove is stand.count</span></span>
<span id="cb11-86"><a href="#cb11-86" tabindex="-1"></a>  EEV_lmer <span class="ot">&lt;-</span> EEV_lmer[<span class="fu">which</span>(<span class="fu">rownames</span>(EEV_lmer) <span class="sc">!=</span> <span class="st">&quot;stand.count&quot;</span>),</span>
<span id="cb11-87"><a href="#cb11-87" tabindex="-1"></a>                         <span class="fu">which</span>(<span class="fu">colnames</span>(EEV_lmer) <span class="sc">!=</span> <span class="st">&quot;stand.count&quot;</span>)]</span>
<span id="cb11-88"><a href="#cb11-88" tabindex="-1"></a>  </span>
<span id="cb11-89"><a href="#cb11-89" tabindex="-1"></a>  <span class="co">#keep only the BLUEs for the id fixed effect</span></span>
<span id="cb11-90"><a href="#cb11-90" tabindex="-1"></a>  placeholder <span class="ot">&lt;-</span> <span class="fu">names</span>(BLUEs)</span>
<span id="cb11-91"><a href="#cb11-91" tabindex="-1"></a>  BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>(BLUEs, models[[enviro]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))])</span>
<span id="cb11-92"><a href="#cb11-92" tabindex="-1"></a>  <span class="fu">names</span>(BLUEs) <span class="ot">&lt;-</span> <span class="fu">c</span>(placeholder, <span class="fu">levels</span>(models[[enviro]]<span class="sc">@</span>frame<span class="sc">$</span>id))</span>
<span id="cb11-93"><a href="#cb11-93" tabindex="-1"></a>  </span>
<span id="cb11-94"><a href="#cb11-94" tabindex="-1"></a>  EEVs[[enviro]] <span class="ot">&lt;-</span> EEV_lmer</span>
<span id="cb11-95"><a href="#cb11-95" tabindex="-1"></a>}</span>
<span id="cb11-96"><a href="#cb11-96" tabindex="-1"></a><span class="fu">colnames</span>(residuals) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;resid&quot;</span>)</span>
<span id="cb11-97"><a href="#cb11-97" tabindex="-1"></a>residuals<span class="sc">$</span>resid <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(residuals<span class="sc">$</span>resid)</span>
<span id="cb11-98"><a href="#cb11-98" tabindex="-1"></a></span>
<span id="cb11-99"><a href="#cb11-99" tabindex="-1"></a></span>
<span id="cb11-100"><a href="#cb11-100" tabindex="-1"></a></span>
<span id="cb11-101"><a href="#cb11-101" tabindex="-1"></a><span class="co">#almost exactly the same BLUEs</span></span>
<span id="cb11-102"><a href="#cb11-102" tabindex="-1"></a><span class="fu">cor</span>(BLUEs, <span class="co">#lmer</span></span>
<span id="cb11-103"><a href="#cb11-103" tabindex="-1"></a>    ans1a<span class="sc">$</span>blues<span class="sc">$</span>BLUE) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>BLUEs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##  A00188-3C  A01143-3C  A09037-6C AAF07847-2 AC01144-1W 
##   66.81192  104.41192   76.55059   83.85059   70.81192</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>ans1a<span class="sc">$</span>blues<span class="sc">$</span>BLUE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1]  66.81187 104.41187  76.55071  83.85071  70.81187</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co">#Almost exactly the same EEV using lmer and StageWise</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>differences <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)) {</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>  differences <span class="ot">&lt;-</span> <span class="fu">c</span>(differences, <span class="fu">mean</span>(<span class="fu">abs</span>(EEVs[[enviro]] <span class="sc">-</span> ans1a<span class="sc">$</span>EEV[[enviro]])))</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>}</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>differences</span></code></pre></div>
<pre><code>## [1] NaN NaN NaN NaN NaN</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>EEVs<span class="sc">$</span>Hancock15[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="co">#lmer</span></span></code></pre></div>
<pre><code>##              idA00188-3C idA01143-3C idA09037-6C idAAF07847-2 idAC01144-1W
## idA00188-3C     45.68736    17.39206    12.77092     12.77092     17.39206
## idA01143-3C     17.39206    45.68736    12.77092     12.77092     17.39206
## idA09037-6C     12.77092    12.77092    78.60379     22.01321     12.77092
## idAAF07847-2    12.77092    12.77092    22.01321     78.60379     12.77092
## idAC01144-1W    17.39206    17.39206    12.77092     12.77092     45.68736</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>ans1a<span class="sc">$</span>EEV<span class="sc">$</span>Hancock15[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co">#example on how to calculate EEV manually</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="co">#For instance, use Hancock15</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>variance_raneff <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">summary</span>(models[[<span class="st">&quot;Hancock15&quot;</span>]])<span class="sc">$</span>varcor<span class="sc">$</span>block)</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models[[<span class="st">&quot;Hancock15&quot;</span>]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a><span class="co">#if the variance for the random effect is 0 it&#39;s equivalent to that random </span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co">#effect not being in the model and it can be ignored</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="cf">if</span> (variance_raneff <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>  only_fixed <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a>  lambda <span class="ot">&lt;-</span> variance_residual<span class="sc">/</span>variance_raneff</span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a>  only_fixed <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a>}</span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a><span class="co">#design matrices</span></span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> pheno1a[pheno1a<span class="sc">$</span>env <span class="sc">==</span> <span class="st">&quot;Hancock15&quot;</span>,]</span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>total.yield),]</span>
<span id="cb23-20"><a href="#cb23-20" tabindex="-1"></a>phenoenv<span class="sc">$</span>stand.count <span class="ot">&lt;-</span> <span class="fu">scale</span>(phenoenv<span class="sc">$</span>stand.count, <span class="at">scale =</span> <span class="cn">FALSE</span>, <span class="at">center =</span> <span class="cn">TRUE</span>) </span>
<span id="cb23-21"><a href="#cb23-21" tabindex="-1"></a>X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(total.yield<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoenv) <span class="co">#remove intercept.</span></span>
<span id="cb23-22"><a href="#cb23-22" tabindex="-1"></a>X.stand.count <span class="ot">&lt;-</span> phenoenv<span class="sc">$</span>stand.count <span class="co">#fixed effect for a covariate is a</span></span>
<span id="cb23-23"><a href="#cb23-23" tabindex="-1"></a><span class="co">#regression. design matrix of a regression contains the values for the variable</span></span>
<span id="cb23-24"><a href="#cb23-24" tabindex="-1"></a><span class="fu">colnames</span>(X.stand.count) <span class="ot">&lt;-</span> <span class="st">&quot;stand.count&quot;</span></span>
<span id="cb23-25"><a href="#cb23-25" tabindex="-1"></a>X.all <span class="ot">&lt;-</span> <span class="fu">cbind</span>(X.id, X.stand.count)</span>
<span id="cb23-26"><a href="#cb23-26" tabindex="-1"></a>Z <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(total.yield<span class="sc">~</span><span class="fu">as.factor</span>(block) <span class="sc">-</span> <span class="dv">1</span>, <span class="at">data=</span>phenoenv)</span>
<span id="cb23-27"><a href="#cb23-27" tabindex="-1"></a><span class="co">#Do all operations needed</span></span>
<span id="cb23-28"><a href="#cb23-28" tabindex="-1"></a></span>
<span id="cb23-29"><a href="#cb23-29" tabindex="-1"></a></span>
<span id="cb23-30"><a href="#cb23-30" tabindex="-1"></a><span class="co">#manually calculate EEV</span></span>
<span id="cb23-31"><a href="#cb23-31" tabindex="-1"></a><span class="co">#Make a design matrix mixing all fixed effects. Then, EEV for only </span></span>
<span id="cb23-32"><a href="#cb23-32" tabindex="-1"></a><span class="co">#id can be subsetted. </span></span>
<span id="cb23-33"><a href="#cb23-33" tabindex="-1"></a><span class="co">#This is equivalent to using a separate matrix for each fixed effect</span></span>
<span id="cb23-34"><a href="#cb23-34" tabindex="-1"></a></span>
<span id="cb23-35"><a href="#cb23-35" tabindex="-1"></a><span class="cf">if</span> (only_fixed) {</span>
<span id="cb23-36"><a href="#cb23-36" tabindex="-1"></a>  <span class="co">#If random effect variance == 0, we can completely ignore them</span></span>
<span id="cb23-37"><a href="#cb23-37" tabindex="-1"></a>  <span class="co">#EEV would be calculated as is there were only fixed effects in the model</span></span>
<span id="cb23-38"><a href="#cb23-38" tabindex="-1"></a>  EEV15 <span class="ot">&lt;-</span> <span class="fu">solve</span>(<span class="fu">t</span>(X.all)<span class="sc">%*%</span>X.all)</span>
<span id="cb23-39"><a href="#cb23-39" tabindex="-1"></a>  EEV15 <span class="ot">&lt;-</span> variance_residual<span class="sc">*</span>EEV15</span>
<span id="cb23-40"><a href="#cb23-40" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb23-41"><a href="#cb23-41" tabindex="-1"></a>  <span class="co">#if random effect variance != 0 we have to consider them for EEV calculation</span></span>
<span id="cb23-42"><a href="#cb23-42" tabindex="-1"></a>  full_mat1 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(X.all)<span class="sc">%*%</span>X.all, <span class="fu">t</span>(X.all)<span class="sc">%*%</span>Z)</span>
<span id="cb23-43"><a href="#cb23-43" tabindex="-1"></a>  full_mat2 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(Z)<span class="sc">%*%</span>X.all, (<span class="fu">t</span>(Z)<span class="sc">%*%</span>Z <span class="sc">+</span> lambda<span class="sc">*</span><span class="fu">diag</span>(<span class="fu">ncol</span>(Z))))</span>
<span id="cb23-44"><a href="#cb23-44" tabindex="-1"></a>  full_mat <span class="ot">&lt;-</span> <span class="fu">rbind</span>(full_mat1, full_mat2)</span>
<span id="cb23-45"><a href="#cb23-45" tabindex="-1"></a>  fullinv <span class="ot">&lt;-</span> <span class="fu">solve</span>(full_mat)</span>
<span id="cb23-46"><a href="#cb23-46" tabindex="-1"></a>  EEV15 <span class="ot">&lt;-</span> fullinv[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(X.all),<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(X.all)]</span>
<span id="cb23-47"><a href="#cb23-47" tabindex="-1"></a>  EEV15 <span class="ot">&lt;-</span> variance_residual<span class="sc">*</span>EEV15</span>
<span id="cb23-48"><a href="#cb23-48" tabindex="-1"></a>}</span>
<span id="cb23-49"><a href="#cb23-49" tabindex="-1"></a></span>
<span id="cb23-50"><a href="#cb23-50" tabindex="-1"></a><span class="co">#keep only EEV entries that correspond to id</span></span>
<span id="cb23-51"><a href="#cb23-51" tabindex="-1"></a>EEV15 <span class="ot">&lt;-</span> EEV15[<span class="fu">rownames</span>(EEV15) <span class="sc">%in%</span> <span class="fu">colnames</span>(X.id),</span>
<span id="cb23-52"><a href="#cb23-52" tabindex="-1"></a>                         <span class="fu">colnames</span>(EEV15) <span class="sc">%in%</span> <span class="fu">colnames</span>(X.id)]</span>
<span id="cb23-53"><a href="#cb23-53" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">abs</span>(EEVs<span class="sc">$</span>Hancock15 <span class="sc">-</span> EEV15)) <span class="co">#the only differences are rounding errors</span></span></code></pre></div>
<pre><code>## [1] 2.100386e-12</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co">#merge EEV matrices from each environmet for their use in the second stage</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="co">#It is done using the direct sum operator</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="fu">library</span>(matrixcalc)</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>()</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(pheno1a<span class="sc">$</span>env))) {</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>  env1 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)[i<span class="dv">-1</span>]</span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>  env2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)[i]</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(<span class="fu">is.na</span>(placeholder_matrix)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEVs[[env1]],EEVs[[env2]])</span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(placeholder_matrix,EEVs[[env2]])</span>
<span id="cb25-12"><a href="#cb25-12" tabindex="-1"></a>  }</span>
<span id="cb25-13"><a href="#cb25-13" tabindex="-1"></a>}</span>
<span id="cb25-14"><a href="#cb25-14" tabindex="-1"></a>EEV_full <span class="ot">&lt;-</span> placeholder_matrix</span>
<span id="cb25-15"><a href="#cb25-15" tabindex="-1"></a><span class="fu">dim</span>(EEV_full)</span></code></pre></div>
<pre><code>## [1] 1300 1300</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">length</span>(BLUEs)</span></code></pre></div>
<pre><code>## [1] 1300</code></pre>
<p>The <code>Stage1</code> function returns a list with several results.
List element ‚Äúblue‚Äù is a data frame of the individual BLUEs per
environment. Element ‚Äúfit‚Äù contains the broad-sense heritability on a
plot basis (H2) and the AIC.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">head</span>(ans1a<span class="sc">$</span>blues)</span></code></pre></div>
<pre><code>##         env          id      BLUE
## 1 Hancock15   A00188-3C  66.81187
## 2 Hancock15   A01143-3C 104.41187
## 3 Hancock15   A09037-6C  76.55071
## 4 Hancock15  AAF07847-2  83.85071
## 5 Hancock15  AC01144-1W  70.81187
## 6 Hancock15 Accumulator  83.36187</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co">#lmer doesn&#39;t return AIC values. Therefore we can&#39;t compare them to StageWise</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>ans1a<span class="sc">$</span>fit <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##            env   H2   AIC
## 1    Hancock15 0.71 531.6
## 364  Hancock16 0.72 436.7
## 668  Hancock17 0.70 284.0
## 1029 Hancock18 0.83 246.4
## 1368 Hancock19 0.76 340.8</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a>H2<span class="co">#lmer</span></span></code></pre></div>
<pre><code>## [1] 0.6624359 0.7232735 0.6480836 0.8157480 0.8022444</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co">#H2 estimation is different, but not by a large margin</span></span></code></pre></div>
<p>To check for outliers and normality of the residuals, use the plots
contained in list element ‚Äúresid‚Äù:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>ans1a<span class="sc">$</span>resid<span class="sc">$</span>boxplot</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAtFBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmtv9uTU1uTW5uTY5ubqtujo5ujshuq+SOTU2OTW6OTY6OyP+QOgCQ2/+rbk2rbm6rbo6rq26rq+SryKur5OSr5P+2ZgC2kDq2///Ijk3I///bkDrbkJDb/7bb///kq27k///r6+v/AAD/tmb/yI7/25D/5Kv//7b//8j//9v//+T///++EhqYAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIUElEQVR4nO2dC1PbRhSFnQApIoHiNK1pRUj6wI1xWwfsmIf+///qrgwTbD3OvdKuuLbOmQyTGR2tV5937z61HmRUrQYvnQHrIiAgAgIiICACAiIgoAaADvqhFoCqL10rkrHgrTETEDATEDALAS2HSZJm2d1Z8u7r5q2qz7PpbQ3o7tfLbPn+8uEizWbHm7e2zJwFb2tAC09lkt79dpUtf7rauLVd5m5ubuRms4C8XClafviaF6bHJv46gG68QiQUS2JADxejbPHuCdBztqovZFM5ILHbcAm6Oxu5UP2BgCq0HKbuL2NQlVZ88mrGVqxMs8QrZT9IJQICIiAgAgIiICA280DsKAIREBABATEGAbEVAyIgIAICIiAgAgIiIKDqSXvNPLzxOXv5pH0124I0nb9edhQJaP3Wggho/daiNM/cy6GGiYc2DMhGtSGggF4Cam4mIGB+cUAmArphQDZKGwE1NxMQML80IMYgIBslqCYPBAQyQUAgE3EAHfQQUL55U7gFzwEKkbcys9TaeQxaJG+vMvGrCDWXilkzUNrat2KToz9dCRJvA5YDilbaOm/mPRjxqwiatxTk3nivLdSkqwIkfhVBUYIiVUfttG+YIC1+FeHlAWlhhgEUIwZZiFfBAIlfRYgEKFZ1DDLU0PSDDADS9Zk637xgAJDl0TwBrd+qu9SZl4CAVICqEyYgkDABgYQJCCRMQCBhAgIJ7ywgXU9aAej2ZPCk118aJmoAkG4s1sMSREBAcQF9e7MDVUy+9FSbcBmg+/PD+/PT25PTpokaAKSZXKtNuAyQRzM+zOZ7/9UnWj0JHmfSPp63xlwFaLqfzbe5ium82hg0zulMUQlqdMmkVwvIBaFsPHj1sXEOLDx0TECtc2DhoQkooFfdim3/UKODEnT7I2NQVlvFYD+o0SWT3oaAWMWyWkBjcQkSn1qt9Cqeo523WZCW94NgXg5K/wu9qmttvLGb+WiA2hS3WtDydI0D0pjllzTpFgA1mXI1AUgVCOXplpagqUejmA+yASiOt2q6I9M08wQkTCRw3swCeqpih4X7KnaY9Q5QNvcxuhiCqnbam2jmOwVUoapdrqrnUHijJazwqgBV7bTHn3fdyHuNv+lmCSu8Jf2g08r5oKqd9v2LQRWq2mlPQI+qjEFx8mYX0KqaFbtBVTvtewdovO+7QtP9wn2m+0FQjdKt6Enfn2tWVk0AUlwKAMh3owno+f82dnfMX330FU2ZiCRDNgDJq2NpDPr2ZrAPp6S3GZDCrGrmqxIlIJBoPECKlqkuHSgtoKkby6PdL00BtXhm1UPHBDTe+3fV0gsTNfnQEQHlzfwpV1bXrhAQuLI+5eqrWNmUqzQHFh46apDOp1wBn14DapsDCw/dAaB/GINKAY0HA9/A354wSGdlgKavv/htwPOydZ/1W5tsXO/Me3OgOUlHsdPew8nme38PUEfadgmK97ZPvu7shvOoDes7IMyn34B+ANvv6hMlIJAoAYFEdxwQ33ouu7JrQw0CAiIgIAICUgLq4QlUPAUPiWeYAREQEAEB9RKQ5d/VsHCWa7Qz7QMAinWmfby+za6caW8CUKjjknf1TPu6hFWAopxpr/uBkFjptjiRfJIkPuxEO9PewE/42T7T3gCgYDFIeqa96vwwC4A67geZ+EEjyz1pGz9otCOAbJQ2wz+G3U9AiswREFIfYxBL0Pqtqs/bFAEBERAQASH1MkizH7R2a0HsSa/fWhABrd9aEAer67cWZeKhDQCKMX9uTi/ezJvwWu4HmfASUHMzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQEzAQGzENBymCSpfKe9hYfuFJDf2rp8fyneaW/hoTsFtPBUJql4l2svV1ZdKZLutFdsiI+2ez6UxID8FmDhTvueLRyudtrfnY2qz7TfUM8A5VoO06z6TPsN9XBldcVHvtPewkN3CmiWeKXsB6lEQEAEBERAQAQEhI/Z3gk1B1THbsu8AjMBAREQUFhAOygCAiIgIAICCgfIz4m4Qe2zUW2NlsP04ULmzWekpFoOk2M3wn57ha1ZPuWVj8XrFBSQZ5Qtf8FevwowSVczcTjd38EzPEv302U+LyNJN8tzMHMZOa4zBQWU5yunhL3+WaRe91XLyoT3zkaydFcml+V6c0BAZ0d/fb4Sfnvuy1uMHldNULp5/l3lFTBSlaD8K3I5WNRW9JBB2kUV93Gy73pV/QV8ZKXhSS4GjcQxaBUxF0eXdSa2YkAEBERAQAQEFLIVS1YSREgLXqE5YAnyq2fb5JWZQ1ax76vT2+EVmRmDgIKWoD/834fPkl6aBa/IHLYn7er0rL5jaskrMgetYg8XPw+lQdKCV2IOHIMmikbEgheb2Q/qrh+0myIgoKCAFkkqbkEseCXm8D3p75sZrXtF5rD9oNT9rZ/BtOQVmYNWsbxdEC65WPBKzAzSQAQEFHiokQhXVm14Jebgg9VsJgymL+8VmYM286slPtE6lgWvyMwS1GEJMhFXLMegnRQBAXGwysFqOzMHqxystjMzSAMREFDgVky6omDDKzEHbcXO0q3yiszhB6vb4xWZg1YxC2ulYddVQ1cxef234BWZ2YoBERBQUEAzRfG24JWYAw9WF8fPTkGx7hWZAzfzT+9EbYVXZA46mv906f7JphoseEXmoDHIfZTrvMs6Iha8EjNbMSACAuIexU570sKpTjNegZmAgIOAgIOAgIOAgIOtGOeD2omAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgI6H9iELoi9D2A/wAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a>ans1a<span class="sc">$</span>resid<span class="sc">$</span>qqplot</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABFFBMVEUAAAAAADoAAGYAOpAAZrYZGT8ZGWIZP2IZP4EZYp8aGhozMzM6AAA6ADo6AGY6kNs/GRk/GT8/GWI/Pxk/P4E/gYE/gZ8/gb1NTU1NTW5NTY5NbqtNjshiGRliGT9iGWJiPxliP4Fin9lmAABmtv9uTU1uTW5uTY5ubqtuq+SBPxmBPz+BP2KBYhmBn4GBvdmOTU2OTW6OTY6OyP+QOgCQkGaQtpCQ2/+fYhmf2b2f2dmrbk2ryKur5P+2ZgC22/+2/7a2//+9gT+92b292dnIjk3I///Zn2LZvYHZ2Z/Z2b3Z2dnbkDrb/7bb///kq27k///r6+v/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///9o9QgmAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOT0lEQVR4nO2dC1sctxWGBwc7zuIY4jg1bWonBtwm0GTTmtSENKQm2CwLbWJMjcn8//9RaTQX3WaOpHPmsjv6noTdHcYfOq+kI40QM0ka1aik7wIMXREQoAgIUAQEKAICFAEBCgC0RqVFsAsCdE6jtUWwi4AAuwgIsIuAALsICLCLgAC7CAiwi4AAu2BAsweH5ddmzfem4k128vFkMrlzaEQUajffndyeKhFh7HjhJpOtngCdZmAOtspv4CIq7U7vnpDZ8TeqHQGg2QaHPnv4Tcb+7DGvUvaV/7zsAyvCfPfuycHtb9nJZYXVRuRld/bFIaUd09kTtUHiAXFH9jLbeMThz3ezKj2QPvyyN2Uf83/CClW0YXtEfnY88MYu5mfHdPyRyjsc0EbWX0W7ZBWZFffBofgxvGJZ0cSH+d6nWQlERJ9My1a0RmC3wVuHpYuF2UkNiDIHHfCSFEXISjsTnUl8mO9+9vREqqMyD9UkDR87fqiIicCuykCUXezxVlE5dXU0PZYbcTMgP7uzpwAg39KJ7kYKKO84eRF4x2YvVS+fPfhhbyo6FT+F19D8bw3DvJ8dZ23tYoF20hhC18XY5OHDL7fyItjHCXF2MQ9qnrj42eXfIbN7QgaIQMs9k6YrwrDtIiDALgIC7CIgwC7+2gewgwC9e7a5+TxN33+3+flvJaA0fVOdEf62iGjQdgCg93//KX33l59+//F5+uufi4PDjqhbQG85lVfP3//jdfrur6/TVopAYJckSU+AuFgrevfVb1ljynv4GxoVEaGNEi46OyFnQL//+HX69vMCUFaEobWgDFBfLej9d1+zVP1VBFSjd8/YGJbGHFQnwSfrZkMdxTI4vY1iv25yPR/wPEh0r94AWRUBAYqAAA0KUM85yCqKIlCNywLP0ADhp6lZxyCY+oo5tDL17XwmbQGEryMFEMIuz0BDa0H4rBoBNUaU0F0bDDMHhf/cig5NRAWfZQGUEAMqe9iSAJL4RECWtyUc/p4iomQQgLJlDoqL1ar1pDRdrOJDldIkZ2dAbzf/9DqlWLSX+aQkVV74kV+LyZMQCNCrP/6LtSCCBTOND0FEJR8aO72gXl0Mv2if/9TsnTaZD7w2KAyJ7ORS+gPCL9onSn2n+CqXBkQKO3mIVaaxvi0osAhFA6p+HDIiFQ8hb+HqO4phc1Dxc6uAkBHpfKh4F66+gLCL9gUf60wjvEEmylEC3qVpx/MgueVSRFT4UfOukPsAMuVbBKXlSieERkTNW0/4aceApJ9OElFCDaixQbYPSOvaZhECeXfUIFsHpLRegoi65t32on3ZfC0Kmfo2+AXNpJNaw25akDk4SCcEVLnqh29BSnruo4tpowM2orZ4NySAVgHZRk+zCO4R6X5UvJsaZJuADD64iBJi3oZf14DMgFAREdu58m4PUDU+1J3rFVFbdhBvR0De12JSAUgiorWrSgfydgPkvSYtVxBBRLR2Eh6Ytxsg3/UgpQETRlR3gpdd4lU6N0Cea9JV+22U69SX1k7uXg52boD81qS1/Ietclo7PTtTtyC4CMbwgItID4jKzrV05DnIKAEqIgMPkV1t8Q0711HMdU06oAgNdhY+JHYOE6WW5kEhRai3S1qygydK3oAU1RYhCSpCR3Zya+wJkNwf8BElw7AjBKTkC3REavbB2iXBdnSAWooI8HCyS5JwOzJA2nCDBJRQ2hl8WgcUfDXgem1AapcEudEu2us1RFPlLh6wndF8/OxIAOGKUGtHAEjuXWF2BICkJhxUhLoqD4xIsktkPr0BQtdRDaDQiCo7lU9fgPBFsDdItJ2Gp3dA4UWwNki0ncGnNUDABip8EWy80XaB44Zph91IrpQCD4igT6y5/dLC1Q65kTwhBUTSJ9YwY7tph9pIXlZTmPSpb/IG5VfaIYul2mE2khv9HDnsaB0W14LCPEy7BkCvNjd52qndSG7mQVxWLSzxduGQTbvgjeT6IBpeBGGXlYXMLtTDtAvdSJ6U4aCLkNvpHWyhABnzoCIgiiJoKY3KLsjDtAufSQfOVswiqCmNjHeQh2kXDKisbaKIyGZ2Sw6IrscGeZh2wwJEZxfoYdpVgK63y4H71su0SWLqi56qajNpqqlvq3d/ueBorrd3GvlQ19Gw7RRAOZpLqAUNOqIIqFs7axdbTxs17IjaHcUueY4GUtDAIxrKME9XhGHbxTuSA3YaoAvWwS4++G/5Od6RXAV09MF/tndu9lfLAzV3JEcsa1kjolzhanmY5yO9NsybdyQnWPGV5qprb2j8OphJWwFZ7kiOWTk264h4EbndeRDvYsU8SKxJ2+5IPl5AYh4kzxPtdyQfaw4ytQB3JG/ZLuiO5LQPhxm2nZakgauwxdMa+qiapB2uxBZLxICYjpJk5QWmSMMSPSCmI2A9aFyytKDIR5Z6LbZc/YtEliXXKFkyoJt9J0DEMw1iO2oFtCDixywS2ynK1/sUKZPexjMNQKm8VrYMgPL1PuWYuvjXdKYJqPjlKvRrH9qIiO1k5et9KgrlwrvpTBOQo4gjIrbTVS3TCKlLN01njgMQX4hQpC7+NZ1pA6Qv2i8wIGm9T1FdCzLPtAAyFu0XGJCg8cxIK/YcZDvTBGRftLcDqh55D4k/8v68/CezjcmdQxMQnR0Utbr4B/LpENApD+XsyfT8+O4JASBh93hLt5NVrPcpss6DrGcagNRFe1dArBInW+ezh9/wF1bkye0p/8orNvvAIprv3j05uP0tP/nhyfnZF4cNgPztnkzrAFEIWLSHAfHysZfZxqPz07sn890t/nIgffhlb8o+5v8EbEF+dp0DcpKIiFU0k0gArE1kxX1wKALlbYSVW3yY7336qEoarBXkfFRAYXZZF7s9TEBV0jjggRURPTzJD7OuID7Mdz97elJW+SfT89M7TV3My46D/cNed4CyLA0umGl9glViXtd1VT49LvsE6yl6n8DY5e2tM0BHq3yT2YXLPKiMKG8XeUQ8T7CXKmnMHvywNxUjj1ML8rPLUtpH510BuhazRM9h/ngy+fDLrTwi+7CTdwf+5XQyyXNGTRfzttNSWsuA+BjvBIhA3VysUgK62V+/XHnBO1oEZAWUXt1LVtkF2VJcrLYCyE3EERHbUSsCAhSyT5o2ImK7VgE57rSn/T0NsV2bgCx/q7F4+6Sp1QxoAfdJU0vpYpf634vV7JMeJaDavzg090nT7kSm3tjcFqA6WfZJj7MF1ci2T3q0gNiVhr5ByL5PeqSAjjI2V/eka9UF3CdNrQpQMbrLo5h9n/Q4Ad3sFxNo6HfPIwVUNZzF/qtnakVAgGIXA9ScpK0a7Uy6GObBXz2PtAXlE0V4n/14ATkqAgIUAQGKgAAND1D2R8F9AQLuJz0EQOLPynsCBNxPevSAgPtJp6MH1Hg/6XBl9+qgmEmL237Idr0Aol6TVqocY1fe/7XHUaz2ftJlKf3fLh0g6hxEBaj8RVXfgMjXpOWJS6hdUuHpvQUNcR4k44kzafNtsryA9D6BBFQeXQJAcq0jAal4lgRQIgkJSMOzHIAUPihAhYs1pVErBFDoJUEpcSz0UkN1KTUgQMFVbh12PO3UbiqdsLiADDj5CSGA9G5q2lGrfUBWPmGANKtlAKTlHuUEX0C2bmraUatVQHrqwQAy09jCA7JmZukEL0C2jjogQP4Xq5aBCwOoLo+ZdtRyAuS7aJ9Y8YQDkr1qz+0TkNeCmU4HiggE1JTHTDtqOQFyX7TX6DTeRttpJu1q1u9M2nXRXseT6Ccob11akNlTB96CuOoiqsk8GEABdtQiy0F1iRkBKMiOWo6jGLRor9NxmSgBgJTs425HLZJ5kNl28IAUw8EDUqRHZOtaSEDOid60oxYekDXz4AC5D12mHbXQgDwzhfS2BhCyQVILB0gNhQCQnuz97aiFWJNWYmme5brOpDGWAwKUz+GM54XjW5DSesLsqBUISMYDT3Ltb01AFD2WWmGAwjOF9NYAFDB0mXbUCgKkZIrqGyhAVLyphQXkG4b0VgFk8lkGQPxTBGRordjCkokYENqOWgPZYRY6tpt21BrETnvKBkmtIey0R0w2TTtq9bPTXr3UCL1Wsdr1Aqjlu78saAsSd81vZad9ujw5qJ2d9qn9UoPEjlr97LRfNkBD3GlfZ0etRd1pX2tHrQgIUAQEKAIC1NFG8oap7+Lc/cVVsQUBioAARUCAIiBAERCgCEhSvCN5M6B4R3IAULwjuUMOinckb1a8I3mdpCe5xjuS1yrekbwZULwjOQDIfkdy2mfNENt1C8hB9nL5HG3TDq8ICFAEBAgNaNkVAQGKgABFQIAiIEBYQPmKmiJlWtl4Ztt2BEICylfUlGPq8lrTmW3bUQgJKF9RU46pl7ZNZ7ZtRyGCHFQthAipiyNNZ3ZhhxUeEL/UV6QurzWd2YEdWghA0oqaoroqN89s045K+FHMyAP2pGE7s3U7AiEB2YqpLq81ndm2HYWQgIoVNUXWiYv1zJbtKBRn0oAiIEAREKAICFAEBCgCAjQQQDf74o9W4QcI/u9n9TP00FOsBgMIevRkrquPX6gHIiBVEdDFyov0env96uPvk4Q/EZd3Ps6Av65mD/JcL49dbycr348NUHq0yv+7unfr5c3+asr/588O5q/8Ebm8BRXHGEd2bCyARJJe553on/dfMkA7WX+6FC1lp+ha/LU4lr1ejAVQlYMukp081zAIFzm4ItXw48Wx7KHUV/dHB+iI5ZoKUP5kbgVQfmykgC5v/ZvnmqyL3X95uVJ1reK1OJZBG90oxjMxC7pK0qyZMCJFwubfLo5db6+OLkmvvDhiwTMS2TC/mn+HtxgxzIsOWBwb0TBvyJgQ9qUICFAEBGiogAajCAhQBAQoAgIUAQGKgABFQID+D0RBZUeqSlV8AAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="co">#lmer</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="co">#A lot of residuals are 0 in lmer. They are not normal</span></span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>residuals,<span class="fu">aes</span>(<span class="at">y=</span>.data<span class="sc">$</span>resid,<span class="at">x=</span>.data<span class="sc">$</span>expt)) <span class="sc">+</span> </span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Residual&quot;</span>) <span class="sc">+</span> </span>
<span id="cb38-5"><a href="#cb38-5" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb38-6"><a href="#cb38-6" tabindex="-1"></a>  <span class="fu">stat_boxplot</span>(<span class="at">outlier.color=</span><span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb38-7"><a href="#cb38-7" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb38-8"><a href="#cb38-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x=</span><span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>,<span class="at">vjust=</span><span class="fl">0.5</span>))</span>
<span id="cb38-9"><a href="#cb38-9" tabindex="-1"></a>p1</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAtFBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmtv9uTU1uTW5uTY5ubqtujo5ujshuq+SOTU2OTW6OTY6OyP+QOgCQ2/+rbk2rbm6rbo6rq26rq+SryKur5OSr5P+2ZgC2kDq2///Ijk3I///bkDrbkJDb/7bb///kq27k///r6+v/AAD/tmb/yI7/25D/5Kv//7b//8j//9v//+T///++EhqYAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIDUlEQVR4nO2de3vTNhTGA3QMdxQadmm3QGEXOtpuC7QhLfX3/16znF6cWNYrRefIjvW+Dw9/wKtj5Rfp6GLFnpSUU5O+KzB0ERAQAQEREBABAREQUAig77LSNoC6/+tLQBhVt1hoT0CLovjhfLNM2JXSuhMDWv54Xs4PNsuEXSmtO3ULKleQ1suEXSmtuwdAqxZU560vGckX0HK6/3ETathX0dbl5WWIfegt6ObXe0JCgC4vwwgNHVB5NtsoE3allsYEaPHyM1uQU/OiYA7yFod5IAIC2gFAcr13nIAE8z8BgdAEBEKPExBzULrQWwHqe4WdUmxBwE1AwE1AwE1AwE1AwN0nIG53uDWqDTNrmbArtZQjoKBPnCGgwI+cXw5SbRPDBrScFsVss0xL+QIyt3yWb/BtH81OM2hAC3NfXvrG4ZgAGa1uHI7y8ELV8Dv/zxvQ7cnRJtSwryKtW2xC5gvo5viBDwFZtJzO2mUsyhXQGh8CamteGM02ykTWa0yA7GUsIqBmGYt2EpBzfktAwE1AwE1AwN3jhhn3g4A0N0fkvqlxAhKsyO4A0gstD8ixbeDYOAjaZYiLHVYRp3tn9qSDG4VQ6HEC6ruLbXUli9RuVI8EkN6sSbAiOzOKKXYxAopxExABxbkJiIDi3N6AHn82T0A2LXweTZExoLP9P9mC3LrrYs7DC/WVZBbRcW7B0MxBw03Sikc+NQFdv5rc6+mnhpEtCIiAgAiooa/P213MUsZ6Ke9q7TCgb29ffHt7eP3qEJTRrFecWxmQQXP6orx69p+7jGa94twJAF3slVfBXSwXQOVpTeeCLajsAFQlofJ08uQdKKNZrzh3H8O8pUzYlVK6CShdaHuStiw1LGU06xXnTtKCrl935qAEuwxxbsHQji7GeZCRCxC7WOkEdMoWVDqTNOdBRhzmCSjO7b3laimjWa84t3YLujBouB9Uq2u7o+QwvxIBRXSxF6CMZr3i3Oqj2JXJ0c0UdHNcvPy8WUazXnHu5MP87cnM570a+QK6+e3c+l4N/1fj+LnLbd16oS3zoMP2ftDyp8/WR1PgKzW3RoLMYW690J4PN/F76H++XeyxBTXLaNYrzq0NaNXNGtOgrhykV684tzag0z0zFbrYe/gH8+gXjmIPqhrQt7frd1Y5D9oAZKbRXGoYdZzuuHryznQ0dxnNesW5tXPQ1+eTve4taQJCIiAgAiovqrV85+kXAqrSz7+rkd5dRrNece4Uw/whh/laBLTVlqvpYo4t162OSaR0C4Z2bLl28WELQiKgO/3DHGQFdDqZmAH++hWTdGkDdPH0kzkGfLV+38dWRrNecW5NQAZOefXs70nnRDpzQPV952o53zmGEVANyMGHgCpA33cdv2uW0axXnJuA0oX2BsSfZN6p4wgeH03hjsFHU/DRFB3iz8KjW9BZUZibzgQEREBABATUN6CyJ/aiO4pBH0K7vflXxOnmliv3pOPcBJQloBxzkOIoNpL3ahCQoHscgBRnTYIV2QqQzC4D36sBNJguxmE+JvTOAAqc2gz0FX5y8zOLRvB+McV3rYTGHmgLUnydUYavESUgIAJCGkoOGiygobxGlICAJAAtp48vCyegtsxPwpdv4M/C8wW0MDefz2YbZSLrNSZARrZHU4xErr0Rb0Dmp/MbUC3ayRYUuRZbne64OX7gI7ZYHQmgWsvprF2mJdW58aABrfEhoLbmhdFso0xL+QKyl2kr3xxkLRN2pbRuAkoVmoCAm4CAm4CAm4CAm4CAW/TwwhjFFgTcBATcBATcBATcBATcBATcBATcBATcBATcPN0B3DwfBNyegBaFz5MX8gVkfrDq8UTyfAEZeTzTPu8ctGpBozy84JIvoOV0//54EIf5dd09msLnvRp5AnrQaA9QCQDyfHVNvoDM6YVGDspKfoB85Whcad3ioQkIiICAhACNVwQEREBABAQkAMi8xu7muGi8iMyp5XR2e+Lrbh7T9gldHMybO1huna0dUe2QDCDDqFz+4uM2r241ixfz+kSf2L+jT9AI/f5jvffgF7quxbyqzIHTJQOorlJNycdtPoi/u/qiPZuEMc+PQkIblsAtAeh4/68P5yFf3OLo7odEOHZd+6oD+zAKa0H111TVYuHu7CJJusop1YXCur4XH8/GcK8qBx3556BV3lw8rjWt4igGREBABAREQEAio1ixkldyVHSrhJZoQY1fbfbr1ggt0sUet617diuEZg4CkmlBf5i/bz94TtD03AqhhWbSVW+egylpCrdCaJkudnvy89Q/Pyq65UNL5aCzkPFD0y0dmvOgBPOgUYuAgGQALYqZ/+Ch6ZYPLTiTbryBoy+3QmihedCsLNHeZQq3QmiZLlaPCL5brppu+dBM0kAEBCS11Ci876yquuVDyy1Wy7l3blRzK4SWGeZXt/c872EpuhVCswWlaEHMQRmLgIC4WOViNc7NxSoXq3FuJmkgAgKSGsW8bybouuVDy4xix7NhuBVCCy5WB+BWCC3TxUZ7X1Wsi4V0fUW3QmiOYkAEBCQDaB40uiq65UNLLVYXB43HxPTmVggtNczf/yaqX7dCaJnV/PuP1R/fXQZFt0JomRxUXaSatvvOQRTd8qE5igEREBDPKKaZSfvucqq7xUMTEBABAREQEAEBcRTjflCcCAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgID+B8NW+7AbIKawAAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>residuals,<span class="fu">aes</span>(<span class="at">sample=</span>.data<span class="sc">$</span>resid)) <span class="sc">+</span></span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>  <span class="fu">stat_qq</span>() <span class="sc">+</span> </span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a>  <span class="fu">stat_qq_line</span>() <span class="sc">+</span> </span>
<span id="cb39-4"><a href="#cb39-4" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>expt) <span class="sc">+</span> </span>
<span id="cb39-5"><a href="#cb39-5" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb39-6"><a href="#cb39-6" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Expected&quot;</span>) <span class="sc">+</span> </span>
<span id="cb39-7"><a href="#cb39-7" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Observed&quot;</span>)</span>
<span id="cb39-8"><a href="#cb39-8" tabindex="-1"></a>p2</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABFFBMVEUAAAAAADoAAGYAOpAAZrYZGT8ZGWIZP2IZP4EZYp8aGhozMzM6AAA6ADo6AGY6kNs/GRk/GT8/GWI/Pxk/P4E/gYE/gZ8/gb1NTU1NTW5NTY5NbqtNjshiGRliGT9iGWJiPxliP4Fin9lmAABmtv9uTU1uTW5uTY5ubqtuq+SBPxmBPz+BP2KBYhmBn4GBvdmOTU2OTW6OTY6OyP+QOgCQkGaQtpCQ2/+fYhmf2b2f2dmrbk2ryKur5P+2ZgC22/+2/7a2//+9gT+92b292dnIjk3I///Zn2LZvYHZ2Z/Z2b3Z2dnbkDrb/7bb///kq27k///r6+v/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T///9o9QgmAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAPMklEQVR4nO2dD1/cthnHTUrS9EgDTdOFrkvaANkWWHvdQhdKN7ZQEuBgayEMQv3+38cs/9UjS3ok65Et3+n5tDmfkX+Svn6kR/bJcpJG01oydAFCtwgIsQgIsQgIsQgIsQgIMRtAK3TmS5Fc0A7QGZWt+FIkF4yAEMEICBGMgBDBCAgRjIAQwQgIEYyAEMEOgE4e7df/6m22Mz1rDjmcTCb39mHuboqz7cndKayPkyAr4WSyAYvYB6DjHMzeRvMXV0C14vH9I6DoJsg2BEEnQCdrjPfJ429z7KdP2fnM/mVZ5V+y3Gfb94/27n6XJa7PlQ6QleLp1/stRSfBzE6fCS7pAoiJZR8na08Y99l2fj73uC8/70yzr+UhWXlq91UBslNk1UaamJ1gZoefiOewA6C1vKkWLpmdxbysj/aLHNhZzUpVfJntfP6kLvDJZ9PGi1ZIFNeYb0ibWDdB3oFo+qA9Vogq97yoJ0VjKr7Mtr94fsSdnqYfUvZBNopsV10jVR9kV8S6B6JpYk83qvOiOj3TQ95/UUB2iqfPUUC2RSyaGxWgsuGUubM2nX00Dfzk0Y8706JRsSTs5Mz+og/zdooMt6KJdRTkAwlFE8vGDR9/s1HmLg8RRepqHCR2qY6K5V90TcxS8BkFIAqb65E0Ze7kiuSCERAiGAEhghEQIogBer++/vt3NSDy31TIFckFEUDXf3qX/vKHGlCaXnB/vLDYFHbXgOwPVWxWgMiLiDcxBslX7nMBqPCg3OsuqKwGRCGWJEmtSF9EDND1iy//WW2H6UEJsyE96MNfK0IRkNzevkp95T56QO+/+jV4D2KEhuukf1lfD70PGjyKNRYBIRZomOcV+w/zAFD0IL0FCCjvoOcMEBeUnQEVIX5IQNcv1terYRBN7nmd5gYQGwNd/5F0HDRfgN6zC1XakTQpoCD6oGIkTXc1zy6+5ynM//bTy2ozwChWbw7mQR++r/lEQBK7fvGq+RIBtQzwIci96lPnBlB2Mb/eDITcck9KIwSU8x48ijXmlHtCD6iQ8wCIH+z3dTVf80n4GOoWlWu5C9owD4Zq/XhQg4ewD0q8eFBVzl4B8XioAEFBGkDceRwEULPbHRAvSQWI59NrJw35kABKPADi6NhEseaX5665C3woAAE+NIBU3SQG6D0/u6NT7iIfAkCQDwUggMcG0Nsv/+HoQYk/QPxeN0AQj5UHVU2s8+2OKmew020cpFF0EgSavfVBVc5gt5MHyVzSzYNaXt5jJ811e5LcuwCSEicBJC2iZ0B1zmSAEnpATWyXFNEvoIQcUN0a6ADJFUcKqOktyAApFHsZScORhSR3W0CJirgzIGURPd7ukAZ4GEPJFe3DvErSvwfB2EnhQUCRyoMwJ/cOSJrCCZA0cVdASuTeASXkgHSCgwH68P36V79WX2xyF/gQANIKdgWkbrRmgH776ZXhowiJ3BTlMwCkEOwGSCVm0E0iszv+9s7sUQSMjzUgjI8VICM+nQBd/7maBlxezSNZGRkM8zSCXJgnEEyMw7zhPGlVNtLEJh6kLLhK0MWD9EU09SB17sr8Nd1lR0CaTlcHyIyPvz4I5oQNZkwAwaIbjPZMAGEqHaPYSzSKJSbVsQGkd33FpgZQoWY+3kotABmMg8yqYwEI6RsUm0MBAqYD1C13HSBE0RgQ2n3pi+h6NV/WxvISWn01X/tPN0VJETvJkV3NV7VRnQfFbqUHtWOLowcZBEB9Ed0AJb4AqQ9VbCoANbwHBdQ5dxGQ7HxTADJR6QpIe0/aNXcBkPR8uwBK/APS/jbvnLsUkP5QxaYUEN+heQKk+22e4PTI6oMcqtiUAQIdvu8m1v5tvmM8BjEUKHaXu5CGeZcCEvw2DwIyhQeB+E7gQcKAgd6D3q6vs6swDJBT7u36oIcqNgVASSLw6T+KkZyeRlFdnS6A2nx6B0RzerwDslChBUTkv7Wi5nx3AFTzMTpUX0T7kXTjvu7+KwAyOVSxWQNK+PL1DYgPng4huR3mHeUaxRX3sglFtPYgmQM5e5DufNt5ECzcEE2M1H9Bn0oKyOJQfRE790EkufOKhocqNvk+yPJQfRG9TqBCc+cAmR6q2NT9quFURAQQ/coL8wXIw8oLij7I+FDF5kCAxJUXqBu47FcN5FDF5lBNLAUrL5AMMPhBxphWXrjZrAP4nTccH37lBcvh8/x50DlDc7O5lX8pbneAlRcWHVCJ5pLzIGHlhQXvg9qAqFdeEHMfGaCqia3WOyhXXpDlPjZA6SXrZLZShUVAiMV1FDFA0YPOswZ2/tF/U7lFQAcf/Wdz63Z3OZWbD0D4wCogQFmYZ5GeHwfBtyLQAwJDzxEC8v5WhHEBSs9ZE+PHQcy8vhVhZICKcRDk4/mtCES3B+JbEbDNAQaK5eSFlHsUgfy9MOSK5IJCJy20rpLUK9neUdiKxV7FbthJi1di8Gmf8Rk1oMwOkmTpdfMVvBVhfOYBUGYH4JbropvEgyIf3uC1GGxf0VLpLddovPGAbnf1gOgHGfSKXgFhHkT/hkV6RWBgZkFp4BFBfdIWoFR9r2yUgODMgsLgUgnapG1A1Y+rqjBGXx16Rd7gzIKSBHhMWZu0DQgz+urQK4omXgfAB921SRcCEDezoDD1xVMrqQwQctOevDr0ipW1ZxYUpvSgdlIJIOymPXl16BUhjRetXkXRB8mStgFJbtpLADVvu8eMve2+OeRkbXKvOmbFmyJWabhUAs6nF0DHrB6nz6Znh/ePaAAVik832oq8wZkFpcnHQdKkLUDym/ZqQNkZnGycnTz+ln1k5Z3cnbJ/2VnNv2TVmW3fP9q7+x1L/Pjo7PTrfT0ge8VnUzUgEsNv2isBscJlHydrT86O7x/NtjfYxx735eedafa1PMTEg+wUhwCkt7I62VnOrGj9mU/kZX20X9SS+UhW6OLLbOfzJw2BzAUqPiKgbop5E7sbIKCmx9hjtaqq8/io3J21g+LLbPuL50f1+f5senZ8D2liVooM7O92egWU99LqG2Zig8jOYHmiVed7elg3iKyZtBuEk2Lpb30COlhmk8zO9eOgujqlX5TVYZ1E9tH0GCePftyZFmHH1IPsFPNe7ZOzHgHdFKNE4zB/OJl8/M1GWR15zCnbAvvneDKpOgxlE7NWbPVqvgGxGI8AorD+LlZJAd3url4uvWYNLQKSAkqvHiTL2QXZPFys+gGEGH116BXJLQJCzGqeNHl16BX9AmrNtBfmKJL/SEOv6BWQ5FkN33MUqSdQ0RvytE/qeY7iuAAVaMSfD/3OUaRW9AdI8cThyOYo0pvpM6tpGgGpjPx986MFlF1pCBOEDFckt6hlMXN8nIAOcjZXD/hrVTBHccEBVdFdPQlmsQHd7lYDaOVvz1QvUh5rmK8cR33DbKGjWAQkN6smttCA5J10L+sHWR2q2OwzzHN3O3paP8jmUMVmfwNF0AGJ6wcFCQiMzcnNav0g8qDsrFivmjgcIH79oOA8qFmzcDBAYP2gCKhlwvpBoQIarA8Kfv2geoHKgQARrx9UL5dIAqh+5n5AQNBcATUrklIAahZtDQcQRUimCvN1gL8AiuTWmwc1/SmNByUhepADIFgfOkDV3nkBVO52B8SvijIsIO3Lj0w3Of+hAZQEA0j78iPTTcBnvgDpXn5kDAjyoQDUEhy+iTlczQsBmSDMtwUHDPPufRA83QQeJBEc3oPS1BEQ2O0EKJkzQGVzALtdALX4REBgs+7RYIKxA0qpANX+EwwgzjoBktbHGVArcRiAOkdkcW/XMC8ZMQwe5jnr4EGKE97Rg/j+OUgPsgbUVIgSkCzxOAEltIDEKzo+wagBtVN0AtS+AOM2hwIEHsO3BaRuEV0AJSECgssRqQAlWhMSGwJC5EIBBJcCUQDS8+kGCOUTCKBmMZnqdgdCw8CEcZC7YDLgOMhknrS+8ELichPzIFQuPA9KUwUgMz52gNS+IkkceB+EnVn5bj0grfMpNgeLYi/1UQw9s/LdWkB651NsBjoOwl1fvtsAEKYYCCBgLUB1T0MJyLDRBglIcffB8p6F/nZHR80wAMk9KLW64kQ8qB3Px+RBQp2bqlABksXzCEgUFDq10AFpbto3bYEYkO5QxWaIkxcSakBc/zMaQKrJC83oP/9KAkghGDYgxeQF/vKoi8nCvIteaJMXEsGBSDwIhPcxeFD58iM9IK6I2s15BFSaBJCIhwQQVBw1IEkopgJkcKhiMxRAfOOKgERbgZ0PKSDIZ6yAmvDeMRgrw7yzahiAWvOdK3P1ILHTH6sHVQ80FUYHqHJLk0MVm6EAsi64fDcEVPnPGAF5X1gAD4tBA/K+sIBBWAwakPeFBcCgfISAmHldWIBo2DBkmPe8sAAeFoP1oPbrg3xFMatDFZvDRbGgFxZoNoea3RH6wgL15kCAiBcWqDfnBhC0CAixANcPaimSW/QgxCIgxCIgxCIgxBBA8K0IEZBo8a0IBk0svhUBsfhWBK3FtyIo/1JOXki5RxHIXzRDr9gjoNqkr2FPVSdNUVKTClgd6o2IYDZP+wgWAaXCWxEEi4CipREQahEQYhEQYhEQYt0BgXkNpYHn7/RJDVPJFc0ESawzIDivoTC4DoE2qZmgQtFMkMY6A4LzGgqDzwBrk5oJKhTNBGnMqQ8SB9nwKXJtUjNBjaKZoLu5AOLmNRSmvjJpJTUTVCuaCRJYJ0DteQ2FKc93O6mZoFIRE6QzlyjWDjnyPkiW1ExQpWgmSGKdAcnKCNch0CY1E1Qo9sinOyA4r6E0+ahFmtQwlVTRTJDG4kgasQgIsQgIsQgIsQgIsQgIsRAA3e4Ws8lVry9t7H//gt+VbzylszAAreKJmF19+hruiICgRUDnS6/Tm83Vq09/SBL2OlzW+BgD9rmcv8Vztd53s5ks/bBogNKDZfbf1YM7b253l1P2P3txMPtk78dlHlTtyzhm+xYFUNFJr7JG9PeHbzJAW3l7uiw8ZatqWuyz2pd/ni8KoKYPOk+2yr4mg3Begqu6Gra/2pe/kfrq4cIBOsj6mgZQ+VpuAKjct6CALu/8m/U1eRN7+OZyqWla1We1L4e2cFGM9cRZpZtOOnOTjEjVYbM/V/tuNpcXrpNeen2QVT4jkYf55fIvzGOKMF80wGrfAoX5lrUGhANaBIRYBIRYkIBCsggIsQgIsQgIsQgIsQgIsQgIsf8DcD3PVakDwkEAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="co">#In lmer the model gets residuals equal to zero for all observations of </span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a><span class="co">#non-repeated genotypes. When a genotype appears only once, it is possible to</span></span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a><span class="co">#fully explain its phenotype using the model effects</span></span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a><span class="co">#As a result, the residuals are not normal. This would be fixed if all genotypes</span></span>
<span id="cb40-5"><a href="#cb40-5" tabindex="-1"></a><span class="co">#were replicated at least twice within each environment.</span></span>
<span id="cb40-6"><a href="#cb40-6" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" tabindex="-1"></a><span class="co">#Example in Hancock15:</span></span>
<span id="cb40-8"><a href="#cb40-8" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> pheno1a[pheno1a<span class="sc">$</span>env <span class="sc">==</span> <span class="st">&quot;Hancock15&quot;</span>,]</span>
<span id="cb40-9"><a href="#cb40-9" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>total.yield),]</span>
<span id="cb40-10"><a href="#cb40-10" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id)) <span class="co">#total number of genotypes tested</span></span></code></pre></div>
<pre><code>## [1] 265</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">duplicated</span>(phenoenv<span class="sc">$</span>id)) <span class="co">#number or replicated genotypes</span></span></code></pre></div>
<pre><code>## [1] 92</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id)) <span class="sc">-</span> <span class="fu">sum</span>(<span class="fu">duplicated</span>(phenoenv<span class="sc">$</span>id)) <span class="co">#non-replicated</span></span></code></pre></div>
<pre><code>## [1] 173</code></pre>
<p>The reserved word ‚Äúexpt‚Äù in the input file, which is short for
‚Äúexperiment‚Äù, directs Stage1 to fit separate models for each experiment
within an environment. Then, in a second step, a single BLUE for each
genotype in that environment is estimated, including the full var-cov
matrix. Compared to simply use ‚Äúexpt‚Äù as a factor in a single step, this
two-step procedure within Stage1 allows for separate spatial models.</p>
<p>The 2020 data file contains two experiments: a preliminary and
advanced trial. Here is a comparison of using random row and range
effects vs.¬†a 2D spline, which requires the additional argument
<code>spline</code> to indicate the names of the variables in the input
file with the x and y coordinates.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a>effects <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">name=</span><span class="fu">c</span>(<span class="st">&quot;row&quot;</span>,<span class="st">&quot;range&quot;</span>,<span class="st">&quot;stand.count&quot;</span>),</span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>                      <span class="at">fixed=</span><span class="fu">c</span>(<span class="cn">FALSE</span>,<span class="cn">FALSE</span>,<span class="cn">TRUE</span>),</span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>                      <span class="at">factor=</span><span class="fu">c</span>(<span class="cn">TRUE</span>,<span class="cn">TRUE</span>,<span class="cn">FALSE</span>))</span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>effects</span></code></pre></div>
<pre><code>##          name fixed factor
## 1         row FALSE   TRUE
## 2       range FALSE   TRUE
## 3 stand.count  TRUE  FALSE</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a>model1 <span class="ot">&lt;-</span> <span class="fu">Stage1</span>(<span class="at">filename=</span>pheno1b.file, <span class="at">traits=</span><span class="st">&quot;total.yield&quot;</span>,</span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>                <span class="at">effects=</span>effects, <span class="at">solver=</span><span class="st">&quot;asreml&quot;</span>)</span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a>model1<span class="sc">$</span>fit</span></code></pre></div>
<pre><code>##           env     expt   H2   AIC
## 196 Hancock20 advanced 0.86 213.3
## 1   Hancock20   prelim 0.79  45.1</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">unique</span>(pheno1b<span class="sc">$</span>env)</span></code></pre></div>
<pre><code>## [1] &quot;Hancock20&quot;</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">unique</span>(pheno1b<span class="sc">$</span>expt)</span></code></pre></div>
<pre><code>## [1] &quot;prelim&quot;   &quot;advanced&quot;</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="co">#this outer for loop is not needed here because there&#39;s only a single environment</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="co">#I leave the outer loop so that the code is generalizable</span></span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a>H2_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a>models_env <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb54-5"><a href="#cb54-5" tabindex="-1"></a>BLUEs_df <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb54-6"><a href="#cb54-6" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno1b<span class="sc">$</span>env)) {</span>
<span id="cb54-7"><a href="#cb54-7" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> pheno1b[pheno1b<span class="sc">$</span>env <span class="sc">==</span> enviro,]</span>
<span id="cb54-8"><a href="#cb54-8" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>total.yield),]</span>
<span id="cb54-9"><a href="#cb54-9" tabindex="-1"></a>  </span>
<span id="cb54-10"><a href="#cb54-10" tabindex="-1"></a>  Envs <span class="ot">&lt;-</span> <span class="fu">c</span>(Envs, <span class="fu">rep</span>(enviro, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))))</span>
<span id="cb54-11"><a href="#cb54-11" tabindex="-1"></a>  </span>
<span id="cb54-12"><a href="#cb54-12" tabindex="-1"></a>  models_expt <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb54-13"><a href="#cb54-13" tabindex="-1"></a>  EEV_expt <span class="ot">&lt;-</span> <span class="fu">matrix</span>()</span>
<span id="cb54-14"><a href="#cb54-14" tabindex="-1"></a>  </span>
<span id="cb54-15"><a href="#cb54-15" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>expt)) <span class="sc">&lt;</span> <span class="dv">2</span>)  {</span>
<span id="cb54-16"><a href="#cb54-16" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">&quot;At least two experiments per environment needed&quot;</span>)</span>
<span id="cb54-17"><a href="#cb54-17" tabindex="-1"></a>  }</span>
<span id="cb54-18"><a href="#cb54-18" tabindex="-1"></a>  </span>
<span id="cb54-19"><a href="#cb54-19" tabindex="-1"></a>  <span class="co">#iterate over all experiments within this environment</span></span>
<span id="cb54-20"><a href="#cb54-20" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>expt))) {</span>
<span id="cb54-21"><a href="#cb54-21" tabindex="-1"></a></span>
<span id="cb54-22"><a href="#cb54-22" tabindex="-1"></a>    expt <span class="ot">&lt;-</span> <span class="fu">unique</span>(phenoenv<span class="sc">$</span>expt)[i]</span>
<span id="cb54-23"><a href="#cb54-23" tabindex="-1"></a>    phenoexpt <span class="ot">&lt;-</span> phenoenv[phenoenv<span class="sc">$</span>expt <span class="sc">==</span> expt,]</span>
<span id="cb54-24"><a href="#cb54-24" tabindex="-1"></a>    phenoexpt<span class="sc">$</span>row <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoexpt<span class="sc">$</span>row) <span class="co">#cofactor</span></span>
<span id="cb54-25"><a href="#cb54-25" tabindex="-1"></a>    phenoexpt<span class="sc">$</span>range <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoexpt<span class="sc">$</span>range) <span class="co">#cofactor</span></span>
<span id="cb54-26"><a href="#cb54-26" tabindex="-1"></a>    phenoexpt<span class="sc">$</span>stand.count <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(phenoexpt<span class="sc">$</span>stand.count) <span class="co">#covariate</span></span>
<span id="cb54-27"><a href="#cb54-27" tabindex="-1"></a>    phenoexpt<span class="sc">$</span>stand.count <span class="ot">&lt;-</span> <span class="fu">scale</span>(phenoexpt<span class="sc">$</span>stand.count, <span class="at">center =</span> T, <span class="at">scale =</span> F) </span>
<span id="cb54-28"><a href="#cb54-28" tabindex="-1"></a>    <span class="co">#it&#39;s important to center the covariates</span></span>
<span id="cb54-29"><a href="#cb54-29" tabindex="-1"></a>    </span>
<span id="cb54-30"><a href="#cb54-30" tabindex="-1"></a>    models_expt[[expt]] <span class="ot">&lt;-</span> <span class="fu">lmer</span>(total.yield <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span> id <span class="sc">+</span> stand.count <span class="sc">+</span></span>
<span id="cb54-31"><a href="#cb54-31" tabindex="-1"></a>                                    (<span class="dv">1</span><span class="sc">|</span>row) <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>range),</span>
<span id="cb54-32"><a href="#cb54-32" tabindex="-1"></a>                                    <span class="at">data=</span>phenoexpt)</span>
<span id="cb54-33"><a href="#cb54-33" tabindex="-1"></a>    </span>
<span id="cb54-34"><a href="#cb54-34" tabindex="-1"></a>      </span>
<span id="cb54-35"><a href="#cb54-35" tabindex="-1"></a>    <span class="co">#extract H2</span></span>
<span id="cb54-36"><a href="#cb54-36" tabindex="-1"></a>    <span class="co">#random effect variances</span></span>
<span id="cb54-37"><a href="#cb54-37" tabindex="-1"></a>    variance_row <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>varcor<span class="sc">$</span>row)</span>
<span id="cb54-38"><a href="#cb54-38" tabindex="-1"></a>    variance_range <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>varcor<span class="sc">$</span>range)</span>
<span id="cb54-39"><a href="#cb54-39" tabindex="-1"></a>    variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models_expt[[expt]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb54-40"><a href="#cb54-40" tabindex="-1"></a>    </span>
<span id="cb54-41"><a href="#cb54-41" tabindex="-1"></a>    <span class="co">#fixed effect variances</span></span>
<span id="cb54-42"><a href="#cb54-42" tabindex="-1"></a>    <span class="co">#1) calculate design matrix</span></span>
<span id="cb54-43"><a href="#cb54-43" tabindex="-1"></a>    X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(total.yield<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoexpt) <span class="co">#remove intercept.</span></span>
<span id="cb54-44"><a href="#cb54-44" tabindex="-1"></a>    X.stand.count <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(phenoexpt<span class="sc">$</span>stand.count) <span class="co">#fixed effect for a covariate is a</span></span>
<span id="cb54-45"><a href="#cb54-45" tabindex="-1"></a>    <span class="co">#regression. design matrix of a regression contains the values for the variable</span></span>
<span id="cb54-46"><a href="#cb54-46" tabindex="-1"></a>    <span class="fu">colnames</span>(X.stand.count) <span class="ot">&lt;-</span> <span class="st">&quot;stand.count&quot;</span></span>
<span id="cb54-47"><a href="#cb54-47" tabindex="-1"></a>    </span>
<span id="cb54-48"><a href="#cb54-48" tabindex="-1"></a>    <span class="co">#calculate fixed effects for each plot</span></span>
<span id="cb54-49"><a href="#cb54-49" tabindex="-1"></a>    <span class="co">#id effects:</span></span>
<span id="cb54-50"><a href="#cb54-50" tabindex="-1"></a>    <span class="co"># y = Xb + ... + epsilon</span></span>
<span id="cb54-51"><a href="#cb54-51" tabindex="-1"></a>    <span class="co"># b (vector of the genotypic fixed effects) can be retrieved from:</span></span>
<span id="cb54-52"><a href="#cb54-52" tabindex="-1"></a>    <span class="co">#&quot;models[[enviro]]@beta[1:length(unique(phenoenv$id))]&quot;</span></span>
<span id="cb54-53"><a href="#cb54-53" tabindex="-1"></a>    <span class="co"># X is the design matrix linking each obervation (each plot in the field) with</span></span>
<span id="cb54-54"><a href="#cb54-54" tabindex="-1"></a>    <span class="co"># its corresponding genotypic effect.</span></span>
<span id="cb54-55"><a href="#cb54-55" tabindex="-1"></a>    id_effects <span class="ot">&lt;-</span> X.id<span class="sc">%*%</span>models_expt[[expt]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoexpt<span class="sc">$</span>id))]</span>
<span id="cb54-56"><a href="#cb54-56" tabindex="-1"></a>    <span class="co">#stand.count value times its regression slope</span></span>
<span id="cb54-57"><a href="#cb54-57" tabindex="-1"></a>    <span class="co">#the slope is the fixed effect found after the id effects</span></span>
<span id="cb54-58"><a href="#cb54-58" tabindex="-1"></a>    <span class="co">#it&#39;s after the id effects because that&#39;s the order in which we wrote it in </span></span>
<span id="cb54-59"><a href="#cb54-59" tabindex="-1"></a>    <span class="co">#the lmer formula</span></span>
<span id="cb54-60"><a href="#cb54-60" tabindex="-1"></a>    stand.count_effects <span class="ot">&lt;-</span> X.stand.count<span class="sc">%*%</span></span>
<span id="cb54-61"><a href="#cb54-61" tabindex="-1"></a>                          models_expt[[expt]]<span class="sc">@</span>beta[<span class="fu">length</span>(<span class="fu">unique</span>(phenoexpt<span class="sc">$</span>id))<span class="sc">+</span><span class="dv">1</span>]</span>
<span id="cb54-62"><a href="#cb54-62" tabindex="-1"></a>    </span>
<span id="cb54-63"><a href="#cb54-63" tabindex="-1"></a>    <span class="co">#calculate variance for the fixed effects</span></span>
<span id="cb54-64"><a href="#cb54-64" tabindex="-1"></a>    id_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(id_effects)</span>
<span id="cb54-65"><a href="#cb54-65" tabindex="-1"></a>    stand.count_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(stand.count_effects)</span>
<span id="cb54-66"><a href="#cb54-66" tabindex="-1"></a>    </span>
<span id="cb54-67"><a href="#cb54-67" tabindex="-1"></a>    <span class="co">#Plot level heritability</span></span>
<span id="cb54-68"><a href="#cb54-68" tabindex="-1"></a>    H2_expt <span class="ot">&lt;-</span> id_variance<span class="sc">/</span>(id_variance <span class="sc">+</span></span>
<span id="cb54-69"><a href="#cb54-69" tabindex="-1"></a>                         variance_row <span class="sc">+</span> </span>
<span id="cb54-70"><a href="#cb54-70" tabindex="-1"></a>                         variance_range <span class="sc">+</span> </span>
<span id="cb54-71"><a href="#cb54-71" tabindex="-1"></a>                         variance_residual <span class="sc">+</span> </span>
<span id="cb54-72"><a href="#cb54-72" tabindex="-1"></a>                         stand.count_variance) </span>
<span id="cb54-73"><a href="#cb54-73" tabindex="-1"></a>    </span>
<span id="cb54-74"><a href="#cb54-74" tabindex="-1"></a>    H2_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(H2_df, <span class="fu">c</span>(enviro, expt, H2_expt))</span>
<span id="cb54-75"><a href="#cb54-75" tabindex="-1"></a>    </span>
<span id="cb54-76"><a href="#cb54-76" tabindex="-1"></a>      </span>
<span id="cb54-77"><a href="#cb54-77" tabindex="-1"></a>    <span class="co">#calculate EEV (estimation error variance-covariance matrix for id effects)</span></span>
<span id="cb54-78"><a href="#cb54-78" tabindex="-1"></a>    EEV_lmer <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>vcov<span class="sc">@</span>x, </span>
<span id="cb54-79"><a href="#cb54-79" tabindex="-1"></a>                       <span class="at">nrow =</span> <span class="fu">sqrt</span>(<span class="fu">length</span>(<span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>vcov<span class="sc">@</span>x))) </span>
<span id="cb54-80"><a href="#cb54-80" tabindex="-1"></a>    <span class="fu">rownames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">1</span>]]</span>
<span id="cb54-81"><a href="#cb54-81" tabindex="-1"></a>    <span class="fu">colnames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models_expt[[expt]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">2</span>]]</span>
<span id="cb54-82"><a href="#cb54-82" tabindex="-1"></a>    <span class="co">#remove EEV entries for all fixed effects other than id</span></span>
<span id="cb54-83"><a href="#cb54-83" tabindex="-1"></a>    <span class="co">#in this case, the only fixed effect to remove is stand.count</span></span>
<span id="cb54-84"><a href="#cb54-84" tabindex="-1"></a>    EEV_lmer <span class="ot">&lt;-</span> EEV_lmer[<span class="fu">which</span>(<span class="fu">rownames</span>(EEV_lmer) <span class="sc">!=</span> <span class="st">&quot;stand.count&quot;</span>),</span>
<span id="cb54-85"><a href="#cb54-85" tabindex="-1"></a>                           <span class="fu">which</span>(<span class="fu">colnames</span>(EEV_lmer) <span class="sc">!=</span> <span class="st">&quot;stand.count&quot;</span>)]</span>
<span id="cb54-86"><a href="#cb54-86" tabindex="-1"></a>    </span>
<span id="cb54-87"><a href="#cb54-87" tabindex="-1"></a>    </span>
<span id="cb54-88"><a href="#cb54-88" tabindex="-1"></a>    <span class="co">#merge EEV of all expts</span></span>
<span id="cb54-89"><a href="#cb54-89" tabindex="-1"></a>    <span class="cf">if</span> (i <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb54-90"><a href="#cb54-90" tabindex="-1"></a>      EEV_expt <span class="ot">&lt;-</span> EEV_lmer</span>
<span id="cb54-91"><a href="#cb54-91" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb54-92"><a href="#cb54-92" tabindex="-1"></a>      EEV_expt <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEV_expt,EEV_lmer)</span>
<span id="cb54-93"><a href="#cb54-93" tabindex="-1"></a>    }</span>
<span id="cb54-94"><a href="#cb54-94" tabindex="-1"></a>    </span>
<span id="cb54-95"><a href="#cb54-95" tabindex="-1"></a>    <span class="co">#merge BLUEs of al expts</span></span>
<span id="cb54-96"><a href="#cb54-96" tabindex="-1"></a>    placeholder <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">id =</span> <span class="fu">levels</span>(models_expt[[expt]]<span class="sc">@</span>frame<span class="sc">$</span>id),</span>
<span id="cb54-97"><a href="#cb54-97" tabindex="-1"></a>                <span class="at">BLUEs =</span>  models_expt[[expt]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoexpt<span class="sc">$</span>id))],</span>
<span id="cb54-98"><a href="#cb54-98" tabindex="-1"></a>                <span class="at">expt =</span> <span class="fu">rep</span>(expt, <span class="fu">length</span>(<span class="fu">unique</span>(phenoexpt<span class="sc">$</span>id))),</span>
<span id="cb54-99"><a href="#cb54-99" tabindex="-1"></a>                <span class="at">Stage1Error =</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(expt, <span class="fu">length</span>(<span class="fu">unique</span>(phenoexpt<span class="sc">$</span>id))),</span>
<span id="cb54-100"><a href="#cb54-100" tabindex="-1"></a>                                     <span class="st">&quot;:&quot;</span>,</span>
<span id="cb54-101"><a href="#cb54-101" tabindex="-1"></a>                                     <span class="fu">levels</span>(models_expt[[expt]]<span class="sc">@</span>frame<span class="sc">$</span>id))</span>
<span id="cb54-102"><a href="#cb54-102" tabindex="-1"></a>                )</span>
<span id="cb54-103"><a href="#cb54-103" tabindex="-1"></a>    </span>
<span id="cb54-104"><a href="#cb54-104" tabindex="-1"></a>    BLUEs_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(BLUEs_df, placeholder)</span>
<span id="cb54-105"><a href="#cb54-105" tabindex="-1"></a>    </span>
<span id="cb54-106"><a href="#cb54-106" tabindex="-1"></a>  }</span>
<span id="cb54-107"><a href="#cb54-107" tabindex="-1"></a></span>
<span id="cb54-108"><a href="#cb54-108" tabindex="-1"></a></span>
<span id="cb54-109"><a href="#cb54-109" tabindex="-1"></a>  models_env[[enviro]][[<span class="st">&quot;within_experiment&quot;</span>]] <span class="ot">&lt;-</span> models_expt</span>
<span id="cb54-110"><a href="#cb54-110" tabindex="-1"></a>  </span>
<span id="cb54-111"><a href="#cb54-111" tabindex="-1"></a></span>
<span id="cb54-112"><a href="#cb54-112" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV_expt) <span class="ot">&lt;-</span> BLUEs_df<span class="sc">$</span>Stage1Error</span>
<span id="cb54-113"><a href="#cb54-113" tabindex="-1"></a>  <span class="fu">colnames</span>(EEV_expt) <span class="ot">&lt;-</span> BLUEs_df<span class="sc">$</span>Stage1Error</span>
<span id="cb54-114"><a href="#cb54-114" tabindex="-1"></a>  EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(EEV_expt),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb54-115"><a href="#cb54-115" tabindex="-1"></a></span>
<span id="cb54-116"><a href="#cb54-116" tabindex="-1"></a></span>
<span id="cb54-117"><a href="#cb54-117" tabindex="-1"></a></span>
<span id="cb54-118"><a href="#cb54-118" tabindex="-1"></a>  <span class="co">#Fix the variance for stage1Error to be 1. We have to scale it dividing it</span></span>
<span id="cb54-119"><a href="#cb54-119" tabindex="-1"></a>  <span class="co">#by the variance of the response variable because Sommer works that way.</span></span>
<span id="cb54-120"><a href="#cb54-120" tabindex="-1"></a>  <span class="co">#We fix the variance to be 1 because we don&#39;t want the model to scale EEV in </span></span>
<span id="cb54-121"><a href="#cb54-121" tabindex="-1"></a>  <span class="co">#any way, but 1 is not the true stage1Error variance component. More on that later</span></span>
<span id="cb54-122"><a href="#cb54-122" tabindex="-1"></a>  Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(BLUEs_df<span class="sc">$</span>BLUEs))</span>
<span id="cb54-123"><a href="#cb54-123" tabindex="-1"></a>  model_Sommer <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> id <span class="sc">+</span> expt <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Fixed effects = id + expt -1 means that</span></span>
<span id="cb54-124"><a href="#cb54-124" tabindex="-1"></a>                     <span class="co">#we remove the intercept (you could leave it if you prefer it)</span></span>
<span id="cb54-125"><a href="#cb54-125" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb54-126"><a href="#cb54-126" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb54-127"><a href="#cb54-127" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb54-128"><a href="#cb54-128" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb54-129"><a href="#cb54-129" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb54-130"><a href="#cb54-130" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb54-131"><a href="#cb54-131" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb54-132"><a href="#cb54-132" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb54-133"><a href="#cb54-133" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb54-134"><a href="#cb54-134" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb54-135"><a href="#cb54-135" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb54-136"><a href="#cb54-136" tabindex="-1"></a>                           </span>
<span id="cb54-137"><a href="#cb54-137" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb54-138"><a href="#cb54-138" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb54-139"><a href="#cb54-139" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb54-140"><a href="#cb54-140" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb54-141"><a href="#cb54-141" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb54-142"><a href="#cb54-142" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb54-143"><a href="#cb54-143" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb54-144"><a href="#cb54-144" tabindex="-1"></a>                           </span>
<span id="cb54-145"><a href="#cb54-145" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb54-146"><a href="#cb54-146" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb54-147"><a href="#cb54-147" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb54-148"><a href="#cb54-148" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb54-149"><a href="#cb54-149" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb54-150"><a href="#cb54-150" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb54-151"><a href="#cb54-151" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb54-152"><a href="#cb54-152" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb54-153"><a href="#cb54-153" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb54-154"><a href="#cb54-154" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb54-155"><a href="#cb54-155" tabindex="-1"></a>                     <span class="at">data=</span> BLUEs_df,</span>
<span id="cb54-156"><a href="#cb54-156" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb54-157"><a href="#cb54-157" tabindex="-1"></a>  )</span>
<span id="cb54-158"><a href="#cb54-158" tabindex="-1"></a>  </span>
<span id="cb54-159"><a href="#cb54-159" tabindex="-1"></a>  models_env[[enviro]][[<span class="st">&quot;across_experiments&quot;</span>]] <span class="ot">&lt;-</span> model_Sommer</span>
<span id="cb54-160"><a href="#cb54-160" tabindex="-1"></a>  </span>
<span id="cb54-161"><a href="#cb54-161" tabindex="-1"></a></span>
<span id="cb54-162"><a href="#cb54-162" tabindex="-1"></a>  </span>
<span id="cb54-163"><a href="#cb54-163" tabindex="-1"></a>  <span class="co">#manually calculate EEV for the sommer model:</span></span>
<span id="cb54-164"><a href="#cb54-164" tabindex="-1"></a>  <span class="co">#design matrices</span></span>
<span id="cb54-165"><a href="#cb54-165" tabindex="-1"></a>  X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>BLUEs_df) <span class="co">#remove intercept.</span></span>
<span id="cb54-166"><a href="#cb54-166" tabindex="-1"></a>  X.expt <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>expt<span class="dv">-1</span>, <span class="at">data=</span>BLUEs_df) <span class="co">#remove intercept.</span></span>
<span id="cb54-167"><a href="#cb54-167" tabindex="-1"></a>  X.all <span class="ot">&lt;-</span> <span class="fu">cbind</span>(X.id, X.expt)</span>
<span id="cb54-168"><a href="#cb54-168" tabindex="-1"></a>  Z <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Stage1Error <span class="sc">-</span> <span class="dv">1</span>, <span class="at">data=</span>BLUEs_df)</span>
<span id="cb54-169"><a href="#cb54-169" tabindex="-1"></a></span>
<span id="cb54-170"><a href="#cb54-170" tabindex="-1"></a>  </span>
<span id="cb54-171"><a href="#cb54-171" tabindex="-1"></a>  </span>
<span id="cb54-172"><a href="#cb54-172" tabindex="-1"></a>  <span class="co">#manually calculate EEV</span></span>
<span id="cb54-173"><a href="#cb54-173" tabindex="-1"></a>  <span class="co">#variance of the random effect is extracted from EEV_expt in this case</span></span>
<span id="cb54-174"><a href="#cb54-174" tabindex="-1"></a>  <span class="co">#beacuse EEV_expt is already the variance covaraince structure multiplied by its</span></span>
<span id="cb54-175"><a href="#cb54-175" tabindex="-1"></a>  <span class="co">#corresponding variance coponent</span></span>
<span id="cb54-176"><a href="#cb54-176" tabindex="-1"></a>  variance_raneff <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(EEV_expt)) <span class="sc">-</span> <span class="fu">mean</span>(EEV_expt)</span>
<span id="cb54-177"><a href="#cb54-177" tabindex="-1"></a>  <span class="co">#EEV_expt is already the variance covaraince structure multiplied by its</span></span>
<span id="cb54-178"><a href="#cb54-178" tabindex="-1"></a>  <span class="co">#corresponding variance coponent. Undo that multiplication to extract the </span></span>
<span id="cb54-179"><a href="#cb54-179" tabindex="-1"></a>  <span class="co">#variance covariance structure</span></span>
<span id="cb54-180"><a href="#cb54-180" tabindex="-1"></a>  EEV_str <span class="ot">&lt;-</span> EEV_expt<span class="sc">/</span>variance_raneff</span>
<span id="cb54-181"><a href="#cb54-181" tabindex="-1"></a>  variance_residual <span class="ot">&lt;-</span> models_env[[enviro]][[<span class="st">&quot;across_experiments&quot;</span>]]<span class="sc">$</span>sigma[<span class="fu">names</span>(model_Sommer<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>]</span>
<span id="cb54-182"><a href="#cb54-182" tabindex="-1"></a>  lambda <span class="ot">&lt;-</span> variance_residual<span class="sc">/</span>variance_raneff</span>
<span id="cb54-183"><a href="#cb54-183" tabindex="-1"></a>  </span>
<span id="cb54-184"><a href="#cb54-184" tabindex="-1"></a>  <span class="co">#reorder EEV_str so that it matches columns in Z</span></span>
<span id="cb54-185"><a href="#cb54-185" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV_str) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;Stage1Error&quot;</span>,<span class="fu">rownames</span>(EEV_str))</span>
<span id="cb54-186"><a href="#cb54-186" tabindex="-1"></a>  <span class="fu">colnames</span>(EEV_str) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;Stage1Error&quot;</span>,<span class="fu">colnames</span>(EEV_str))</span>
<span id="cb54-187"><a href="#cb54-187" tabindex="-1"></a>  </span>
<span id="cb54-188"><a href="#cb54-188" tabindex="-1"></a>  EEV_str <span class="ot">&lt;-</span> EEV_str[<span class="fu">colnames</span>(Z), <span class="fu">colnames</span>(Z)]</span>
<span id="cb54-189"><a href="#cb54-189" tabindex="-1"></a>  </span>
<span id="cb54-190"><a href="#cb54-190" tabindex="-1"></a>  </span>
<span id="cb54-191"><a href="#cb54-191" tabindex="-1"></a>  <span class="co">#if the variance for the random effect is 0 it&#39;s equivalent to that random </span></span>
<span id="cb54-192"><a href="#cb54-192" tabindex="-1"></a>  <span class="co">#effect not being in the model and it can be ignored</span></span>
<span id="cb54-193"><a href="#cb54-193" tabindex="-1"></a>  <span class="cf">if</span> (variance_raneff <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb54-194"><a href="#cb54-194" tabindex="-1"></a>    only_fixed <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb54-195"><a href="#cb54-195" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb54-196"><a href="#cb54-196" tabindex="-1"></a>    lambda <span class="ot">&lt;-</span> variance_residual<span class="sc">/</span>variance_raneff</span>
<span id="cb54-197"><a href="#cb54-197" tabindex="-1"></a>    only_fixed <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb54-198"><a href="#cb54-198" tabindex="-1"></a>  }</span>
<span id="cb54-199"><a href="#cb54-199" tabindex="-1"></a>  </span>
<span id="cb54-200"><a href="#cb54-200" tabindex="-1"></a>  </span>
<span id="cb54-201"><a href="#cb54-201" tabindex="-1"></a>  <span class="cf">if</span> (only_fixed) {</span>
<span id="cb54-202"><a href="#cb54-202" tabindex="-1"></a>    <span class="co">#If random effect variance == 0, we can completely ignore them</span></span>
<span id="cb54-203"><a href="#cb54-203" tabindex="-1"></a>    <span class="co">#EEV would be calculated as is there were only fixed effects in the model</span></span>
<span id="cb54-204"><a href="#cb54-204" tabindex="-1"></a>    EEV <span class="ot">&lt;-</span> <span class="fu">solve</span>(<span class="fu">t</span>(X.all)<span class="sc">%*%</span>X.all) </span>
<span id="cb54-205"><a href="#cb54-205" tabindex="-1"></a>    EEV <span class="ot">&lt;-</span> variance_residual<span class="sc">*</span>EEV</span>
<span id="cb54-206"><a href="#cb54-206" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb54-207"><a href="#cb54-207" tabindex="-1"></a>    <span class="co">#if random effect variance != 0 we have to consider them for EEV calculation</span></span>
<span id="cb54-208"><a href="#cb54-208" tabindex="-1"></a>    full_mat1 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(X.all)<span class="sc">%*%</span>X.all, <span class="fu">t</span>(X.all)<span class="sc">%*%</span>Z)</span>
<span id="cb54-209"><a href="#cb54-209" tabindex="-1"></a>    full_mat2 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">t</span>(Z)<span class="sc">%*%</span>X.all, (<span class="fu">t</span>(Z)<span class="sc">%*%</span>Z <span class="sc">+</span> lambda<span class="sc">*</span><span class="fu">solve</span>(EEV_str)))</span>
<span id="cb54-210"><a href="#cb54-210" tabindex="-1"></a>    full_mat <span class="ot">&lt;-</span> <span class="fu">rbind</span>(full_mat1, full_mat2)</span>
<span id="cb54-211"><a href="#cb54-211" tabindex="-1"></a>    <span class="co">#fullinv &lt;- solve(full_mat) #doesn&#39;t work, matrix singular</span></span>
<span id="cb54-212"><a href="#cb54-212" tabindex="-1"></a>    fullinv <span class="ot">&lt;-</span> pracma<span class="sc">::</span><span class="fu">pinv</span>(full_mat) <span class="co">#use pseudoinverse to invert singular matrix</span></span>
<span id="cb54-213"><a href="#cb54-213" tabindex="-1"></a>    <span class="fu">rownames</span>(fullinv) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(full_mat)</span>
<span id="cb54-214"><a href="#cb54-214" tabindex="-1"></a>    <span class="fu">colnames</span>(fullinv) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(full_mat)</span>
<span id="cb54-215"><a href="#cb54-215" tabindex="-1"></a>    <span class="co">#nonsingularity</span></span>
<span id="cb54-216"><a href="#cb54-216" tabindex="-1"></a>    EEV <span class="ot">&lt;-</span> fullinv[<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(X.all),<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(X.all)]</span>
<span id="cb54-217"><a href="#cb54-217" tabindex="-1"></a>    EEV <span class="ot">&lt;-</span> variance_residual<span class="sc">*</span>EEV</span>
<span id="cb54-218"><a href="#cb54-218" tabindex="-1"></a>  }</span>
<span id="cb54-219"><a href="#cb54-219" tabindex="-1"></a>  </span>
<span id="cb54-220"><a href="#cb54-220" tabindex="-1"></a>  <span class="co">#keep only EEV entries that correspond to id</span></span>
<span id="cb54-221"><a href="#cb54-221" tabindex="-1"></a>  EEV <span class="ot">&lt;-</span> EEV[<span class="fu">rownames</span>(EEV) <span class="sc">%in%</span> <span class="fu">colnames</span>(X.id),</span>
<span id="cb54-222"><a href="#cb54-222" tabindex="-1"></a>                           <span class="fu">colnames</span>(EEV) <span class="sc">%in%</span> <span class="fu">colnames</span>(X.id)]</span>
<span id="cb54-223"><a href="#cb54-223" tabindex="-1"></a>  </span>
<span id="cb54-224"><a href="#cb54-224" tabindex="-1"></a>  models_env[[enviro]][[<span class="st">&quot;EEV&quot;</span>]] <span class="ot">&lt;-</span> EEV</span>
<span id="cb54-225"><a href="#cb54-225" tabindex="-1"></a>  </span>
<span id="cb54-226"><a href="#cb54-226" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">colnames</span>(H2_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;env&quot;</span>, <span class="st">&quot;expt&quot;</span>, <span class="st">&quot;H2&quot;</span>)</span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a>goodbetas <span class="ot">&lt;-</span> <span class="fu">rownames</span>(models_env[[enviro]][[<span class="st">&quot;across_experiments&quot;</span>]]<span class="sc">$</span>b) <span class="sc">%in%</span> pheno1b<span class="sc">$</span>id</span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>BLUEs2 <span class="ot">&lt;-</span> models_env[[enviro]][[<span class="st">&quot;across_experiments&quot;</span>]]<span class="sc">$</span>b[goodbetas]</span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a><span class="fu">names</span>(BLUEs2) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(models_env[[enviro]][[<span class="st">&quot;across_experiments&quot;</span>]]<span class="sc">$</span>b)[goodbetas]</span>
<span id="cb56-6"><a href="#cb56-6" tabindex="-1"></a></span>
<span id="cb56-7"><a href="#cb56-7" tabindex="-1"></a><span class="co">#Extremely similar BLUEs</span></span>
<span id="cb56-8"><a href="#cb56-8" tabindex="-1"></a><span class="fu">cor</span>(BLUEs2[<span class="fu">match</span>(model1<span class="sc">$</span>blues<span class="sc">$</span>id,<span class="fu">names</span>(BLUEs2))], <span class="co">#open source</span></span>
<span id="cb56-9"><a href="#cb56-9" tabindex="-1"></a>    model1<span class="sc">$</span>blues<span class="sc">$</span>BLUE) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 0.999999</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a>BLUEs2[<span class="fu">match</span>(model1<span class="sc">$</span>blues<span class="sc">$</span>id,<span class="fu">names</span>(BLUEs2))][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## A13125-3C  AF6165-9  AF6188-9  AF6197-8  AF6200-4 
##  60.75104  49.49585  41.13264  49.15104  67.10104</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a>model1<span class="sc">$</span>blues<span class="sc">$</span>BLUE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 61.51620 50.34224 41.92488 49.91621 67.86620</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="co">#slightly different heritability estimation</span></span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a>model1<span class="sc">$</span>fit <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##           env     expt   H2   AIC
## 196 Hancock20 advanced 0.86 213.3
## 1   Hancock20   prelim 0.79  45.1</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a>H2_df <span class="co">#Open source</span></span></code></pre></div>
<pre><code>##         env     expt                H2
## 1 Hancock20   prelim  0.83782148150161
## 2 Hancock20 advanced 0.863069291325532</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a><span class="co">#small differences in EEV</span></span>
<span id="cb66-2"><a href="#cb66-2" tabindex="-1"></a><span class="co">#The matrix that generates EEV happened to be not invertible</span></span>
<span id="cb66-3"><a href="#cb66-3" tabindex="-1"></a><span class="co">#Therefore, we had to use the pseudoinverset to approximate EEV, but as a result</span></span>
<span id="cb66-4"><a href="#cb66-4" tabindex="-1"></a><span class="co">#we incurred in some error</span></span>
<span id="cb66-5"><a href="#cb66-5" tabindex="-1"></a><span class="co">#Still, our EEV has a very similar structure as the one returned by StageWise</span></span>
<span id="cb66-6"><a href="#cb66-6" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">abs</span>(models_env[[<span class="st">&quot;Hancock20&quot;</span>]][[<span class="st">&quot;EEV&quot;</span>]] <span class="sc">-</span> model1<span class="sc">$</span>EEV<span class="sc">$</span>Hancock20))</span></code></pre></div>
<pre><code>## [1] NaN</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a>models_env[[<span class="st">&quot;Hancock20&quot;</span>]][[<span class="st">&quot;EEV&quot;</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##             idA13125-3C idAF6165-9 idAF6188-9 idAF6197-8 idAF6200-4
## idA13125-3C   37.147886   2.006341   3.049659   3.571317   3.606392
## idAF6165-9     2.006341  40.258768   3.564960   2.006341   2.041416
## idAF6188-9     3.049659   3.564960  36.797995   3.049659   3.084733
## idAF6197-8     3.571317   2.006341   3.049659  37.147886   3.606392
## idAF6200-4     3.606392   2.041416   3.084733   3.606392  25.538111</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a>model1<span class="sc">$</span>EEV<span class="sc">$</span>Hancock20[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a><span class="co">#I&#39;m not doing the equivalence for this because the &quot;spats&quot; solver is already</span></span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a><span class="co">#open source</span></span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a>model2 <span class="ot">&lt;-</span> <span class="fu">Stage1</span>(<span class="at">filename=</span>pheno1b.file, <span class="at">traits=</span><span class="st">&quot;total.yield&quot;</span>,</span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a>                <span class="at">effects=</span>effects[<span class="dv">3</span>,], <span class="at">solver=</span><span class="st">&quot;spats&quot;</span>, <span class="at">spline=</span><span class="fu">c</span>(<span class="st">&quot;row&quot;</span>,<span class="st">&quot;range&quot;</span>))</span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a>model2<span class="sc">$</span>fit</span></code></pre></div>
<pre><code>##           env     expt   H2
## 196 Hancock20 advanced 0.85
## 1   Hancock20   prelim 0.78</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a>compare <span class="ot">&lt;-</span> <span class="fu">merge</span>(model1<span class="sc">$</span>blues,model2<span class="sc">$</span>blues,<span class="at">by=</span><span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;env&quot;</span>))</span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>compare,<span class="fu">aes</span>(<span class="at">x=</span>BLUE.x,<span class="at">y=</span>BLUE.y)) <span class="sc">+</span> </span>
<span id="cb74-3"><a href="#cb74-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb74-4"><a href="#cb74-4" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;i.i.d. Random Effects&quot;</span>) <span class="sc">+</span></span>
<span id="cb74-5"><a href="#cb74-5" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;2D Spline&quot;</span>) <span class="sc">+</span> </span>
<span id="cb74-6"><a href="#cb74-6" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb74-7"><a href="#cb74-7" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept=</span><span class="dv">0</span>,<span class="at">slope=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb74-8"><a href="#cb74-8" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">25</span>,<span class="dv">90</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">25</span>,<span class="dv">90</span>)) <span class="sc">+</span> </span>
<span id="cb74-9"><a href="#cb74-9" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;2020 Yield BLUEs (Mg/ha)&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABJlBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYzMzM6AAA6ADo6AGY6OgA6OmY6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmOgBmOjpmZjpmkJBmkLZmkNtmtpBmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQZgCQZjqQZmaQkDqQkGaQkLaQtpCQttuQ27aQ2/+rbk2rbm6rjk2r5OSr5P+2ZgC2Zjq2kDq2tpC2ttu225C229u22/+2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb27bb29vb/7bb/9vb///kq27k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///++K1L8AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOJ0lEQVR4nO2di5/bRhHH13nUOG2h9TXBpBSIU0ig9Mw7OJSohTZNziQQ7hLF4RxH//8/wc4+pF1ppJVWT1vz+7SxLc2+vjczu5JsiUWkQrG+OzB0ESCHCJBDBMghAuQQAXKIADlEgByqDOj5B4xdewzv3txm7MPHedu4AraM3+xX18+TSpJP+9VUbdswoR9B6c2VJ3qrerdbzGMTxqZRRrufrKPgwwe6l5/O7d3h9f9aPdBdnGc2pVQVUCg6OFlDl+Ed9B/bJgcl34V8QKUBidIegGCogegE13bGUiMPpnsM0PbG2jHgioD2q8mDaP8N9HDDPj7n7+boNqlQjGS/ioebVIMAElb7r8Ht8gBlKooVwq5g8oFsenN1ZgPaLZYoIA7OMeKKgLaiYWhLtgf/Ytt082K08I/cqCNQfno2Y9e+SwGSLZQAtP89d9o78SdRTXDlj6Lp/eoz0SfewORnYg/nx9t8cZtNfg7mz3hOkO+cLuSVpKE7u4UYWaD7jG0TQbadSTfiXeeeL2NRfBIBc+12yoO+gShxA9qvRKypJMdrXoqGvxXj3b7/7SyJSGhgA0F+7bYqonYsI9ODmwQEPiH9JtqoqEe3QZC9JwNMOhj8yZ/PZEbaLXhkvlmxdA6aR84cxOvfzj4C4Dbd4Mr3KygeTKEnRgP7FWQCxv8q/xL58MrjSBcO8sPWG9AzaDMFA9sm2ld/5SQWYTDJJ2OMavR84KUAXf3su6RLgQgtPlh4wxMOVB6yuAEIJJkKlZP/5x9/mMk/jdnZZgDxNAq9sWBg2yI1MAkAkITx+OQn28UVCkG6RA4K4jVBFKc9DijkbYcQ13PdNWgg5Lt1ghR/yGQuDOMwbQgQ99R53GPlodg2X0CitAEonrjTs9jz28mKIgEEXeAzUwoQrAESQHwx8sM///PFohVAPLDvGH0S/2LbMEAykyaf5sb+PEC6/xt79pf69+/U3gQQJOqZzIZJiPGYMwHJfqjgbhpQEIdPAGuer6ET2DYtE5BYMPG0KXOQwLo3krQOMZjgYhLbmYD/bGZPbnxcH5+rKU80HAMKJz/l/9pJOpRungCansOKQ5RpOAep6AXXjlfN2DYUkFpyi7Wl/mRM83qZvLRmNGty0yaAWLzRviqHGcgu6OkinuY38UJDAFJlWQy1QUDm8QCs+uAIDNuGA4q2fPfVO/oTzyLXvk8fany4jmwqzz+FrQ8shrwpsVD8SMeyXgc9UYvTbbxQ/MViul/NIxOQ7OUDlT7bWAcNT/njVBNHntpZSQ9QYTaVbGfv8TQVFCfhpo/FhqvsiYtUmkLV+NH8cAXng1Ky0xSq5s8HjU4EyCEC5BABcogAOUSAHKoI6AfjkSeg/F2vXGVrG7TfgmFAgPIN+PI7YgQo10AcoJAH5RsQIIcBAXIYCD4EKNdAESFAOQYaCAHCDWIevoBeHbdY/I48CBOjlXShGB1qFO6V87sSAUpLXZPUHwlQSkwvoZUIkC3gQIDyDQQGykG5BjEfAoRKQiBAeQaKAQHKMYgRcD60UMyK4QYESInlGBAgIcbyDAgQKDN6AmQpO3gCZAoZOwEyhA2dACVCR06AlFh6+srUMFpAAo19VIHWMFZAEg0ByjVgiYprGDUgykEFBrnOk6qhBKDLuyc3z6Lo7a9Pbr1MF/PsXT2DZlooJFQF0NvfPIoubr189/A0uvhxuphv72oZ1G+hMP/YNbgBXd57Gb394oz/F13+8ixVzKd3/QMqTtB2DaU9SHDib9VXgHv8YkF9xXxK2JbIQTL5vL6lAZlcC+G3ZdCYB5WpoUSIff4oen3zLPEgs5hH74YCqGQNbkDKdY4nBzndJ/LyoHcP7x/HLKYuLpetoUQOen1y8smjo1kHsYo1jG0lzarWMDJArHIN4wEkTm9Ur+GoAZm52Jq6CJCQNZtbRxcESAgHxKp14YgBWUHFCFDagOU4EAEyAektmg3loMTACjCaxRADk0/qBAcBssQyp1jHCwjLwSzKnCEbLSBFwcpB9vxVuQtHC8iYtzJ8CJCJBeEzXkA6Bzn4jBiQMkioZOFU7MJRAjK9BsEzVkAxCjysPLtwPICQaZ0AmVJEyvAhQORBmLIR1ieg1r+BUV0Mvq1hr30aqPV4PMixNvTtAgFyGBAgh8HxAMoSaqQLRwSo5BqoYheOFFBzXcABbRhbbopuAj9oQI12AQUUXH8BT3IreFDAoADpFXTsQW0D2i2W8BScsOBZJUMCZAZWuQgbLSDWEaBoAyFW+CiT4QBClj/tA5LPIyp61MRgAKF8Wk/Sbg0CkJmXLfchQEKo81T9dovTAAWknhQ28CSdy6d1QM5nbY0dkHgUYLEGACifTwcedJCA2ukCmoOKloh2sXqNexr0Dkg+wnHYSTovuhrvwmFN85LDq8y1C5vPeAEpHrme00YXMoDEkepAQ6xg4mqtCwfoQW5CowVU1oVaDrG4YRVi7x4O6CeZ/QPK6ukp/PJ5ILemGGAOgt+Dq5cB/Cy8BJ4uAImnTavjjct7X0GIDeTWFCabThrET7lC9tGnXC/vngKdYdyawuU7zXeh4GheHZEp1+nt1hQmijJ0Gu6CG9DbL3u9NYWJw3SfIkjtH83DM8njqxpPRYj1dWuKHECFbtTV0bxaCvEFUI+36DJRDAVQCXWbgxh++YIACWFkHIm6ZUAbnoEgygZx4bAynKa7kAUEGRquawzi0rMPnpYB7VdzuDC2hCv0+V8Q6gZQ9eBqvAvoCTM1z/f+7Q5fPh0AEs4zWEDtXxYoDrGlurQa9B1ied7TLyDwHpGCQlZw+bC/HNRsC24DZJoPYIbfryAN5aqjWYwx66YJrPlL706DIS4UGXZo0WgLhw1I8iiYugiQY24fM6C8md0iNGJARXwIUHQwgMSXgItWQb0AaqaFygZZQPuVOMTYLfr4MUsZPn0D0gfx4rA+T50CSp+g7/9YTKrz0x1u36nbgocBfjQv1OnRfKnkU6sFT4MBAGI5p+Tz+IwNUAEb54N3GupCkcEQAb0aOKC4d0WAmvv6RJoJf1/pySAtawAraZa+acsrx08re19Jl1C7s5jjl5V9AwpUcHWXgyIUUGMt1DLIAoJz9eJ8dDeA0LlryIDkLLZbTHsEVPztn0EAgis/fQKqUEHLBniIRWK6bxNQQuHgAEWBvOCjTnvgqgsoxuDhQL0DKqOmAKF8hu5BZUSAHPIHZICJCFBWavy5gKLiGb5ECw0b9AQoxoB5jqv7IwAUs0Aja9SAkvOH+V97HjcgMIjdiADhBgTI1TgaYQRItYTOX+bkPm5AxjyP4yFAaAKqUMNRA8rEFcJnzIDy+BAg3QqChQAZwtzmGADBXRfq/iQzO3fZe9w1lN7fPaCLk9Oo5q0pXHnHXUP5/Z0DuvzVb0/r3pqiAp+DA/Tub39/eFr31hRZQG1806AFlQB0cR/Cq+6tKSzfKXSgQ/MgcVMB04PMYlUaLxFcjhpK7u8Y0MUJ6H4Tt6ZQsVW+d377+5nmm7g1hXSfIwXUwK0pdHZ2RdnhAcqqGqA4M5eZ5McByEKALBLHDsiezQlQpm7JQJNIrxApB6UApbLPEc5iWZXKQSigo5zmsyo3izHznFh8xZkA2cpkZQJkKTtnESBD2KROgBKhqx4ClIgAZYQeY3j3zm//IAGZKx5zYxs/OT1EQPZRRXqrd+/89h8WoFq989s/ZEA2qdq9a6mCHnNQ+oviNXvXUgUDmOYJkKNuAuSou4APAXLVRoAclRGgtn/NdBCA8r4MIc7KH5Ma9qCi2UtrHB6Us50ApYqlRYBSxVJy3DIh3binweECYhUb9zQ4WECsauOeBocKiFVu3NPgQAHpGgiQWUyXNnIzATKLqcIEKC0C5BABcsgCZC8PCZBZDClKgMxiSEkCZBZDChIgsxhSjgA5ihEgoxhWigDFxfBTPwRIF2vruwnHAqi1K8tHAqi9K8vHAWg88gNURwVXZTuqwKcGAuQQAXKoQ0CHKQLkEAFyiAA5RIAc6ghQ9h4OFYuffPKoTgXR5d2Tm2c+NXQEKHsvq0p6Km7TU6MCuH/NhVcN3QBC7mVVRVBSvfhVIG/y88WZRw2dAErfy6qqLu99BSHmX4H2II8aOgGE3cuqii7vCrz+Fejk41FDF4DQe1lVrABK1vCgy88fRa9vng3Ug9B7WVXR2y/FuGrkIOU6Q81B2L2sKunpqXRD7wqUB3nUcBjrIF7SbxUT6/WJ50qKVtIOESCHCJBDBMghAuQQAXKobUDbG2vjRch4yOR+Je9IMM8pXfQ8Sl2Wm2zYZA3/p03ePPbud6wePMgCJNBsZ0unaUaqbCQf8Jk8yzuR+Vfx1iAARcHUaZpRAghAYDAOAlAqxHYLNvkTDmg7E7G2vfEX/maZmEIowW6xfb6VO42youDVGRiBKVQui8D2uaw2x0HLqFtAu8Wc/58BBE9QFjGyufJkO7t+Dq/adL+CgcNwYTuTO42yUeJBYBZtrp/DK1QXe1ZuCJdQt4BEyGyySZr3/3/n0kwMhr9qU/EaArilGqmqK0nwGoQw5WS0u4rt7+cHaSl1C2gDTwM2+iy8YDuTrhDy0U7W8k9+Y61NQ/F6I95uAEp70EYR05lLGAYiQP01AEDyGeW7xWRtgvABJB/GHFmAXA9ndqmHEMvOYgHfIkCEiQdp0xCWNxBibkDhZG22mcxi2BqgrLpO0lMsSW9nUzG67SwBpE3jJO0GtF9xyLweXQS4CM515vsOAMFI1Iueu/Xg9OuGp1qeLCZ/VRkW/slM8xlAMuVIpmIjbJqs9TQP+WeqMpv/AOhYzCEC5BABcogAOUSAHCJADhEghwiQQwTIIQLk0P8B6cRG1Ufhjj4AAAAASUVORK5CYII=" /><!-- --></p>
<p>The above figure shows that the BLUEs are similar with the two
different models. A figure showing the 2D spline and spatial
distribution of residuals is also returned when SpATS is used:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a>model2<span class="sc">$</span>resid<span class="sc">$</span>spatial<span class="sc">$</span>advanced</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAIAAACb4TnXAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO19f5BcWXXeuff1jMRCCAaxEGYHS3GPiBURYlCVlh4wVmL/MbO1EzmWFYNjyUmqekquVHpU8cZJactJVdhK7CVoOk6VosFVLi3ldayI8tY424NDjEg200ZhMMSWFVvdZgSjYQEvYIN3l11p3s0f795zz73vvZ7uef2m33Sfr7qk2+9H9+3Xfea755zvnCeUUsBgMPKBHPQEGIxhBhsYg5Ej2MAYjBzBBsZg5Ag2MAYjR7CBMRg5gg2MwcgRbGAMRo4oloG169MiCdPT8yvtPr38dD3lldr1aSHmV/ryLt0fvFKfn54mn7Pejw9qJjI/jZ+n82d3T0s9cmU+8evR6MO16wL9+pp2CcUysDQ0m0uzU3vnonaLdn16anZhqdk0G5rNpYXZqa7MoItXf/bqUnP7wxi5ooAGVm0oD61GtQKw9OGsP7xybVWp1Vq5P/PMjJUnF5oAlcVGy37QxQpAc+HJ/v8t6ctnn7lMvpTFCkBlsWW3XJ7p01yHCQU0sDjKM5cfr3rb2ivz03Zh5a6r2vV5u+oi+/zFj32N+Cv4yyR/ZdJeIUs7MT3fu/G3b98EqCxeqc3gr748U1ttVAGWnoneaGU+mgWZp/c+KdNo16fF1EITYGlWf46kz55t/mkfK7pQZrVvLmzq96UntrJSn075kOlf016AzxYDRWuxksBgrVajCkD/XLYWK/7nwJ0J+8wrOn90/eMq1ap98/ifZ2dqCe/h7uzq0zaq4L1J0hHRvJI+aYdpOHsqi61tPnv6VUpH2mHR9kqFTrbH74tMpvPXtAdQQANLg3fNK1WztIoWVnp/9P3afWSX85swNhsdqF+iSwMzL9pydtK9XX7ehradSqVaXWw0Wv6vtVE1H92ZZVfT8P5a0U+0/fyzGpizp+P3hd+52d1yPmPnr2kPYE8YWKVCfgyJ36z9Melvs1JdXIz9YMmZKRbUJYPpTa1Wo7G4iAzTu4EppVqtxmK1Qj42/awJHNeo+r+vlGl0MLDt55/ZwMgMO39f8ePpR9zma9oDKKAPZi9eFNyoVE9fuWz9FGjdagI0F6ZohHhqoQlw83YboFx7vBqF4xZmp6aiFX3Cwr11qwlw9DDx+cuPnO5Anx7a9WkhpqamZmcXFpYyxerK5Zna5dVVpYytwZIXR3RmCVNHKvqTZplG/+afjMqRKRx3/r7ix0cfkZy786+pACiggVmUZy6vNqrNpYWeItczl1X0S43QXFrod4h/ZX5qoQlQqVYXFxuNRquFa7lMiGztymIFmlef7eLz7ngaOc2fkYBCGxgAwMzlRhWguXAWTWzqSIqXa4PQ5ZnaZUsKYMNyCIcIIrRu2b/k5cNHAZq3Wol727dvAlQbavXy5VptZmamXG7fvtnzB1uZF4n53NhbO7NsP3vV/E3f8TT6M/+u0cX31fnc9K9pD6DwBmZNzCSHyo+crsDS7LRd+bVX6tNGSaB/t3pfuTx1+EjFW4IAmJ/xwlmMItenZ5f8d156pp66d+kZ8/btlfmzC03wfwjbfqyT0R8OZwXbbq/Mf3jJnW9z4aw+pL1SP7vQBKiexIzTdtNInVPm+XeNzt/XNud28zUVG3k5dztCcpjejyypxBVNpxA0jZ6hy+y9iBf/7bQ3LRZj36W7j5u2MHOnWKm475YawU67ELEw/fbzzxymd7dv+311iCd1/poKjz3AYABQrl1ZrAAszeo/ejOXVct6WVGU0Sw4yrXV1H0OZi5bV61Sbaw+dsTba3cutpy95dqVBnmH6mKj1ah6S8puYLxF+3OvVKqL/nSPPr7asDNptFAwsc00yrXHo73NWy2Xmvo1/67R6fva/txOX1PhIRR3lSouVubF7FK1wSKkvYu9wWAMxh4FGxiDkSPYwBiMHME+GIORI5jBGIwcwQbGYOQINjAGI0ewgTEYOYINjMHIEaVcXnXzWu3cxoeWzx/HZ0+tm30nLpjtcbzwwgtpL3ngwIFi7k3b1RkdXrPXOfTllF14i15P2fG1LRRyMLAbF+eeuA5wAjfc3VjvZFUMxvCiz0vEGxfn5p64c+bCmUN22+bGnUOTD/X3fRiMvYE+M9jx88vL5wE2rz1nt93dWF+/fm7uKQCILRCPHTuG47W1tc6rgmLuZTA6IB8fjGJz4w4cOnOpfmoCADav1WrXHorGAABra2v02GJ6WXn4YIwRQf4GNnGqvnwKn0weXG/eBZjodAaDMTTgMD2DkSPyN7AbF+dq1zb1k82NOycqHE5kjAzyXyIeP39po2ZiHIfOXKqzfTFGB/kY2MSp+rL77FT6wQzG8CJ/BusRH/jox3H8wHdtrdrLb7bjB75hj3/pzfZud+XjX8Vx+7Nvw/GBL97D8VdPbeFYhPbc8bvjOH72Z/4Ljo//3E/h+MV32HM/89P/eJtP0jU++OA5KElRkhA9AgmBFOMlkBICif8qKWGsBFKClAq3SKlKJRUEIAGkUBJAgpIAJaEkKCkgABVtkSIsgZLRdlASQimUhK0ShFKEEpQAkCCkAgkQKCEBJED0VICSCkoK9GFKv4tUSoAqKSVBCaUkKKmkVONCvTbYKomwJMOSUCUZBiIsSTUe3A+EkiIMZBiIUMowEGosuC9lKEX4r8sf6dclLQ4KZ2CjiDFjXYH+V7imhRYVbdF2Zf5NsC49FtoeBIAQSoDS/4I71gNtXQJAgLYuoZ8qsx0i64qOFwr/VeTfEqhAKKkfII2l65OEkkIJoYTQ8wIztUF/DbmAo4gMRo5gBhs8xFiAK0MIpAgkBAJilEV5TMlAk1gg4/SFW8xGuyw0D7pQhFACyIhMzLJQAAizOMTtmoMofekx8ltJqEBAIMibRGSlWQspE+xTs2UowQzGYOQIZrACYMylL81LAZBIhiE0S1zEE/PpC9kphb40cel/I6fL8JX/sN6XsmRljlRChz2UIbpAqECoQLthIIUSlMoofYHhtIi82AdjMBi9ghmsACgFhLuE5qsAXSz0uxJCiGCZytIX4TSR5HeJUIASEEYkIkEgd0kAqYQJxIOJLibEDy2JmYeEMaGcECLoQCISl/XH0O+ye5nBGAxGj2AGKwBKiYmvAKRIy30R+hK90VdEFkhlmogUmMSXoC4WOmASE2I0fki8MqEkRA4YSPMvITH9EMhXnifGeTAGg7EDMIMNHiKIAoaCMpjvfelHoKRU+shU+vIyXXqj8bhM/FDLo4TE+KEShNBI+ksZ7rIPhSIPQ2VjQpVMCNEoSRRJf2m+MrFEoMQVqTqGEsxgDEaOKByDPfCev8Dxgf3fw/GXN9+A4yM/vIHj/3P77Tj+3v0xHKsHrcD3hXfZ7YsPfwrHX7m3H8dPf/IDOJ79rZ/EsXyt/Rv0yptf6fJT9IbAJy5LWZ73pZ8K9L5AAgQm9yUEDSr69OWkv6z+0M93uTIOw2ZEeWg9NJQgQimiL7DxQ4FaRDDiQ+tuGeYzzAagOA/GYDB6RuEYbBRRCkDzElXNB4bHhBXOx+hLkZih9pSIAxbRVyhszFBJCIVQgugPjfhQoAoRNfgercnE+CEII+CQkQ+GMUMSMCT+WCyQKDgPxmAwdgRmsAIgiEcLhXI3Ipt59EUHSldG2kAiEPpSSF/SpS9h6AvFHI5vRhX0Vn8IRr2hhCoZAYcR0UNMzBHFDyl3KcBAoi5YG/S3kA+YwRiMHMEMVgAEkUZegGWwSDgvfOlGoD0uIDFDRRT0INw8mEtfKk5feuzJNZSnPEQqU2ajccOwBix6eJVgkZreln4REiNsJpQOJA4jmMEYjBzBDDZ4qIASl6Yy6oMpIYlWw3e6LH1JqjMEw2DG+/KKl42XJWgVMwo4rHrD5sHAOmNOsTPxvlBzaKOI1gGjqg6grKVgeLWIbGCDATe1j2Morwkb2GDg3FAiIHkwKUHE44fCYa0k5SFYyqIOmBUfKqG185a+bFc2Ql+SBg+NGyZt0yhawqwEjGsJItAMmFVygMNmVsBBFYmmw5R3k43hsDf2wRiMHMEMVgD4qnlhuh1GhEacrkCYNqBJ9GXihF7MENNfQLXzWP3lyg6pbsMrAEO/C6uYSfwwXv0FJFqoo5tu/NBWOg9xPRgbWAEQU0URe7MFlBjPoDF6V93rFFNGeiiaXPaty4RMtAyKWJdygxxR7N6NzisZpZiNXbk2pjPezvowUSQVmdbwhumLZWAHDhyYeO1f4tM7n7Ltr4O3hDj+0uWDdvt77Ol3X3wjjt/0Obv90Qv/E8f//uZ7cSx+46/g+Fs/GOB463Wv4vj7nr+P4we+u89O9UPD4CEw8kaxDKzXu9YPENFU++KIY3eAKMJBwvQ6NZug65W62ER3n6cKKQFE4KsrU0KHrAh9CUfFi8F62sKNRDtIn3paYUnDG748SlNWbH2IAQ8AE/nIfiULCA5yMBg5olgMNqIIMDov4g1tHB4LqEKKCKNisY04fdnO2JS+pKEvn7JMeMPEVFDyiw0CAkpfNrxhK1Ns11FacEmiGm7AYzjBDMZg5AhmsMFD+bpeAYJG5J1SFCO6pZlltxObwNYA5K5fVtqLjdlirUX9uLwprzQUp3sESN0goGRbZMdi9Nr7omJfr122aRZgCI1bBjAYjJ7BDFYA0M42NK2MqijhRxEJpWiaipeoODetFMn0BVLF6Ys6ZqQ9mya0AOkrHj90lb62TQAkOmDKxg+5ZQCDwdgBmMEGDxUpOYS9cZ7rg+k6FCcb5lVSkuAh8cGMNooKo1z6oukvWpxi6cvUWUZemcAObSR+GNf4ugIOqu61PAYmfmh4bMDfQk5gBmMwcgQz2OChgsC6XiikwPJKQ00RrSnszUZzX0R5qL2yGH0ZfYZLXzKBviLHjNxcz7lDrMdgvg+G0xFkgvE7P/i+GbAPxmAwegYzWAFAXC/kMRU4iS+gHpeTELMBQyrdEF6okPYVpfRFRBtIX8haSjhu2JhQYyLEDm2p8UNPwGFSXpbEAKwDZvyxYf1LXzgD+/on3orjT/+bKzj+yZ8+heOXy6/BsXrJesfqjVs4/uYP2yXHlevTOH7dTaua/9ZPvIzjfS27Pbhre9l/9y3E+x5SR5yRHwpnYCMIDBha+qLhRE92SHUbwtCXDh7a3FcCfdmiSXsDB1MG5tAX0hp1w7zyyqQ6S/DYzAo4vFSYuwVs57YhxLAyM4NRCDCDFQDobmHNsgAVYDLKlR2Soi/aFAAlHY7yUJh76lmacvsCiAT6UlL59CV19VdcwOE1C3Bkhx5l+RsjBwzDosOJYf1cDEYhwAw2eGjNIf4lJwkuENQTE2GA3pHhLuJ6JUg3nCY2MfrC+KRLX7Ynh6Evw10h8b5CP4QIXlCR3LnWoTIAZyNgg7dBfw+5gBmMwcgRzGAFQEDunScSKr4c3QZ6Yqb3hm6ILfzcl1Ue2oovZ4slLlc7T6qYNX15DljkcSXkwZL8LpMZA5/HXAdMDmkKhBmMwcgRzGCDh+0VZYhFSVRyaO4Cl7tsDVjUENu9nwNI433FlYf0vg3uLqSviLi84uUO+kPtRKFXZmad8ADyMDyGfR+HEmxgBUBgIxmmYaebXzZtZ0K7StTtRJ2SypTYBgjHuhSJ0ZMlqLUucK1rzI1wRLYUuKvEuC3hTIWAxAPQ0kw5Dgc5GAxGj2AGGzz8nLJeIprKFLJQVIHpKIppZdpuXiigt6v0loIksIFsSbPPrrQ3gb6czLJ9+PF6n9P8ZaGVR4HmOk20Q4nCGdhfHrThpA88/TM4ftvGn+L4L/7BQRz/6Ltv4fjT19+JY/WK/crUkRdx/MqXXovjN/2uFQ3/4IfaOP7cZ38Axy+++xUcv+7z+7v7EAyGRuEMbASRkFOmcl7ilYWEu3RaOVnU6/S49psCuIME+iLcFcsvK4lsRoopUz0x3+8CABOyR69Tx/0zXcPwa4ezfxHyrbezv4gHNjDGMOB+PwrKxrO/RAxsYINH5G4R4ZIumrT5ZZtTpq4XTSIrUpDi3c/BOUy/ix2gm2e0UVJ3xt6GvjQ7hdS5ojyWHDlMojjskJDlGt5Tffgl52Fg+fiWm9dqcxdv2Oc3Ls7Nzc3NzdWubebyfoxRx6sgsz/ymFgODHbj4twT1wFOmOeb12pPwIXl5eOwea127uLk8vnj/X/TvQZ63yPb00YKZBUVOPSlBIS0m42XVo6popRLaA5rubpeMKwFSfRVMmHDOH3Fe92ImPPoPoDSHRDuinLsWe4FdU8F2x80CPTZwG5cnHvi+qEzF84898SG2XZ3Y/1E5TgAwMR733/oqeaN88fZwvbQndB2DVlugj4qBnb8/PLyeYDNa8/hps2NO4cmT0fjicmD8NzGJhyf0DuPHTuGB66tre2hG8v3caomDyaQrJSAkBZ7YPBQeq5XTLdBqYnINWjw0JFuIH2ZLWM6fhh6tZVJXW5CxxmLxRJlijOG5SqG+iB7FPGeKqhYOP8gx92NdZhM27m2tkaf7qG/6328wyUjO0aFwRLw0OSh3N9jb4OWn+iiDu2D6VY2OqjoiDZM4otyF5Ud0pJKEjw0b+G0A9DHWO8rgb7SFBsyIX4YdhM/DEyhKYqcs1zDV0eXwSYmD6437wJMAMDmxh04WJnY9hwGozcUlsF2QQL20OSh609f2wSAzd97zsQ7GBZhYMKGUoRShIGuoQylCAOh6ykjyXyAjbWViPwx8oia1ZAOcKCkMn2wsUNbtJFKN5QSSgg1LsIxEY659GW6A8TjhyGhrzDN3UovV4Ho7u3GAcuaJ35VyeyP/nydLnYh0Txxqn5hY+7c3FMAh85cqrN9MfqPezBSS8SJU/Vl+lwHFxmJsCWKJGzYQbRhXC9no6fbcCOH+gBlc2I09wWBVc2bf2XkeoUycpZSg4e+dMNxyWKpMOwxGlXA2dsMAmSsB7sXFlSOXzip1NHjX8HxN19+AMd3//lBHL/+C3baLxz+qzhW+8iXNG7baItbVkH/avkejh95n1WbfOM7Nh74tiWr3P/Sv7Uh0A/81BfJTM90+hiM3UVOOozsKJyBjSBCU/pFs15bgY4WJog2qOaQSuapbiMu4CA5MeMGgYo1tCkJwl2eViOJvmL6w9D6Wmn6Q/QZdRRRZW96k5MHlR1sYIxhwKu8RGSkIQwc7kJ6EY5ig0rmHSUH0I2Jug2TAfOrvwRg7stTzVPuCtK9L6rhwDih6XjjbrdRRDCN32xbumiyWTBiQQ4GY3fxSsgGxkgBdsC2bZacDtg0cui4W0q4zlisltnqNqjs0Ag4Al257FV8xWWHCRoO9LjsU6Og7JT+IvHDQJOY7jqaNYrIQQ4GIz8wgzFSoaOI1vUyRESzXjbNFesVRTWHGB6U1vUyZVhKEcdsjKS8sCF2N8pDVG84STBIyInFw4ko3QgwCcZRREbeIPf1srEKCFzTMianhL90VDJplUgKUkhfUaUESKekUpWMDMo2ZgNniSh86wrdVSJaV5iwLMTAvT4epL6hJylUMW2Ls1zDERb7Mhj543vMYIw0qMDlLici7zUCcNqw2ZyyTSV7t/lyeYz0mg8MU5Vc+upUkCJDuhQk8QwntqGDNV6MHpQAFQAEuDIEextqIbLGKHaFwdr16albj6vLMz2c0+3natenhRBCiPkVaNenxXS9vf1JDMYuYSuU2R+d36JdP7vQ7HliXRnYyryYunq6pRpVAAAo164swsLU/ErP78ZIgjBFJbpQMlAiMONoexAVm5BjrPtCq1GUcWtMWjnaK0FJFQo1LtS4UGPmUZI2wuE/ZBhVqbiPZFFvEn2FHndZksQUMyExmTnRvBUG2R+d3qBdP3v1aLXS88S6WSKuPLME1UatDGhR5drj1YXZZ1Yuz/TClgxGXvjld/8mjn/+8x/s8qyPvOc3ujurXT979fSVK3B26Vb6QYkonA9281MHcUz/pgiilP/2j76E4//7Bat2f9PUd3A89h/sRzvwi9/G8e9/7UEc/87H3ofj773Fvr76oF3QK/Eyjj+3aecGf73Tp+gNRPQU1/LayGFSH5vUnLLtjK0ds5JQ4zJE18vmlGWIN630ZVCQvMVpDQDJ9KUPgxB3OaWhQAKJAiSAEJmiiOc/9w/Js26rm7s8KzKv1TLUe59YNwY289hiZerD9cdmsP33yvzsUmWxxfTFKAjCzgu8LEDzgp2EHbpisHJtVR2eF2IBAGBWLAFUFlurtfIO3o+RgHjKS5g/9W5okd5/CLNeyg0wes1tlAAp1JjQt1qmKa+okrJEicuvrbSUZZ66+WUvYOh7XyHmlwOAAEyGDWxyGct0MvtgeYXp289ebTabU9GPHwBmxc1efvxdLxFnLit1uffpMRi7gW1CFBlQrq2qWjTcSZi+cD7YCMJPeUVEFLjche6WRJoyWS8r1CBumCG0MaeY0hdtJGp5BaiA+FEk9xUm5LisVxaiPyaJbxaAClDaK0CaO+aiVyaMbCrLNdzay4nmlXkxu5S2s9rozaIZjByQ3xKRoFxb7fmvQDfTmnlssQJQbShEowpQWWwp1Vq8OSs4IZYRXsorUCqgKS+by4oanUXpr4Ssl9AboybbgVT7ZIjN2MZEGCW+aHeAklABSYVpLaJMFXOQOkuv5jJEZrOEJhxRr567cAOJJsCYUYhxPwyyP/rzbbroxsBWnlxoVhuUp2YuN6rNhSdXoFx7vApLz7CFMQaLMJTZH3lMjH2wAsC6W1jhHxPI2yIUewxJdvn92MaJ64XVKCXaysawR9Tfhjha4FGWX41CA4k0CUa73ADSl+ODWR19dLd3U3ApMt/84f5e7skx89hiZWp2/qR1tmwerH37JlROT+U5RQZjW+zpIAfmwewfmWpDrc5Auz49tQCLLc6IZYPLS07KC7mLiDaoYgMpi2S9YEyofTLEZBc6VyUZ6vCdEzBMEGpIGZLwoO1jI13WshFFu1E7Ywn0pRuEk1JLcu+ijPVge5rBACA5D0ZSBAzGIJFTiCI72AcbPJzOodbdoj5YLOWV5IPRrNeYk/LS8Tqb/nL0GSiTB8fdIgwmyCkx6YYj2sCNpgcBpS+Uz2NRM2ivrw9KjoJWNHf3ubAazAGH5xlFwb1QZn/kMbFuGKxdP7vQ3JWE8oEDB17/sFW+f/sbtqc83LfDv/1GeyPMP/rTh3D8wldfh+N3/uLzOP5/n3w7jv/45/+zfZ19Vkz94CW7xvjKP7F/Dl+zvh/HL//Za+xUfzzTvS2dW2MGrobD8cHcsCFxt6gPBgLsrRtASaHGpDJcYT2ukr3RKyBHBdIRHBpxRig8EiMbPbKiGg7jgCXQl4le0r7ZyjqYQmW5Xej9vRzkaN1qQvXx3ZBrjM4tZPfQJ901ZLoJ+l4OckwdqUCvdWaMHkDuKuRXeXVOeQkFAmxVsi70goisHCmGS1aWr8D4YGDdsHgGzHW3HAbz4orS6OUT6CuqzAbDY6SWObsPtqcNrFx7vCqcPBiDUSzs6b6IK/OzS6ALwShY59sfqAAAYjFD6XJXkg8GWKdMuEuLNmRoiMsqM6xG3jpdECRJMWyTQ1fhkeiDGTmi5qVAtxa1oUIin1dYv0wyYCAy58Fe2csMxqVgOSO2RFReASU4S0QlQJioRiDUuAipaWE8QxqTwAh7oNPHQC0qyimTfDE41SheEjktyCHIOtAsEU1amRie13JUkBVjtku4pw2MwSg8wkFPIAWcBxs8TH81ldhuTUmlAhWV3YeBEhLGpdovw30y3CfCcRGOSTUmwzERlkRYkmFJRgPdeq0kwkCGpg1bVIoSBiIMRCjNRilNNF8q3aFN+g3b0h7CCLJKGGWx3QE0fQVm3egpfQVmnzOLfWUosj/69H26E+vimHb97AJE1V+VqCysUQVwC1gYjEFChH145IFuDKx1qwlHD5cByoePws3b7agebOnD3Ny3TzAEZUgsKppUSqqodDKMOodKNS7UfhmORw8RjslwXCqfu/QNUyh3Re3ZosqUMJCaoAJDWQGlLBn6t3HAksroIc0AQiH065QERJRVoj3hbJcb4xw6XUdVFO0Qhs2yXEI6wR0/+vV9UvTmg00dqTRvtQDKAAB2xGAMGDmZR3Z01TLgpKlaLh8+Go3at29C5QiXgfUFxNcC87DOWPSI6v/3yXA8+leEY9J4XzIcI65XSYYB8bt0a1HcaGouA3S34o5WtMvltLjrhT1tSpGXZUjMNGlza1IE9hhN2ZjxIm7145EDuvpcM5dbizdnp+vtaG04K8TUAixe4TIwRlGQsw+2Mh/F9Xq/50mXS0Rb+ZV3UuzP174Px5L+Uflr93C49mdvxvHnz17B8ft+7Wdx/Dfe8mUcb35tAsdzv38Sx+pPrJj46yftm+3/gzEc/52f+AKOf/d33tXNR+gVKlAKknPKEmy+S4Iak06yKyKBkimOxHoTqXPKQLJbLvkQ3VOkn/JSXjI63d/uZp/BBgmxCba587JpMOryVQC2OAWlUviJs1zDPJeI7fr0LDSUiiqM5w/3JK/oLoo4zSF5RrER9uORjNatZvXkDABA+ZHTlR5bPHWrpmfkByKAAjCiDd1NzVR52CIU2urMKUIBWnUSUDGuHkMgvbqSSEXl6jPAKjkAt5MiS7BRQcC+ANJTRZHsluduCZe++pUH+6Of+wUcH/1Pv9zlWTf/6b/Y/qz27ZuVI49F4/Lho3D1dhtmuvaOerr5AztdjILi6H/s1qh6Pqt1qwlHdvDiEboS+z650AQg/e81WOzbH4TunYci4hqXobRiCJ1ZKqF4wqh4he67RugrKnkknpW0CsOQ0BGYQUi5SwgF4LlbmK1yNIfSulu+W0WdMVpYibkvpC87vWw+GOTng00d6f22exYs9mUMBfLTIpYPH20+o1O+7ds34ejJXhZyBdUgjxRUoCKRYUmq/TLcL8L9mO8yD5vsimSHMizJMNC5rzAQ+qnJgCkpMA8WosIwehpo5aFJcEmT4zISDWcjkXHYbtsQDaAE+G+0nSbElOmVHZVaqvg9Y4X2JPuQB53JWWEAABDfSURBVMtTyTF1pKJlS+1nr5p4R7dgNT1jKJBPmhgAAMq11cYtMSUWovvi9eYVsYENHiWT3dJlXWDdLVtdb3JWEuu4jBMVpbxovksAkCovSM5u0VsNab/L3QgKaBIMoEQqJmVC5JDmuCxZCecUu8VI6ZXsRx4s49nbYedeEhsYYyiwVVAtIhvY4LFfhjabZKQYJcMkmO/SDEZoytFnkESWm93CGCDNidFUWAjCz4M5xcumHrkUE2cIJw+GZKWEobUYWVmiE2Ys+pEHU/ky2M7BBsYYCjCDMdKwzzaosYIMvMkQ9a8CIiaUhpqkoTVaW2V8MIh7Vt6RAuUdmI+yBIhNbEAKVQIn6Ie9NAJkJGFpDZ0x4ea+hGU/JDeV/eYPhW0ZwAbGGAqwgXWJ13/DBlx/+9/9LI4f/pSN4jz+Q6s4/oX1h3H83g/8CY7/28fej+OX3m+V+D822cLxmLRfy5f/+9twvBXYv6ZvfeA7dnL5yAXG5RYSF3pcAW38ZAOGVGFopRiSMhUQ0YYbM5RUXkjdLb1F05fAhtvgZK48xQZSVkAdLS3agAC5ywkn6r2Ob2YGWaCKWnBZOANjMHaAnDpqZAcb2OAxJkNpHB50rnR2i8TxhBcwdBSGhs0gEryDFDEfDJSUXr4LUGEoAJDcZHLKK2KqGCn5AUPX0SKHCQH+YaYnR3RklmvIBsZIhTEwKsCNLIQ20wUBUedQSIpnOMEMAUAlvECWiCSnHP2mdTgE3OVfvFwySjR7phVNOHDvk2I2KrQoQQ8wERGyEfQ8MyDYKmicfjcMbPNa7dxT6+bZiQvL54/vwrsyRgkjzWB3N9bZqjpgTIa0asNdCtq7ThIBFCSQFWE2vUR0I+/REhHHSG5IlbT2sZQUhLCSKBp/N8EPYSL4glAf3UgDG4Lym6HELNdQjrCBbW7cOTR5Ov/3YYwwRtnA7m6sr18/N/cUAMQWiMeOHcPx2tqad8+1LLc8zBt9nFvJhOktEdFeNISaHJcswbOyQXk3noHpY0fCCwndaRz9rskOa/IJPO4yAY8gybMKSHNsYSnL+mMCj7fdEnYOEY6sD7a5cQcOnblUPzUBAJvXarVrD0VjAIC1tTV6rHePwyLfBjLjHS4Z/YXIsVwlE/I3sIlT9eVT+GTy4HrzLsBEpzNGAdQySyZM7+hrndggULKizAaxwD10CBiK0B7gpomljaTTyDut9nfyy5Z5kNmMctfLOAtCZVTd69GayHaP5pEOcjDiKDI5DwpZ7tE8wkvEGxfnnp7UK0TY3LhzonI+9/fcWyjJLZpTRncLfK2TTo4J1+mivOQKoICSWFrA0Ck2IR4UTQdjvgsjfriRuFs2bKgPdjnNJvoIDQqbbs4EObpLxOPnL23UTIzj0JlLdQ7XM/qPUV4iOl4YI4aSDIXjg9kebFGszwkYRlsctVRIHS0z9gVQglRMCjdgKMwdGNCVCohLhoQjIZ7aAiEMWRnukjGCQq0GkUeB3WuOzHINR3iJ2CO++3a7WPiRj/8ajvd94wEc//qvvAfHf3fpD+z2hlXQv/H5+zh+6W/a17/6CXvM/YOv4PjBr9pFxtcfsef+7y+/A8dbD73a3Ydg7DY4yMFIhWEwGwOk+gygtOZmtzBgGN/oBCRtZaQjxaC1/R7JYBQRtbmRYsMyEvGvghgpkapK3OVHGkVsbxYU1sC4LyJjGCC2+vDoBd3eEYUZbPAI5Jbng4Hngzk1JonulhUcRqcH2L/NRAutHD7GYEhK0iUlj8RcHYZDVnF2SiArN4qIpIeOWaaLuLs+WLt+dqEJ1S6OZANjDAN2dYnYrp+9erRa6eqeQ2xgg0eUB4upLrxqLsCNNLVlM2Yo+NANajRj0G5qlJecWGJMNIhZLEkoKHAyV1RMmOBZSUgnK5L46lce7DOfsDci+pEf7/ZOK5/5rR2c1a6fvXr6yhU4u3Sri6PZwBjDgBNzv4Tj7m11B2dF5rVahnp3x7OBDR6BDD05POgoIqSSFVhmE64+A5XvTtc004CNOlpRpiswHGIZzLSscekImSrZB0P2Q7KCeMrLHACE3/qSB8uvN/3KvJhdAgCAakM9dlubF3R7r2Y2MMZQILcYB21L365/uNkkN8qbFTcXW6sd70vJBjZ4YBTRc7fSyIqKMzrUbsWlGCi5kCSjRdtpIEEFhItiTIVHdsNLADEmpFuo55YFYld6cpRrq6oWDdv16albj29/B0o2MMZQoKiJZjawwcNlMBtFBFfbgXJ49LiE2w+UpKRsxI+msALiEUl0lpIifiTc1zFaaCKN4G5xjzGfCx0z4psBWFrLcg3F1i5bWLm22tWE2cAYwwARFrRehQ1s8AjkFpFiUBeLyjiIuxXjpXihse1cHVO4W75KKNyymTTkH9wrjWLQVx46jGS9OyAvAsZbA+Fsoa+fCWFB14iFM7B3TX8Zx1/8H4dwPPHJr+L4pXc+iOOP/y97D3j5qv2avvm3Ahxv7bcq+In32LLZ143Z7fIfWcb/1qptafDSF96I49LDuVyuINjSPzjXwDxbkib/6wiXaCfQ5Ip9XPX5cXbaeTdtNUjMICluAQAkvg+xjXgkxJaO4K4nM15DscUMxmDkB64HY6QhkA6DRcQSEDUTTQr7qz63xSddDdLEca+rQZG0fsPIh7f2iyoy4qtETU2xCD4lNLo9E9gHYzDyw65HEbsFG9jgMSbCdA1uUqZYAGUnDL5LNwqfFHzv5G5h1WOauyViG8ENeIAlqJSDdeLBOz6r0heAgxwMRq7gIAcjDftkSB0tQRp90nCfABTmWl9LpFSRJJTxx3kp1gMUkAABINWtsj4Y3Y7Hg0tQkEJcgOO+BBJ5ichg5AfBS0RGGvZH9SaEf/zW004Y0HeonDCgX4fvBPRkPBNF8mCQFEg0GyFKbUGKrwUd3a3YWZiwdnZlAi8RGYwcwXkwRhr2ob6WUEfgkg+GFmO85N8NiHho8boSANcxA/O+4G/0HS0wHciielDKWtiZDH0qcAkQKHEJl8TMrqwXkRmMwcgRHORgpGGfrvlP5yWq5+hCigG03iQmxQA3POhtBJfKdHzPVehCUhgQ3L3ISx3cLTeimA1sYN3gwIEDz//S2/BpecEKfL/zh/ZmNq+efBHHf+8H/hDHn/sh20f17sdse+3v/5U/x/EL/8q24H7+S2/AsSJXIny7FQG//PUxe8w++y3yrfcKBcVLxG6wh+6a1cc7XO7TnUNJiM/pJNOP2GDM9RLGm8LYoBdIhBhluS6WPgscCkqmLO94fEqnkRW70jJgByiWgTEYO8R9ZjBGCvZF9zt3/SiHgmx+zA8MprGTdDt74kaIJaOkGxX0E1bd8JXoQFAeg0XQL+vuykZB99kHYzDyAxsYg4I6b/u+HYLDSBgwdMJ0GEUEz2FzSAxiTdHAeXGnQAtsNVd0ouGrWLgP3CwWIAGCw07xswD8d8S9Pt1lcWjVPTYwBsEeCufsGrLcBP1T936z39PpD9jABo9x04yN8lVShxkkIl97gXzlbXQLh6k/1r3koqtIIABQpoIUsvK2e7uGEnwDPgYjRzCDDR6UwZCaknp0Jqsu7JGue2Y2AnGuknQYSTmrFIfK4Siy1y9J7kxT/l79fDipjBmMwcgRzGCDxz4RRRFTGAwAYrXJQEhMOukph6w8t6qDuiIxT0Woxo8Zgn3aiZ1o1DG2ZyTABjZ4REtE6CJQIZMWftK3ELQlf0HoWJFIWC5Cxy3RKG4wHUwltithHTjclsZLRAYjRxSOwb7+qLX51xLn+R3/0t5T8I8/MoXj3374vTj+9Mav4/jEr9rtz/9922r7wbFv2e0P7MNxqWXH8g02a/ni95PJjefiiBsGSw2pxwsiaTA9mcGil06iKbLYc2v1RUIInoy7oqyejvEIbSijHMxgDEaOKByDjSDGRehVLjr0FQ38ODstDEmlKYejRJzBwIvFQ4yvkoko+81QCJT5jxmMwWD0BmawwWNcbBFpki/bhVgsMR5nJyznhBmhM1+JZBKjLw7mgAxwXjv+WkiwBZXrZgMzGIORI5jBBg8dRYy5UnG+irdMgzS+EnH6ghRNU8dM8Y6gEp5peo2//lC6XghmMAYjRzCDDR7aB0uQX3gCJRU/DBKyW+AxVUSPZK+D7AUjKumFYy+qQPndbSy1Da+cgxmMwcgRzGCDBzIYDSRCzLNK4DTf0YqLLToUkqQSl9j2CHMIOUCR/+J79VS8V46eKiWG2BFjBmMwcgQz2OBhGSzGSzFSipeNaEUiPcY9AOGzWfcwk3AYiJAOca4UACjkMUpMygQRlfNCIgp4DiuFMYMxGDlCKFWgvx0dei0dOHCgmHvTdnVGT12lOs+hL6fswlv0espwdP9nBmMwcgQbGIORI9jAGIwcsTsGduPi3Nzc3Nxc7drmrrwfg1EQ7IKBbV6rPQEXlpeXly+9/7lzF2/k/44MRlGwCwZ2d2P9ROU4AMDEe99/6HqTLYwxOsg/0by5cefQ5OloPDF5EJ7b2ITjE3rnsWPH8MC1tbXOkdli7mUwOiB/A7u7sQ6TaTvX1tbo02JmuvLIgzFGBPkb2EOTh7o78NixY569eejway7suTt4zb4cv4OJ7cJb7OBd9jry98EmJg+ub9yNxpsbd+Dg5ETnExiM4cEuBDkemjx0/elrmwCw+XvPmXgHgzES2AU1/cSp+oWNuXNzTwEcOnOpzvbFGCEUS+zLYAwZWCrFYOQINjAGI0ewgTEYOYINjMHIEUXoyXHj4twT1yGKMZ7qKUm2ea127ql18+zEheXz3cYoN6/Vzm18yB7fyxzcc3c+hxyQ+2SyXLfRhBow7v7Xf/boRz/rjrrGZz/a6xnmtEcffdSe2ssc/HN3Ood8kO9ksly3UcXAl4hZtPabG3cOTT7U2/vduDg398SdMxfOEAFXt3NIOncnc8gNOU4my3UbZQzawOhPYmLyINzZ6KEk8+7G+vpT56JazrkuC82On19eXq6for/CrueQcO6O5pAbcpxMlus2yhi0gd3dWN/+oBRsbtyBQ2cuLS8vLy8vX5p8eqf10kWYQ1+wy5PJct1GBoMOcnSttU/AxKn68il8MnlwvXkXYAeOdhHm0Bfs8mSyXLeRwaAZrAha+yLMYS+Cr1sXGLSBZdHa37hIuuhsbtzZsVC/CHPoB3Z7MlwnsT0GvUTMorU/fv7SRu3c3FMAkE2oX4Q59AO7PRmuk9gerKZnMHLEwJeIDMYwgw2MwcgRbGAMRo5gA2MwcgQbGIORI9jAGIwcwQbGYOQINjAGI0fsfQNr16fFfL0+LYQQYn7FbNKINkC7Pi2m6219ysq83RM9wydDCb5EA8SgKz4zo7VYAYBqw3mun9pxa7EClcWWUkqpRhUqlYp51lqskLOHEnyJBochMTD79Teq9Jl9an8krcVKZXGxqn9Mo/Dj4Us0OOz9JSIAQOXIVMqzqSMVuHm7DVB+5HRl6ZkVgPazV+H0I7WT1eatFkD72atN9+whBV+iwWA4DKwblA8fhaVnVqB1q3n0cBmmjlSWnllpP3u1WTn9SHnQkysG+BL1H8NoYM2rzxpfHVq3mnD0cBkAYOZkFW7erj+zVD05Y/5cP3lrNH88fIl2DYNeo2YG9c3N87gHr5SKvA2gjjx5NszgSzQ4DLzgsu8o11ZbMD0lBAAAVBvq8ozZNXOyCks3zZ9j99kogS/R7oELLhmMHDGMPhiDURiwgTEYOYINjMHIEWxgDEaOYANjMHIEGxiDkSPYwBiMHMEGxmDkCDYwBiNH/H+4yalmBVvNxQAAAABJRU5ErkJggg==" /><!-- --></p>
<p>Before proceeding to Stage 2, the BLUEs and variance-covariance
matrices from the 2015-19 analysis and 2020 analysis need to be
combined. (If other software is used for Stage 1, it can be incorporated
in a similar manner.) The model with i.i.d. row and column effects is
selected based on the higher estimate for H2.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb76-2"><a href="#cb76-2" tabindex="-1"></a>stage1.blues <span class="ot">&lt;-</span> <span class="fu">rbind</span>(ans1a<span class="sc">$</span>blues,model1<span class="sc">$</span>blues)</span>
<span id="cb76-3"><a href="#cb76-3" tabindex="-1"></a>stage1.EEV <span class="ot">&lt;-</span> <span class="fu">c</span>(ans1a<span class="sc">$</span>vcov,model1<span class="sc">$</span>vcov)</span>
<span id="cb76-4"><a href="#cb76-4" tabindex="-1"></a></span>
<span id="cb76-5"><a href="#cb76-5" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb76-6"><a href="#cb76-6" tabindex="-1"></a>stage1.blues.open <span class="ot">&lt;-</span> <span class="fu">c</span>(BLUEs,BLUEs2)</span>
<span id="cb76-7"><a href="#cb76-7" tabindex="-1"></a>stage1.EEV.open <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEV_full,models_env[[<span class="st">&quot;Hancock20&quot;</span>]][[<span class="st">&quot;EEV&quot;</span>]])</span>
<span id="cb76-8"><a href="#cb76-8" tabindex="-1"></a></span>
<span id="cb76-9"><a href="#cb76-9" tabindex="-1"></a><span class="fu">cor</span>(stage1.blues<span class="sc">$</span>BLUE, stage1.blues.open)</span></code></pre></div>
<pre><code>## [1] 0.9997544</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a><span class="fu">length</span>(stage1.blues.open)</span></code></pre></div>
<pre><code>## [1] 1631</code></pre>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a><span class="fu">dim</span>(stage1.EEV.open)</span></code></pre></div>
<pre><code>## [1] 1631 1631</code></pre>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a><span class="fu">length</span>(Envs)</span></code></pre></div>
<pre><code>## [1] 1631</code></pre>
</div>
<div id="marker-data" class="section level3">
<h3>Marker data</h3>
<p>The <code>read_geno</code> function reads bi-allelic marker data as a
CSV file. If you intend to run GWAS, the option <code>map=TRUE</code>
indicates the first three columns of the input file are the marker name,
chromosome, and position, followed by columns for the individuals. Map
information is not used for genomic prediction and can be omitted by
using <code>map=FALSE</code>, in which case the first column is the
marker name and subsequent columns are individuals. The marker data
should represent allele dosage, with numeric values between 0 and
ploidy. (For compatibility with other software, the coding {-1,0,1} is
also allowed for diploids.)</p>
<p>The potato marker data were generated using an Infinium SNP array.
Most clones were genotyped with Version 3 (V3) of the array, but some
were genotyped with an earlier version (V2). Data from the two different
versions were combined via BLUP using the function
<code>merge_impute</code> from R package <a href="https://github.com/jendelman/polyBreedR">polyBreedR</a>.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a><span class="co">#I don&#39;t provide equivalent codes to these functions because they are already</span></span>
<span id="cb84-2"><a href="#cb84-2" tabindex="-1"></a><span class="co">#open source and don&#39;t require ASReml license</span></span>
<span id="cb84-3"><a href="#cb84-3" tabindex="-1"></a>geno.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;geno1.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb84-4"><a href="#cb84-4" tabindex="-1"></a>geno <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(geno.file,<span class="at">check.names=</span>F)</span>
<span id="cb84-5"><a href="#cb84-5" tabindex="-1"></a>geno[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>##                marker chrom position AF5392-8 AF5393-1 AF5429-3
## 1 solcap_snp_c2_51460 chr01   449027        0        1     1.03
## 2 solcap_snp_c2_36608 chr01   508800        1        0     2.00
## 3 solcap_snp_c2_36615 chr01   510745        1        0     2.03
## 4 solcap_snp_c2_36658 chr01   527068        4        3     3.00</code></pre>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a>geno <span class="ot">&lt;-</span> <span class="fu">read_geno</span>(<span class="at">filename=</span>geno.file, <span class="at">ploidy=</span><span class="dv">4</span>, <span class="at">map=</span><span class="cn">TRUE</span>, <span class="at">min.minor.allele=</span><span class="dv">5</span>, </span>
<span id="cb86-2"><a href="#cb86-2" tabindex="-1"></a>                  <span class="at">dominance=</span>T)</span></code></pre></div>
<pre><code>## Minor allele threshold = 5 genotypes
## Number of markers = 12242
## Number of genotypes = 943</code></pre>
<p>The function <code>read_geno</code> computes genomic relationship
matrices from the markers. At a minimum, the additive (G) matrix is
computed. When <code>dominance=TRUE</code>, the dominance (D) matrix is
also computed. These matrices and other information needed for the
<code>Stage2</code> function are stored in the returned object as an S4
class.</p>
<p>The command <code>inbreeding</code> returns genomic inbreeding
coefficients, estimated from either the diagonal elements of the G
matrix or the average dominance coefficient. As shown below, the two
methods are very similar and have the same population mean. The negative
inbreeding coefficient indicates excess heterozygosity relative to
panmictic equilibrium, which may be expected when there is inbreeding
depression.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">inbreeding</span>(geno)</span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a><span class="fu">head</span>(x)</span></code></pre></div>
<pre><code>##                  F.G         F.D
## AF5392-8 -0.07650919 -0.07976518
## AF5393-1 -0.10542949 -0.10912896
## AF5429-3 -0.06456956 -0.06159264
## AF5445-2 -0.08590067 -0.08484250
## AF5450-7 -0.09444887 -0.09295317
## AF5484-3 -0.11489437 -0.11669210</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="fu">apply</span>(x,<span class="dv">2</span>,mean)</span></code></pre></div>
<pre><code>##         F.G         F.D 
## -0.07786684 -0.07786684</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a><span class="fu">ggplot</span>(x,<span class="fu">aes</span>(<span class="at">x=</span>F.G,<span class="at">y=</span>F.D)) <span class="sc">+</span> <span class="fu">geom_hex</span>()</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAACPVBMVEUAAAAAADoAAGYAOpAAZrYTK0MTLEQULUYULkYULkcVL0gWMEoWMUsXMk0XM00XM04YNE8YNVEZNlIZN1QaOFQaOFUaOVYbO1gbO1kcPVscPVwdPlwdP10eQF8eQWAfQmIfQ2MfQ2QgRGUgRWYhRmghR2kiSGoiSWsiSmwjS24kTG8kTXEkTnIlTnIlT3QmUHUmUXYnUngnU3koVHooVXspVn0pV38qWIAqWYIrWoMsXIUsXYYtXoctX4kuYIouYYwvY44wZI8wZJAwZZExZpIxZ5QyaZYyaZczMzMzapgza5k0bJo0bZw1bp41b582cKA2caE3cqM3c6Q4dKY4dac5d6k6AAA6ADo6AGY6OmY6OpA6eKs6eaw6kNs7eq47e688fbE9frM9f7Q+gLY+gbc/grg/g7lAhLtAhbxBhr5Bh79CicFDisNDi8REjcZFjshFj8lGkMtGkc1Hk89IlNBIldJJl9RJmNVKmddLm9lLm9pMndxNTU1NTW5NTY5NbqtNjshNnt1Nn99OoeFOouNPo+VQpOZQpuhRp+pSqOtSqu1Tq+5Tq+9UrPBUrvJVr/RWsPZWsfdmAABmADpmAGZmOgBmtrZmtv9uTU1uTW5uTY5ubqtuq+SOTU2OTW6OTY6OyP+QOgCQOjqQtpCQ29uQ2/+rbk2rbm6rbo6ryKur5P+2ZgC22/+2///Ijk3I///bkDrb///kq27k///r6+v/tmb/yI7/25D/27b/5Kv//7b//8j//9v//+T////whHWEAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMBElEQVR4nO2diV8jSRWAwzjRaDSKomi8UBRF0WgUDzxQNOp44TnjhQeK93jrerDu4jmKF6zKru7q4MqqMzKT3YVBGKb/Nut1d/X5kteV6qM6ed/vBzmqOyQfle46Xr2uWExfKkW/AdNhQQQsiIAFEbAgAhZEoCXowCNwN0x6Bcl2SMuLBwsiYEEELIiABRGwIAJC0M371u94OHjvofX19R9fZ0Eutx64bD10Z/DetcuBYoMEVauFCLr59+vW8V+u+/du/euKkYKqVddQzoKO//qwdfNvV/x74ou2vm5XoncIUn83AwOCsnnl/oJu3CEFufeO77li+bWIa1CsBtnPeschFhQ7BrGgMLceuNc7i9n34Jt269/mneaLEuS2fqDq+O2g270TmTmCijrNExgkyHuQNiyIgAURsCACFkRQPkFwvqo6v1gQ8nw1CAtiQSyIBbGg3hQgqFZzz2KeHxYUfL4mgNuAH5MF5Q8IgttqdmOsEbgGEbAgAhZEUDJBB46fA78nxoLUdkjLiwcLImBBBCyIgAURlESQOGdZ9onLO4uxoLAfF9kOYkEsiAWxIBZkIVGu/hM5CoJhRM+P3wkzQFAsytV/IkdBjhbPT9RQkYJiEWbBQDMWhMQo+k/kGOUaE5TT37VRjHL1nwC4BvWpQSwIMOIYBGIaDddQcCjRAEGxKFf/idwEgZhGwzVE7pCzoHiUa/7tILMF9Wd4BP3vv71KWBBwdGGvVxELAljQ6W6lcta/tYWIX0cX7j9fqewciV8rJRRk+wAzExONhjvkOpig092z1sn2jn0LP56g84940Oqe2S9nDZJVpzEh0Bu0lwIOz+zbv3xBO25NGnFBthi4FRXGExK9ZUEsqKcg7ys2thf+irEgB3lwlj8n2yvidswTBAfwUgmq1x0/k5OTjcbUFPjREhQ9zVtwYv/zXZ4g66r9bGkE1QWun8nJKUHyV9L5OCgsiMDIKNeYoALfC9cgAhZEYKKgRqNen5mpVlst4adadfywIInd9JkRtASBLjwLYkHFCdrCYUGSP+CwIMnvcUolSHpoNpuNxtLSzMzUVKtln+6VXgl/o7/DKZEgr6Y0gSUBNH+gPVRXeyX8jf4WhwVJfoPDgiS/xmFBkl/hsCDJJZwSCQJF0OfqdISf5eWlJeikVqsBP1qCfomTgqAcc7lClekIloFJu5saminUEfQLHH1BeeZyzVLQz3H0BeWZyzVLQT/D0ReUZy5XVFBKr/1TnCS7mpPLNcsa9BMcf4PT3R3LOtmuwMxrUkHX1tfvzCeXq+1hfn5q6ty5Tkf0UJeXx8cnJ6PhmjqCNnD8DbqVHVtSNzY9ZkAuV6gp88A5AXThW+OCQV6pp6Af4XjlRx/904518sd96+gj+0qCcsnlmoMg4IchQkWn//inqD1HH3vQOrkrOklvQC7XHAT9AEcWd1fg6wWhH8qC+lMaQd/HcUtF1TkdtAYNiaDbcNzSbgVYGeQYlIegiYlqVbzZ+flLl86dm58Xfmo1xI+WoO/h+BtADTrdXVE+i+UgCIJ/7H8ndK+hHtWwZZeagr6LExak2g4aJkHfwUnyGUdD0LdxWJDkWzgsSPJNnFIIagtBW8JPpyME1YUgdG28pqBv4JRBULvdntja2roN+vGX6vX6vLuuMF1BX8cpgSDhpw1hBOCnI+OmEizJUBT0NRwWJLmIw4IkX8XJXJA+uKDU/8xXcJLsOho16Ms4Rguy+6MbG+325ubW1uLiYqdTq7krEBRfKYGgL+GoCjq0u/09l3WkKghGNOxx4U3BIiCbP7320BH0RRw1Qae79qKqk+14j7b0gr6Aoyao64qBYZFhE/R5HCVB9ohIyNQQCfocjpIgf02ZXL84RII+iyOLxcEXPnP/AbP8BMFpanZ2fFyc3Tc2Op3NzVpN+PG6YFkIWsVxS2EgunuWmjjMTRA0dGYFW7L946UH6v9SOoI+gxPYQkgiBu2HWhDw6RDRUlF1iGkf8QWUDJ+gT+F45Ufnx/aMmTgsQtAncQJbCDOmTBwWIegTOMFNru4knjjM9hgEyUpWV2dn19aEn3bb8XMQiHXJQtDHceQHdr5bRkwc2itV4Ay7JoCufDs2Ap2FoA/jyOJupQIpGQaZOAymLLPDgzRzmBUk6EM4xP8/gaBg6tYbEP6rm8u1IEEfxFEShKavCMSVXbv9PxAlpJlHsSBBH8BRFxQ9PocSb4IY3VyuuCD111Hk/ThJdu0vKJS6FQRp5nJttVqNxsWLq6ut1tqa6K622/HrHGRRg96HoycoHOVqRWvQIILs6CiYb5FRUkgYUCaC3oujJwgIHXKOtY9BhQl6D46+oFDqVhCjl8u1MEHvxlEUhHVW/SjXNNpBhQl6F46SoEEojaB34pgkCLpa09PCy8KC46daFX6wObAsBL0dxyBB0FmfFiwAXsYAdBY1C0Fvw2FBkg4OC5K8FYcFSd6Cw4Ikb8YxSJAwNDc3Pd1sNhcWms1Wyx1BzOss9iYcWQxpJwuPtJ8TNCWJP29KghZx3FIYqT+6sDfIisNhEfRGHLf0EKwoDNoPoSDgDSGipcVP+xQq6PU4/gYwo1HwxGGhgl6H45VD+uQsVhwmQ/S5LGt6em5ufNzOv9VsZv83IyzgyGI7tf1Aq54JEv177V47tIHg1h6TVqgQKdWg1+KE/BS24tAEQa/BcUu7bvBqQe0gEwS9GifJZxwNQa/CYUGSeRwWJHkljhGC7KFVGG51V8O7fvIV9AocEwR5IxrROfhcBb0chwVJXobDgiQtHBYkeSkOC5K8BKdgQc5VY10/sNJggJxtKQl6MU6xgkLXHY4P0ecq6EU4LEgyi5OCoFiUa/JcriYJeiGOvqBYlKtCLleTBL0AR19QLMpVIZerSYKej6MvKBajqJLL1fHjPoA7GWQMSMjzcPwN7KHWAQbMYlGuSrlcbT+qFSWTGjSN45UfQlSd6sQhGuXqlCRMVWqOoOfiyOKrY/eLGjTIoH0syrWkgoDnhIgUgphBpn1iUa5KuVzNEfRsnLCggSYOY1GuKrlczRH0LJywoNynnu3OlxmCnokTFpT3xKHTfTdD0DNwwoLynjg0SdDTccKC8p44NEnQ03CSfMbREPRUHBYkmcQpVFC9bg+xWnoZt1IS9BScAgV513VCI31zF/RkHBYkeRIOC5I8EYcFScZxWJDkCTjFCarCxeVsM2acxR6Pk7mgXngrVbJ48UF4HE6SXTOpQYGlPGY0FB+Lw4IkdRwWJHkMDguSPBqnMEFOzmPnLGaEoEfhFCcoiwIdQY/EYUEpwIIIWBABCyLIQNAAV0cdKUFVvVjNoRdUZUE+LIiABQWiXI/vtmPvyBxmoyXIj3KFOLPje67QuVxHS5AfYXYDrFy73C+PomtmkIugl1ZQOPGmuNcnl6tRY6zpoRDlCvF4fXK5RuKgh78GRaNcb953b7RKjbYgW4p/yDm++7Jl9cvlOpKC/ChXx0+/XK4jKciPcoVVPtAQQttBTjCr60dzYWHJBPXH8xOoOpGllyyIBfWBBRGwIAIWRCDfVcSQloehFNQ7TIoFsaB+sCACFkTAgghoPyyI8DPiglyGc6zVhWsQAQsiYEEEKQiK5nVhQR4BP9wOwmBBBCyIgAURsCCCA9cQfpEeFmT1eJMsSMKCCFgQAQsiYEEELIiABSWGyFqaxh7qfyIVWBABCyJgQQQ5XKGu3LAgAhZEwIIIWBCBtqDYkjuFPQJJvJNsHgpizwtdQbEldwp7yEtRJNw8vJgvL3QFxZbcKezhXooi6ebhhTR5oSsotuRObQ/6A/ubh/9UXugKii25U9uDFuRvHl7Mlxc6gtAld0p7jEANii25U9kjiaCyH4NiS+4U9rCSCPI3Dy/my4uU2kGBJXfJ97AU2kHuhYXK1w4aelgQAQsiYEEELIiABREUJ+h0t2Kz4z7uBh+YQ5GCVkKPzuzDpfRWem5fEKYIOjxjX7eqG7/6UMEYIihsyyQMEXSybd7Rx6Hwg7Rj6ejCnvh1vlIZi15Cr2gMq0G2KKMwRJB8wIJ8wsflrnMWY0E+kRPXVacdxKd5D0+Qe+eQW9KlhAURsCACFkTAgghYEAELImBBBCyIgAUR/B9E12g62wqrrwAAAABJRU5ErkJggg==" /><!-- --></p>
</div>
<div id="stage-2" class="section level3">
<h3>Stage 2</h3>
<p>The <code>Stage2</code> function uses the BLUEs from Stage 1 as the
response variable, as well as their variance-covariance matrix to
partition micro-environmental variation from GxE. It is also possible to
run <code>Stage2</code> without the covariance of the BLUEs from Stage 1
(EEV), e.g., when there are no replicated entries within environment. In
this case, use <code>vcov=NULL</code>. The benefit of including the EEV
information is reflected in the lower value of AIC, which is a penalized
likelihood to measure goodness-of-fit.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb93-2"><a href="#cb93-2" tabindex="-1"></a><span class="co">#1) no EEV</span></span>
<span id="cb93-3"><a href="#cb93-3" tabindex="-1"></a>ans2a <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues, <span class="at">vcov=</span><span class="cn">NULL</span>)</span>
<span id="cb93-4"><a href="#cb93-4" tabindex="-1"></a><span class="co">#2) EEV</span></span>
<span id="cb93-5"><a href="#cb93-5" tabindex="-1"></a>ans2b <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues, <span class="at">vcov=</span>stage1.EEV)</span>
<span id="cb93-6"><a href="#cb93-6" tabindex="-1"></a></span>
<span id="cb93-7"><a href="#cb93-7" tabindex="-1"></a><span class="co">#data.frame(EEV=c(FALSE,TRUE), AIC=c(ans2a$aic,ans2b$aic))</span></span>
<span id="cb93-8"><a href="#cb93-8" tabindex="-1"></a></span>
<span id="cb93-9"><a href="#cb93-9" tabindex="-1"></a></span>
<span id="cb93-10"><a href="#cb93-10" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb93-11"><a href="#cb93-11" tabindex="-1"></a><span class="co">#1) Prepare data</span></span>
<span id="cb93-12"><a href="#cb93-12" tabindex="-1"></a>Stage1Output <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">BLUEs =</span> stage1.blues.open,</span>
<span id="cb93-13"><a href="#cb93-13" tabindex="-1"></a>                       <span class="at">id =</span> <span class="fu">names</span>(stage1.blues.open),</span>
<span id="cb93-14"><a href="#cb93-14" tabindex="-1"></a>                       <span class="at">Env =</span> Envs)</span>
<span id="cb93-15"><a href="#cb93-15" tabindex="-1"></a></span>
<span id="cb93-16"><a href="#cb93-16" tabindex="-1"></a></span>
<span id="cb93-17"><a href="#cb93-17" tabindex="-1"></a><span class="co">#We need to add to Stage1Output a column for stage 1 estimation error </span></span>
<span id="cb93-18"><a href="#cb93-18" tabindex="-1"></a>Stage1Error <span class="ot">&lt;-</span> <span class="fu">paste0</span>(Stage1Output<span class="sc">$</span>id,<span class="st">&quot;:&quot;</span>,Stage1Output<span class="sc">$</span>Env)</span>
<span id="cb93-19"><a href="#cb93-19" tabindex="-1"></a><span class="fu">rownames</span>(stage1.EEV.open) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb93-20"><a href="#cb93-20" tabindex="-1"></a><span class="fu">colnames</span>(stage1.EEV.open) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb93-21"><a href="#cb93-21" tabindex="-1"></a></span>
<span id="cb93-22"><a href="#cb93-22" tabindex="-1"></a><span class="co">#model:</span></span>
<span id="cb93-23"><a href="#cb93-23" tabindex="-1"></a><span class="co">#BLUEs = XE + Zg + s + gE</span></span>
<span id="cb93-24"><a href="#cb93-24" tabindex="-1"></a><span class="co">#We need a column for BLUEs, environment name, genotypic id and stage1error</span></span>
<span id="cb93-25"><a href="#cb93-25" tabindex="-1"></a>sommerdf <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Stage1Output,</span>
<span id="cb93-26"><a href="#cb93-26" tabindex="-1"></a>                  Stage1Error)</span>
<span id="cb93-27"><a href="#cb93-27" tabindex="-1"></a><span class="fu">head</span>(sommerdf)</span></code></pre></div>
<pre><code>##       BLUEs          id       Env           Stage1Error
## 1  66.81192   A00188-3C Hancock15   A00188-3C:Hancock15
## 2 104.41192   A01143-3C Hancock15   A01143-3C:Hancock15
## 3  76.55059   A09037-6C Hancock15   A09037-6C:Hancock15
## 4  83.85059  AAF07847-2 Hancock15  AAF07847-2:Hancock15
## 5  70.81192  AC01144-1W Hancock15  AC01144-1W:Hancock15
## 6  83.36192 Accumulator Hancock15 Accumulator:Hancock15</code></pre>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" tabindex="-1"></a><span class="co">#When you want use a known variance-covariance matrix in Sommer mmec() function,</span></span>
<span id="cb95-2"><a href="#cb95-2" tabindex="-1"></a><span class="co">#what you actually have to provide is the inverse of the </span></span>
<span id="cb95-3"><a href="#cb95-3" tabindex="-1"></a><span class="co">#matrix and it must be of the &quot;dgCMatrix&quot; class (sparse matrix)</span></span>
<span id="cb95-4"><a href="#cb95-4" tabindex="-1"></a>EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(stage1.EEV.open),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb95-5"><a href="#cb95-5" tabindex="-1"></a></span>
<span id="cb95-6"><a href="#cb95-6" tabindex="-1"></a><span class="co">#Fix the variance for stage1Error to be 1. We have to scale it dividing it</span></span>
<span id="cb95-7"><a href="#cb95-7" tabindex="-1"></a><span class="co">#by the variance of the response variable because Sommer works that way.</span></span>
<span id="cb95-8"><a href="#cb95-8" tabindex="-1"></a><span class="co">#We fix the variance to be 1 because we don&#39;t want the model to scale EEV in </span></span>
<span id="cb95-9"><a href="#cb95-9" tabindex="-1"></a><span class="co">#any way, but 1 is not the true stage1Error variance component. More on that later</span></span>
<span id="cb95-10"><a href="#cb95-10" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf<span class="sc">$</span>BLUEs))</span></code></pre></div>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="co">#without EEV</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a>ans2a_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Fixed effects = Environment. -1 means that</span></span>
<span id="cb96-3"><a href="#cb96-3" tabindex="-1"></a>                     <span class="co">#we remove the intercept (you could leave it if you prefer it)</span></span>
<span id="cb96-4"><a href="#cb96-4" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id)),  <span class="co">#vsc() is just an auxiliary function</span></span>
<span id="cb96-5"><a href="#cb96-5" tabindex="-1"></a>                       <span class="co">#used every time you want to specify details for a random</span></span>
<span id="cb96-6"><a href="#cb96-6" tabindex="-1"></a>                       <span class="co">#effect.</span></span>
<span id="cb96-7"><a href="#cb96-7" tabindex="-1"></a>                       <span class="co">#isc(id) means that we assume identity variance-covariance</span></span>
<span id="cb96-8"><a href="#cb96-8" tabindex="-1"></a>                       <span class="co">#structure for the &quot;id&quot; effect (genotypic effect)</span></span>
<span id="cb96-9"><a href="#cb96-9" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb96-10"><a href="#cb96-10" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb96-11"><a href="#cb96-11" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb96-12"><a href="#cb96-12" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf,</span>
<span id="cb96-13"><a href="#cb96-13" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb96-14"><a href="#cb96-14" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb97-2"><a href="#cb97-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf)</span>
<span id="cb97-3"><a href="#cb97-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2a_open<span class="sc">$</span>b</span>
<span id="cb97-4"><a href="#cb97-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb97-5"><a href="#cb97-5" tabindex="-1"></a></span>
<span id="cb97-6"><a href="#cb97-6" tabindex="-1"></a></span>
<span id="cb97-7"><a href="#cb97-7" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb97-8"><a href="#cb97-8" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb97-9"><a href="#cb97-9" tabindex="-1"></a>  ans2a_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2a_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:isc:isc&quot;</span>], <span class="co">#genetic</span></span>
<span id="cb97-10"><a href="#cb97-10" tabindex="-1"></a>  ans2a_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2a_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>]) <span class="co">#residual </span></span>
<span id="cb97-11"><a href="#cb97-11" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb97-12"><a href="#cb97-12" tabindex="-1"></a>variances_ans2a <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb97-13"><a href="#cb97-13" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb97-14"><a href="#cb97-14" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2a) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2a<span class="sc">$</span>vars))</span>
<span id="cb97-15"><a href="#cb97-15" tabindex="-1"></a><span class="co">#almost same variances with Sommer and ASReml</span></span>
<span id="cb97-16"><a href="#cb97-16" tabindex="-1"></a>variances_ans2a</span></code></pre></div>
<pre><code>##            Sommer ASReml
## env      61.42390   60.5
## genotype 82.03648   82.0
## residual 65.14920   65.1</code></pre>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" tabindex="-1"></a><span class="co">#almost same aic</span></span>
<span id="cb99-2"><a href="#cb99-2" tabindex="-1"></a>ans2a<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 9634.154</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" tabindex="-1"></a>ans2a_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 9642.584</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" tabindex="-1"></a><span class="co">#with EEV</span></span>
<span id="cb103-2"><a href="#cb103-2" tabindex="-1"></a>ans2b_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Fixed effects = Environment. -1 means that</span></span>
<span id="cb103-3"><a href="#cb103-3" tabindex="-1"></a>                     <span class="co">#we remove the intercept (you could leave it if you prefer it)</span></span>
<span id="cb103-4"><a href="#cb103-4" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id)) <span class="sc">+</span> <span class="co">#vsc() is just an auxiliary function</span></span>
<span id="cb103-5"><a href="#cb103-5" tabindex="-1"></a>                       <span class="co">#used every time you want to specify details for a random</span></span>
<span id="cb103-6"><a href="#cb103-6" tabindex="-1"></a>                       <span class="co">#effect.</span></span>
<span id="cb103-7"><a href="#cb103-7" tabindex="-1"></a>                       <span class="co">#isc(id) means that we assume identity variance-covariance</span></span>
<span id="cb103-8"><a href="#cb103-8" tabindex="-1"></a>                       <span class="co">#structure for the &quot;id&quot; effect (genotypic effect)</span></span>
<span id="cb103-9"><a href="#cb103-9" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb103-10"><a href="#cb103-10" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb103-11"><a href="#cb103-11" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb103-12"><a href="#cb103-12" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb103-13"><a href="#cb103-13" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb103-14"><a href="#cb103-14" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb103-15"><a href="#cb103-15" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb103-16"><a href="#cb103-16" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb103-17"><a href="#cb103-17" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb103-18"><a href="#cb103-18" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb103-19"><a href="#cb103-19" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb103-20"><a href="#cb103-20" tabindex="-1"></a>                           </span>
<span id="cb103-21"><a href="#cb103-21" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb103-22"><a href="#cb103-22" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb103-23"><a href="#cb103-23" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb103-24"><a href="#cb103-24" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb103-25"><a href="#cb103-25" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb103-26"><a href="#cb103-26" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb103-27"><a href="#cb103-27" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb103-28"><a href="#cb103-28" tabindex="-1"></a>                           </span>
<span id="cb103-29"><a href="#cb103-29" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb103-30"><a href="#cb103-30" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb103-31"><a href="#cb103-31" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb103-32"><a href="#cb103-32" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb103-33"><a href="#cb103-33" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb103-34"><a href="#cb103-34" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb103-35"><a href="#cb103-35" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb103-36"><a href="#cb103-36" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb103-37"><a href="#cb103-37" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb103-38"><a href="#cb103-38" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb103-39"><a href="#cb103-39" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf,</span>
<span id="cb103-40"><a href="#cb103-40" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb103-41"><a href="#cb103-41" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb104-2"><a href="#cb104-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf)</span>
<span id="cb104-3"><a href="#cb104-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2b_open<span class="sc">$</span>b</span>
<span id="cb104-4"><a href="#cb104-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb104-5"><a href="#cb104-5" tabindex="-1"></a></span>
<span id="cb104-6"><a href="#cb104-6" tabindex="-1"></a><span class="co">#extract stage1error variance component from EEV</span></span>
<span id="cb104-7"><a href="#cb104-7" tabindex="-1"></a>Stage1Errorvar <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(stage1.EEV.open)) <span class="sc">-</span> <span class="fu">mean</span>(stage1.EEV.open)</span>
<span id="cb104-8"><a href="#cb104-8" tabindex="-1"></a></span>
<span id="cb104-9"><a href="#cb104-9" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb104-10"><a href="#cb104-10" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb104-11"><a href="#cb104-11" tabindex="-1"></a>  ans2b_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2b_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:isc:isc&quot;</span>], <span class="co">#genetic</span></span>
<span id="cb104-12"><a href="#cb104-12" tabindex="-1"></a>  ans2b_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2b_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>], <span class="co">#residual</span></span>
<span id="cb104-13"><a href="#cb104-13" tabindex="-1"></a>  Stage1Errorvar) <span class="co">#stage1 estimation error</span></span>
<span id="cb104-14"><a href="#cb104-14" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2b<span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb104-15"><a href="#cb104-15" tabindex="-1"></a>variances_ans2b <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb104-16"><a href="#cb104-16" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb104-17"><a href="#cb104-17" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2b) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2b<span class="sc">$</span>vars))</span>
<span id="cb104-18"><a href="#cb104-18" tabindex="-1"></a><span class="co">#almost same variances with Sommer and ASReml</span></span>
<span id="cb104-19"><a href="#cb104-19" tabindex="-1"></a>variances_ans2b</span></code></pre></div>
<pre><code>##                Sommer ASReml
## env          61.31918   60.0
## genotype     71.21687   71.3
## g x env      32.47563   32.6
## Stage1.error 39.00490   38.8</code></pre>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" tabindex="-1"></a><span class="co">#almost same aic</span></span>
<span id="cb106-2"><a href="#cb106-2" tabindex="-1"></a>ans2b<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 9535.647</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" tabindex="-1"></a>ans2b_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 9544.137</code></pre>
<p>Several other pieces of information are in the list output from
<code>Stage2</code>. The variance components are contained in ‚Äúvars‚Äù as
an S4 class, which is used by the <code>blup_prep</code> function (see
below). As shown above, the <code>summary</code> command returns a
matrix with two columns: the first is the variance, in units of the
trait; the second is the proportion of variance excluding the
environment effect, which makes the result for genotype comparable to
heritability (environment basis). Including the Stage1 var-cov matrix
for the BLUEs enables partitioning of the residual into GxE and Stage1
error.</p>
<p>The above Stage 2 analysis did not include the marker data. To
partition the genotype effects into additive and non-additive effects,
the output from <code>read_geno</code> is included in the function call.
StageWise has two ways of modeling non-additive effects. The argument
non.add=‚Äúg.resid‚Äù leads to a genetic residual, with independent and
identically distributed (iid) effects. When non.add=‚Äúdom‚Äù (which
requires that <code>read_geno</code> was run with dominance=TRUE), the
covariance of the non-additive effects follows the D matrix. To omit
non-additive effects, use non.add=‚Äúnone‚Äù. The AIC can be used to assess
which model is better.</p>
<p>The following example also shows the <code>silent=FALSE</code> option
(default is TRUE), which shows the convergence progress from
ASReml-R.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" tabindex="-1"></a>ans2c <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues,<span class="at">vcov=</span>stage1.EEV,<span class="at">geno=</span>geno,</span>
<span id="cb110-2"><a href="#cb110-2" tabindex="-1"></a>                <span class="at">silent=</span><span class="cn">FALSE</span>, <span class="at">non.add=</span><span class="st">&quot;g.resid&quot;</span>)</span>
<span id="cb110-3"><a href="#cb110-3" tabindex="-1"></a>ans2d <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues,<span class="at">vcov=</span>stage1.EEV,<span class="at">geno=</span>geno,</span>
<span id="cb110-4"><a href="#cb110-4" tabindex="-1"></a>                <span class="at">silent=</span><span class="cn">FALSE</span>, <span class="at">non.add=</span><span class="st">&quot;dom&quot;</span>)</span>
<span id="cb110-5"><a href="#cb110-5" tabindex="-1"></a></span>
<span id="cb110-6"><a href="#cb110-6" tabindex="-1"></a><span class="co">#data.frame(non.add=c(&quot;g.resid&quot;,&quot;dom&quot;),AIC=c(ans2c$aic,ans2d$aic))</span></span>
<span id="cb110-7"><a href="#cb110-7" tabindex="-1"></a></span>
<span id="cb110-8"><a href="#cb110-8" tabindex="-1"></a></span>
<span id="cb110-9"><a href="#cb110-9" tabindex="-1"></a><span class="co">#not all individuals are genotyped. </span></span>
<span id="cb110-10"><a href="#cb110-10" tabindex="-1"></a><span class="co">#Sommer can&#39;t handle missing levels of the effect (id) in the variance-covariance</span></span>
<span id="cb110-11"><a href="#cb110-11" tabindex="-1"></a><span class="co">#matrix (G).</span></span>
<span id="cb110-12"><a href="#cb110-12" tabindex="-1"></a><span class="co">#We have to remove all entries whose id is not in G</span></span>
<span id="cb110-13"><a href="#cb110-13" tabindex="-1"></a>G <span class="ot">&lt;-</span> geno<span class="sc">@</span>G</span>
<span id="cb110-14"><a href="#cb110-14" tabindex="-1"></a>D <span class="ot">&lt;-</span> geno<span class="sc">@</span>D</span>
<span id="cb110-15"><a href="#cb110-15" tabindex="-1"></a><span class="fu">dim</span>(G)</span></code></pre></div>
<pre><code>## [1] 943 943</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id))</span></code></pre></div>
<pre><code>## [1] 1294</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" tabindex="-1"></a><span class="co">#this is needed later to scale variance components</span></span>
<span id="cb114-2"><a href="#cb114-2" tabindex="-1"></a>meanG <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno<span class="sc">@</span>G))<span class="sc">-</span><span class="fu">mean</span>(geno<span class="sc">@</span>G)</span>
<span id="cb114-3"><a href="#cb114-3" tabindex="-1"></a>meanD <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno<span class="sc">@</span>D))<span class="sc">-</span><span class="fu">mean</span>(geno<span class="sc">@</span>D)</span>
<span id="cb114-4"><a href="#cb114-4" tabindex="-1"></a></span>
<span id="cb114-5"><a href="#cb114-5" tabindex="-1"></a><span class="co">#create a relationship matrix that includes non-genotyped individuals</span></span>
<span id="cb114-6"><a href="#cb114-6" tabindex="-1"></a><span class="co">#We assume that nongenotyped individuals are i.i.d. (variance 1, covariance</span></span>
<span id="cb114-7"><a href="#cb114-7" tabindex="-1"></a><span class="co">#0 with all other genotypes.)</span></span>
<span id="cb114-8"><a href="#cb114-8" tabindex="-1"></a>genotyped_id <span class="ot">&lt;-</span> <span class="fu">unique</span>(sommerdf<span class="sc">$</span>id)[<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id) <span class="sc">%in%</span> <span class="fu">rownames</span>(G)]</span>
<span id="cb114-9"><a href="#cb114-9" tabindex="-1"></a>non_genotyped_id <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id), genotyped_id)</span>
<span id="cb114-10"><a href="#cb114-10" tabindex="-1"></a></span>
<span id="cb114-11"><a href="#cb114-11" tabindex="-1"></a><span class="co"># NonGenoEEV &lt;- diag(length(non_genotyped_id))</span></span>
<span id="cb114-12"><a href="#cb114-12" tabindex="-1"></a><span class="co"># rownames(NonGenoEEV) &lt;- non_genotyped_id</span></span>
<span id="cb114-13"><a href="#cb114-13" tabindex="-1"></a><span class="co"># colnames(NonGenoEEV) &lt;- non_genotyped_id</span></span>
<span id="cb114-14"><a href="#cb114-14" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-15"><a href="#cb114-15" tabindex="-1"></a><span class="co"># #additive matrix</span></span>
<span id="cb114-16"><a href="#cb114-16" tabindex="-1"></a><span class="co"># GoodG &lt;- direct.sum(as.matrix(G)[genotyped_id,genotyped_id], NonGenoEEV)</span></span>
<span id="cb114-17"><a href="#cb114-17" tabindex="-1"></a><span class="co"># colnames(GoodG) &lt;- rownames(GoodG)</span></span>
<span id="cb114-18"><a href="#cb114-18" tabindex="-1"></a><span class="co"># #now we have all individuals in G matrix</span></span>
<span id="cb114-19"><a href="#cb114-19" tabindex="-1"></a><span class="co"># dim(GoodG)</span></span>
<span id="cb114-20"><a href="#cb114-20" tabindex="-1"></a><span class="co"># length(unique(sommerdf$id))</span></span>
<span id="cb114-21"><a href="#cb114-21" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-22"><a href="#cb114-22" tabindex="-1"></a><span class="co"># #Repeat for dominance matrix</span></span>
<span id="cb114-23"><a href="#cb114-23" tabindex="-1"></a><span class="co"># GoodD &lt;- direct.sum(as.matrix(D)[genotyped_id,genotyped_id], NonGenoEEV)</span></span>
<span id="cb114-24"><a href="#cb114-24" tabindex="-1"></a><span class="co"># colnames(GoodD) &lt;- rownames(GoodD)</span></span>
<span id="cb114-25"><a href="#cb114-25" tabindex="-1"></a><span class="co"># #now we have all individuals in G matrix</span></span>
<span id="cb114-26"><a href="#cb114-26" tabindex="-1"></a><span class="co"># dim(GoodD)</span></span>
<span id="cb114-27"><a href="#cb114-27" tabindex="-1"></a><span class="co"># length(unique(sommerdf$id))</span></span>
<span id="cb114-28"><a href="#cb114-28" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-29"><a href="#cb114-29" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-30"><a href="#cb114-30" tabindex="-1"></a><span class="co"># #Extract inbreeding coefficients for all genotypes from G:</span></span>
<span id="cb114-31"><a href="#cb114-31" tabindex="-1"></a><span class="co"># ploidy &lt;- 4</span></span>
<span id="cb114-32"><a href="#cb114-32" tabindex="-1"></a><span class="co"># Fg &lt;- (diag(GoodG)-1)/(ploidy-1)</span></span>
<span id="cb114-33"><a href="#cb114-33" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-34"><a href="#cb114-34" tabindex="-1"></a><span class="co"># #include Fg into sommerdf:</span></span>
<span id="cb114-35"><a href="#cb114-35" tabindex="-1"></a><span class="co"># #1) reorder</span></span>
<span id="cb114-36"><a href="#cb114-36" tabindex="-1"></a><span class="co"># Fg &lt;- Fg[match(sommerdf$id, names(Fg))]</span></span>
<span id="cb114-37"><a href="#cb114-37" tabindex="-1"></a><span class="co"># identical(sommerdf$id, names(Fg))</span></span>
<span id="cb114-38"><a href="#cb114-38" tabindex="-1"></a><span class="co"># sommerdf &lt;- cbind(sommerdf, Fg)</span></span>
<span id="cb114-39"><a href="#cb114-39" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-40"><a href="#cb114-40" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb114-41"><a href="#cb114-41" tabindex="-1"></a><span class="co"># #Precalculate the inverse of G and D matrices for Sommer</span></span>
<span id="cb114-42"><a href="#cb114-42" tabindex="-1"></a><span class="co"># Ginv &lt;- as(solve(as.matrix(GoodG)),Class=&quot;dgCMatrix&quot;)</span></span>
<span id="cb114-43"><a href="#cb114-43" tabindex="-1"></a><span class="co"># Dinv &lt;- as(solve(as.matrix(GoodD)),Class=&quot;dgCMatrix&quot;)</span></span>
<span id="cb114-44"><a href="#cb114-44" tabindex="-1"></a></span>
<span id="cb114-45"><a href="#cb114-45" tabindex="-1"></a></span>
<span id="cb114-46"><a href="#cb114-46" tabindex="-1"></a>Ginv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(G)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb114-47"><a href="#cb114-47" tabindex="-1"></a>Dinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(D)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb114-48"><a href="#cb114-48" tabindex="-1"></a></span>
<span id="cb114-49"><a href="#cb114-49" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> sommerdf[<span class="sc">-</span><span class="fu">which</span>(sommerdf<span class="sc">$</span>id <span class="sc">%in%</span> non_genotyped_id),]</span>
<span id="cb114-50"><a href="#cb114-50" tabindex="-1"></a></span>
<span id="cb114-51"><a href="#cb114-51" tabindex="-1"></a><span class="co">#There are more genotypes in the genomic relationship matrix than </span></span>
<span id="cb114-52"><a href="#cb114-52" tabindex="-1"></a><span class="co">#with BLUEs. BLUPs for unphenotpyed genotypes will be predicted</span></span>
<span id="cb114-53"><a href="#cb114-53" tabindex="-1"></a><span class="fu">dim</span>(Ginv)</span></code></pre></div>
<pre><code>## [1] 943 943</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(sommerdf_smaller<span class="sc">$</span>id))</span></code></pre></div>
<pre><code>## [1] 926</code></pre>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> geno<span class="sc">@</span>Fg</span>
<span id="cb118-2"><a href="#cb118-2" tabindex="-1"></a></span>
<span id="cb118-3"><a href="#cb118-3" tabindex="-1"></a><span class="co">#include Fg into sommerdf:</span></span>
<span id="cb118-4"><a href="#cb118-4" tabindex="-1"></a><span class="co">#1) reorder</span></span>
<span id="cb118-5"><a href="#cb118-5" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> Fg[<span class="fu">match</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))]</span>
<span id="cb118-6"><a href="#cb118-6" tabindex="-1"></a><span class="fu">identical</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_smaller, Fg)</span>
<span id="cb120-2"><a href="#cb120-2" tabindex="-1"></a></span>
<span id="cb120-3"><a href="#cb120-3" tabindex="-1"></a><span class="co">#leave only records for genotypes with marker data!</span></span>
<span id="cb120-4"><a href="#cb120-4" tabindex="-1"></a>EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(stage1.EEV.open[sommerdf_smaller<span class="sc">$</span>Stage1Error,</span>
<span id="cb120-5"><a href="#cb120-5" tabindex="-1"></a>                                     sommerdf_smaller<span class="sc">$</span>Stage1Error]),</span>
<span id="cb120-6"><a href="#cb120-6" tabindex="-1"></a>              <span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb120-7"><a href="#cb120-7" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs))</span>
<span id="cb120-8"><a href="#cb120-8" tabindex="-1"></a></span>
<span id="cb120-9"><a href="#cb120-9" tabindex="-1"></a>Stage1Error_geno <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(stage1.EEV.open[sommerdf_smaller<span class="sc">$</span>Stage1Error,</span>
<span id="cb120-10"><a href="#cb120-10" tabindex="-1"></a>                                     sommerdf_smaller<span class="sc">$</span>Stage1Error])) <span class="sc">-</span></span>
<span id="cb120-11"><a href="#cb120-11" tabindex="-1"></a>  <span class="fu">mean</span>(stage1.EEV.open[sommerdf_smaller<span class="sc">$</span>Stage1Error,</span>
<span id="cb120-12"><a href="#cb120-12" tabindex="-1"></a>                                     sommerdf_smaller<span class="sc">$</span>Stage1Error])</span></code></pre></div>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" tabindex="-1"></a><span class="co">#without dominance</span></span>
<span id="cb121-2"><a href="#cb121-2" tabindex="-1"></a>ans2c_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Fixed effects = Environment. -1 means that</span></span>
<span id="cb121-3"><a href="#cb121-3" tabindex="-1"></a>                     <span class="co">#we remove the intercept (you could leave it if you prefer it)</span></span>
<span id="cb121-4"><a href="#cb121-4" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Ginv) <span class="sc">+</span></span>
<span id="cb121-5"><a href="#cb121-5" tabindex="-1"></a>                     <span class="co">#additive genotypic effects, following G matrix</span></span>
<span id="cb121-6"><a href="#cb121-6" tabindex="-1"></a>                     <span class="fu">vsc</span>(<span class="fu">isc</span>(id)) <span class="sc">+</span> <span class="co">#Genotypic effects not captured by the </span></span>
<span id="cb121-7"><a href="#cb121-7" tabindex="-1"></a>                     <span class="co">#addtivie effect, i.i.d.</span></span>
<span id="cb121-8"><a href="#cb121-8" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb121-9"><a href="#cb121-9" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb121-10"><a href="#cb121-10" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb121-11"><a href="#cb121-11" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb121-12"><a href="#cb121-12" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb121-13"><a href="#cb121-13" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb121-14"><a href="#cb121-14" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb121-15"><a href="#cb121-15" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb121-16"><a href="#cb121-16" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb121-17"><a href="#cb121-17" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb121-18"><a href="#cb121-18" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb121-19"><a href="#cb121-19" tabindex="-1"></a>                           </span>
<span id="cb121-20"><a href="#cb121-20" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb121-21"><a href="#cb121-21" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb121-22"><a href="#cb121-22" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb121-23"><a href="#cb121-23" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb121-24"><a href="#cb121-24" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb121-25"><a href="#cb121-25" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb121-26"><a href="#cb121-26" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb121-27"><a href="#cb121-27" tabindex="-1"></a>                           </span>
<span id="cb121-28"><a href="#cb121-28" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb121-29"><a href="#cb121-29" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb121-30"><a href="#cb121-30" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb121-31"><a href="#cb121-31" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb121-32"><a href="#cb121-32" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb121-33"><a href="#cb121-33" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb121-34"><a href="#cb121-34" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb121-35"><a href="#cb121-35" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb121-36"><a href="#cb121-36" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb121-37"><a href="#cb121-37" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb121-38"><a href="#cb121-38" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb121-39"><a href="#cb121-39" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb121-40"><a href="#cb121-40" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" tabindex="-1"></a><span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp</span></code></pre></div>
<pre><code>##                                      VarComp  VarCompSE   Zratio Constraint
## id:Ginv:isc:isc                     52.68572 10.8186562 4.869895   Positive
## id:isc:isc                          33.29483 12.8250744 2.596073   Positive
## Stage1Error:Variance_scaled:EEVInv:  1.00000  0.4196304 2.383050      Fixed
## units:isc:isc                       31.94601 17.2890829 1.847756   Positive</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb124-2"><a href="#cb124-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb124-3"><a href="#cb124-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2c_open<span class="sc">$</span>b</span>
<span id="cb124-4"><a href="#cb124-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb124-5"><a href="#cb124-5" tabindex="-1"></a></span>
<span id="cb124-6"><a href="#cb124-6" tabindex="-1"></a></span>
<span id="cb124-7"><a href="#cb124-7" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb124-8"><a href="#cb124-8" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb124-9"><a href="#cb124-9" tabindex="-1"></a>  ans2c_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2c_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Ginv:isc:isc&quot;</span>]<span class="sc">*</span>meanG, <span class="co">#additive</span></span>
<span id="cb124-10"><a href="#cb124-10" tabindex="-1"></a>  ans2c_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2c_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:isc:isc&quot;</span>], <span class="co">#other genetic</span></span>
<span id="cb124-11"><a href="#cb124-11" tabindex="-1"></a>  ans2c_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2c_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>],</span>
<span id="cb124-12"><a href="#cb124-12" tabindex="-1"></a>  Stage1Error_geno) <span class="co">#residual </span></span>
<span id="cb124-13"><a href="#cb124-13" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb124-14"><a href="#cb124-14" tabindex="-1"></a>variances_ans2c <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb124-15"><a href="#cb124-15" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb124-16"><a href="#cb124-16" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2c) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2c<span class="sc">$</span>vars))</span>
<span id="cb124-17"><a href="#cb124-17" tabindex="-1"></a><span class="co">#Extremely similar variances</span></span>
<span id="cb124-18"><a href="#cb124-18" tabindex="-1"></a>variances_ans2c</span></code></pre></div>
<pre><code>##                Sommer ASReml
## env          52.73437   51.7
## additive     40.37831   40.4
## g.resid      33.29483   33.3
## g x env      31.94601   32.1
## Stage1.error 33.91894   33.6</code></pre>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" tabindex="-1"></a><span class="co">#similar aic</span></span>
<span id="cb126-2"><a href="#cb126-2" tabindex="-1"></a>ans2c<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 6958.828</code></pre>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" tabindex="-1"></a>ans2c_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 6965.393</code></pre>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2c_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Ginv)</span><span class="st">`</span> <span class="sc">+</span> <span class="fu">mean</span>(ans2c_open<span class="sc">$</span>b)</span>
<span id="cb130-2"><a href="#cb130-2" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb130-3"><a href="#cb130-3" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb130-4"><a href="#cb130-4" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb130-5"><a href="#cb130-5" tabindex="-1"></a>              <span class="at">geno=</span>geno,</span>
<span id="cb130-6"><a href="#cb130-6" tabindex="-1"></a>              <span class="at">vars=</span>ans2c<span class="sc">$</span>vars)</span>
<span id="cb130-7"><a href="#cb130-7" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">geno=</span>geno, <span class="at">what=</span><span class="st">&quot;BV&quot;</span>)</span>
<span id="cb130-8"><a href="#cb130-8" tabindex="-1"></a></span>
<span id="cb130-9"><a href="#cb130-9" tabindex="-1"></a></span>
<span id="cb130-10"><a href="#cb130-10" tabindex="-1"></a><span class="co">#Extremely similar BLUPs</span></span>
<span id="cb130-11"><a href="#cb130-11" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb130-12"><a href="#cb130-12" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9999949</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" tabindex="-1"></a>ASRemlBLUPs<span class="sc">$</span>value[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 55.20110 55.67120 61.11249 63.81276 63.59544</code></pre>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" tabindex="-1"></a>SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 55.04386 55.50965 60.93716 63.61817 63.41638</code></pre>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" tabindex="-1"></a><span class="co">#with dominance</span></span>
<span id="cb136-2"><a href="#cb136-2" tabindex="-1"></a>ans2d_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb136-3"><a href="#cb136-3" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb136-4"><a href="#cb136-4" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Ginv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb136-5"><a href="#cb136-5" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb136-6"><a href="#cb136-6" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic</span></span>
<span id="cb136-7"><a href="#cb136-7" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb136-8"><a href="#cb136-8" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb136-9"><a href="#cb136-9" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb136-10"><a href="#cb136-10" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb136-11"><a href="#cb136-11" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb136-12"><a href="#cb136-12" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb136-13"><a href="#cb136-13" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb136-14"><a href="#cb136-14" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb136-15"><a href="#cb136-15" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb136-16"><a href="#cb136-16" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb136-17"><a href="#cb136-17" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb136-18"><a href="#cb136-18" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb136-19"><a href="#cb136-19" tabindex="-1"></a>                           </span>
<span id="cb136-20"><a href="#cb136-20" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb136-21"><a href="#cb136-21" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb136-22"><a href="#cb136-22" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb136-23"><a href="#cb136-23" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb136-24"><a href="#cb136-24" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb136-25"><a href="#cb136-25" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb136-26"><a href="#cb136-26" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb136-27"><a href="#cb136-27" tabindex="-1"></a>                           </span>
<span id="cb136-28"><a href="#cb136-28" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb136-29"><a href="#cb136-29" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb136-30"><a href="#cb136-30" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb136-31"><a href="#cb136-31" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb136-32"><a href="#cb136-32" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb136-33"><a href="#cb136-33" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb136-34"><a href="#cb136-34" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb136-35"><a href="#cb136-35" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb136-36"><a href="#cb136-36" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb136-37"><a href="#cb136-37" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb136-38"><a href="#cb136-38" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb136-39"><a href="#cb136-39" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb136-40"><a href="#cb136-40" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb137-2"><a href="#cb137-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb137-3"><a href="#cb137-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2d_open<span class="sc">$</span>b[<span class="fu">unique</span>(sommerdf_smaller<span class="sc">$</span>Env),]</span>
<span id="cb137-4"><a href="#cb137-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb137-5"><a href="#cb137-5" tabindex="-1"></a></span>
<span id="cb137-6"><a href="#cb137-6" tabindex="-1"></a></span>
<span id="cb137-7"><a href="#cb137-7" tabindex="-1"></a><span class="co">#heterosis effects for all observations </span></span>
<span id="cb137-8"><a href="#cb137-8" tabindex="-1"></a>XFg <span class="ot">&lt;-</span> sommerdf_smaller<span class="sc">$</span>Fg</span>
<span id="cb137-9"><a href="#cb137-9" tabindex="-1"></a>FgEffects <span class="ot">&lt;-</span> ans2d_open<span class="sc">$</span>b[<span class="st">&quot;Fg&quot;</span>,]</span>
<span id="cb137-10"><a href="#cb137-10" tabindex="-1"></a>Fg_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(XFg<span class="sc">*</span>FgEffects)</span>
<span id="cb137-11"><a href="#cb137-11" tabindex="-1"></a></span>
<span id="cb137-12"><a href="#cb137-12" tabindex="-1"></a><span class="co">#heterosis effects for each genotype</span></span>
<span id="cb137-13"><a href="#cb137-13" tabindex="-1"></a><span class="co">#extract XFg from geno, not from the model</span></span>
<span id="cb137-14"><a href="#cb137-14" tabindex="-1"></a><span class="co">#In the model, there are repeated Fg coefficients for genotypes that had more</span></span>
<span id="cb137-15"><a href="#cb137-15" tabindex="-1"></a><span class="co">#than one BLUE (genotypes present in more than one environment) and there are </span></span>
<span id="cb137-16"><a href="#cb137-16" tabindex="-1"></a><span class="co">#no Fg coefficients for genotypes that were not phenotyped and had no BLUEs</span></span>
<span id="cb137-17"><a href="#cb137-17" tabindex="-1"></a><span class="co">#If we take Fg from geno, there are no such problems</span></span>
<span id="cb137-18"><a href="#cb137-18" tabindex="-1"></a>XFg <span class="ot">&lt;-</span> geno<span class="sc">@</span>Fg</span>
<span id="cb137-19"><a href="#cb137-19" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> XFg<span class="sc">*</span>FgEffects</span>
<span id="cb137-20"><a href="#cb137-20" tabindex="-1"></a><span class="co">#reorder heterosis effects to be in the same order as BLUPs</span></span>
<span id="cb137-21"><a href="#cb137-21" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> heterosis_effects[<span class="fu">match</span>(<span class="fu">rownames</span>(ans2d_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Ginv)</span><span class="st">`</span>),</span>
<span id="cb137-22"><a href="#cb137-22" tabindex="-1"></a>                        <span class="fu">names</span>(heterosis_effects))]</span>
<span id="cb137-23"><a href="#cb137-23" tabindex="-1"></a></span>
<span id="cb137-24"><a href="#cb137-24" tabindex="-1"></a> </span>
<span id="cb137-25"><a href="#cb137-25" tabindex="-1"></a></span>
<span id="cb137-26"><a href="#cb137-26" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb137-27"><a href="#cb137-27" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb137-28"><a href="#cb137-28" tabindex="-1"></a>  ans2d_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2d_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Ginv:isc:isc&quot;</span>]<span class="sc">*</span>meanG, <span class="co">#additive</span></span>
<span id="cb137-29"><a href="#cb137-29" tabindex="-1"></a>    ans2d_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2d_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Dinv:isc:isc&quot;</span>]<span class="sc">*</span>meanD, <span class="co">#dominance</span></span>
<span id="cb137-30"><a href="#cb137-30" tabindex="-1"></a>  Fg_variance, <span class="co">#heterosis</span></span>
<span id="cb137-31"><a href="#cb137-31" tabindex="-1"></a>  ans2d_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2d_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>], <span class="co">#residual</span></span>
<span id="cb137-32"><a href="#cb137-32" tabindex="-1"></a>  Stage1Error_geno) <span class="co">#stage1 estimation error</span></span>
<span id="cb137-33"><a href="#cb137-33" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2d<span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb137-34"><a href="#cb137-34" tabindex="-1"></a>variances_ans2d <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb137-35"><a href="#cb137-35" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb137-36"><a href="#cb137-36" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2d) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2d<span class="sc">$</span>vars))</span>
<span id="cb137-37"><a href="#cb137-37" tabindex="-1"></a><span class="co">#almost same variances </span></span>
<span id="cb137-38"><a href="#cb137-38" tabindex="-1"></a>variances_ans2d</span></code></pre></div>
<pre><code>##                 Sommer ASReml
## env          52.080736   51.0
## additive     48.972937   49.1
## dominance    17.844183   17.8
## heterosis     3.757561    3.8
## g x env      41.746397   42.0
## Stage1.error 33.918939   33.6</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" tabindex="-1"></a><span class="co">#almost same aic</span></span>
<span id="cb139-2"><a href="#cb139-2" tabindex="-1"></a>ans2d<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 6940.182</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" tabindex="-1"></a>ans2d_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 6948.604</code></pre>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2d_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Ginv)</span><span class="st">`</span> <span class="sc">+</span></span>
<span id="cb143-2"><a href="#cb143-2" tabindex="-1"></a>              ans2d_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span> <span class="sc">+</span> </span>
<span id="cb143-3"><a href="#cb143-3" tabindex="-1"></a>              heterosis_effects <span class="sc">+</span></span>
<span id="cb143-4"><a href="#cb143-4" tabindex="-1"></a>              <span class="fu">mean</span>(Env_effects)</span>
<span id="cb143-5"><a href="#cb143-5" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb143-6"><a href="#cb143-6" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb143-7"><a href="#cb143-7" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb143-8"><a href="#cb143-8" tabindex="-1"></a>              <span class="at">geno=</span>geno,</span>
<span id="cb143-9"><a href="#cb143-9" tabindex="-1"></a>              <span class="at">vars=</span>ans2d<span class="sc">$</span>vars)</span>
<span id="cb143-10"><a href="#cb143-10" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">geno=</span>geno, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb143-11"><a href="#cb143-11" tabindex="-1"></a></span>
<span id="cb143-12"><a href="#cb143-12" tabindex="-1"></a></span>
<span id="cb143-13"><a href="#cb143-13" tabindex="-1"></a><span class="co">#Extremely similar BLUPs</span></span>
<span id="cb143-14"><a href="#cb143-14" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb143-15"><a href="#cb143-15" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9999938</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" tabindex="-1"></a>ASRemlBLUPs<span class="sc">$</span>value[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 53.54244 56.23127 62.28803 70.11746 67.53857</code></pre>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" tabindex="-1"></a>SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 53.37766 56.06199 62.11953 69.95073 67.37771</code></pre>
<p>Based on the AIC values, we select the model with dominance. To
compare AIC values for models with marker data to models without marker
data, one needs to ensure that all individuals analyzed in Stage1 are in
the marker data file. Because Stage2 excludes ungenotyped individuals
from the analysis, the populations will not be the same. In this potato
dataset, there are 1294 clones in the phenotype file but only 943 with
marker data.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" tabindex="-1"></a><span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>##              Variance   PVE
## env              51.7    NA
## additive         40.4 0.290
## g.resid          33.3 0.239
## g x env          32.1 0.230
## Stage1.error     33.6 0.241</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" tabindex="-1"></a><span class="fu">summary</span>(ans2d<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>##              Variance   PVE
## env              51.0    NA
## additive         49.1 0.335
## dominance        17.8 0.122
## heterosis         3.8 0.026
## g x env          42.0 0.287
## Stage1.error     33.6 0.230</code></pre>
<p>The output for the dominance model includes rows for ‚Äúdominance‚Äù and
‚Äúheterosis‚Äù; the former is based on the variance of a random effect with
zero mean, while the latter is due to the mean (<a href="https://doi.org/10.3389/fgene.2018.00078">Varona et
al.¬†2018</a>).</p>
</div>
<div id="pedigree-data" class="section level3">
<h3>Pedigree data</h3>
<p>Estimating additive relationships using both marker and pedigree data
is often beneficial because they have complementary properties. Unlike
the G matrix, the A matrix is ‚Äúsparse‚Äù, meaning it has zero covariance
between unrelated individuals. However, the A matrix does not account
for segregation within biparental families or capture LD between
founders. The G and A matrices can be combined into an ‚ÄúH‚Äù matrix, which
also allows ungenotyped individuals to be included in the relationship
matrix (<a href="https://doi.org/10.3168/jds.2009-2061">Legarra et
al.¬†2009</a>; <a href="https://doi.org/10.1186/1297-9686-42-2">Christensen and Lund
2010</a>). (The inclusion of ungenotyped individuals is only available
with the genetic residual model.)</p>
<p>To illustrate this feature, a three-column pedigree file for the
potato population is included with the package. When combining the G and
A matrices, their relative weights must be specified using the argument
<code>w</code> in <code>read_geno</code>, such that H = (1-w)G + wA.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" tabindex="-1"></a>ped.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;ped.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb153-2"><a href="#cb153-2" tabindex="-1"></a>ped <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(ped.file)</span>
<span id="cb153-3"><a href="#cb153-3" tabindex="-1"></a><span class="co">#When you use read_geno with pedigree information output@G is actually </span></span>
<span id="cb153-4"><a href="#cb153-4" tabindex="-1"></a><span class="co">#the H matrix!</span></span>
<span id="cb153-5"><a href="#cb153-5" tabindex="-1"></a>geno2 <span class="ot">&lt;-</span> <span class="fu">read_geno</span>(geno.file,<span class="at">ploidy=</span><span class="dv">4</span>,<span class="at">map=</span><span class="cn">TRUE</span>,<span class="at">ped=</span>ped,<span class="at">w=</span><span class="fl">0.1</span>,<span class="at">dominance =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Minor allele threshold = 5 genotypes
## Number of markers = 12242
## Number of genotypes = 943</code></pre>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" tabindex="-1"></a><span class="co">#read_geno doesn&#39;t return the A and H matrices. We can calculate them manually</span></span>
<span id="cb155-2"><a href="#cb155-2" tabindex="-1"></a>ped2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">id=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ped),</span>
<span id="cb155-3"><a href="#cb155-3" tabindex="-1"></a>                    <span class="at">parent1=</span><span class="fu">match</span>(ped<span class="sc">$</span>parent1,ped<span class="sc">$</span>id,<span class="at">nomatch=</span><span class="dv">0</span>),</span>
<span id="cb155-4"><a href="#cb155-4" tabindex="-1"></a>                    <span class="at">parent2=</span><span class="fu">match</span>(ped<span class="sc">$</span>parent2,ped<span class="sc">$</span>id,<span class="at">nomatch=</span><span class="dv">0</span>))</span>
<span id="cb155-5"><a href="#cb155-5" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">as</span>(AGHmatrix<span class="sc">::</span><span class="fu">Amatrix</span>(ped2,<span class="at">ploidy=</span>geno2<span class="sc">@</span>ploidy),<span class="st">&quot;symmetricMatrix&quot;</span>)</span></code></pre></div>
<pre><code>## Verifying conflicting data... 
## Organizing data... 
## Your data was chronologically organized with success. 
## Processing a large pedigree data... It may take a couple of minutes... 
## Constructing matrix A using ploidy = 4 and proportion of double reduction = 0 ;as in Kerr et al. (2012) 
## Completed! Time = 0.002  minutes</code></pre>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" tabindex="-1"></a><span class="fu">rownames</span>(A) <span class="ot">&lt;-</span> ped<span class="sc">$</span>id</span>
<span id="cb157-2"><a href="#cb157-2" tabindex="-1"></a><span class="fu">colnames</span>(A) <span class="ot">&lt;-</span> ped<span class="sc">$</span>id</span>
<span id="cb157-3"><a href="#cb157-3" tabindex="-1"></a>Asmall <span class="ot">&lt;-</span> A[<span class="fu">rownames</span>(geno2<span class="sc">@</span>G),<span class="fu">colnames</span>(geno2<span class="sc">@</span>G)] <span class="co">#leave only common individuals in </span></span>
<span id="cb157-4"><a href="#cb157-4" tabindex="-1"></a><span class="co">#matrices A and G</span></span>
<span id="cb157-5"><a href="#cb157-5" tabindex="-1"></a></span>
<span id="cb157-6"><a href="#cb157-6" tabindex="-1"></a><span class="co">#Calculate H matrix:</span></span>
<span id="cb157-7"><a href="#cb157-7" tabindex="-1"></a>w <span class="ot">&lt;-</span> <span class="fl">0.1</span> <span class="co">#blending parameter</span></span>
<span id="cb157-8"><a href="#cb157-8" tabindex="-1"></a>H <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">-</span>w)<span class="sc">*</span>G <span class="sc">+</span> w<span class="sc">*</span>Asmall</span>
<span id="cb157-9"><a href="#cb157-9" tabindex="-1"></a><span class="co">#Manually calculated H matrix is similar to the matrix calculated by read_geno</span></span>
<span id="cb157-10"><a href="#cb157-10" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">abs</span>(geno2<span class="sc">@</span>G <span class="sc">-</span> H))</span></code></pre></div>
<pre><code>## [1] 5.499464e-07</code></pre>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" tabindex="-1"></a>ans2e <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues,<span class="at">vcov=</span>stage1.EEV,<span class="at">geno=</span>geno2,<span class="at">non.add=</span><span class="st">&quot;dom&quot;</span>)</span>
<span id="cb159-2"><a href="#cb159-2" tabindex="-1"></a>ans2e<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 6935.89</code></pre>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" tabindex="-1"></a><span class="fu">summary</span>(ans2e<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>##              Variance   PVE
## env              51.3    NA
## additive         56.9 0.385
## dominance        13.8 0.093
## heterosis         3.5 0.024
## g x env          40.1 0.271
## Stage1.error     33.6 0.227</code></pre>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb163-2"><a href="#cb163-2" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb163-3"><a href="#cb163-3" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb163-4"><a href="#cb163-4" tabindex="-1"></a>              <span class="at">geno=</span>geno2,</span>
<span id="cb163-5"><a href="#cb163-5" tabindex="-1"></a>              <span class="at">vars=</span>ans2e<span class="sc">$</span>vars)</span>
<span id="cb163-6"><a href="#cb163-6" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">geno=</span>geno2, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb163-7"><a href="#cb163-7" tabindex="-1"></a></span>
<span id="cb163-8"><a href="#cb163-8" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id)) <span class="co">#there are 1294 genotypes with stage1 BLUEs</span></span></code></pre></div>
<pre><code>## [1] 1294</code></pre>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" tabindex="-1"></a><span class="fu">dim</span>(A) <span class="co">#there are 1372 genotypes in the pedigree</span></span></code></pre></div>
<pre><code>## [1] 1372 1372</code></pre>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" tabindex="-1"></a><span class="fu">dim</span>(geno2<span class="sc">@</span>G) <span class="co">#there are 943 genotpes with marker data</span></span></code></pre></div>
<pre><code>## [1] 943 943</code></pre>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" tabindex="-1"></a><span class="co">#only 942 out of 1294 phenotyped individuals are present in the pedigree</span></span>
<span id="cb169-2"><a href="#cb169-2" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id) <span class="sc">%in%</span> <span class="fu">rownames</span>(A)) </span></code></pre></div>
<pre><code>## [1] 942</code></pre>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" tabindex="-1"></a><span class="co">#all 943 genotyped individuals are in the pedigree</span></span>
<span id="cb171-2"><a href="#cb171-2" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">rownames</span>(A) <span class="sc">%in%</span> <span class="fu">rownames</span>(geno2<span class="sc">@</span>G)) </span></code></pre></div>
<pre><code>## [1] 943</code></pre>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" tabindex="-1"></a><span class="co">#we still only have predictions for the genotypes with marker</span></span>
<span id="cb173-2"><a href="#cb173-2" tabindex="-1"></a><span class="co">#data in spite of using pedigree information</span></span>
<span id="cb173-3"><a href="#cb173-3" tabindex="-1"></a><span class="fu">dim</span>(ASRemlBLUPs)</span></code></pre></div>
<pre><code>## [1] 943   3</code></pre>
<p>The above result shows that blending G and A at w=0.1 slightly
reduced the AIC and shifted variance from the non-additive to additive
component. One way to select the blending parameter is based on AIC.
When a vector of w values is provided to <code>read_geno</code>, the
function returns a list output corresponding to those values. For
numerical conditioning, a minimum threshold of 1e-5 is used for w.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" tabindex="-1"></a>w.vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">1e-5</span>, <span class="fu">seq</span>(<span class="fl">0.2</span>,<span class="fl">0.8</span>,<span class="at">by=</span><span class="fl">0.2</span>))</span>
<span id="cb175-2"><a href="#cb175-2" tabindex="-1"></a>geno <span class="ot">&lt;-</span> <span class="fu">read_geno</span>(geno.file,<span class="at">ploidy=</span><span class="dv">4</span>,<span class="at">map=</span><span class="cn">TRUE</span>,<span class="at">ped=</span>ped,<span class="at">w=</span>w.vec,<span class="at">dominance=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Minor allele threshold = 5 genotypes
## Number of markers = 12242
## Number of genotypes = 943</code></pre>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">w=</span>w.vec, <span class="at">aic=</span><span class="fu">numeric</span>(<span class="dv">5</span>), <span class="at">h2=</span><span class="fu">numeric</span>(<span class="dv">5</span>))</span>
<span id="cb177-2"><a href="#cb177-2" tabindex="-1"></a>ans2 <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>,<span class="dv">5</span>)</span>
<span id="cb177-3"><a href="#cb177-3" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb177-4"><a href="#cb177-4" tabindex="-1"></a>  ans2[[i]] <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>stage1.blues,<span class="at">vcov=</span>stage1.EEV,<span class="at">geno=</span>geno[[i]],<span class="at">non.add=</span><span class="st">&quot;dom&quot;</span>)</span>
<span id="cb177-5"><a href="#cb177-5" tabindex="-1"></a>  result<span class="sc">$</span>aic[i] <span class="ot">&lt;-</span> ans2[[i]]<span class="sc">$</span>aic</span>
<span id="cb177-6"><a href="#cb177-6" tabindex="-1"></a>  result<span class="sc">$</span>h2[i] <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2[[i]]<span class="sc">$</span>vars)[<span class="dv">2</span>,<span class="dv">2</span>]</span>
<span id="cb177-7"><a href="#cb177-7" tabindex="-1"></a>}</span>
<span id="cb177-8"><a href="#cb177-8" tabindex="-1"></a></span>
<span id="cb177-9"><a href="#cb177-9" tabindex="-1"></a><span class="co">#The process above can be easily replicated in Sommer, but as it&#39;s slower than</span></span>
<span id="cb177-10"><a href="#cb177-10" tabindex="-1"></a><span class="co">#ASReml it could take a while to run. Instead, I will just show here how to </span></span>
<span id="cb177-11"><a href="#cb177-11" tabindex="-1"></a><span class="co">#do the modelling in Sommer with the best value for w found by ASReml. </span></span>
<span id="cb177-12"><a href="#cb177-12" tabindex="-1"></a><span class="co">#To find the best w value with Sommer, just do a different Sommer model </span></span>
<span id="cb177-13"><a href="#cb177-13" tabindex="-1"></a><span class="co">#for each w value and select the one with smallest AIC.</span></span>
<span id="cb177-14"><a href="#cb177-14" tabindex="-1"></a></span>
<span id="cb177-15"><a href="#cb177-15" tabindex="-1"></a></span>
<span id="cb177-16"><a href="#cb177-16" tabindex="-1"></a></span>
<span id="cb177-17"><a href="#cb177-17" tabindex="-1"></a><span class="co">#with H matrix</span></span>
<span id="cb177-18"><a href="#cb177-18" tabindex="-1"></a><span class="co">#When reading the geno.file using pedigree,</span></span>
<span id="cb177-19"><a href="#cb177-19" tabindex="-1"></a><span class="co">#the G matrix in geno is actually the H matrix</span></span>
<span id="cb177-20"><a href="#cb177-20" tabindex="-1"></a>Hinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]]<span class="sc">@</span>G)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb177-21"><a href="#cb177-21" tabindex="-1"></a><span class="co">#needed later to scale variances</span></span>
<span id="cb177-22"><a href="#cb177-22" tabindex="-1"></a>meanH <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]]<span class="sc">@</span>G)) <span class="sc">-</span></span>
<span id="cb177-23"><a href="#cb177-23" tabindex="-1"></a>          <span class="fu">mean</span>(geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]]<span class="sc">@</span>G)</span></code></pre></div>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" tabindex="-1"></a>ans2_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb178-2"><a href="#cb178-2" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb178-3"><a href="#cb178-3" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Hinv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb178-4"><a href="#cb178-4" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb178-5"><a href="#cb178-5" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic</span></span>
<span id="cb178-6"><a href="#cb178-6" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb178-7"><a href="#cb178-7" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb178-8"><a href="#cb178-8" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb178-9"><a href="#cb178-9" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb178-10"><a href="#cb178-10" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb178-11"><a href="#cb178-11" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb178-12"><a href="#cb178-12" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb178-13"><a href="#cb178-13" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb178-14"><a href="#cb178-14" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb178-15"><a href="#cb178-15" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb178-16"><a href="#cb178-16" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb178-17"><a href="#cb178-17" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb178-18"><a href="#cb178-18" tabindex="-1"></a>                           </span>
<span id="cb178-19"><a href="#cb178-19" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb178-20"><a href="#cb178-20" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb178-21"><a href="#cb178-21" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb178-22"><a href="#cb178-22" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb178-23"><a href="#cb178-23" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb178-24"><a href="#cb178-24" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb178-25"><a href="#cb178-25" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb178-26"><a href="#cb178-26" tabindex="-1"></a>                           </span>
<span id="cb178-27"><a href="#cb178-27" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb178-28"><a href="#cb178-28" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb178-29"><a href="#cb178-29" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb178-30"><a href="#cb178-30" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb178-31"><a href="#cb178-31" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb178-32"><a href="#cb178-32" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb178-33"><a href="#cb178-33" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb178-34"><a href="#cb178-34" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb178-35"><a href="#cb178-35" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb178-36"><a href="#cb178-36" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb178-37"><a href="#cb178-37" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb178-38"><a href="#cb178-38" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb178-39"><a href="#cb178-39" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb179-2"><a href="#cb179-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb179-3"><a href="#cb179-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2_open<span class="sc">$</span>b[<span class="fu">unique</span>(sommerdf_smaller<span class="sc">$</span>Env),]</span>
<span id="cb179-4"><a href="#cb179-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb179-5"><a href="#cb179-5" tabindex="-1"></a></span>
<span id="cb179-6"><a href="#cb179-6" tabindex="-1"></a></span>
<span id="cb179-7"><a href="#cb179-7" tabindex="-1"></a><span class="co"># #heterosis effects for all observations </span></span>
<span id="cb179-8"><a href="#cb179-8" tabindex="-1"></a><span class="co"># #needed to calculate heterosis variance component</span></span>
<span id="cb179-9"><a href="#cb179-9" tabindex="-1"></a><span class="co"># XFg &lt;- sommerdf_smaller$Fg</span></span>
<span id="cb179-10"><a href="#cb179-10" tabindex="-1"></a><span class="co"># FgEffects &lt;- ans2d_open$b[&quot;Fg&quot;,]</span></span>
<span id="cb179-11"><a href="#cb179-11" tabindex="-1"></a><span class="co"># Fg_variance &lt;- var(XFg*FgEffects)</span></span>
<span id="cb179-12"><a href="#cb179-12" tabindex="-1"></a></span>
<span id="cb179-13"><a href="#cb179-13" tabindex="-1"></a></span>
<span id="cb179-14"><a href="#cb179-14" tabindex="-1"></a><span class="co">#heterosis effects for each genotype</span></span>
<span id="cb179-15"><a href="#cb179-15" tabindex="-1"></a><span class="co">#extract XFg from geno, not from the model</span></span>
<span id="cb179-16"><a href="#cb179-16" tabindex="-1"></a><span class="co">#In the model, there are repeated Fg coefficients for genotypes that had more</span></span>
<span id="cb179-17"><a href="#cb179-17" tabindex="-1"></a><span class="co">#than one BLUE (genotypes present in more than one environment) and there are </span></span>
<span id="cb179-18"><a href="#cb179-18" tabindex="-1"></a><span class="co">#no Fg coefficients for genotypes that were not phenotyped and had no BLUEs</span></span>
<span id="cb179-19"><a href="#cb179-19" tabindex="-1"></a><span class="co">#If we take Fg from geno, there are no such problems</span></span>
<span id="cb179-20"><a href="#cb179-20" tabindex="-1"></a>XFg <span class="ot">&lt;-</span> geno2<span class="sc">@</span>Fg</span>
<span id="cb179-21"><a href="#cb179-21" tabindex="-1"></a>FgEffects <span class="ot">&lt;-</span> ans2_open<span class="sc">$</span>b[<span class="st">&quot;Fg&quot;</span>,]</span>
<span id="cb179-22"><a href="#cb179-22" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> XFg<span class="sc">*</span>FgEffects</span>
<span id="cb179-23"><a href="#cb179-23" tabindex="-1"></a><span class="co">#reorder heterosis effects to be in the same order as BLUPs</span></span>
<span id="cb179-24"><a href="#cb179-24" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> heterosis_effects[<span class="fu">match</span>(<span class="fu">rownames</span>(ans2_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span>),</span>
<span id="cb179-25"><a href="#cb179-25" tabindex="-1"></a>                        <span class="fu">names</span>(heterosis_effects))]</span>
<span id="cb179-26"><a href="#cb179-26" tabindex="-1"></a>Fg_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(heterosis_effects)</span>
<span id="cb179-27"><a href="#cb179-27" tabindex="-1"></a> </span>
<span id="cb179-28"><a href="#cb179-28" tabindex="-1"></a></span>
<span id="cb179-29"><a href="#cb179-29" tabindex="-1"></a></span>
<span id="cb179-30"><a href="#cb179-30" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span> <span class="sc">+</span></span>
<span id="cb179-31"><a href="#cb179-31" tabindex="-1"></a>              ans2_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span> <span class="sc">+</span> </span>
<span id="cb179-32"><a href="#cb179-32" tabindex="-1"></a>              heterosis_effects <span class="sc">+</span></span>
<span id="cb179-33"><a href="#cb179-33" tabindex="-1"></a>              <span class="fu">mean</span>(Env_effects)</span>
<span id="cb179-34"><a href="#cb179-34" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb179-35"><a href="#cb179-35" tabindex="-1"></a>ans2best <span class="ot">&lt;-</span> ans2[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]]</span>
<span id="cb179-36"><a href="#cb179-36" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb179-37"><a href="#cb179-37" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb179-38"><a href="#cb179-38" tabindex="-1"></a>              <span class="at">geno=</span>geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]],</span>
<span id="cb179-39"><a href="#cb179-39" tabindex="-1"></a>              <span class="at">vars=</span>ans2best<span class="sc">$</span>vars)</span>
<span id="cb179-40"><a href="#cb179-40" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">geno=</span>geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]], <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb179-41"><a href="#cb179-41" tabindex="-1"></a></span>
<span id="cb179-42"><a href="#cb179-42" tabindex="-1"></a></span>
<span id="cb179-43"><a href="#cb179-43" tabindex="-1"></a><span class="co">#Extremely similar BLUPs</span></span>
<span id="cb179-44"><a href="#cb179-44" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb179-45"><a href="#cb179-45" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9999815</code></pre>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" tabindex="-1"></a>ASRemlBLUPs<span class="sc">$</span>value[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 54.38557 54.94530 62.97290 71.99530 67.85179</code></pre>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" tabindex="-1"></a>SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 54.22447 54.79992 62.78910 71.78185 67.67393</code></pre>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb185-2"><a href="#cb185-2" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb185-3"><a href="#cb185-3" tabindex="-1"></a>  ans2_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Hinv:isc:isc&quot;</span>]<span class="sc">*</span>meanH, <span class="co">#additive</span></span>
<span id="cb185-4"><a href="#cb185-4" tabindex="-1"></a>    ans2_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Dinv:isc:isc&quot;</span>]<span class="sc">*</span>meanD, <span class="co">#dominance</span></span>
<span id="cb185-5"><a href="#cb185-5" tabindex="-1"></a>  Fg_variance, <span class="co">#heterosis</span></span>
<span id="cb185-6"><a href="#cb185-6" tabindex="-1"></a>  ans2_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>], <span class="co">#residual</span></span>
<span id="cb185-7"><a href="#cb185-7" tabindex="-1"></a>  Stage1Error_geno) <span class="co">#stage1 estimation error</span></span>
<span id="cb185-8"><a href="#cb185-8" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2best<span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb185-9"><a href="#cb185-9" tabindex="-1"></a>variances_ans2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb185-10"><a href="#cb185-10" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb185-11"><a href="#cb185-11" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2best<span class="sc">$</span>vars))</span>
<span id="cb185-12"><a href="#cb185-12" tabindex="-1"></a><span class="co">#Extremely similar variances with Sommer and ASReml </span></span>
<span id="cb185-13"><a href="#cb185-13" tabindex="-1"></a>variances_ans2</span></code></pre></div>
<pre><code>##                 Sommer ASReml
## env          52.594901   51.5
## additive     73.168752   74.2
## dominance     5.483771    5.0
## heterosis     3.075492    2.9
## g x env      34.822580   34.5
## Stage1.error 33.918939   33.6</code></pre>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" tabindex="-1"></a><span class="co">#almost same aic</span></span>
<span id="cb187-2"><a href="#cb187-2" tabindex="-1"></a>ans2best<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 6927.393</code></pre>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" tabindex="-1"></a>ans2_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 6935.901</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" tabindex="-1"></a>axis.scaling <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">range</span>(result<span class="sc">$</span>h2))<span class="sc">/</span><span class="fu">diff</span>(<span class="fu">range</span>(result<span class="sc">$</span>aic))</span>
<span id="cb191-2"><a href="#cb191-2" tabindex="-1"></a>result<span class="sc">$</span>y2 <span class="ot">&lt;-</span> (result<span class="sc">$</span>aic<span class="sc">-</span><span class="fu">min</span>(result<span class="sc">$</span>aic))<span class="sc">*</span>axis.scaling <span class="sc">+</span> <span class="fu">min</span>(result<span class="sc">$</span>h2)</span>
<span id="cb191-3"><a href="#cb191-3" tabindex="-1"></a>y2lab <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">seq</span>(<span class="fu">min</span>(result<span class="sc">$</span>aic),<span class="fu">max</span>(result<span class="sc">$</span>aic),<span class="at">length.out=</span><span class="dv">5</span>))</span>
<span id="cb191-4"><a href="#cb191-4" tabindex="-1"></a>y2axis <span class="ot">&lt;-</span> y2lab<span class="sc">-</span><span class="fu">min</span>(result<span class="sc">$</span>aic) <span class="sc">+</span> <span class="fu">min</span>(result<span class="sc">$</span>h2)<span class="sc">/</span>axis.scaling</span>
<span id="cb191-5"><a href="#cb191-5" tabindex="-1"></a></span>
<span id="cb191-6"><a href="#cb191-6" tabindex="-1"></a><span class="fu">ggplot</span>(result) <span class="sc">+</span> </span>
<span id="cb191-7"><a href="#cb191-7" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>w,<span class="at">y=</span>h2)) <span class="sc">+</span></span>
<span id="cb191-8"><a href="#cb191-8" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>w,<span class="at">y=</span>y2),<span class="at">colour=</span><span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb191-9"><a href="#cb191-9" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name=</span><span class="st">&quot;Genomic h2&quot;</span>,</span>
<span id="cb191-10"><a href="#cb191-10" tabindex="-1"></a>                     <span class="at">sec.axis=</span><span class="fu">sec_axis</span>(trans<span class="sc">~</span>.<span class="sc">/</span>axis.scaling,<span class="at">name=</span><span class="st">&quot;AIC&quot;</span>,</span>
<span id="cb191-11"><a href="#cb191-11" tabindex="-1"></a>                                       <span class="at">breaks=</span>y2axis,</span>
<span id="cb191-12"><a href="#cb191-12" tabindex="-1"></a>                                       <span class="at">labels=</span>y2lab)) <span class="sc">+</span> </span>
<span id="cb191-13"><a href="#cb191-13" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span>  </span>
<span id="cb191-14"><a href="#cb191-14" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y.right=</span><span class="fu">element_text</span>(<span class="at">colour=</span><span class="st">&quot;red&quot;</span>),</span>
<span id="cb191-15"><a href="#cb191-15" tabindex="-1"></a>        <span class="at">axis.title.y.right=</span><span class="fu">element_text</span>(<span class="at">colour=</span><span class="st">&quot;red&quot;</span>)) <span class="sc">+</span></span>
<span id="cb191-16"><a href="#cb191-16" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Blending G and A for Yield&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABSlBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZmYAZpAAZrYzMzM6AAA6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOgBmOjpmkLZmkNtmtpBmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQOjqQZgCQZjqQZmaQkGaQkLaQttuQ2/+rbk2rbm6rbo6rjk2r5OSr5P+2ZgC2Zjq2kDq2tpC2ttu225C229u22/+2/9u2///Ijk3I///bkDrbtmbbtpDb27bb29vb/7bb///kq27k///r6+v/AAD/ADr/AGb/OgD/Ojr/Omb/OpD/ZgD/Zrb/kDr/kLb/kNv/tmb/tpD/tv//yI7/25D/27b/29v/2///5Kv//7b//8j//9v//+T///+rKx9gAAAACXBIWXMAAA7DAAAOwwHHb6hkAAARDElEQVR4nO2d+WPUxhXH5QQauglpg52mbklbTA/Sw+4JpC2bpA0FDC0QlEAxLireNF7W8///2rkkzf1G0kg7Ws33B6/9PDsz+uxcK703k6Ekp7J1VyB2JUCAEiBACRCgBAhQAgQoAQKUAAFqD+gwo3rvNkKrg/OP7AlXBxeAFDjN/Ys4r3c/h4sluVUqsj3hX59k2QVTKfMyFf5Frkj9l5StqM6AsjfudAe0vMwz2waLla5knkm0ssx4lctdXEeaQK1Iv4BoqatP8McDA3JrdZC9hxvP6sFs6wZUrJjbYnZReIPcnEQVFNzqgHGSMusdEK7kNi/na9wM3mUd7rN3sq0fk//en2XnPuctSLFv/aiqU1F+8otZVc3772Q0rSm3ugp/n1Vtbs5aM6nFOVqLC4fZG7fL/+2RJr+H9Ko+0rMNCmj1Kf4MaTmLGekj9C/WXfZ4Nzx3uQQk2+sOMdc+2jLFnjE3ruXuBaHtMkBFXYs3Z1n5T9LJGHy1qvgvNdtggOphg13+1hWEHswIDDJWHuLLX+5uXUdfH2QVIM1eXqiaOe4Ot1mDsr6L9SmhXxUM5xVSNVqaMKAV2Vusg6lVPf9IyzY0oEuPaDkL1thxu2J9nVw1sy1mJaDSXmSlXQREM6wHs3999ocZRavnxkUangCXAOL/nW/dUEacecZIqlUt/xKyDQaIVeA+bx8F50WqRq6SjHrs413VYxC3H9KhtRoXTYBYL6imHTk3KUnFgQGiGIqyFpVKkHpV1WxDAyKfY0dA9YddXdVyN/vun/72cNcFSOjjYwDEq1ZdJSmRtd7lrgpI6WL470sKIJZb2TnV3MS0y90ShNrFrIDkqirZBgd0P2OXjke+63j+PGC42CXh8e8KHWUVQOq4iJPQRfQ/P8wqQHhkxpOx8i6eG+fBms68XAqpg7QVkFxVJduAgHhDLReKRdng60titnqar+zKNF+vpM/dqJBlbEgy5SaBKWpiuGEU0gxuBqRUVc42OKB3b5StfYGv8s0rYhfDc+nl7Nw/tDGILRR/IrbqBx+SvK5Xf9P13vW52B7r3IiqDlqNYGwwWVQLRTsgpapitgEBddZyF/7mtX6tBdBi9hZePM2t351i0loA8RHGfQckEq2ni61+jwfSS2Pgk+4oQkqAACVAgBIgQAkQoAQIUGtA3x6RegL0+qOdD76gvz3e2dn53r3yhQPir0/0NxpMnsnav9OVrCdAZzf30ePv01/v7osvcqnTBfT6V/fQ6c9Igzn7+BaqX5RSpwvo9BdfoNe/JExwX9vZ2S9fEB9/noxGPQE6+aAEdPrTW6T58Bf+79SC6hZEpY5DCVA9BlElQKrObl7jsxjpbGd/vsdflFKnC4ivg0gjwgug92+h8kUudUSAVk/zL79BaJHnX/G/XyG0zPPnhuwqtV9J89fcXjHA1jegTAV09AoVzzGRV+iYMjnGvxFIxy8M+VW5OP7nVOSAsixTW9CStZsFhrOkLenfGA4xLr4y5Fdl5PifU1UX0wmtHRChI5j417HF8yPSxUpAq2f/w4AqXDZtHiBOB2lj0CKnPGgXw0iOX5LuVUwMUE0H6YA4CzxI/+fZN/iv1TAtSB+F1gVIooM0QMuaBR55jnOilwOMQZEAUukgDRCZxUgXI92LzlukBa2evux5FosCEKbjsQ5asgVQUa58hlkHPdFn+oEBsbazhpW0W5EAqnpWAmQwieNOzIBUQsMAUkblBEgy6XNWAiRIn9EjB6QQ6huQaUpPgGplTQuYFiDSvcYBSHywkg/2CCdr8Z4IWpDchHpsQVmbAiYEyL5qToCI1DthnrnFAEgi1BOgevWTAJlshluFnrlNA1Dml8xkigKQSKgPQNAX06kDMt9s9sxtAoAyv2QWUxyABEKhAfnc2pgyIL2KCZAov3s/IwBUEwoKyPU8xzO3jQZkvPmTAJWy3PwZKaCKUDBAmV8yKLeNBWS9+ZMAUcGPTD1ziwZQSSgMIMfNn0gAadE+tQH1D8h18ycOQFq0j2BAfQPye+jumdtQ0T6y673RTzq31bUhIODmTxyAtGif2mCN9gn0+KfNsx2Hhor2qQ1EPbYg8OZPbC2I6u6+bDCHIuS6yZTMaYJv/sQBSIv28RiDAgBq4tfimWyoaJ/aIJYaGJDXzR/PAsqFqzmYpaAvzBv4lSFDribRPuA6iFeqA6Bmjj+ATQUkBbMQ52ju/1u43FyDrqRRV0Cm2rQGlCtdTA1mKT3InY72cQFq6hnlttWAjMEsqGxBTj/yaQBiL0owC/6bYnI3oOCACKG2l9nYdcxt0wBJwSzEQtk4R6CoADV3HXPacnWal4NZqPCwjY5eGnKTatVO4wMkB7MUVVyds4eFByRUTBR8mZaAi4CA5GAWvAIizQkYguIBZLv/3BaQ0JLjuaNIlQDJpTpmD1HQZbbyznTZEiDA1gOgOd/+vNDPKDAoLKB27qsOm7hgDQRoTrdW3usOyLTXAHCZLd1XHbbwgJa7e4ifWZAACVIAkZ2thwfU1r/XbpPuSwXsYoieR9AVkGGvAedlOh/xxAOo3CTecEaQSRME1EyuTd6aPf0J/JBHr0Fs6yBk2GvA0Q46eIhbbfLjp3CA1MNgXJoiIPikNEGhAHVxobfaegLEZ3o/OQF57efxRC0/GCDFiSJgC0qANEkV9FsiyqV2A9QtxsBm6wUQXiCW6j5Ie+3n0Rsg1RUwwmneF1DHIAyLbXMAWfeMAt/ptI0CkMd+HgmQZlNMncN4zDbN43+sgLqH8ZhtCdD6AK0Otr2/b0CAwA1PAsQ5GW163Fo4QHN22K8Xoa6AQsQ5GW09AuLfxTrfck2AxFLt9Xfu5xEkEMxo67OLHbJjlbd93tgNkMVTIQAgQ3h6wFmsEI8GdmuagCSJTq0kkKXx2T72/TwChRIChTYEtDQ6A3tF+xA2+83P9lkHIFOZvoCM296KDw7LOx5skBYd609//uv95mf7WAGFirU02DoAoq5Vmk+5V6zG2cd/vdnmbB/L458+nvO4imwwBh3lajPyifZBj6+R/tb8bB/LhifBglENtk4tCAN6oTYj+2OfugXh3+ggTdTo6BozoHDBqLrNOHM2A4RVCHu42x/71GMQmb12dq5Ra7OzfUz7eQQMRtVtTkCmYBbENybnwoAWZKv7OhP7Yx8puIe0oBZn+8QFyBLMciz0KDIGyUcAOB771NE+9Tqo4dk+BkAho3U1m/n7nzuYhR7QwlTIjUepLwr52KeUdm8mMyZzmEIAcgWzsANaiJZq41ErHPDZfKU4ALEXczALO6CFaGmM+pG7mN/XMLHUhoDCxntrNjcgUzALP6DFrp6ezVeSnzBAngodAVlu87qCWfgBLdSYV7JO8xsNyBzMghq0ILR4O/QgLQMKHRCv2gBAxmCWJoDkL6uAPAFJjgQ9A7I9zY3zsU8pAVDwHQMUW3dAZBWdy/FRAwIKv2OAYusKaPWURUHn4nwv1/ow5C1XrsodLnpAx/xGBzlSq1KPN+25SkA+ngqdAFmdtjwB1VyOhVtCPT724eKAQu450Q+gZfWl1Xa7ox9A3Ot9EwD108VYxYNuymFMZnc99h2Djl6ov6AhBunxACqnr2Nxou99mmeAPF051gsID9NkdC6kmx4DACKExgHIpPb3g5o8iunzQU9ViFURxmrIyr19XTq0IEeMWoyxGrLM+3mMD1Af94OoBgBkikUPPwaFv2nPytgUQL3cD6JlGPfzGB+gRvIHlFk2PAkJyFlAAjQcIPxVY++Qxc9DagTI2QPcpqgAzc8/3N0L5SctlbAZgKiT2V7o2x1DAHJPk3ED4gU03IyhYbKBAKFD0sUC3w/aKEBh/aSl/B1ftt0mz9sFo53mEyB3/QV3Mu9Q+ubJEiDANBAgMoPN8RgUKKBOzv2Jdyi9h00x5UCyQIAWs20aUud5V2h6gGi4Ifnh912jOSCvSHEv21oAsYZDAJULRS3ax2dPe8mkOCT2BCiHkoUERMQBadE+8Nk+qkn12ITifKMGVHtw8i6mRfv4nKshmTSPzVEDYs+dUd2UtGgfj7N9ZOkPewKd/dMw01DTPNsmcDHj07wW7eNxto9kMnls2mPsbCYwWQ4mC7ZQpM8N2WaKyBDt43O2j2iKDpAlmGVh8K+vZV9Ja9E+Dccgs8emJYTMbgoIyBzMUpgCEGq5YlaVaB+fs30E00CA1L1wDMmcwSxHX/63JSA92qfROsjm0mqKkHKZQgByn8zStosBGhMg9mI5mSU2QHoAkDGV09YSkPlkljUBsvv8hgXkk5srmAXFB8jnMwds7QDZglnWAsjlFL02QJZglugAeaxcAJuhvw6ykm4kJyC3U3QCBACCv38DttEDAp2ioXuAgC0BAmyGVeeoAJnyVJIBD2oAWwIE2CYAyN/x1ZXM8+5SbIC8DgNNgBRpNl+3xc0D5HlaagIkS7d5euU5rty51SdaOyDLMxbvwJ7uj4B8c4irBWXe7cDPK2+8LUiphZCf72V6OZ3Zr9z7OW0C5Ew2YkBeTmcbBijTTaZk3NQJkL8zzXgB+TidbRagTDeZklWmBKiSpf7wVW4UoEw3IUMywdQeUAOX0DEDgqeiTQKU6aZKCRBqBwj8vuA5vI8BkBB2oMvvG1UCZDAlQDZbaXLfed8cQHLYgarAgBpFn40dkPvx6MYAUsMOFCVAHQA5nXyM72wWItwXoNqp9YSd6eM620eLy1C0gYDq4B7iCUx+c53t0wmQy+ksXkCyYz3+zXW2jynsAAG2sQOSQzNwC3Kd7dN1F7dmz4AaPjHqCZAY3HN69f1brrN9zGEHgE002b3yDLbc0IDW3oLK38xH13QGZPfKU215bnxovfYxqCTTFyCr05k0VOW5TzLV1NssVgb38M5mP9vHGnbgtDUElOeNVwP8td91EJ3j2aE+1rN9QgCyOZ0xW563+9JPtfaVtCPswGXzBpSro/I0AZmdznKNTvMC1g3IGXbgsMGAKByPdwIFbAogxemsbDqjBwSEHdhtLkDuKWuqgLhXXg5OWaMCBIYdWG1GQPqYPHZA3esvyPQtKwEK8s4EqEOyBAhIZt67Y0GPLGZH+DlOYJsSIGnvDhpBz6J6C8Mp6pUmBEjeu6Mg21LQuPmlfOKaokkAsu3dwV6OjeerlpoEIPai7t3BDxRzN6BJAVL27liy41WdI9CkACl7dyz43HWkHn8tqz2gEYlXWdq7o+SzeubsYe0BVaBCJguamZZM2rujXAABQ9CkALVSAgSoM6BNVwIEKAEClAABSoAAtQZUO+hJ+0/ak51e5d5FzlR8T0so2dlN4Sm4s8zKZ7Cl2gKqHfTkfbityYj/DPEwcqdCbG9vKDPiZHJi+1TkMh/bPzwvtQVUO8cobjK2ZCekynfN167t7e1TpkfVqIuTK6mH2gKSNlMWHK2syYhsybS9veEy/2LvYnWydbag2kFP3ofbmgwxjyMoFdvbG0p2epVtbw3l5hwe/TRgC3r9kYWPvrd3oDLJmHfScZQebAyinzqYGd/bGyzzNw5Awrjnatyeaj+LXatmsWuuWaz8p4OPvrc3mOyuo4sJZa6xBQkOeh7rIJyMtQ3LtWt7e3uUab/wOtnJjmO55Ke0kgaUAAFKgAAlQIASIEAJEKAECFACBCgBAhQLIHr+W5Ht8bMW41EsgOi5ePOL25iU1yGdgykaQIu376x++8fzj/DruqsiKRpAuOUsvvPwh3f8DjEdTtEAwmNPcWF1cIWMRTEpHkDF+d/tocNv/eAGnHRIxQNoubt1A89jkfWwiAChOWaz3I2sh8UEKE4lQIASIEAJEKAECFACBCgBApQAAUqAACVAgP4Pslbt6DMQr0QAAAAASUVORK5CYII=" /><!-- --></p>
<p>Based on the above figure, the blending parameter w=0.4 is chosen, at
which there is no longer much dominance variance. The optimal value for
w will not be the same for all traits.</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" tabindex="-1"></a>w.vec[<span class="dv">3</span>]</span></code></pre></div>
<pre><code>## [1] 0.4</code></pre>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1" tabindex="-1"></a>genoH <span class="ot">&lt;-</span> geno[[<span class="dv">3</span>]] </span>
<span id="cb194-2"><a href="#cb194-2" tabindex="-1"></a>ans2H <span class="ot">&lt;-</span> ans2[[<span class="dv">3</span>]]</span>
<span id="cb194-3"><a href="#cb194-3" tabindex="-1"></a><span class="fu">summary</span>(ans2H<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>##              Variance   PVE
## env              51.5    NA
## additive         74.2 0.494
## dominance         5.0 0.033
## heterosis         2.9 0.019
## g x env          34.5 0.230
## Stage1.error     33.6 0.224</code></pre>
<p>To include ungenotyped individuals in the H matrix, put a fourth
column in the pedigree data frame with binary (0/1) values to indicate
which individuals should be included.</p>
</div>
<div id="blup-reliability" class="section level3">
<h3>BLUP Reliability</h3>
<p>The calculation of BLUPs is split into two functions:
<code>blup_prep</code> and <code>blup</code>. The computationally
intensive steps occur in <code>blup_prep</code>, which combines the
phenotype and genotype information used in <code>Stage2</code> with the
variance component estimates to estimate the var-cov matrix of the
predicted random effects. The <code>blup</code> command extracts the
appropriate linear combination of predictions based on the argument
‚Äúwhat‚Äù, which has 5 possible values:</p>
<ul>
<li><p>AV = additive values</p></li>
<li><p>BV = breeding values</p></li>
<li><p>GV = genotypic values</p></li>
<li><p>AM = additive marker effects</p></li>
<li><p>DM = dominance marker effects</p></li>
</ul>
<p>The ‚Äúvalues‚Äù are properties of individuals, as opposed to markers.
Breeding values should be used for parent selection, while genotypic
values should be used for clone selection. For diploids, the AV and BV
are equivalent. For polyploids, if the dominance model was used, the BV
includes a portion of the dominance. GV is the sum of additive and
non-additive values. When predicting values, the software also returns
the predicted reliability r2, which is the squared correlation between
the true and predicted values (assuming the model is correct).</p>
<p>The following code illustrates the prediction of genotypic
values:</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" tabindex="-1"></a>prep1 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb196-2"><a href="#cb196-2" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb196-3"><a href="#cb196-3" tabindex="-1"></a>              <span class="at">geno=</span>genoH,</span>
<span id="cb196-4"><a href="#cb196-4" tabindex="-1"></a>              <span class="at">vars=</span>ans2H<span class="sc">$</span>vars)</span>
<span id="cb196-5"><a href="#cb196-5" tabindex="-1"></a>GV1 <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep1, <span class="at">geno=</span>genoH, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb196-6"><a href="#cb196-6" tabindex="-1"></a><span class="fu">head</span>(GV1)</span></code></pre></div>
<pre><code>##         id    value        r2
## 1 AF5392-8 54.38557 0.5597140
## 2 AF5393-1 54.94530 0.7015017
## 3 AF5429-3 62.97290 0.7407578
## 4 AF5445-2 71.99530 0.6968320
## 5 AF5450-7 67.85179 0.7011862
## 6 AF5484-3 58.92143 0.7781236</code></pre>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" tabindex="-1"></a><span class="co">#Sommer computes reliability for all random effects</span></span>
<span id="cb198-2"><a href="#cb198-2" tabindex="-1"></a>reliabilities <span class="ot">&lt;-</span> sommer<span class="sc">::</span><span class="fu">r2</span>(ans2_open)</span>
<span id="cb198-3"><a href="#cb198-3" tabindex="-1"></a><span class="fu">head</span>(reliabilities<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span>)</span></code></pre></div>
<pre><code>##                [,1]
## variance0 0.6620011
## variance0 0.6709527
## variance0 0.6349565
## variance0 0.6272746
## variance0 0.6194958
## variance0 0.6703176</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" tabindex="-1"></a><span class="fu">head</span>(reliabilities<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span>)</span></code></pre></div>
<pre><code>##                   [,1]
## variance0  0.243591452
## variance0  0.245685594
## variance0  0.205725661
## variance0 -0.005006784
## variance0  0.012181875
## variance0  0.141301463</code></pre>
<p>The figure below illustrates how the reliability of genotypic value
predictions is typically higher when using marker data because it
improves estimation of the additive component.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1" tabindex="-1"></a><span class="co">#predict genotypic values without marker data</span></span>
<span id="cb202-2"><a href="#cb202-2" tabindex="-1"></a>prep2 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb202-3"><a href="#cb202-3" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb202-4"><a href="#cb202-4" tabindex="-1"></a>              <span class="at">vars=</span>ans2b<span class="sc">$</span>vars)</span>
<span id="cb202-5"><a href="#cb202-5" tabindex="-1"></a>GV2 <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep2, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb202-6"><a href="#cb202-6" tabindex="-1"></a></span>
<span id="cb202-7"><a href="#cb202-7" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(GV2,GV1,<span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb202-8"><a href="#cb202-8" tabindex="-1"></a><span class="fu">ggplot</span>(plot.data,<span class="fu">aes</span>(<span class="at">x=</span>r2.x,<span class="at">y=</span>r2.y)) <span class="sc">+</span></span>
<span id="cb202-9"><a href="#cb202-9" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb202-10"><a href="#cb202-10" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;GV Reliability&quot;</span>) <span class="sc">+</span></span>
<span id="cb202-11"><a href="#cb202-11" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb202-12"><a href="#cb202-12" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Without markers&quot;</span>) <span class="sc">+</span> </span>
<span id="cb202-13"><a href="#cb202-13" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;With markers&quot;</span>) <span class="sc">+</span> </span>
<span id="cb202-14"><a href="#cb202-14" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="at">ratio=</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb202-15"><a href="#cb202-15" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fl">0.4</span>,<span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb202-16"><a href="#cb202-16" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fl">0.4</span>,<span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb202-17"><a href="#cb202-17" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="fl">0.4</span>,<span class="dv">1</span>),<span class="at">y=</span><span class="fu">c</span>(<span class="fl">0.4</span>,<span class="dv">1</span>)),</span>
<span id="cb202-18"><a href="#cb202-18" tabindex="-1"></a>            <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y),</span>
<span id="cb202-19"><a href="#cb202-19" tabindex="-1"></a>            <span class="at">linetype=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABKVBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYzMzM6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmOgBmOjpmOmZmZmZmkNtmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQOmaQZgCQZmaQkDqQkGaQkLaQkNuQtpCQttuQ27aQ2/+rbk2rbm6rjk2ryKur5OSr5P+2ZgC2Zjq2kDq2tpC2ttu225C229u22/+2///Ijk3I///bkDrbtmbbtpDb27bb29vb/7bb/9vb///kq27k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T////dyooFAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOSElEQVR4nO2dC3vbthWGYTeeqmTp2kxOOyXd1irdmq1btPtSZ+uiXZ1UyZbatazUioL//yOGC0lcCOBAJEWC9PmeJ5HMQ4A8rw8OQJAGCUUFRbo+gdSFgAAhIEAICBACAoSAACEgQAgIUE1A2xe3CSHv/YfSzfToTG6aHz4TX5ZE6r1To8T86Iz/U1s205FlE/b1j5/VO7dmVA/Q5n4GYcJ5zOS23N8cEDnUPY0GtDzsP6DtnHzAgmf7cnxwQtdj6ecqA5V7uP07x6eV0eFwKUCGfQiAViRzTcBZSI8WuWO5h5spB/T6vmxsKoJefJ+Qg58KQKsxuXFqRNA3cx6XErasoCPVArQwfskrESkqHjJA3y349vWYNzYWaKoJyfY3YyXeGZu2ApCsbMVMnakOILNtZI2jaGFGDtrODz6l9CULNAXh8FSG3mZKfni2/QcZOXKQ+BUs7EbZphoAJEBwH5b8N70svMkAHdw54yBEK2FOa0n6f//83ZiMsv6Pd35lQCsRYR22sEYBcU+288Ib4eGL8QecxSoLpoMT1Y2LRicASXjKpgDxQ3TawmoBKkY8edfDGsT6ZuGNzEErIhueDYg1rPf/+K9X0zAgXmWnLaxuL3ZHfskArccz1cLyJJ3l6Dwz5QDkFpmDRtTXxNghPum0hdUeB/FBNP3vxyJM2M/f+63yJgO0mcok/Zh19RoEhuaM9/0illjTLCdp0bK289udtrCmRtI3pBMronmzLC45RkUbm2gQ5BYO6P1x3tfpgMT4nH10O2Cse7H68mN+tfU4+6m4IOPKATEUrDWtGct3PtUGinzkeOMxyzEsB7GB4g9OrUsNBn90RosBeldK/mpejau6UeqAXt/vtA9LHRDL36TTPix1QCx93en4FNIGlIAQECAEBAgBAUJAgBAQoDqA3h2gmgVUfLsA9kzcrswICDAjIMCMgAAzAgLMCAgwVwd09bOvxeebXxx/9G2ptsQBtADo8vhDAejtk0f0/Eel2hIHsH9Az+9+JSPozRdfF8GEgHRlVK4+/5a++eVTKofRF4NTbUCXH+WAKEaQoVIEUQRk6GqwOYg0Cujtk4cD68UIbQ4Q/ze0cRAxzTiSBswICDAjIMCMgHSRshkBaSIOMwJSIi4zAgLMCAgwIyDAjICkiM+MgIRspxGQaS/5jIAwB9WzIyAEFG13+ouACrvbXQSU2z3eIiDMQfXsCAgBgXZCAp4ioAvxx40RxdMAVD5XBGQcsnyyCMg4ZAeAOspB1Z7BEYCafaynsUMmEUGt56B8kZWI4mkAatXOwSAgv51oiih+TQG5+gV38esKCC81/CYROgjIs50A9pL5mgHSnm/BHOSQ9nwL9mKAHQF5lGFBQMZPikXOBXOQxST/UQGKrH6ogCwkBS4ElMkHqMhBsdVfO0Bx5YcPSPVSCMhtL6ggILfdBrRb+YED4tPNJqBd668AaDOdbabQIqFpADLnDNsCtBjxlUaX4VVCuwJkIiAGoZYAsQDazkd0FQ6hjgBZDEgphHauvxIgvgx2yoDM0bK784qtf3dA2/mEL5S+SLKJmQ0q6r5F8zlozVfjh9ZK7zIH2WmnbUBR6rIX6xjQdh6ziHOn3bzdutoFxJK08bP6a/mrB9lCObQDQBoCZw6qWn+FJrY00o9aNYiv23GeryzQNiDVkevjnQg++4ggecism1crdojFTb7oaO2OUsopeAF89p6k1ZovKoKaWT+IxD22I3crpRyubGv9UxGqCEhbNajZtTvy3zxQPt/NBQh6MCHm/KpE0JIQ9ZYVFUFXnz2llx821sTiAJEQIEGoZhOucrF69EpejgmpHNTwClS7APKlnogI3NO12Ky4FlOrBjUcQXGT6k5AxuRh14CyzMOD6PL4+O7TUm377cUAQPDjLXsZB72SF/QhtQTInXmMNtY+IPmyIuCdH+0AcvLpHlCMEgFU//i9BuTmY93AaD9J/0l8bH+z7xlFp4NEG/iBfDoCJNLzAritEQJEdC93GwcRvWwkIBIeTO9jRnGyJNC78wKAiOHlPgDpfMDLjT3kIHY9D86ZdQNItkDz/kUHgOzX8LrkB0QiATlzkFnU1bpKu7ULKJ8MIqRyDooG5LJHRZC5W/s5KErdAMrNBqBd6g+ZW5u0t2K+chPz8CFa2WD9cfb6k/ZuBcdBUYByZy/sbcTPxmIfrD/SXnvS3iNoHFStF1OZxo3GwaeLCJJnU2+gGAJk5pELq6AHkF7UUA+vxQBAejMy7EaglOgUBqu6HgICclAsIGrzMWIpdPxd7JUuNWo2Md8Z5IeNA6RvoCkB2s4nrKeH+rIGASmfbUC09C2FHMTRLCZ0Fe7LavzdvAXIiAovIJJODuKAlvt6BI+YgIpN+g4WO2IXD9UPHT9k3uUhTkYHGA1VAUQ0WXcEqYFGj7EEAbEkRBfQhFAsIBUmZuetAF3onABASeSgOEV286arZUCkKO8FdAGcZw8Bqd+7CcMDyBtBNLOHz6SH4yAPIFcOIlYOMpN06CwDx9/JXmUcBE4n0vD6QbJlFN9IaUOxjW9VRkcFbaqL6Q49SqiRisoRZJZvbL4nwQkz/Qg6GaCJOYqB9Tdgr5CDgCGiXVs0IKfCgOz5Ipf6maS1r3bs7ACoPCXrUBcXq8AhzdqCc9J+MiFAeY/mrt9S/+akNY9NFk4+TkBuD/Zi7/auBhhAaqAY4cFe7FVy0K2aSVoRCoBRgIAjJQeo9qR90XjqAApN+pvq37WYAuRmov/gKm+d2kABFV/CgDwRpJ/ZMAEVX3Qm7kuN3uWgOEEPL6hvVSJI10ABuQi5wgkB+QB5I6jhOefErsWI3j9FjBPLgJqec+5swsx5Aobn6h5YiI993yzggVu9uhbTeFANl58Q7EB6gOpci7kjyEkm36l/gGpNmMUDKgj1C1DNp1yJ3j3FdPNGeXVvw+2BW33q5q3sEsrSjnGQZ3IoOUBZkq7y8IIFyIYRBqSQ+jxwawiAXHzcI+n0AS0LH6qsH0SMHFTmUoqgkgMlPqkBampOeqi9WLRiplzDgPJguQDOJC1AcolAIfPPwrnePon7s3AbUCAHyfI9mrTfTO3crJbHoc8fidUXrNoCk/YmIK1X0/fa8/M/TTexhR48XGppimJpHLO2ACD3jVVrrx7mIDHfkUPSFjf5/K9ZEwOWxxGuq29EX9zGtVci2jFJLzNCakmTqwePBC4L924RVN6rhxHELujVygt6BEWu/rILIPDdOwkCYmlIewRYy0G/3hlQ+cEgczfavzlpNpI2HwBWy+PwXiymidmpmOojIms32IHEADmG0Wp5HPYtYpnAUl9OhwSIp5+Y+bJIQPnXAQGi2VqukCoAsi5E+wuI8kxUfWkKFyB1mW/t5yhvK0lAvKk19viLsV07OOnrpL05kvaoLiB17H4B4kNEEI5Zmz1h5gVkT4b1MoLKV/MeRU3aiy2+E+x5DoIUCYj6TxABle3Gca8joFIOsu3GYXuZg6LljyA7NZfGSd4zdOvaAbKFgICarhOgPK0YAx4EBOxpAATsUPk92JMC5DwgAgKOh4DStiMgBFTPngog/7EQEHAoBAQcCQGlbUdACKiefW+Aoh+7SeoZoJA6iqC6b2/qbQRFnkDnz/+kDih5OwJCQPXsXQIigB0q34q9Q0AEsEPl27F3B4gAdqh8S3bMQQionh0BJQlIrxkBlY1GxQgIqBcBAXsmbkdACKievW1A5ToREFAlAgJqREDAnonbERACqmdvD5CvNgQEVIaAgLoQELBn4nYEtHdAav0guVaOXVviAPYOSFs/iNLzYxegcD1DB6TW7qD06ue/cgACqhk6IG3Nl7d/+ZtsYsb6Qb15/gdSRUBq/SB6/hBzUFnG+kEIqCyVg86PuR7atSUOoIVe7KHqxTCCHFLrByGgHYSAACEgQAgIEAIC9O4A1SggDVW/y4eKIyCgOAICijcDaMBCQIAQECAEBAgBAaoHyH23o0p5/W0LVcpfPVCrFO+gfKZdd8RULUCeux1VyutvW6hQns93nu9cnl5mVA1HTNUC5LnbUaG88baFCuXFpPDOdTy/+5U8e90RS7UAue52VCuv3rZQrXzFCMqpGEv1m6oFyHO3o0J5420LVY7vTyJBZYC0imw1FEHG3Y5q5X1nGFf+s6f0skKW3nMEee52VChvvG2hQvlACAR1td8c5LnbUaX88ypNTJWvGUGGI6aaGAfZdzuqlNfftlCl/OVxlXGUKJodfh/joOsgBAQIAQFCQIAQECAEBKhtQAvxVq6leDHOdj5a3zyhr08p/3CK2cLylmxKbQMSL//dzn/COW2m/A1v3EWfm7D7gwO0Hs8YmXtf3mN+rcTLgxCQoe2cRc3q6Bv+sTw6W9/885iQyfrmHwghM/lWmFHmdmbjpfLdxmIn8VqdCV3f+j05/Dfbc8WK8IIsOMW2Z2KHmHcNR6j1JM2o0MWELhkG9sFIiAgas63Lw2csK/HMlAM6yeND2oncSbwgbskxSJQ8KHkhAXxc4B03Q6h1QCuOYcZ+1cLRAtBM+CUyFPuvDGhGi52+42k+AyNia5KlNlah3HYr7tU7UWod0ObeCXeAO8M+ckAZkBXP3TkYHVD2c/bfivC3MGaBIhpT9oLziSywgN5zvoPaHwctJrx5ZR9VAG2mBydqp/GMjxyW2Ts98wKbadxbCGG1D2g5WkyKDxsQ79dcTUwHJCiuDk4UsEnWIeqdWtwb4WG1D2h9WzghP4R/s8L3PElzn7fzg5PcSxMQD6CxAkSXB6wgo5ZDEwmpqf6/fUCbqWgOwiXhxoKMCl+zbl6klk/uCRsvY+YglmIOvmQ5TBU6OuMFD/IBlcxRzZwuXosBQkCAEBAgBAQIAQFCQIAQECAEBAgBAUJAgP4PnxQTn2wb/QQAAAAASUVORK5CYII=" /><!-- --></p>
<p>The <code>blup_prep</code> function allows for masking the phenotypes
of some individuals before making the predictions, which allows for
cross-validation. All else being equal, predictions for individuals
without phenotypes, which is called marker-based selection, have lower
reliability than predictions for individuals with phenotypes, which is
called marker-assisted selection. To illustrate, we will mask the
phenotypes for the most recent cohort of breeding lines in the dataset
(which have names beginning with ‚ÄúW17‚Äù) and compare with the previous
prediction.</p>
<p>When masking, we only directly use training set data for the BLUP
calculation. However, finding the BLUPs requires the variances estimated
by the model, and the model was calibrated using all data (training set
+ test set). As a result, we are not performing a truly fair
cross-validation, but it is much faster than estimating new variances
for each cross-validation fold. If you still want to make this kind of
cross-validation, you can do it in sommer as seen in the following
code:</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb203-2"><a href="#cb203-2" tabindex="-1"></a>id <span class="ot">&lt;-</span> stage1.blues<span class="sc">$</span>id</span>
<span id="cb203-3"><a href="#cb203-3" tabindex="-1"></a>mask <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">id=</span><span class="fu">unique</span>(id[<span class="fu">substr</span>(id,<span class="dv">1</span>,<span class="dv">3</span>)<span class="sc">==</span><span class="st">&quot;W17&quot;</span>]))</span>
<span id="cb203-4"><a href="#cb203-4" tabindex="-1"></a><span class="fu">head</span>(mask)</span></code></pre></div>
<pre><code>##          id
## 1  W17037-1
## 2 W17037-10
## 3 W17037-11
## 4 W17037-12
## 5 W17037-13
## 6 W17037-15</code></pre>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" tabindex="-1"></a>prep3 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb205-2"><a href="#cb205-2" tabindex="-1"></a>              <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb205-3"><a href="#cb205-3" tabindex="-1"></a>              <span class="at">geno=</span>genoH,</span>
<span id="cb205-4"><a href="#cb205-4" tabindex="-1"></a>              <span class="at">vars=</span>ans2H<span class="sc">$</span>vars,</span>
<span id="cb205-5"><a href="#cb205-5" tabindex="-1"></a>              <span class="at">mask=</span>mask)</span>
<span id="cb205-6"><a href="#cb205-6" tabindex="-1"></a>GV3 <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep3, <span class="at">geno=</span>genoH, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb205-7"><a href="#cb205-7" tabindex="-1"></a></span>
<span id="cb205-8"><a href="#cb205-8" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(GV3, GV1, <span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb205-9"><a href="#cb205-9" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> plot.data[plot.data<span class="sc">$</span>id <span class="sc">%in%</span> mask<span class="sc">$</span>id,]</span>
<span id="cb205-10"><a href="#cb205-10" tabindex="-1"></a></span>
<span id="cb205-11"><a href="#cb205-11" tabindex="-1"></a><span class="fu">ggplot</span>(plot.data,<span class="fu">aes</span>(<span class="at">x=</span>r2.x,<span class="at">y=</span>r2.y)) <span class="sc">+</span></span>
<span id="cb205-12"><a href="#cb205-12" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb205-13"><a href="#cb205-13" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb205-14"><a href="#cb205-14" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Reliability&quot;</span>) <span class="sc">+</span></span>
<span id="cb205-15"><a href="#cb205-15" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;MBS&quot;</span>) <span class="sc">+</span> </span>
<span id="cb205-16"><a href="#cb205-16" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;MAS&quot;</span>) <span class="sc">+</span> </span>
<span id="cb205-17"><a href="#cb205-17" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="at">ratio=</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb205-18"><a href="#cb205-18" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>),<span class="at">y=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>)),</span>
<span id="cb205-19"><a href="#cb205-19" tabindex="-1"></a>            <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y),</span>
<span id="cb205-20"><a href="#cb205-20" tabindex="-1"></a>            <span class="at">linetype=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABFFBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZrYzMzM6AAA6ADo6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmAGZmOgBmOmZmOpBmZmZmkNtmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQOjqQOmaQZmaQkGaQkLaQttuQ2/+rbk2rbm6rbo6rjk2r5OSr5P+2ZgC2Zjq2Zma2kDq2tpC2ttu22/+2///Ijk3I///bkDrbtmbbtpDb29vb/7bb/9vb///kq27k///r6+v/tmb/trb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///9QidVoAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALhElEQVR4nO2dDXvktBHHlYNbti9QSApNr1BuA72jLdm2lDtKE/pyCXstTTg3oZvb6Pt/j0qy15JsyWNbL5bt+T/PBWdWkeQfo5F2VisTimoUGboDqQsBAUJAgBAQIAQECAEBQkCAEBAgJ0AbkuunL1Trbv3wiv+TlrvjReU18frtry5cWo8jL4DIA/VOWwPaPJg8oPwOd9+QQ8Wqw+GSgLTXZwOIAeCAfniUDzbpQS9/TMjBrwWgbEnefKF50H/WzPUOM7IqK0hSPgD975x70O2SD7aDZxJQMQJX7P7fWOqvlYBy58rYS4nKWwzarQ8eU/qv5UKB8OAF57ZggMh7V7u/k4UhBp2LH9VBmY58ADp494qDEKOE3bQSpP/9jz8tCQfEf2PALuqAMuFhyY4wD0Ps5fJnnEVWONPBMzmNi0EnAOXw5GsSEB9jCY8wHzEoI9w/6oDYwHrnz//87rgZEB9jCY8wL0G6iNGrwroHkFvyGLSgtiHGyH6c8AjzAujuOA/SX7KpXoHA0FzxuV/40iE1BGkxsnbrnyQ8wvysgzbs1vdj7FCBkFs4oHeW+7lOBUTEAnNDUl4wegHEULDRdMsWim88VhaKfOX45pcsxrAYxBaKb7+ovNW4e0QWV/kgTFcJvJvPF9OpanhAPzxKeA4bHhCL3yThOWx4QCx8vTtwF5o1NKDkhYAAISBACAgQAgKEgAC5AHprgvILqLx6VX+xt8lrZV2bREAIyK1JBISA3JpEQGEAvf7s6IPvxdX25Oj9b+u1zRzQ/fNTev0LfvX6t2f0umCFgEq9/t23dPspd5ztk+/Fb9XaZg5IYGG+o3qQWHa+mpz6Abr5YA9IiUYUPaiU9KDtJ2f05n0cYhXJGKT4Ep0MIOJhFntazGJT9CBCnQEVkYc70c3R0c/P6rWNGBDRbLiSxrcabk0iIATUxURqNgSkmkjdhoAUEzHYEBDGILcmERACamcilmIIKDdV7xoB6abaTSOgtk16BjR0gtS/0INwFoNMBAE1mgh6UKOJ4BBza3KsgAixL128NjlSQOJrZ77qnxCgkoofQKRFk6MCJLF4AUTspUYGSMNCNJND/fX8s6nYGACVWIgkhLOY2iME1GwiFfmuf2qAfNRPgFIjAmRyH+f6CVTKGVC5r+z6iOu0VpsvQBY+bvXb8s8mm+suV64bj7tc9enbhif5GCR3mFGqbDBzB6RiqNLR0CUOSO5RZCpcyWmXK7t78UNob6nJWx61g5x3uaoO1MODhF/UHcWEB5p5Wk6JrUr59CAZgXoAMnMA8TjEuPgx6PKptLsD4qWMfPwAMuZXg8xi+12u9P6rM2nvBIgHlarj0MCAoq2k5S5XNQR1ApTfdfFPW+BMAZBF3QHpVxY2dT4zAERaYbHx6XPrXd7EDQ8ov+1X+q9BAdnzq+kBKm/7VdXQlk93QN3e5Q4ISLvtohQEJ3DASQQQqS2RTcyMzjMLQMXNmu4cAYlqWwCqstr/keutd08DDAWoWBmWaZ/KqrD2rtUPoB7vcgeKQYXfqN1o4uMJEJhfTQRQ8c5UbcnIQxuDZD4xqA6oBZ+GysKaBloHybuu8SiMRj5TAtQyj1l1njybml+Uv/qQe0XDvBerAyqttUgFVtZgapdfTWeIFQ2YVL7gE1DL/GpigGoxWd/dokYquLJQpvCA6nepAZJXlbDd3NtopuCADOOkDqiWAdFNtb+MaBoSEI0DyG2TzOCAauk934AcN8nEikHEtB42AiqL+gHUKb863Cym+1EzoFrPmm1jj0FNgKgpBtVKGf+yXiyQKUYMolZAjV0zzHUTBLS/PWMMau6aOyAfaYBYgKB+hADkJQ2QPqDeMcjPLqJIMQjsx3RnMXl64v3zqZ1hptucd7lennrd5ToRQHKHmTzHVa9tEEA+9oBUba57FLdPvi6G2PBnuQbZBeu6y3V7cipwVXEP4EG986uBPWiCR5VSnzHocwRkkLLL9TKZITYoIP4AVPXZesou189SONPe9wfUXQFl4rnnC/AZ1UMBcsuvugParfdPJt40P79yIECO+VV3QHfHq9yLaNb8CN25xiABSDgPArIMsZUIQZQCjzmfKyDuPSIEQc8YHgBQmMx152n+PH/oN/SU8/iAAm2zcl0o2hQdUKhtVv0B/Q2DdAMgtpbGWcwOaEMI9Bz4GQPib8TIwviSoqiAQm6z6jzNE8ImsPOkAPnLr3paSdO0AHn8WMmDB2Ui/LQAFCIrPKzaxqDdOrkYFNLUaxbLcBYTalgHMTdKYR0U4fst/VfS/x0ekPf04cTei/lPH04MUBRTR0BiCsuVQgyKYOqxDgLmr2ptMwNERcoMZhQaULSNr71iEB9ow66D4n1Ls2+Q3q2HjEERv6XZ14PSy0kHMnUH1IIOnTGgrA0dOl9ACayDIn9L03klLbcB89Nuwz/rOWj6MAAgfRuwVChAYdOHAQDJLXjaMYqzjUE1ya2bbKwVx0kPvw04iJy3AX9ypngRelAhffOvjEMhAJExAtIO3A4LKEJ+Ncgstt8GzAfb/V/CTfMx8qvh1kHcidg6CL8O1VoICBACAuQX0IS/Fu6lt5M+WMBDb+PlV0cKyMdfIqBQJgTUusk0AZF2xcKZEgcUOb86OkCx86ujA+TvLxFQKFMwQEMnSP0rNQ8yHb86JQ/q2I+aaTZHlfbsmjn/jIAAEwICTAgIMCEgtV0EZDE1nF6KgGhz/hkBASYEBJgQEGCaPSDo+NW5AyJQqZkDgvPPYwQkd7nmOzprtc08Bim7XPlGYARUlbrDbPub3yOgqpQ9ivdf/fV5712uQY6n9SvXXa70+mn/GNTy8Vaj9iB21RuQMb86DUAyBvGvahwdPa3VNvMYpBx2i9O8UXKXKwLqovaAupyfMEdA9vwqAlLbQEBAPxAQ0A8EBPQDARna7H5+wrwAQfnVuQMC86tzB9THhIAAEwICTLMB1Pf8hCkBakpejiC/alA8D2qZX524B3XsBwIC+oGAgH4gILVOBGQF5Hg+7eQBdcqvzhGQqwkBASYEBJgmDcjH+QlTBuTlfNoJA+qRX50XIKgfCAjox3QByV2uN0cpPMo4XJOuu1zFOW/7nUIaIH/nJ4wQkH6OorxSAHk8n3aEgPSTOHMP0ne5jjO/apDzLle6PcFzFGvSPUheIaBCEc9y9V1ZrFlMPcsVPaguPMu1nxAQIAQECAEBQkCA3pqgvAJSUHks5bUy9yYREFAMAQHF/ACasBAQIAQECAEBQkCA3ABZPu6wltK+8movpj0Hz1rq/rmSWbCVyr+nbWtTVrY9sbToBMj2cYetFNWO/rAXu7RCrJS6KcE3NGktpRTjSa5rYzEnQLaPO+yl1KM/rMX05+DZSvGrVh1TnvxlLSayysYKnQCZPu5oLFUe/dFYTD4Hr6nU9snX1iGmdczWL63JMB5k+7jDWko7+sNaTH8OnrXUyam4Q6hjdgdSi6lxUpNHD7L1xHb0R2NltjikVdaiyYYIpFXG/p/cGKO0xxhkuyfb0R+Olb3+3A5IrevS0p5WTP80QpXjLGb+uMNWijZN87bn4Fkru7QPMaVUU8SXxQJ5kO3jDmupFuug9pWxK2DpJUrZQ5Ba7MbWJK6kASEgQAgIEAIChIAAISBAaQK6XT64YP+5O354tVsTrhU3b9jFwbO4XUkVkOBwu+SADtlVxgltOLUsZxVNqQL6iGPZfLQHxH/u1gLN+cOrmF1JFdDHH17R3R++0AEdDtCVVAE9/uMFvX37ZTnEuNtkhMRHlCqg1WZFs8OsDNIiaIsovYjblWQBZQt6vsr2HsTns/wleRVHyQK6+/C7X16UgGiWuxB/7UdRJ/pkAdFvvlhQDVBBBgHRHNCGReRMncXoOV8c7dZxg1C6gLinyCAt3GhTXsVTmoASEgIChIAAISBACAgQAgKEgAAhIEAICBACAvR/AzMNlTPY1CUAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" tabindex="-1"></a><span class="co">#Sommer</span></span>
<span id="cb206-2"><a href="#cb206-2" tabindex="-1"></a><span class="co">#Mask the data (replace BLUEs with NAs)</span></span>
<span id="cb206-3"><a href="#cb206-3" tabindex="-1"></a>sommerdf_mask <span class="ot">&lt;-</span> sommerdf_smaller</span>
<span id="cb206-4"><a href="#cb206-4" tabindex="-1"></a>sommerdf_mask<span class="sc">$</span>BLUEs[<span class="fu">which</span>(sommerdf_mask<span class="sc">$</span>id <span class="sc">%in%</span> mask<span class="sc">$</span>id)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb206-5"><a href="#cb206-5" tabindex="-1"></a></span>
<span id="cb206-6"><a href="#cb206-6" tabindex="-1"></a></span>
<span id="cb206-7"><a href="#cb206-7" tabindex="-1"></a><span class="co">#To calculate BLUPs with known variances, just fix all variance terms </span></span>
<span id="cb206-8"><a href="#cb206-8" tabindex="-1"></a><span class="co">#and perform a single REML iteration:</span></span>
<span id="cb206-9"><a href="#cb206-9" tabindex="-1"></a>ans2mask <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb206-10"><a href="#cb206-10" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb206-11"><a href="#cb206-11" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id,</span>
<span id="cb206-12"><a href="#cb206-12" tabindex="-1"></a>                                  <span class="at">theta =</span> <span class="fu">matrix</span>(ans2_open<span class="sc">$</span>sigma[[<span class="st">&quot;id:Hinv:isc:isc&quot;</span>]]<span class="sc">/</span></span>
<span id="cb206-13"><a href="#cb206-13" tabindex="-1"></a>                                         <span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb206-14"><a href="#cb206-14" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)), </span>
<span id="cb206-15"><a href="#cb206-15" tabindex="-1"></a>                                  <span class="at">Gu =</span> Hinv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb206-16"><a href="#cb206-16" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb206-17"><a href="#cb206-17" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id,</span>
<span id="cb206-18"><a href="#cb206-18" tabindex="-1"></a>                                  <span class="at">theta =</span> <span class="fu">matrix</span>(ans2_open<span class="sc">$</span>sigma[[<span class="st">&quot;id:Dinv:isc:isc&quot;</span>]]<span class="sc">/</span></span>
<span id="cb206-19"><a href="#cb206-19" tabindex="-1"></a>                                         <span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb206-20"><a href="#cb206-20" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)),</span>
<span id="cb206-21"><a href="#cb206-21" tabindex="-1"></a>                                  <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic</span></span>
<span id="cb206-22"><a href="#cb206-22" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb206-23"><a href="#cb206-23" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb206-24"><a href="#cb206-24" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb206-25"><a href="#cb206-25" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb206-26"><a href="#cb206-26" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb206-27"><a href="#cb206-27" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb206-28"><a href="#cb206-28" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb206-29"><a href="#cb206-29" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb206-30"><a href="#cb206-30" tabindex="-1"></a>                               <span class="at">theta =</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)), <span class="co">#initial </span></span>
<span id="cb206-31"><a href="#cb206-31" tabindex="-1"></a>                               <span class="co">#value for the variance.</span></span>
<span id="cb206-32"><a href="#cb206-32" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb206-33"><a href="#cb206-33" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb206-34"><a href="#cb206-34" tabindex="-1"></a>                           </span>
<span id="cb206-35"><a href="#cb206-35" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb206-36"><a href="#cb206-36" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb206-37"><a href="#cb206-37" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb206-38"><a href="#cb206-38" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb206-39"><a href="#cb206-39" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb206-40"><a href="#cb206-40" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb206-41"><a href="#cb206-41" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb206-42"><a href="#cb206-42" tabindex="-1"></a>                           </span>
<span id="cb206-43"><a href="#cb206-43" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb206-44"><a href="#cb206-44" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb206-45"><a href="#cb206-45" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb206-46"><a href="#cb206-46" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb206-47"><a href="#cb206-47" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb206-48"><a href="#cb206-48" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb206-49"><a href="#cb206-49" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb206-50"><a href="#cb206-50" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units,</span>
<span id="cb206-51"><a href="#cb206-51" tabindex="-1"></a>                                   <span class="at">theta =</span> <span class="fu">matrix</span>(ans2_open<span class="sc">$</span>sigma[[<span class="st">&quot;units:isc:isc&quot;</span>]]<span class="sc">/</span></span>
<span id="cb206-52"><a href="#cb206-52" tabindex="-1"></a>                                         <span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb206-53"><a href="#cb206-53" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>))), <span class="co">#residuals. isc(units) means</span></span>
<span id="cb206-54"><a href="#cb206-54" tabindex="-1"></a>                     <span class="co">#that we assume identity variance-covariance structure for the </span></span>
<span id="cb206-55"><a href="#cb206-55" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb206-56"><a href="#cb206-56" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_mask,</span>
<span id="cb206-57"><a href="#cb206-57" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span>,</span>
<span id="cb206-58"><a href="#cb206-58" tabindex="-1"></a>                     <span class="at">nIters=</span><span class="dv">1</span> <span class="co">#as we already know the variances, a single </span></span>
<span id="cb206-59"><a href="#cb206-59" tabindex="-1"></a>                    <span class="co">#REML iteration is enough</span></span>
<span id="cb206-60"><a href="#cb206-60" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Version out of date. Please update sommer to the newest version using:
## install.packages(&#39;sommer&#39;) in a new session
##  Use the &#39;dateWarning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;BLUEs&#39; 
## Adding additional levels of Gu in the model matrix of &#39;BLUEs&#39; 
## Adding additional levels of Gu in the model matrix of &#39;BLUEs&#39; 
## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -2989.72   13:12:8      23           0</code></pre>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1" tabindex="-1"></a><span class="co">#rename BLUP names to be shorter:</span></span>
<span id="cb208-2"><a href="#cb208-2" tabindex="-1"></a><span class="fu">names</span>(ans2mask<span class="sc">$</span>uList) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Additive&quot;</span>, <span class="st">&quot;Dominance&quot;</span>, <span class="st">&quot;Stage1Error&quot;</span>)</span>
<span id="cb208-3"><a href="#cb208-3" tabindex="-1"></a></span>
<span id="cb208-4"><a href="#cb208-4" tabindex="-1"></a></span>
<span id="cb208-5"><a href="#cb208-5" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb208-6"><a href="#cb208-6" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_mask)</span>
<span id="cb208-7"><a href="#cb208-7" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2mask<span class="sc">$</span>b[<span class="fu">unique</span>(sommerdf_mask<span class="sc">$</span>Env),]</span>
<span id="cb208-8"><a href="#cb208-8" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb208-9"><a href="#cb208-9" tabindex="-1"></a></span>
<span id="cb208-10"><a href="#cb208-10" tabindex="-1"></a></span>
<span id="cb208-11"><a href="#cb208-11" tabindex="-1"></a><span class="co">#heterosis</span></span>
<span id="cb208-12"><a href="#cb208-12" tabindex="-1"></a>XFg <span class="ot">&lt;-</span> geno[[<span class="fu">which.min</span>(result<span class="sc">$</span>aic)]]<span class="sc">@</span>Fg</span>
<span id="cb208-13"><a href="#cb208-13" tabindex="-1"></a>FgEffects <span class="ot">&lt;-</span> ans2mask<span class="sc">$</span>b[<span class="st">&quot;Fg&quot;</span>,]</span>
<span id="cb208-14"><a href="#cb208-14" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> XFg<span class="sc">*</span>FgEffects</span>
<span id="cb208-15"><a href="#cb208-15" tabindex="-1"></a><span class="co">#reorder heterosis effects to be in the same order as BLUPs</span></span>
<span id="cb208-16"><a href="#cb208-16" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> heterosis_effects[<span class="fu">match</span>(<span class="fu">rownames</span>(ans2mask<span class="sc">$</span>uList<span class="sc">$</span>Additive),</span>
<span id="cb208-17"><a href="#cb208-17" tabindex="-1"></a>                        <span class="fu">names</span>(heterosis_effects))]</span>
<span id="cb208-18"><a href="#cb208-18" tabindex="-1"></a></span>
<span id="cb208-19"><a href="#cb208-19" tabindex="-1"></a></span>
<span id="cb208-20"><a href="#cb208-20" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2mask<span class="sc">$</span>uList<span class="sc">$</span>Additive <span class="sc">+</span></span>
<span id="cb208-21"><a href="#cb208-21" tabindex="-1"></a>              ans2mask<span class="sc">$</span>uList<span class="sc">$</span>Dominance <span class="sc">+</span> </span>
<span id="cb208-22"><a href="#cb208-22" tabindex="-1"></a>              heterosis_effects <span class="sc">+</span></span>
<span id="cb208-23"><a href="#cb208-23" tabindex="-1"></a>              <span class="fu">mean</span>(Env_effects)</span>
<span id="cb208-24"><a href="#cb208-24" tabindex="-1"></a></span>
<span id="cb208-25"><a href="#cb208-25" tabindex="-1"></a><span class="co">#Extremely similar BLUPs after masking</span></span>
<span id="cb208-26"><a href="#cb208-26" tabindex="-1"></a><span class="fu">cor</span>(GV3<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb208-27"><a href="#cb208-27" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(GV3<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9999812</code></pre>
<p>Using the <code>mask</code> argument, one can also specify that
individuals are masked only in some environments, which is useful for
assessing the accuracy of prediction into new environments based on
phenotypes in other environments. This idea is revisited in Vignette
2.</p>
</div>
<div id="marker-effects-and-gwas" class="section level3">
<h3>Marker effects and GWAS</h3>
<p>Using what=‚ÄúAM‚Äù or ‚ÄúDM‚Äù leads to the prediction of additive or
dominance marker effects, respectively, in <code>blup</code>. This can
be a convenient way to save the results of a training set analysis to
predict future individuals. Multiplying the additive marker effects by
the matrix of (centered) marker dosages to obtain additive values is
equivalent (up to a constant) to directly predicting the additive values
using what=‚ÄúAV‚Äù, provided there has been no blending with the pedigree
relationship matrix:</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" tabindex="-1"></a><span class="co">#w = 0</span></span>
<span id="cb210-2"><a href="#cb210-2" tabindex="-1"></a>prep <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>stage1.blues,</span>
<span id="cb210-3"><a href="#cb210-3" tabindex="-1"></a>                  <span class="at">vcov=</span>stage1.EEV,</span>
<span id="cb210-4"><a href="#cb210-4" tabindex="-1"></a>                  <span class="at">geno=</span>geno[[<span class="dv">1</span>]],</span>
<span id="cb210-5"><a href="#cb210-5" tabindex="-1"></a>                  <span class="at">vars=</span>ans2[[<span class="dv">1</span>]]<span class="sc">$</span>vars)</span>
<span id="cb210-6"><a href="#cb210-6" tabindex="-1"></a></span>
<span id="cb210-7"><a href="#cb210-7" tabindex="-1"></a>marker.effects <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;AM&quot;</span>)</span>
<span id="cb210-8"><a href="#cb210-8" tabindex="-1"></a><span class="fu">head</span>(marker.effects)</span></code></pre></div>
<pre><code>##                marker chrom position       effect
## 1 solcap_snp_c2_51460 chr01   449027 -0.008699362
## 2 solcap_snp_c2_36608 chr01   508800  0.023544993
## 3 solcap_snp_c2_36615 chr01   510745  0.003188104
## 4 solcap_snp_c2_36658 chr01   527068  0.044723421
## 5 solcap_snp_c1_10930 chr01   566972  0.023206037
## 6       PotVar0120126 chr01   603013  0.001992423</code></pre>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="#cb212-1" tabindex="-1"></a>AV1 <span class="ot">&lt;-</span> <span class="fu">predict</span>(geno[[<span class="dv">1</span>]], marker.effects)</span>
<span id="cb212-2"><a href="#cb212-2" tabindex="-1"></a></span>
<span id="cb212-3"><a href="#cb212-3" tabindex="-1"></a><span class="co">#compare with G-BLUP</span></span>
<span id="cb212-4"><a href="#cb212-4" tabindex="-1"></a>AV2 <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;AV&quot;</span>)</span>
<span id="cb212-5"><a href="#cb212-5" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(AV1, AV2, <span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb212-6"><a href="#cb212-6" tabindex="-1"></a><span class="fu">ggplot</span>(plot.data,<span class="fu">aes</span>(<span class="at">x=</span>value.x,<span class="at">y=</span>value.y)) <span class="sc">+</span> </span>
<span id="cb212-7"><a href="#cb212-7" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb212-8"><a href="#cb212-8" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;RR-BLUP&quot;</span>) <span class="sc">+</span> </span>
<span id="cb212-9"><a href="#cb212-9" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;G-BLUP&quot;</span>) <span class="sc">+</span> </span>
<span id="cb212-10"><a href="#cb212-10" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAolBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6kNtNTU1NTW5NTY5NbqtNjshmAABmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQ2/+rbk2rbm6rbo6rjk2ryKur5OSr5P+2ZgC2///Ijk3I///bkDrb///kq27k///r6+v/tmb/yI7/25D/5Kv//7b//8j//9v//+T///+ysajdAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAI/UlEQVR4nO2da3vTOBCFDWXbsED3QveWUvZC6A1KC43//1/byLET2ZY043QkjdE5H4Kf+kSWXnQZTY2oaiioKncFtAuACAEQIQAiBECEAIjQREA/lKMDAQXvfuYUMRMTABEmACJMAESYAIgwARBhAiDCBECECYBcpqra8aAB3S6MlvW33xavPg2/JlAZhaaq2hPi9aAvrz6tL5b17evuBwDU07ffP9Tf/riqH365Gnzt6ZXRaJoMyPSch7efGlDtRvXzd60Nn+6SA6jhshllLSCba/AvIqyZmDiADBurB9lfk62MRhMH0OXZ5qOkOcgWA9D6vek364uzYlYxWwxA7cAqKA6yhUiaMAEQYQIgwgRAhAmACBMANWr3XgDkUbc7BSCPAAiADjc1aDAHeU12cgyAHAIgwgRAlMniA0AHmACIMAEQYSoYkD35+EsqF1Bv+fKXVCygCoBCpqoCoJDJwweAWvn4ANBWPjwAtJW3/wBQowAfAKpD48tdUmmAgnwAKDi+3CWVBYjgUzwgio8koJyvWB6qFs+0LxXUg6ju4y6pHEAMPkUD4vApGRCLT8GAeHzKBcTkUzqgg0oqARAZH4ZKKgAQn0+ZgLr4+cCSvntAFQAFTbvxBUBO7ecfAHLJmp8ByCF7/QKgsXrrOwCN1I9/AGioQXwIQAMN42cA6mu0vwCgnsb7LwDqabw/BSBLrv07AO3lzG8A0E7u/A8AdfLkxwColS9/CEBbefOrANTIn38GIKNAfh6A6snvZ7JM3xOgEB8Aon7/FRHQ+mLx8oP+E6jCfGICulzO4SROgk9EQOb4u1r9KXgUn4iAHt7+Y4aY7pM4Oz7yJTMA/bQ0dHSfxEn1n7g9qCGj+iROmk/MOeivhozmOYjBRxLQqqqef7Qsl80Q03sSJzk/P+lxY0Cro5v62ia0CYB+vNIbB/H4yAF6PD9tP0LSA4jJRw7Q1zcNoJNwSWoAcfmUCojNp3BAER83b0B8PpKAum7bX+zdX5OtzETTbn8R83Ezzgft5h8Acmo/P6cFtBtigQGmAZC1fmXqQXfP3vlLyg7IXt9zDbHVC39JuQH14p9cgO70rmL9+BCARurHPwA01CA+xBzU12j/hVWsp/H+NFMcFOKTEZBj/45I2pIrv4FJei9n/geAdnLnxwCokyd/CECtfPlVAGrlyx8C0Fbe/GqmOEhbytXLB3FQIz8fADIK8EkOaFVVJ/fHVRXYqyYHFOKTGtDq6ObxvDoN/24sMaAgn8SAmvcWmgXs7ujGW1JaQGE+yVexHSAtqxjBRycg+bclvWr5JHyiLf09iOo/SntQlMq4TDSfsiNpBp+iA0UOn5IB+RIckR43BdB1IAiqUwHi8SkXEJNPsYC4fEoFxOZT6CTN51MmoAl8kgMysfSqypsP6vhEbTvLNAZ0f3zSvNnR7Dl8igxo1380AmreejEfoXk6LqD9+FIIaNtxDKBsm1Vr/lELyCgXIHt+Vghon4rONMR665dCQLt/bZhpku6v7xoB1avm3bL74yzL/CD+UQmoNr8UC7+BFw3QMD7UCYihSIBG8TMA9TTeXwCQLcf+C4AsufanALSXc/8OQDu58xsA1MmT/wGg7vGe/BgAtU/35Q8BaPtwb34VgJpn+/PPAFSH8/MARPz+YhaAbhcLcwRVpBOognzmAehyaT4jncQZ5jMLQOv3zdmAcU7BI/jMAtBmaC0WyzgncXZ8ZEqLIhrQw88fTC+KcRIn1X/m0YMaXS4jnMRJ85kTIPk5iMFnFoDM2Fr/fSV+EieHzywAmTgowonkLD7zAOTSkwHx+JQLiMmnWEBcPqUCYvMpFBCfT5mAJvApEtBu/8V5Tn5TckC7/pO/7SxTakD78ZW/7SxTYkDW/JO/7SxTWkD2/Jy/7SxTUkC99St/21mmlID663v+trNMCQEN4p/8bWeZ0gEaxof5284yJQM0ip/zt51lSgVovL/I33aWKREgx/4rf9tZpjSAXPvT/G1nmZIAcu7f87edZUoJiKwMXVJ6UwpA7vxP/razTAkAefJj+dvOMkUH5M0f5m87yxQbkD+/mr/tLFNkQIH8c/62s0xxAYXy8/nbzjIlAcStjEZTCkDsymg0xQQU6j8a2s4yxQMUmn88ldFoigaIwKOh7SzToYCol0O7/pPiRdSoitSDqPFVa+gcLFMcQAw+CtrOMkUFNLUyGk1RAHH4KGg7yxQDEIuPgrazTPEAHVAZjSZ5QLz+o6HtLJM4IC4fBW1nmeIAOrAyGk0ARJhkARk21Wzez2SZRAF1vSd/s+RMAESYAIgwyQFq55+nVEajSQyQvXrlb5acSQpQBUDOr3WqAMj9tfbLg/1F/mbJmSQAjfKH+ZslZxIE9PTKaDQJABrv3/M3S84kBkiiMhpNTwbkyv/kb5ac6amAnPmx/M2SM4kAkqqMRhMAESaJOUisMhpNuQ43mY0JgAgTABEmACJMAESYWIDMKZyxjirVbmIBul0sYx1Vqt7EAfTw65/LSEeV6jcxAK3f/3cxOqq0HNGAbs/M8OofVRpWuH9NkZ6S/IA2XWfd60HxK6OxJD8gc972YnHWn4PiVkZjSeQy3z+qtDxNjIPK08RIujwBECEAIiQIqFnrJGYssVlPokZygL6Y/4Sjv3M7TBJlyNVIDNDly383f19ToiafJMqQq5HwEJsSd3vLEShDrkbCgKbs3HySKEOuRhKALheL1zV6EKs6muYgmRoJA5LYucnt/iRqhDiIECJpQgBECIAIARAhACIEQIQUAHo8377veNK7tG+dmqvtz76+OTV/3D3/2N579i5u7VQAatp+f3xqX1q37qpTJ6DmJ9fPP0atnR5A9eqFfWndMp9eQO0PogmACOkBZA0kc2nduju68QNalTDEqv1UvLu0b20YeCfpyHx0ADox8/JJ/3J/awPl6Mbbg2JLC6DtsLIur81q3w25PY7H8wbQ9R5ZZKkB1Ewmw3llDKhebR37CT2y9AC6P35hX1q37FVs272smCm29ADaDKoT67K9NYixmylqGz6XA0i3AIgQABECIEIARAiACAEQIQAiBECEAIjQ/xP3T934CzqtAAAAAElFTkSuQmCC" /><!-- --></p>
<p>This equivalency also holds for genotypic values (A + D):</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" tabindex="-1"></a>DM <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;DM&quot;</span>)</span>
<span id="cb213-2"><a href="#cb213-2" tabindex="-1"></a>DV1 <span class="ot">&lt;-</span> <span class="fu">predict</span>(geno[[<span class="dv">1</span>]], DM)</span>
<span id="cb213-3"><a href="#cb213-3" tabindex="-1"></a>GEGV1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">id=</span>AV1<span class="sc">$</span>id, <span class="at">value=</span>AV1<span class="sc">$</span>value <span class="sc">+</span> DV1<span class="sc">$</span>value)</span>
<span id="cb213-4"><a href="#cb213-4" tabindex="-1"></a></span>
<span id="cb213-5"><a href="#cb213-5" tabindex="-1"></a>GEGV2 <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb213-6"><a href="#cb213-6" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(GEGV1,GEGV2,<span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb213-7"><a href="#cb213-7" tabindex="-1"></a></span>
<span id="cb213-8"><a href="#cb213-8" tabindex="-1"></a><span class="fu">ggplot</span>(plot.data,<span class="fu">aes</span>(<span class="at">x=</span>value.x,<span class="at">y=</span>value.y)) <span class="sc">+</span> </span>
<span id="cb213-9"><a href="#cb213-9" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb213-10"><a href="#cb213-10" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;RR-BLUP&quot;</span>) <span class="sc">+</span> </span>
<span id="cb213-11"><a href="#cb213-11" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;GD-BLUP&quot;</span>) <span class="sc">+</span> </span>
<span id="cb213-12"><a href="#cb213-12" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAmVBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6kNtNTU1NTW5NTY5NbqtNjshmAABmtv9uTU1uTW5uTY5ubqtuq+SOTU2OTW6OTY6ObquOyP+QOgCQ2/+rbk2rbm6rbo6rjk2ryKur5P+2ZgC2///Ijk3I///bkDrb///kq27k///r6+v/tmb/yI7/25D/5Kv//7b//8j//9v//+T///9KocHCAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAI3klEQVR4nO2dC3ubNhSGWZPFbrsmXdNtzrqtcZxL68S58P9/3CwMSIgjHUCAJOv7nj0pj79IYm91OToYNcshqzLfNxC6AIgRADECIEYAxAiAGPUE9Gs6Ggio/dFPy69P4c3UIAAxHgAxHgAxHgAxHgAxHgAxHgAxHgBRXpbVPDoA2l0sP9zm+cvX5ccferHx7qm3N2GDWSYJ8YBe/rjOHz7+ePu2yh9+qz4EIKndlx/5y5+3+//y3e+3WrHR7qm/FwygsgcVnPaX5Ub151Frz6e67DAHHSafx48VIJVrr7+Ycb1gVrHd5+v88cOt7EFqsUnuqZsXDKCy66Q0B6nq3IPevl0ms4qp6jAHPS6X769TiIPE1Nz6EJF0LXVxlwKgWgDEeADEeZiDhngAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHgAxHjJAqJyPwAkPTo7BkC1B0BWL8smB+Txwbm7Sj6dfjfJHmToPmS5BAEZxxdZLj1AGQDZPRsfALLjASCOT/KArPMPXS4pQCyfxAFV8WGvcgkBqroPABm8angBkMGrph8AMnnl9AxAo3ppAGos7QDU8prBDwDpXgZAVk8PnwGo6bW2FwDU8Nr7LwDKCUCDGzx2QMT+HYAUj8pvAJD0yPwPAFWeIT82NqCHpdAqvhfqTPnDKXrQY4QHCxjzqxMAEm87x/ZSrzn/PAEg0XMiO1ig4uNcURdABZe4DhawPb8YvwcJNnlUBwtYn++MD2hzuf8R0xxkf/41OqC3f0S/iehgAeb54OiAyoEVTRzEPT9NPZJmny8nDojlkzggnk/agOr4cLQGjwtQ3X8AiDTl+AIgylTmHwAiTHV+BqC22Vi/AKhlNtf3SQGts+zdje3O8gABafHPlIDWJ/f5HUcoNEB6fDghoNer8/KHTYEBasXPEwJ6/lQAWthuLTRA7f0FAKkmsf8CIMWk9qcAJE1y/z4poKpF22IfDiA6v4FAsTIN+R8AKk1TfmyOIWaNFQMBZMwfztGDtr/8ba4oDEDm/OosQ2x9aq4oiGfzdX51SlkAbQNfxcz9Z6YeFDggGx8AGvf5u80bOgdNeU9dPDsfrGIMn1niIBufUADN0GCUkTTXf1KfpHk+AFTEh7M0GCGgqv8AEOnJ8QVAlKfMPwBEeOr87AlQ0CnXxvrlvQfZ5BfQfA1GBqgZ/3gDtM6yxdNZlln2qn4AafGhL0Drk/vXq+zc/mzMByA9fvYEqPjeQrGAbU/ujRV5ANTaX3hbxWpAQa1i7f0XAKkesT8FIMWj9u8AJD0yv4FIuvbo/E8wgeLbN8//EK0hPxYMoM3K73vzpvyhX0B3dRAkXlbNPb6zakywhgJo9+U/McR8vTdf8ZmtQU0dAF2sBB1P781bEvTh9KCCjJ/35i18gpmkX/4qyHiZg2x8PALaFjdVf9d+UwwxH+/NW/l4A/R6VQSIz5+qQbYPgD7c+oiD7Hy8AapmH+/5IIaP13xQoTu/+SCOj9fNaiG/m1UOT+qA2P6TOKAOfJIGVOGxlUs5H5SFDKibJgUkxxcAUZ4y/wAQ4anzc5iAxDS9znw9em6sX0ECejpbFF8il8sZockANdf3IAEVX7AXP3xsNbT4J0RAh44jAHmIg/T4MFxAQvMDasXPIQKSWY7Zh1h7fxEioPpgk9knaWL/FSSgfF28xvJ0NvMyT+1PwwSUi+/f2V/2mQAQuX8PFFAHjQ6Izm8AUOUZ8j8AVHqm/BgAHTxj/jBiQGN+P6DiM2adoymAHmTJP0fcg1zblbLwAaDc4fsJiQCy8gEgl+8nJAGI4ZM8II5P6oBYPokD4vmkDaiOn3uWG9zgMM8boLr/ABDpyfEFQJSnzD8ARHjq/AxAba+xfgFQy2uu7wCke1r8A0Cap8eHANT0WvEzADW89v4CgFSP2H8BkOJR+1MAkh65fweg2qPzG/EDelguxRt1zi/UGfI/8QParMRP54MFTPmx6AEd/qln54MFjPnD6AHth9ZyuXI9WKDOH8YmHtDu87XoRW4HC1jyz9H3oEKbldPBArb8/LEAcpmDbHziByTG1tu/ty4HC9j4xA9IxEFuByxZ+RwBIEo9AFnHl6Uc5x0LII5P6oBYPokD4vkAUBE/97qnbt4xAKr7DwCRdcvxBUBU3cr0A0BE3RkA2etW1y8AatXdXN8BSK9bi38ASK9biw8BSKtbj58BqFF3e3sBQGrdxPYLgJS6qf0pAMm6yf07AOU6IJd76uZFCojO/wBQVbchPQZAZd2m/GGCgMjn2FX+8Jg0Zg8y558T7EFE3Zb8PADlNaAR7qmbFymgMe6pmxcfoNz8/AuAerfr7gEQ4wEQ4wEQ48UDaMJnX0cBqFrdAYgqpuzfAYgqpuy/AIgqpiSAAIgqlsv9BQBRxXK5vwAgqtjgdt09AGI8AGI8AGI8AGI8AGI8AGI8AGK8wAGlo2GAKGYzl5u5QQBiBECM3AEduQCIEQAxAiBGAMTIFVBxakXj7IFO6l/CpbndhTjHZlijjoAexRlVzTOYuqh/CZfmxNksu8/Xwxp1A7R5/33/V9o8/6SL+pdwae5RUNGPaemqMYZY8wSdTqV6l3BqLi960bBGxwDUPIOpi/qXcGpOjOnLgY0OBrRZLkXPjaMHvXy9HNroGD1otjloaHO7i9XgRscA1DyDqYv6l3Bp7sBnYKMpxEHixEwRCPmIg45fAMQIgBgBECMAYgRAjAIA9Hp1+D7fonGpWufi6vDZ86dz8cf23U3p2f91c3cFAaj4f386O1cvFWubnZOAik/u3t1MenfhAMrXp+qlYomfRkDlB5MJgBiFA0gZSOJSsbYn92ZA6xSGWCan4vpStfYMjJP0xHzCALQQ8/KieSmtPZSTe2MPmlqhADoMK+XyTqz21ZCTOF6vCkB3EtnECgZQMZno80obUL4+/Iac0CdWOICezk7VS8VSV7FD91JipqkVDqD9oFool6WlxdjFFHUIn9MBFLYAiBEAMQIgRgDECIAYARAjAGIEQIwAiNH/jM+H2JJJSBcAAAAASUVORK5CYII=" /><!-- --></p>
<p>The marker effects can be standardized to compute GWAS -log10(p)
scores that are equivalent to the traditional fixed effect method (<a href="http://www.biomedcentral.com/1471-2105/15/246">Duarte et
al.¬†(2014)</a>; <a href="https://doi.org/10.1111/age.12378">Bernal Rubio
et al.¬†2016</a>). This calculation is easily parallelized, and the
argument <code>gwas.ncore</code> specifies how many cores to use (the
default is 0, which skips computing the GWAS scores). The
<code>gwas_threshold</code> command computes the -log10(p) threshold for
QTL discovery based on an effective number of markers (<a href="https://doi.org/10.1002/gepi.20331">Moskvina and Schmidt,
2008</a>), and <code>manhattan_plot</code> displays the result. For
yield there were no significant QTL, so results for the vine maturity
trait are shown instead.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1" tabindex="-1"></a>effects <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">name=</span><span class="st">&quot;block&quot;</span>,<span class="at">fixed=</span><span class="cn">FALSE</span>,<span class="at">factor=</span><span class="cn">TRUE</span>)</span>
<span id="cb214-2"><a href="#cb214-2" tabindex="-1"></a>ans1vm <span class="ot">&lt;-</span> <span class="fu">Stage1</span>(<span class="at">filename=</span>pheno1a.file,<span class="at">traits=</span><span class="st">&quot;vine.maturity&quot;</span>,</span>
<span id="cb214-3"><a href="#cb214-3" tabindex="-1"></a>                <span class="at">effects=</span>effects,<span class="at">solver=</span><span class="st">&quot;asreml&quot;</span>)</span>
<span id="cb214-4"><a href="#cb214-4" tabindex="-1"></a>ans2vm <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>ans1vm<span class="sc">$</span>blues, <span class="at">vcov=</span>ans1vm<span class="sc">$</span>vcov, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], </span>
<span id="cb214-5"><a href="#cb214-5" tabindex="-1"></a>                 <span class="at">non.add=</span><span class="st">&quot;dom&quot;</span>)</span>
<span id="cb214-6"><a href="#cb214-6" tabindex="-1"></a></span>
<span id="cb214-7"><a href="#cb214-7" tabindex="-1"></a>prep <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(ans1vm<span class="sc">$</span>blues, ans1vm<span class="sc">$</span>vcov, geno[[<span class="dv">1</span>]], ans2vm<span class="sc">$</span>vars)</span>
<span id="cb214-8"><a href="#cb214-8" tabindex="-1"></a>gwas.ans <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep, geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;AM&quot;</span>, <span class="at">gwas.ncore=</span><span class="dv">2</span>)</span>
<span id="cb214-9"><a href="#cb214-9" tabindex="-1"></a><span class="fu">head</span>(gwas.ans)</span></code></pre></div>
<pre><code>##                marker chrom position        effect     score
## 1 solcap_snp_c2_51460 chr01   449027 -0.0009276395 0.2699447
## 2 solcap_snp_c2_36608 chr01   508800  0.0030442837 0.7265478
## 3 solcap_snp_c2_36615 chr01   510745  0.0010404326 0.3693143
## 4 solcap_snp_c2_36658 chr01   527068  0.0022755583 0.5325289
## 5 solcap_snp_c1_10930 chr01   566972  0.0013878767 0.2930203
## 6       PotVar0120126 chr01   603013  0.0023887891 0.5896128</code></pre>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" tabindex="-1"></a><span class="fu">gwas_threshold</span>(geno[[<span class="dv">1</span>]], <span class="at">alpha=</span><span class="fl">0.05</span>, <span class="at">n.core=</span><span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] 5.103417</code></pre>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1" tabindex="-1"></a><span class="co">#Sommer alternative</span></span>
<span id="cb218-2"><a href="#cb218-2" tabindex="-1"></a><span class="co">#Sommer function GWAS() is based on the mmer() solver, which uses the </span></span>
<span id="cb218-3"><a href="#cb218-3" tabindex="-1"></a><span class="co">#Direct-Inversion Newton-Raphson or average information algorithm for variance </span></span>
<span id="cb218-4"><a href="#cb218-4" tabindex="-1"></a><span class="co">#estimation. It works fine, but it&#39;s not exactly equivalent to ASReml. That&#39;s the reason </span></span>
<span id="cb218-5"><a href="#cb218-5" tabindex="-1"></a><span class="co">#why in the previous examples we have been using mmec() solver, which uses the </span></span>
<span id="cb218-6"><a href="#cb218-6" tabindex="-1"></a><span class="co">#Henderson mixed model equations and the Average Information algorithm and it&#39;s </span></span>
<span id="cb218-7"><a href="#cb218-7" tabindex="-1"></a><span class="co">#equivalent to ASReml. </span></span>
<span id="cb218-8"><a href="#cb218-8" tabindex="-1"></a></span>
<span id="cb218-9"><a href="#cb218-9" tabindex="-1"></a><span class="co">#There is no option to use GWAS() with the mmec() solver, but it&#39;s possible to</span></span>
<span id="cb218-10"><a href="#cb218-10" tabindex="-1"></a><span class="co">#perform a regular GBLUP with mmec() to estimate the variances and subsequently</span></span>
<span id="cb218-11"><a href="#cb218-11" tabindex="-1"></a><span class="co">#fixing these variances in GWAS() function, which is equivalent to ASReml</span></span>
<span id="cb218-12"><a href="#cb218-12" tabindex="-1"></a></span>
<span id="cb218-13"><a href="#cb218-13" tabindex="-1"></a></span>
<span id="cb218-14"><a href="#cb218-14" tabindex="-1"></a></span>
<span id="cb218-15"><a href="#cb218-15" tabindex="-1"></a><span class="co">#Repeat stage1 for vine.maturity trait</span></span>
<span id="cb218-16"><a href="#cb218-16" tabindex="-1"></a><span class="fu">library</span>(lme4)</span>
<span id="cb218-17"><a href="#cb218-17" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb218-18"><a href="#cb218-18" tabindex="-1"></a>BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb218-19"><a href="#cb218-19" tabindex="-1"></a>EEVs <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb218-20"><a href="#cb218-20" tabindex="-1"></a>Envs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb218-21"><a href="#cb218-21" tabindex="-1"></a>H2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb218-22"><a href="#cb218-22" tabindex="-1"></a>residuals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb218-23"><a href="#cb218-23" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)) {</span>
<span id="cb218-24"><a href="#cb218-24" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> pheno1a[pheno1a<span class="sc">$</span>env <span class="sc">==</span> enviro,]</span>
<span id="cb218-25"><a href="#cb218-25" tabindex="-1"></a>  <span class="co">#1) remove entries with missing values for yield</span></span>
<span id="cb218-26"><a href="#cb218-26" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>vine.maturity),]</span>
<span id="cb218-27"><a href="#cb218-27" tabindex="-1"></a>  </span>
<span id="cb218-28"><a href="#cb218-28" tabindex="-1"></a>  Envs <span class="ot">&lt;-</span> <span class="fu">c</span>(Envs, <span class="fu">rep</span>(enviro, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))))</span>
<span id="cb218-29"><a href="#cb218-29" tabindex="-1"></a>  </span>
<span id="cb218-30"><a href="#cb218-30" tabindex="-1"></a>  phenoenv<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>id)</span>
<span id="cb218-31"><a href="#cb218-31" tabindex="-1"></a>  phenoenv<span class="sc">$</span>block <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>block) <span class="co">#cofactor</span></span>
<span id="cb218-32"><a href="#cb218-32" tabindex="-1"></a>  <span class="co">#Important to center!</span></span>
<span id="cb218-33"><a href="#cb218-33" tabindex="-1"></a>  <span class="co">#If you don&#39;t center the data, EEV calculation will be massively off!!!</span></span>
<span id="cb218-34"><a href="#cb218-34" tabindex="-1"></a>  </span>
<span id="cb218-35"><a href="#cb218-35" tabindex="-1"></a>  <span class="co">#remove the intercept in the model!</span></span>
<span id="cb218-36"><a href="#cb218-36" tabindex="-1"></a>  <span class="co">#It&#39;s important for correct EEV calculation</span></span>
<span id="cb218-37"><a href="#cb218-37" tabindex="-1"></a>  </span>
<span id="cb218-38"><a href="#cb218-38" tabindex="-1"></a>  <span class="co">#Make sure the first fixed effect in the formula is id!</span></span>
<span id="cb218-39"><a href="#cb218-39" tabindex="-1"></a>  models[[enviro]] <span class="ot">&lt;-</span> <span class="fu">lmer</span>(vine.maturity <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span> id <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>block) ,</span>
<span id="cb218-40"><a href="#cb218-40" tabindex="-1"></a>                        <span class="at">data=</span>phenoenv)</span>
<span id="cb218-41"><a href="#cb218-41" tabindex="-1"></a>  <span class="co">#If you get this message:</span></span>
<span id="cb218-42"><a href="#cb218-42" tabindex="-1"></a>  <span class="co">#boundary (singular) fit: see help(&#39;isSingular&#39;)</span></span>
<span id="cb218-43"><a href="#cb218-43" tabindex="-1"></a>  <span class="co">#It means that the estimated variance for the random effect was 0. It&#39;s fine</span></span>
<span id="cb218-44"><a href="#cb218-44" tabindex="-1"></a>  <span class="co">#because we are interested in the fixed effects of id.</span></span>
<span id="cb218-45"><a href="#cb218-45" tabindex="-1"></a>  </span>
<span id="cb218-46"><a href="#cb218-46" tabindex="-1"></a></span>
<span id="cb218-47"><a href="#cb218-47" tabindex="-1"></a>  <span class="co">#extract H2</span></span>
<span id="cb218-48"><a href="#cb218-48" tabindex="-1"></a>  <span class="co">#random effect variances</span></span>
<span id="cb218-49"><a href="#cb218-49" tabindex="-1"></a>  variance_block <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>varcor<span class="sc">$</span>block)</span>
<span id="cb218-50"><a href="#cb218-50" tabindex="-1"></a>  variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models[[enviro]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb218-51"><a href="#cb218-51" tabindex="-1"></a>  </span>
<span id="cb218-52"><a href="#cb218-52" tabindex="-1"></a>  <span class="co">#fixed effect variances</span></span>
<span id="cb218-53"><a href="#cb218-53" tabindex="-1"></a>  <span class="co">#1) calculate design matrix</span></span>
<span id="cb218-54"><a href="#cb218-54" tabindex="-1"></a>  X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(vine.maturity<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoenv) <span class="co">#remove intercept.</span></span>
<span id="cb218-55"><a href="#cb218-55" tabindex="-1"></a></span>
<span id="cb218-56"><a href="#cb218-56" tabindex="-1"></a>  <span class="co">#calculate fixed effects for each plot</span></span>
<span id="cb218-57"><a href="#cb218-57" tabindex="-1"></a>  <span class="co">#id effects:</span></span>
<span id="cb218-58"><a href="#cb218-58" tabindex="-1"></a>  <span class="co"># y = Xb + ... + epsilon</span></span>
<span id="cb218-59"><a href="#cb218-59" tabindex="-1"></a>  <span class="co"># b (vector of the genotypic fixed effects) can be retrieved from:</span></span>
<span id="cb218-60"><a href="#cb218-60" tabindex="-1"></a>  <span class="co">#&quot;models[[enviro]]@beta[1:length(unique(phenoenv$id))]&quot;</span></span>
<span id="cb218-61"><a href="#cb218-61" tabindex="-1"></a>  <span class="co"># X is the design matrix linking each obervation (each plot in the field) with</span></span>
<span id="cb218-62"><a href="#cb218-62" tabindex="-1"></a>  <span class="co"># its corresponding genotypic effect.</span></span>
<span id="cb218-63"><a href="#cb218-63" tabindex="-1"></a>  id_effects <span class="ot">&lt;-</span> X.id<span class="sc">%*%</span>models[[enviro]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))]</span>
<span id="cb218-64"><a href="#cb218-64" tabindex="-1"></a></span>
<span id="cb218-65"><a href="#cb218-65" tabindex="-1"></a>  <span class="co">#calculate variance for the fixed effects</span></span>
<span id="cb218-66"><a href="#cb218-66" tabindex="-1"></a>  id_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(id_effects)</span>
<span id="cb218-67"><a href="#cb218-67" tabindex="-1"></a>  </span>
<span id="cb218-68"><a href="#cb218-68" tabindex="-1"></a>  <span class="co">#Plot level heritability</span></span>
<span id="cb218-69"><a href="#cb218-69" tabindex="-1"></a>  H2 <span class="ot">&lt;-</span> <span class="fu">c</span>(H2, id_variance<span class="sc">/</span>(id_variance <span class="sc">+</span></span>
<span id="cb218-70"><a href="#cb218-70" tabindex="-1"></a>                       variance_block <span class="sc">+</span> </span>
<span id="cb218-71"><a href="#cb218-71" tabindex="-1"></a>                       variance_residual) )</span>
<span id="cb218-72"><a href="#cb218-72" tabindex="-1"></a>  </span>
<span id="cb218-73"><a href="#cb218-73" tabindex="-1"></a>  </span>
<span id="cb218-74"><a href="#cb218-74" tabindex="-1"></a>  <span class="co">#extract residuals</span></span>
<span id="cb218-75"><a href="#cb218-75" tabindex="-1"></a>  resid <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>residuals)</span>
<span id="cb218-76"><a href="#cb218-76" tabindex="-1"></a>  expt <span class="ot">&lt;-</span> <span class="fu">rep</span>(enviro, <span class="fu">length</span>(resid))</span>
<span id="cb218-77"><a href="#cb218-77" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">cbind</span>(expt, resid))</span>
<span id="cb218-78"><a href="#cb218-78" tabindex="-1"></a>  residuals <span class="ot">&lt;-</span> <span class="fu">rbind</span>(residuals, tmp)</span>
<span id="cb218-79"><a href="#cb218-79" tabindex="-1"></a>  </span>
<span id="cb218-80"><a href="#cb218-80" tabindex="-1"></a>  </span>
<span id="cb218-81"><a href="#cb218-81" tabindex="-1"></a>  <span class="co">#calculate EEV (estimation error variance-covariance matrix for id effects)</span></span>
<span id="cb218-82"><a href="#cb218-82" tabindex="-1"></a>  EEV_lmer <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>x, </span>
<span id="cb218-83"><a href="#cb218-83" tabindex="-1"></a>                     <span class="at">nrow =</span> <span class="fu">sqrt</span>(<span class="fu">length</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>x))) </span>
<span id="cb218-84"><a href="#cb218-84" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">1</span>]]</span>
<span id="cb218-85"><a href="#cb218-85" tabindex="-1"></a>  <span class="fu">colnames</span>(EEV_lmer) <span class="ot">&lt;-</span> <span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>vcov<span class="sc">@</span>Dimnames[[<span class="dv">2</span>]]</span>
<span id="cb218-86"><a href="#cb218-86" tabindex="-1"></a>  <span class="co">#remove EEV entries for all fixed effects other than id</span></span>
<span id="cb218-87"><a href="#cb218-87" tabindex="-1"></a>  <span class="co">#in this case, id is the only fixed effect. Nothing to remove</span></span>
<span id="cb218-88"><a href="#cb218-88" tabindex="-1"></a>  </span>
<span id="cb218-89"><a href="#cb218-89" tabindex="-1"></a>  <span class="co">#keep only the BLUEs for the id fixed effect</span></span>
<span id="cb218-90"><a href="#cb218-90" tabindex="-1"></a>  placeholder <span class="ot">&lt;-</span> <span class="fu">names</span>(BLUEs)</span>
<span id="cb218-91"><a href="#cb218-91" tabindex="-1"></a>  BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>(BLUEs, models[[enviro]]<span class="sc">@</span>beta[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))])</span>
<span id="cb218-92"><a href="#cb218-92" tabindex="-1"></a>  <span class="fu">names</span>(BLUEs) <span class="ot">&lt;-</span> <span class="fu">c</span>(placeholder, <span class="fu">levels</span>(models[[enviro]]<span class="sc">@</span>frame<span class="sc">$</span>id))</span>
<span id="cb218-93"><a href="#cb218-93" tabindex="-1"></a>  </span>
<span id="cb218-94"><a href="#cb218-94" tabindex="-1"></a>  EEVs[[enviro]] <span class="ot">&lt;-</span> EEV_lmer</span>
<span id="cb218-95"><a href="#cb218-95" tabindex="-1"></a>}</span>
<span id="cb218-96"><a href="#cb218-96" tabindex="-1"></a><span class="fu">colnames</span>(residuals) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;resid&quot;</span>)</span>
<span id="cb218-97"><a href="#cb218-97" tabindex="-1"></a>residuals<span class="sc">$</span>resid <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(residuals<span class="sc">$</span>resid)</span>
<span id="cb218-98"><a href="#cb218-98" tabindex="-1"></a></span>
<span id="cb218-99"><a href="#cb218-99" tabindex="-1"></a></span>
<span id="cb218-100"><a href="#cb218-100" tabindex="-1"></a></span>
<span id="cb218-101"><a href="#cb218-101" tabindex="-1"></a><span class="co">#almost exactly the same BLUEs</span></span>
<span id="cb218-102"><a href="#cb218-102" tabindex="-1"></a><span class="fu">cor</span>(BLUEs, <span class="co">#lmer</span></span>
<span id="cb218-103"><a href="#cb218-103" tabindex="-1"></a>    ans1vm<span class="sc">$</span>blues<span class="sc">$</span>BLUE) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 0.9972429</code></pre>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1" tabindex="-1"></a>BLUEs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##  A00188-3C  A01143-3C  A09037-6C AAF07847-2 AC01144-1W 
##   7.936207   8.436207   5.127587   8.127587   5.936207</code></pre>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="#cb222-1" tabindex="-1"></a>ans1vm<span class="sc">$</span>blues<span class="sc">$</span>BLUE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 7.936207 8.436207 5.127587 8.127587 5.936207</code></pre>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" tabindex="-1"></a><span class="co">#Almost exactly the same EEV</span></span>
<span id="cb224-2"><a href="#cb224-2" tabindex="-1"></a>differences <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb224-3"><a href="#cb224-3" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)) {</span>
<span id="cb224-4"><a href="#cb224-4" tabindex="-1"></a>  differences <span class="ot">&lt;-</span> <span class="fu">c</span>(differences, <span class="fu">mean</span>(<span class="fu">abs</span>(EEVs[[enviro]] <span class="sc">-</span> ans1vm<span class="sc">$</span>EEV[[enviro]])))</span>
<span id="cb224-5"><a href="#cb224-5" tabindex="-1"></a>}</span>
<span id="cb224-6"><a href="#cb224-6" tabindex="-1"></a>differences</span></code></pre></div>
<pre><code>## [1] NaN NaN NaN NaN NaN</code></pre>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="#cb226-1" tabindex="-1"></a>EEVs<span class="sc">$</span>Hancock15[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##              idA00188-3C idA01143-3C idA09037-6C idAAF07847-2 idAC01144-1W
## idA00188-3C   0.62994543  0.12284961  0.05815088   0.05815088   0.12284961
## idA01143-3C   0.12284961  0.62994543  0.05815088   0.05815088   0.12284961
## idA09037-6C   0.05815088  0.05815088  1.20173998   0.18754833   0.05815088
## idAAF07847-2  0.05815088  0.05815088  0.18754833   1.20173998   0.05815088
## idAC01144-1W  0.12284961  0.12284961  0.05815088   0.05815088   0.62994543</code></pre>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" tabindex="-1"></a>ans1vm<span class="sc">$</span>EEV<span class="sc">$</span>Hancock15[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" tabindex="-1"></a><span class="co">#merge EEV matrices from each environment for their use in the second stage</span></span>
<span id="cb230-2"><a href="#cb230-2" tabindex="-1"></a><span class="co">#It is done using the direct sum operator</span></span>
<span id="cb230-3"><a href="#cb230-3" tabindex="-1"></a><span class="fu">library</span>(matrixcalc)</span>
<span id="cb230-4"><a href="#cb230-4" tabindex="-1"></a>placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>()</span>
<span id="cb230-5"><a href="#cb230-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(pheno1a<span class="sc">$</span>env))) {</span>
<span id="cb230-6"><a href="#cb230-6" tabindex="-1"></a>  env1 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)[i<span class="dv">-1</span>]</span>
<span id="cb230-7"><a href="#cb230-7" tabindex="-1"></a>  env2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno1a<span class="sc">$</span>env)[i]</span>
<span id="cb230-8"><a href="#cb230-8" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(<span class="fu">is.na</span>(placeholder_matrix)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb230-9"><a href="#cb230-9" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEVs[[env1]],EEVs[[env2]])</span>
<span id="cb230-10"><a href="#cb230-10" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb230-11"><a href="#cb230-11" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(placeholder_matrix,EEVs[[env2]])</span>
<span id="cb230-12"><a href="#cb230-12" tabindex="-1"></a>  }</span>
<span id="cb230-13"><a href="#cb230-13" tabindex="-1"></a>}</span>
<span id="cb230-14"><a href="#cb230-14" tabindex="-1"></a>EEV_full <span class="ot">&lt;-</span> placeholder_matrix</span>
<span id="cb230-15"><a href="#cb230-15" tabindex="-1"></a><span class="fu">dim</span>(EEV_full)</span></code></pre></div>
<pre><code>## [1] 1335 1335</code></pre>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="#cb232-1" tabindex="-1"></a><span class="fu">length</span>(BLUEs)</span></code></pre></div>
<pre><code>## [1] 1335</code></pre>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="#cb234-1" tabindex="-1"></a><span class="co">#Stage 2 GWAS</span></span>
<span id="cb234-2"><a href="#cb234-2" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb234-3"><a href="#cb234-3" tabindex="-1"></a><span class="co">#1) Prepare data</span></span>
<span id="cb234-4"><a href="#cb234-4" tabindex="-1"></a>Stage1Output <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">BLUEs =</span> BLUEs,</span>
<span id="cb234-5"><a href="#cb234-5" tabindex="-1"></a>                       <span class="at">id =</span> <span class="fu">names</span>(BLUEs),</span>
<span id="cb234-6"><a href="#cb234-6" tabindex="-1"></a>                       <span class="at">Env =</span> Envs)</span>
<span id="cb234-7"><a href="#cb234-7" tabindex="-1"></a></span>
<span id="cb234-8"><a href="#cb234-8" tabindex="-1"></a></span>
<span id="cb234-9"><a href="#cb234-9" tabindex="-1"></a><span class="co">#We need to add to Stage1Output a column for stage 1 estimation error </span></span>
<span id="cb234-10"><a href="#cb234-10" tabindex="-1"></a>Stage1Error <span class="ot">&lt;-</span> <span class="fu">paste0</span>(Stage1Output<span class="sc">$</span>id,<span class="st">&quot;:&quot;</span>,Stage1Output<span class="sc">$</span>Env)</span>
<span id="cb234-11"><a href="#cb234-11" tabindex="-1"></a><span class="fu">rownames</span>(EEV_full) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb234-12"><a href="#cb234-12" tabindex="-1"></a><span class="fu">colnames</span>(EEV_full) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb234-13"><a href="#cb234-13" tabindex="-1"></a></span>
<span id="cb234-14"><a href="#cb234-14" tabindex="-1"></a><span class="co">#sommer input dataframe</span></span>
<span id="cb234-15"><a href="#cb234-15" tabindex="-1"></a>sommerdf <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Stage1Output,</span>
<span id="cb234-16"><a href="#cb234-16" tabindex="-1"></a>                  Stage1Error)</span>
<span id="cb234-17"><a href="#cb234-17" tabindex="-1"></a><span class="fu">head</span>(sommerdf)</span></code></pre></div>
<pre><code>##      BLUEs          id       Env           Stage1Error
## 1 7.936207   A00188-3C Hancock15   A00188-3C:Hancock15
## 2 8.436207   A01143-3C Hancock15   A01143-3C:Hancock15
## 3 5.127587   A09037-6C Hancock15   A09037-6C:Hancock15
## 4 8.127587  AAF07847-2 Hancock15  AAF07847-2:Hancock15
## 5 5.936207  AC01144-1W Hancock15  AC01144-1W:Hancock15
## 6 5.936207 Accumulator Hancock15 Accumulator:Hancock15</code></pre>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" tabindex="-1"></a><span class="co">#remove non-genotyped individuals</span></span>
<span id="cb236-2"><a href="#cb236-2" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> sommerdf[<span class="sc">-</span><span class="fu">which</span>(<span class="sc">!</span>(sommerdf<span class="sc">$</span>id <span class="sc">%in%</span> </span>
<span id="cb236-3"><a href="#cb236-3" tabindex="-1"></a>                                      <span class="fu">rownames</span>(<span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>G)))),]</span>
<span id="cb236-4"><a href="#cb236-4" tabindex="-1"></a>EEV_smaller <span class="ot">&lt;-</span> EEV_full[sommerdf_smaller<span class="sc">$</span>Stage1Error,</span>
<span id="cb236-5"><a href="#cb236-5" tabindex="-1"></a>                          sommerdf_smaller<span class="sc">$</span>Stage1Error]</span>
<span id="cb236-6"><a href="#cb236-6" tabindex="-1"></a>stage1errorvar2 <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(EEV_smaller)) <span class="sc">-</span> <span class="fu">mean</span>(EEV_smaller)</span>
<span id="cb236-7"><a href="#cb236-7" tabindex="-1"></a></span>
<span id="cb236-8"><a href="#cb236-8" tabindex="-1"></a><span class="co">#heterosis coefficients</span></span>
<span id="cb236-9"><a href="#cb236-9" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> geno[[<span class="dv">1</span>]]<span class="sc">@</span>Fg</span>
<span id="cb236-10"><a href="#cb236-10" tabindex="-1"></a></span>
<span id="cb236-11"><a href="#cb236-11" tabindex="-1"></a><span class="co">#include Fg into sommerdf:</span></span>
<span id="cb236-12"><a href="#cb236-12" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> Fg[<span class="fu">match</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))]</span>
<span id="cb236-13"><a href="#cb236-13" tabindex="-1"></a><span class="fu">identical</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_smaller, Fg)</span>
<span id="cb238-2"><a href="#cb238-2" tabindex="-1"></a></span>
<span id="cb238-3"><a href="#cb238-3" tabindex="-1"></a></span>
<span id="cb238-4"><a href="#cb238-4" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs))</span>
<span id="cb238-5"><a href="#cb238-5" tabindex="-1"></a></span>
<span id="cb238-6"><a href="#cb238-6" tabindex="-1"></a></span>
<span id="cb238-7"><a href="#cb238-7" tabindex="-1"></a><span class="co">#2) Find variance components with mmec() solver:</span></span>
<span id="cb238-8"><a href="#cb238-8" tabindex="-1"></a>Dinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>D)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb238-9"><a href="#cb238-9" tabindex="-1"></a>Hinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>G)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb238-10"><a href="#cb238-10" tabindex="-1"></a><span class="co">#Scaling factors for the variance components:</span></span>
<span id="cb238-11"><a href="#cb238-11" tabindex="-1"></a>meanH <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>G))<span class="sc">-</span><span class="fu">mean</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>G)</span>
<span id="cb238-12"><a href="#cb238-12" tabindex="-1"></a>meanD <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>D))<span class="sc">-</span><span class="fu">mean</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>D)</span>
<span id="cb238-13"><a href="#cb238-13" tabindex="-1"></a></span>
<span id="cb238-14"><a href="#cb238-14" tabindex="-1"></a></span>
<span id="cb238-15"><a href="#cb238-15" tabindex="-1"></a>EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(EEV_smaller)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="#cb239-1" tabindex="-1"></a>modelGWASvar <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb239-2"><a href="#cb239-2" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb239-3"><a href="#cb239-3" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Hinv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb239-4"><a href="#cb239-4" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb239-5"><a href="#cb239-5" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic</span></span>
<span id="cb239-6"><a href="#cb239-6" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb239-7"><a href="#cb239-7" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb239-8"><a href="#cb239-8" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb239-9"><a href="#cb239-9" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb239-10"><a href="#cb239-10" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb239-11"><a href="#cb239-11" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb239-12"><a href="#cb239-12" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb239-13"><a href="#cb239-13" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb239-14"><a href="#cb239-14" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb239-15"><a href="#cb239-15" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb239-16"><a href="#cb239-16" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb239-17"><a href="#cb239-17" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb239-18"><a href="#cb239-18" tabindex="-1"></a>                           </span>
<span id="cb239-19"><a href="#cb239-19" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb239-20"><a href="#cb239-20" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb239-21"><a href="#cb239-21" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb239-22"><a href="#cb239-22" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb239-23"><a href="#cb239-23" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb239-24"><a href="#cb239-24" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb239-25"><a href="#cb239-25" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb239-26"><a href="#cb239-26" tabindex="-1"></a>                           </span>
<span id="cb239-27"><a href="#cb239-27" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb239-28"><a href="#cb239-28" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb239-29"><a href="#cb239-29" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb239-30"><a href="#cb239-30" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb239-31"><a href="#cb239-31" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb239-32"><a href="#cb239-32" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb239-33"><a href="#cb239-33" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb239-34"><a href="#cb239-34" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb239-35"><a href="#cb239-35" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb239-36"><a href="#cb239-36" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb239-37"><a href="#cb239-37" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb239-38"><a href="#cb239-38" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb239-39"><a href="#cb239-39" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="#cb240-1" tabindex="-1"></a><span class="co">#extract variances</span></span>
<span id="cb240-2"><a href="#cb240-2" tabindex="-1"></a>Addtive_var <span class="ot">&lt;-</span>  modelGWASvar<span class="sc">$</span>sigma[<span class="fu">names</span>(modelGWASvar<span class="sc">$</span>sigma)<span class="sc">==</span><span class="st">&quot;id:Hinv:isc:isc&quot;</span>]</span>
<span id="cb240-3"><a href="#cb240-3" tabindex="-1"></a>Scaled_additive <span class="ot">&lt;-</span> Addtive_var<span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs) <span class="co">#sommer convention.</span></span>
<span id="cb240-4"><a href="#cb240-4" tabindex="-1"></a><span class="co">#variances have to be scaled this way</span></span>
<span id="cb240-5"><a href="#cb240-5" tabindex="-1"></a></span>
<span id="cb240-6"><a href="#cb240-6" tabindex="-1"></a>Dominance_var <span class="ot">&lt;-</span>  modelGWASvar<span class="sc">$</span>sigma[<span class="fu">names</span>(modelGWASvar<span class="sc">$</span>sigma)<span class="sc">==</span><span class="st">&quot;id:Dinv:isc:isc&quot;</span>]</span>
<span id="cb240-7"><a href="#cb240-7" tabindex="-1"></a>Scaled_dominance <span class="ot">&lt;-</span> Dominance_var<span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs) <span class="co">#sommer convention.</span></span>
<span id="cb240-8"><a href="#cb240-8" tabindex="-1"></a><span class="co">#variances have to be scaled this way</span></span>
<span id="cb240-9"><a href="#cb240-9" tabindex="-1"></a></span>
<span id="cb240-10"><a href="#cb240-10" tabindex="-1"></a>Residual_var <span class="ot">&lt;-</span>  modelGWASvar<span class="sc">$</span>sigma[<span class="fu">names</span>(modelGWASvar<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>]</span>
<span id="cb240-11"><a href="#cb240-11" tabindex="-1"></a>Scaled_residual <span class="ot">&lt;-</span> Residual_var<span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs) <span class="co">#sommer convention.</span></span>
<span id="cb240-12"><a href="#cb240-12" tabindex="-1"></a><span class="co">#variances have to be scaled this way</span></span>
<span id="cb240-13"><a href="#cb240-13" tabindex="-1"></a></span>
<span id="cb240-14"><a href="#cb240-14" tabindex="-1"></a></span>
<span id="cb240-15"><a href="#cb240-15" tabindex="-1"></a></span>
<span id="cb240-16"><a href="#cb240-16" tabindex="-1"></a></span>
<span id="cb240-17"><a href="#cb240-17" tabindex="-1"></a><span class="co">#GWAS data preparation</span></span>
<span id="cb240-18"><a href="#cb240-18" tabindex="-1"></a><span class="co">#marker matrix, centered</span></span>
<span id="cb240-19"><a href="#cb240-19" tabindex="-1"></a>Mc <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>coeff)</span>
<span id="cb240-20"><a href="#cb240-20" tabindex="-1"></a></span>
<span id="cb240-21"><a href="#cb240-21" tabindex="-1"></a><span class="co">#duplicate column for id, as we cannot use the same column for additive </span></span>
<span id="cb240-22"><a href="#cb240-22" tabindex="-1"></a><span class="co">#and dominance effects in mmer() solver</span></span>
<span id="cb240-23"><a href="#cb240-23" tabindex="-1"></a><span class="co">#Each effect must be associated to its own column in the dataframe</span></span>
<span id="cb240-24"><a href="#cb240-24" tabindex="-1"></a>dom_id <span class="ot">&lt;-</span> sommerdf_smaller<span class="sc">$</span>id</span>
<span id="cb240-25"><a href="#cb240-25" tabindex="-1"></a>sommerdf_smaller_GWAS <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_smaller, dom_id)</span>
<span id="cb240-26"><a href="#cb240-26" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller_GWAS<span class="sc">$</span>BLUEs))</span></code></pre></div>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="#cb241-1" tabindex="-1"></a><span class="co">#3) Perform GWAS using the variances found by mmec()</span></span>
<span id="cb241-2"><a href="#cb241-2" tabindex="-1"></a>GWASsommer_additive <span class="ot">&lt;-</span> <span class="fu">GWAS</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span> , <span class="co">#fixed effects</span></span>
<span id="cb241-3"><a href="#cb241-3" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span> <span class="fu">vsr</span>(id, <span class="co">#additive effect</span></span>
<span id="cb241-4"><a href="#cb241-4" tabindex="-1"></a>                           <span class="at">Gu=</span><span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>G), <span class="co">#known additive relationship</span></span>
<span id="cb241-5"><a href="#cb241-5" tabindex="-1"></a>                           <span class="co">#matrix</span></span>
<span id="cb241-6"><a href="#cb241-6" tabindex="-1"></a>                           <span class="at">Gti=</span><span class="fu">matrix</span>(Scaled_additive), <span class="co">#known variance component</span></span>
<span id="cb241-7"><a href="#cb241-7" tabindex="-1"></a>                           <span class="co">#for additive effect (estimated by mmec())</span></span>
<span id="cb241-8"><a href="#cb241-8" tabindex="-1"></a>                           <span class="at">Gtc=</span><span class="fu">matrix</span>(<span class="dv">3</span>)) <span class="sc">+</span> <span class="co">#tell mmer() not to estimate the </span></span>
<span id="cb241-9"><a href="#cb241-9" tabindex="-1"></a>                           <span class="co">#variance component and instead use the one privided </span></span>
<span id="cb241-10"><a href="#cb241-10" tabindex="-1"></a>                           <span class="co">#in Gti argument</span></span>
<span id="cb241-11"><a href="#cb241-11" tabindex="-1"></a>                       <span class="co">#for all other random effects and the residual we fix</span></span>
<span id="cb241-12"><a href="#cb241-12" tabindex="-1"></a>                       <span class="co">#their variance as shown for the additive effect.</span></span>
<span id="cb241-13"><a href="#cb241-13" tabindex="-1"></a>                       <span class="fu">vsr</span>(dom_id, <span class="co">#dominance effect</span></span>
<span id="cb241-14"><a href="#cb241-14" tabindex="-1"></a>                           <span class="at">Gu=</span><span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>D),</span>
<span id="cb241-15"><a href="#cb241-15" tabindex="-1"></a>                           <span class="at">Gti=</span><span class="fu">matrix</span>(Scaled_dominance),</span>
<span id="cb241-16"><a href="#cb241-16" tabindex="-1"></a>                           <span class="at">Gtc=</span><span class="fu">matrix</span>(<span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb241-17"><a href="#cb241-17" tabindex="-1"></a>                       <span class="fu">vsr</span>(Stage1Error, <span class="co">#stage1error effect</span></span>
<span id="cb241-18"><a href="#cb241-18" tabindex="-1"></a>                           <span class="at">Gu=</span>EEV_smaller,</span>
<span id="cb241-19"><a href="#cb241-19" tabindex="-1"></a>                           <span class="at">Gti=</span>Variance_scaled, </span>
<span id="cb241-20"><a href="#cb241-20" tabindex="-1"></a>                           <span class="at">Gtc=</span><span class="fu">matrix</span>(<span class="dv">3</span>)),</span>
<span id="cb241-21"><a href="#cb241-21" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsr</span>(units, </span>
<span id="cb241-22"><a href="#cb241-22" tabindex="-1"></a>                           <span class="at">Gti=</span><span class="fu">matrix</span>(Scaled_residual),</span>
<span id="cb241-23"><a href="#cb241-23" tabindex="-1"></a>                           <span class="at">Gtc=</span><span class="fu">matrix</span>(<span class="dv">3</span>)),</span>
<span id="cb241-24"><a href="#cb241-24" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller_GWAS,</span>
<span id="cb241-25"><a href="#cb241-25" tabindex="-1"></a>                     <span class="at">gTerm =</span> <span class="st">&quot;u:id&quot;</span>, <span class="co">#focus on additive effects (&quot;id&quot; column</span></span>
<span id="cb241-26"><a href="#cb241-26" tabindex="-1"></a>                     <span class="co">#in the dataframe) for the GWAS score calculation</span></span>
<span id="cb241-27"><a href="#cb241-27" tabindex="-1"></a>                     <span class="at">M =</span> Mc, <span class="co">#provide centered marker matrix</span></span>
<span id="cb241-28"><a href="#cb241-28" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb241-29"><a href="#cb241-29" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="#cb242-1" tabindex="-1"></a><span class="co">#marker effects from GBLUP:</span></span>
<span id="cb242-2"><a href="#cb242-2" tabindex="-1"></a>MMT<span class="ot">&lt;-</span>Mc<span class="sc">%*%</span><span class="fu">t</span>(Mc) <span class="co">#MM&#39;=additive relationship matrix </span></span>
<span id="cb242-3"><a href="#cb242-3" tabindex="-1"></a>MMTinv<span class="ot">&lt;-</span><span class="fu">solve</span>(MMT) <span class="co">#inverseof MM&#39; </span></span>
<span id="cb242-4"><a href="#cb242-4" tabindex="-1"></a>MTMMTinv<span class="ot">&lt;-</span><span class="fu">t</span>(Mc)<span class="sc">%*%</span>MMTinv <span class="co">#M&#39;%*%(M&#39;M)</span></span>
<span id="cb242-5"><a href="#cb242-5" tabindex="-1"></a></span>
<span id="cb242-6"><a href="#cb242-6" tabindex="-1"></a>a.from.g<span class="ot">&lt;-</span>MTMMTinv<span class="sc">%*%</span><span class="fu">matrix</span>(modelGWASvar<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span>,<span class="at">ncol=</span><span class="dv">1</span>) </span>
<span id="cb242-7"><a href="#cb242-7" tabindex="-1"></a><span class="co">#Proving the forumula above:</span></span>
<span id="cb242-8"><a href="#cb242-8" tabindex="-1"></a><span class="co">#t(M)%*%solve(M%*%t(M))%*%BLUPs = Marker_effects</span></span>
<span id="cb242-9"><a href="#cb242-9" tabindex="-1"></a><span class="co">#Multiply both sides in the left by M matrix:</span></span>
<span id="cb242-10"><a href="#cb242-10" tabindex="-1"></a><span class="co">#(M%*%t(M))%*%solve(M%*%t(M))%*%BLUPs = M%*%Marker_effects</span></span>
<span id="cb242-11"><a href="#cb242-11" tabindex="-1"></a><span class="co">#in the left side, (M%*%t(M)) is multiplied by its inverse, they cancel out</span></span>
<span id="cb242-12"><a href="#cb242-12" tabindex="-1"></a><span class="co">#BLUPs = M%*%Marker_effects</span></span>
<span id="cb242-13"><a href="#cb242-13" tabindex="-1"></a><span class="co">#in the right side, M%*%Marker_effects are the BLUPs (for each individual, we</span></span>
<span id="cb242-14"><a href="#cb242-14" tabindex="-1"></a><span class="co">#are multiplying its marker coefficients by the corresponding marker effects </span></span>
<span id="cb242-15"><a href="#cb242-15" tabindex="-1"></a><span class="co">#and summing them)</span></span>
<span id="cb242-16"><a href="#cb242-16" tabindex="-1"></a><span class="co">#BLUPs = BLUPs</span></span>
<span id="cb242-17"><a href="#cb242-17" tabindex="-1"></a><span class="co">#The equality holds</span></span>
<span id="cb242-18"><a href="#cb242-18" tabindex="-1"></a></span>
<span id="cb242-19"><a href="#cb242-19" tabindex="-1"></a></span>
<span id="cb242-20"><a href="#cb242-20" tabindex="-1"></a></span>
<span id="cb242-21"><a href="#cb242-21" tabindex="-1"></a></span>
<span id="cb242-22"><a href="#cb242-22" tabindex="-1"></a><span class="fu">cor</span>(Mc<span class="sc">%*%</span>a.from.g, <span class="co">#BLUPs from marker effects</span></span>
<span id="cb242-23"><a href="#cb242-23" tabindex="-1"></a>    modelGWASvar<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span>) <span class="co">#BLUPs from model</span></span></code></pre></div>
<pre><code>##            isc
## [1,] 0.9999994</code></pre>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="#cb244-1" tabindex="-1"></a><span class="co">#you can do the same for dominance effects</span></span>
<span id="cb244-2"><a href="#cb244-2" tabindex="-1"></a><span class="co">#Just replace the additive relationship matrix with dominance coefficients</span></span>
<span id="cb244-3"><a href="#cb244-3" tabindex="-1"></a>Dc <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(geno[[<span class="dv">1</span>]]<span class="sc">@</span>coeff.D)</span>
<span id="cb244-4"><a href="#cb244-4" tabindex="-1"></a>DDT<span class="ot">&lt;-</span>Dc<span class="sc">%*%</span><span class="fu">t</span>(Dc) <span class="co">#DD&#39;=dominance relationship matrix </span></span>
<span id="cb244-5"><a href="#cb244-5" tabindex="-1"></a>DDTinv<span class="ot">&lt;-</span><span class="fu">solve</span>(DDT) <span class="co">#inverse of DD&#39; </span></span>
<span id="cb244-6"><a href="#cb244-6" tabindex="-1"></a>DTDDTinv<span class="ot">&lt;-</span><span class="fu">t</span>(Dc)<span class="sc">%*%</span>DDTinv <span class="co">#D&#39;%*%(D&#39;D)</span></span>
<span id="cb244-7"><a href="#cb244-7" tabindex="-1"></a>d.from.g<span class="ot">&lt;-</span>DTDDTinv<span class="sc">%*%</span><span class="fu">matrix</span>(modelGWASvar<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span>,<span class="at">ncol=</span><span class="dv">1</span>) </span>
<span id="cb244-8"><a href="#cb244-8" tabindex="-1"></a><span class="fu">cor</span>(Dc<span class="sc">%*%</span>d.from.g, <span class="co">#BLUPs from marker effects</span></span>
<span id="cb244-9"><a href="#cb244-9" tabindex="-1"></a>    modelGWASvar<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span>) <span class="co">#BLUPs from model</span></span></code></pre></div>
<pre><code>##      isc
## [1,]   1</code></pre>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="#cb246-1" tabindex="-1"></a><span class="co">#You can use StageWise function for finding the GWAS significance threshold</span></span>
<span id="cb246-2"><a href="#cb246-2" tabindex="-1"></a><span class="co">#It uses a Bonferroni-type correction based on an effective number of markers </span></span>
<span id="cb246-3"><a href="#cb246-3" tabindex="-1"></a><span class="co">#that accounts for LD. Described by:</span></span>
<span id="cb246-4"><a href="#cb246-4" tabindex="-1"></a><span class="co">#Moskvina V, Schmidt KM (2008) On multiple-testing correction in genome-wide </span></span>
<span id="cb246-5"><a href="#cb246-5" tabindex="-1"></a><span class="co">#association studies. Genetic Epidemiology 32:567-573. doi:10.1002/gepi.20331</span></span>
<span id="cb246-6"><a href="#cb246-6" tabindex="-1"></a>threshold <span class="ot">&lt;-</span> <span class="fu">gwas_threshold</span>(geno[[<span class="dv">1</span>]], <span class="at">alpha=</span><span class="fl">0.05</span>, <span class="at">n.core=</span><span class="dv">2</span>)</span>
<span id="cb246-7"><a href="#cb246-7" tabindex="-1"></a></span>
<span id="cb246-8"><a href="#cb246-8" tabindex="-1"></a><span class="co">#Get chromosome and position for each marker</span></span>
<span id="cb246-9"><a href="#cb246-9" tabindex="-1"></a>Mraw <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(geno.file,<span class="at">check.names=</span>F)</span>
<span id="cb246-10"><a href="#cb246-10" tabindex="-1"></a><span class="co">#remove markers eliminated by quality control by read_geno() function</span></span>
<span id="cb246-11"><a href="#cb246-11" tabindex="-1"></a>Mraw <span class="ot">&lt;-</span> Mraw[<span class="fu">which</span>(<span class="fu">colnames</span>(Mc)<span class="sc">%in%</span>Mraw<span class="sc">$</span>marker),]</span>
<span id="cb246-12"><a href="#cb246-12" tabindex="-1"></a>GWAS_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">marker =</span> Mraw[,<span class="dv">1</span>],</span>
<span id="cb246-13"><a href="#cb246-13" tabindex="-1"></a>                      <span class="at">chrom =</span> Mraw[,<span class="dv">2</span>],</span>
<span id="cb246-14"><a href="#cb246-14" tabindex="-1"></a>                      <span class="at">position =</span> Mraw[,<span class="dv">3</span>],</span>
<span id="cb246-15"><a href="#cb246-15" tabindex="-1"></a>                      <span class="at">score =</span> <span class="fu">c</span>(GWASsommer_additive<span class="sc">$</span>scores))</span>
<span id="cb246-16"><a href="#cb246-16" tabindex="-1"></a><span class="co">#very similar GWAS scores</span></span>
<span id="cb246-17"><a href="#cb246-17" tabindex="-1"></a><span class="fu">cor</span>(GWAS_df<span class="sc">$</span>score, <span class="co">#sommer</span></span>
<span id="cb246-18"><a href="#cb246-18" tabindex="-1"></a>    gwas.ans<span class="sc">$</span>score) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 0.9977387</code></pre>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="#cb248-1" tabindex="-1"></a><span class="co">#Extremely similar GWAS results with Sommer and StageWise</span></span>
<span id="cb248-2"><a href="#cb248-2" tabindex="-1"></a><span class="co">#Sommer</span></span>
<span id="cb248-3"><a href="#cb248-3" tabindex="-1"></a><span class="fu">manhattan_plot</span>(GWAS_df, <span class="at">thresh=</span>threshold, <span class="at">rotate.label=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABI1BMVEUAAAAAADoAAGYAOjoAOmYAOpAAZrYhkIwzMzM6AAA6AGY6OgA6Ojo6OmY6OpA6ZpA6ZrY6kNtEAVRNTU1NTW5NTY5Nbm5NbqtNjshmAABmADpmOgBmOjpmOmZmZgBmZmZmZpBmkJBmkLZmkNtmtttmtv9uTU1uTW5uTY5ubm5ubqtujo5uq+R/f3+OTU2OyP+QOgCQZjqQZmaQZraQkGaQtraQttuQ2/+rbk2r5P+2ZgC2Zjq2ZpC2kGa2kJC2tpC2tra2ttu229u22/+2/7a2/9u2///Ijk3I///bkDrbkGbbtmbbtpDbtrbb27bb29vb2//b///kq27k5P/k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///8ng05XAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOXElEQVR4nO2dfWPUuBHGZ9OEvS1QYHNtkyZ3UDjStxCSa0toaHscuSOlwIWkFzbkZfH3/xSVJcvWmz2W117L63n+yFpjWRr9Io1ke72GiFQoaNuB0AVtOxC6oG0HQhe07UDogrYdCF3QtgOhC2Y49ouFVh2AougsTZxZG5bpLC+rfWzOEdhniQ13yXZ2AuRMEiACVLJIAoQUSYCQIrsKaDweFx3Re0DjcUKIALkNBMiddWEBndUuxqf+QispzB5Es5g7KwFyOkWACBABKhYBQkSAEBEgRAQIEQFCRIAQESBEBAgRAUJEgBARIEQECBEBQkSAEBEgRARI0eUf/8P+Xu+sMq1lZRCgRNc7v44BXf7hH4qRAKWarK5yQBP+VymDAHFNVrcFmtM11UyAFAlAR39iIWhbFMDU9t3PJlUJ0PXOb39mlLYzyNSDpJTok24SIEUKoMutp2kZBEhKBSTnegKkiAMSfYeGmEvJLBZP8xSkXUq6zRGb5p8qZRCgIhEgRAQIEQFCRIAQESBEBAgRAUJEgBARIEQECBEBQkSAEBEgRAQIEQFCRIAQ0VPPiOiZVUT131nlgGovtZKoBzmToQ+x8GPQu29vjWBw68GrqIR6N4tNX4yA6Vb8Z/DwJMLUN0BvR3Dn2Qe+efHdJgy+iRD1C9B0V+8003+NVpBO1C9AV381cUz/+ToqVL8AVVA/AU3fv3yFx2dZRu8AfdqIpzD41WFUQj0EdD6CXzx8+bebsISEH1lG3wAdwA3ts1j9A3S1PtjnG+ejMl2oj4ASLulGofoHKHoDj/nnRxpiObPYm8Gjk/ikY+XV8fHxh6hY/QN0tQ6qsGHWP0DTb++rejD/c7Hgrwd5ia4oIuoXoKuvzNOLT1/NPQaFDIhN8XdURJ824a6SFM+sxo8efpk+ldm3GHS+AYMH3x+zWH38fm8Ey/vZLvnM6oTRmaSE+jeLvdtIp/jlZ+qO5JnVz8/jxzSO+vzU8/S7+5u3bj/4uz7By2dWxdM+pw0+7RP0ECuUAMSfFGvwcaigg3SxOBURfsTfRh7q7cit5wMRhu4VAor624MOxO2e6a56Rj+fIdaJGHT1u2R+P/+lslCcU5BGjggdUPPTfBd6UDbE1Durk7ksFLsRg9hJR06QZsOr2VONjgAqKwKEqK8xqLR6OouVV08BZVfu6aJ9lHOqUeamWNRjQNPdew6rrd4CKisChIgAIeofIPXO6u/x75L3D5B+b/4G9lXF/gGKpi8Gjz5EF29H8OinXcDmsh4Cyr4ftHJy3sYXyQM/F9O+YYZ/y6x/Z/MECBti6rdc2xhiwQNKvie9AYN9NqO1EKQDj0HpN+2XD9kowzpQH2cxpulPL78v+6xG23c/m1Q+oOOXx/QwS/4Q25NDrIR6CIhF5gE7DxvRwyz503w8vKb0MAv6MAs6hUX9BEQPs8gN9/WgXepBciPvbJ6Dme5SDMqZxdhC+vb9TTbP0yyWfEKkafqCr4Pulloq9vFcjOkCfU5MKaNmQEGfzV8cqyoDqV+APB+mS8roESDPh+mSMvoYg8qrj7NYrIv2gnQnAJU7zUjKIEBF6mkMahPQrLOY5EuA3J/p4QTI/TkfQNP3+Xc1mn4RbScAFanpF9F2IgYVqekX0YZ8qlFKxotonzx5En8+iT/Pkk+28US1e37GgGY5Xn6e5fl1pu8/yzuuGqCmX0TbkWdWc9X4i2g7P8S4+v7MKqoGX0S7GD2owRfRdh1Q4y+i7Tqgxl9E23lATb+ItvuANBEgRAQIEQFC1ElAegWtAeJuhAhI1tAyIOEGASJAxaob0HCIAnJHkO7FoEqAhjignP+/DyB3yZ0CNM4/ogZAYfSgSrOYN6AS9zBCjUFuU8kYpBHQMugjMM2sAcx1ZwEAWQTiPqVmwACN8T6lAWKbwQLSxmBOkB7GUo+sGVBh9pYB6VE8/uQ0EECRDsRq3yIDGpYFpA2QbgPSImbc1AJAQxegSPLRAKRjU2ZmuTRAzhk0QEBa9bz5VgzKOojCZ6xkMaK4E1DWz0SpCnvFhVxA+jRYM6DCW5O8dpkQgAwNM6sKyHW8YtIzjK2isyx8y6hXKUAvrZY7qyYgvAfJ/4/WgxJzMqyGWQxKe0g2CrUyHT1oXNCDEj5aHNMrMNcR8xhiov6xlAg+wk29mWMV0BAFFGd0AoqGKQRhSHPYgJI6ZAXZkfMDlPqoUvAExDuA6P5p4XqM0iNM2k6FaZJjqAFS5gEOMHWgPUDDDFDqu7HTBWh8lg0UBVA6y5mzGAJIdhId0NgEpPwT5wVomElx3uxBeh4FUDbEjCHoGGLDSKsDA6S6EClL1TnFoDKAhrbGlpLG6fvdgIYWIA2BBsiqoGlA4yxcRGd69UOzcfkdyAaUtM4qIwPk6IESkM4nAWSVn81iDQJS/nNKiNEJSYsJyOoB1tEFgNw9ULRYK8ENSOtejQOyYrAbkMM9DJAN4MzJR/knyH5cBtAcYpDiWB6fIkAyBrsB2QCNAeSow+imbQNS/PIAlHlnDiEzhz8gcx4tBUhbk4cDyBFjzBzOGFQMyPFfKBeDUkLNAMqF4PDebF0BIMc07ywieEB5/jmdnwFQXhHa0fY0b2XJkHQaUO4QLA1omAEaOo6P1EqUxi04IL0I7aJugQtd7EGV94/zRykBKpUlbEAO9wgQAfIA5PBtAQA5X0Q7L0B+rW8FkPv9YgRIKucNdQRIKucdh2X49CMG5bwlsy1A5VfS8wKU8yLaxQI0y63nvBfRliFUBpB9SR0BUB6Q83zO5YJsa51DrDyhyN4XH242MffOkNhp33uUANwOCB/N+9tuQrNdcqUX0SKi981javpFtJ0H1PSLaLsPSBMBQkSAEH2x0KoBUA421GTlsA9BjsDSuAVzKRXk7qkqAoSIAPVL0LYDoQvadiB0QeM1nMofjElPcvVL/0Lpz1pd72yLHKsy+xHbuNySh4jfUVN1tLq69vm58rM0PKH+TE1q0X7lsMBBRVC2nVV1ytp1vbMm62et2b7888/MJNwXv+2puM8BncYQkiycT3ydRaSvd/6yta1VwPIexb8OKtAmF2Xiyo5kg0+TX+5j3LVDXQ6agjogSFmtlRcAPj9nreA+x63hXeE06QiJz1oPStAkmLblBYQ0fbqa/Z4jr2CiFClq5D8DlVx3SCyyCN1Lw0FLUB8eXo9Zh+wpjAyrn6eMy25izGiAkgTPEh+StDBJJ0NwTalAFDZJumBKN6lD+iCzaF7qDtpNguo0XPr8fM00JP+9ozUOiP+7/hdlPSjexf7/uT2Id4NTvQdxTUQvze1Ba+mYMXqQ7qXuoN0iqIKhQBMlNAqXklpZx2Zbp1m/UaIBixEKIPmDw0mW651kHAgK2oH8YBmDZFPjoMK7qYgsvPyn0rtty0vdQUtQvu0VJb0THftUzEUGx8ut32xlBuZqfGEuDagTES/Wkp1mnLVmsTj/lwqfLDiaE6DDQVNQspkNizXwKZ6rDUHbDoQuqLc4x4pMN9kZLItp8E0XLwt9BTWUkcmxItNNdgbLYhp808XLQm/B7EVksuZT02RncEzBusE37XDCWL/qScfiVhP4Y8iXuSKzTHYGy2IafNMuJ4z5SU+6J69UULDPW4H2IGv9qietxa0mKNjnL2tFZprsDJbFNPimXU6Y6y49aS1uVUH+ripyrMh0k53BspgG33TxstBXUEMZCy2ouTwlArhNdgbLYhp803gdmEuKIH9XJSlziNtkZ7AspsE3jdeBuaQI8ndV08SeNHWTncGymAbfNF4H5lImyN1TTem6K6tTN9kZLItp8E3jdWAuKYI66SyioG0HQhfUXWCg5/OGBXMpE8xMxNCRXYtusjNYFtPgm8brwFzKBLl7qsm+ImqY7AyWxTT4pvE6MJcUQf6uSqJ1ECLHqTF27uy4VbQ2Uxqvw+N0HvJ3VdPllvXv0E12BstiGnzTeB2YS5kgd08FOa7OYZfvHPfmdYNv2uFEOFcUF1HQtgOhC2ovMf4Ch3E/QTfZGSyLafBN43VgLqWC4tb664hfx1Nu+5omO4NlMQ2+abwOzKVMULrl5STXFMoVBN1kZ7AspsE3jdeBuaQIyra8pNKvWKinQarJzmBZTINvGq8Dc0kRlG15WYl7Cpfq1+R0k53BspgG3zReB+ZSJijd8rKKv29pfEdTN9kZLItp8E3jdWAupYLi1pKgbQdCF7TtQOiCth0IXdC2A6EL2nYgdEELdb7bBBjcOWRb092l1y044COYe43TXeAa7BMgpw5g+VUUXbwAxoYA2foIKyd84wDuESCHDuCx2Di//SwG9OMeDB6x5JulHzZgaT+avhjB4O4JH34/vgC4G0332N8YarqLbwGPYsX52UB+eDKbvzDb4d7S+8x0d3Azjkf3YkBfA+tcVxs8QC3z4fd1vPnNbpIj28Uox4pLKsq/znetzEYIZmuvt67WVYdZwF45jN7ADQaIR23WdGb4tMssbN/SYfQWYsN/42Zmu67Wlw+jKe+LhflZP2K96N5MDsNMR/vLBBRT4bY3vCFX67yDxR+swY+zHEuvlV3no6VDWVxB/huiitm6EMxycAWZgPgwEYDi2HQ+4q2KDgb7oqkiR/xX28WGzp1XWH6xnIDZ5gGY5eAKUmLQ+xMEkGyqBYjHYRZ49ovyf4ROAspmMTEsqgFiln9vxvG3RP4ZBXUU4qPzkbUOUgBpMUVvsLJLlHS1oRlz8s8qqKEMP7GV9DO2kt7LVtIKIG1W0hus7PoYb0XvRtrU5s4fvR11K0i7zsVUQMniJYWn/FV3iSIel8nPR2R1QR1t9tP07U39bF4FpK2MXxt/9ZV03BGL8++l6+3qgtkOX3xB2w6ELmjbgdAFbTsQuqBtB0IXtO1A6IK2HQhd0LYDoQvadiB0QdsOhC5o24HQ9X9moKVmJjj4GQAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="#cb249-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb249-2"><a href="#cb249-2" tabindex="-1"></a><span class="fu">manhattan_plot</span>(gwas.ans, <span class="at">thresh=</span>threshold, <span class="at">rotate.label=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABI1BMVEUAAAAAADoAAGYAOjoAOmYAOpAAZrYhkIwzMzM6AAA6AGY6OgA6Ojo6OmY6OpA6ZpA6ZrY6kNtEAVRNTU1NTW5NTY5Nbm5NbqtNjshmAABmADpmOgBmOjpmOmZmZgBmZmZmZpBmkJBmkLZmkNtmtttmtv9uTU1uTW5uTY5ubm5ubqtujo5uq+R/f3+OTU2OyP+QOgCQZjqQZmaQZraQkGaQtraQttuQ2/+rbk2r5P+2ZgC2Zjq2ZpC2kGa2kJC2tpC2tra2ttu229u22/+2/7a2/9u2///Ijk3I///bkDrbkGbbtmbbtpDbtrbb27bb29vb2//b///kq27k5P/k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///8ng05XAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOhElEQVR4nO2d+WPTyBXHn9MErwsUcLZt0mQXCkt6hZBsW8KGtssmXdIU2BB3jUPiGP3/f0Xn0PHmkJ4OyxpZ8/3B1ry53nw8mhlJlgSBV6agaQdcFzTtgOuCph1wXdC0A64LmnbAdUHTDrguqJD3i6XWPAAFwTgOjI0NwzROS2rmTclBfefYsJdsJveArEEPaLGAxkss34OsQb+LtR3QcDjMytF5QMNhSMgDshs8IHtSD8jqlB+DSgAicnhARA4PiMjhARE5PCAihwdE5PCAiBweEJHDAyJyeEBEDg+IyOEBETk8ICKHB0Tk8ICIHA4Duv7jf9jnzd4600ZShgcU6mbv1xzQ9R++Q0YPKNZkfV0AmohPVIYHJDRZ35VoRhvY7AEhSUCnf2JD0K4sgKnpy8N1qhSgm73f/swo7SaQfQ+KhEafeNMDQkKArneex2V4QJEwoGiu94CQBCDZd/wuZlM4i/Fp3g/SNoXd5pRN889RGR5QljwgQh4QIQ+IkAdEyAMi5AER8oAIeUCEPCBCHhAhD4iQB0TIAyLkARHygAh5QIQ8IEIeECEPiJC/HYrQ/K/Nixvq5l5qKbnZg/wdh/akHpDVqQ6MQX4Wy5YHRKiLgN59e2cAvTuPjoMc6hyg2asBMN3hH73HFwGlrgF6O4B7Lz6Izavvt6H3TUCoW4Bm+2qnmf1zsEZ0om4Bmv5VxzH7x5sgU90CVELdBDR7//qYHp+jMjoH6NMWn8LgVydBDnUQ0OUAfvH49d9uwwox/ERldA3QEdxSvrPVPUDTzd6h2Lgc5OlC3Tuan26GXOKNTHXwfNAZPBXfHxvaxZwHFJz1nlzwg4614/Pz8w9BtroHaLoJWNRu1j1As28fYj1CC0Z51zO/efXL71AZHRuk0xXe9TxhdCYxoe5N86kK73r+/JLf6HPazfvmp1/phxefvorGoOiuZ3m/2Kij94udwT2M6NM23EdBCUjca9jZG+out6D36IdzNlafvz8YwOphgCSoyOFHftZzW7gzV57tl57fbcVT/OqLIMgEFHRxmmed5/uH23fuPvq7cUZoMbuY+4BStZhBuu2A6p/m27FQPJLD0ANsmyxkodiOHnQkL/fM9vERfTjwjOo91GgFoOnvwvn98pf+YNU1QO0Yg+JdzF9ZDeyD9JllkLaps4DyygMi5AER8oAIdRRQcuZ+8SftWwGIzfN5LooFHQY02ydn+KiMbgLKKw+IkAdEqHuA8JXV39P/Je8eIPXa/C3qr4odPJqfveo9+RBcvR3Ak5/2yaPVDp4PSv4ftHZx2cAfyV0HpPzDjP6XmQfkzyhm/cs1xy7W9OXhOpX5P+kt6B2yGW3xgzSVo/EeFP3TfvWE7WVUB+okIDbT//T6B3+vRhag89fnHlD6LnYQ7WI51EFAbGTuseOwgb+ZJX2a57vXzN/MQt7MQk5hQTcBNXwzi+uAZvu+B0UbaUfzAsxsv6kxyPVjsSlbSN99uM3m+WZmMdeP5vkZM7EOup9rqdhFQExX5H1iqIwuAbo6x8oDqVtjUMGb6cIy5g2IyNEkoIyb6dLULUAl1FVAV80N0i6PQbHyHWaEZcwZkNOzWCQPSMgDsgarAqr7ZditB1T3y7CrAorG+FoBzd6nX9Wo+2XYFQHF2WsFlCXtZdjPnj3j38/49zj8ZhvPsL3gN29htfzye5zm11iNH2v543A5QHW/DLviA7fn+bzuUoBqfxl2K8YgUv6uZ0I1vgx7SQDV9zLstgOq/WXYbQdU+8uwhxGhtgKq+2XY7QekyAMi5McgQh4QIQ+IUAYg0U4PKBWQbKgHNHdAC7iqof4CbQNUfw/S1hEe0DICqnWh6AigcrNYv08Cso8g7RuD7CYCUJ8GpEbnOEW/lIDSc6j84pACMNUdt3Yxq6kwIGZRElKAhnSfag0gZZAS330DELf0cU41erkA8aYip9Rpjn/3cwAKFCAmjqKA2GaTgPCAINqaBaifF5C2iyl1FgSUmbx+QEr1FKB+LkDDaK/KBGRdYrQNkGgFar8N0HBoH6R1QKjkkE/MPnGhGUD0deD4QjBvhSbJBG8rWZTsuEgtASpFPK8bJeFbWsVarFFFtUvPOiC6B0VjqugL2szRj3ehMR6Coi6Cft+MHjREO2IUH/cgyUfZTVEPikuzeV/fLiarTDwV05d0U0vSjwnpgIYWQDxlLkCBkkIH1E/2YRy7OECxj6HkDqAAUuNk1+qrhKIhQhvFrYBQT0RMwxR9FZAsIwopP1ErACXZxngP1RIEGFAQIEJpgKJOopegAcK9fFGA+oli3/U4K6C4L8nCcQ8LrD2oL3efXIB4H1R+oWSpuqAxyM5HJZTwUdNoEjnUeG3AtdahjEEWQMiHKPmCAAVj1Do7oLEelw5I+q/F23uQASipJLJYq4i9rxPQMBlPA1y5pfGFAGldzAAk4+2AUCWpgLhhEYCQYyogRAH9Yia9FEB9G4AIkK0E1CWUElIAJbH1AkKdIyB7kBGZMQZlAbIhTvsRsgHVP0gjv1I7UMruE/5+WYBMgDSgYRlAgbuAiB5kHYOyAdl+hVw9qO8coKwGRg2wALKWQAEK6DEoJlQPoAwIJQD1Y5cNQPYSlNx4GA8Z6EmS2AUByoJQBZAWnzoNqoBUBOFyfRkBUWNURh2oE2YBwpWgxrUGUGoZ+QH1A1sSiwtt7EHpZRQAZE3iAbUaEH2o4QFVBEQCbDcgi29LAMj6MmwPKOFjfcdhU4DcG4NS3pLZFCAS4MIBpbxnNQ+feQAq1vpGANnf1OsBRUp5GXZTgOoZg6pcm097GXYeQnMA5P6xWOrLsFGlWWfMrHGB7aQ0PuGjX9exXncLrPEyhfRxHCiH8/ZTbhWP5mt/GTaVg/rOsWEveU7X5ut/GTaRw3VAtb8Mu/WA6n4ZdvsBKfKACHlAhL5Yas0BUAo20mSkMLMQOagwbaFcigWpMWXlARHygLolaNoB1wVNO+C6oPYaRtEjh+KDXPXUv1T8YLSbvV2ZYj1Kfso2rneiLPJJfFin6+sbn1+iBxuJAH7QUWxRnpOZ4SAS5G1nWY1Yu272NqL6WWt2r//8MzNJ9+XTYZH7AtCIQwiTCD78PIsM3+z9ZWdXqYClPeXPl5Vow5MyvLLTqMGj8NmPjLuS1eagLpgHhEhGa6MTAJ9fslYIn3lrRFcYhR0h9FnpQSGaENNudAIhDo/WkyeCigomqEhZo3iQWHjeIbRERaheag4agvnhEfXodUQ9hZFh9YuQdtpN7jMKoDAgkvAsYQvDcLgLbqAKZGGTsAvGdMM6Ih+iJIqXqoNmk6A8DZs+v9zQDeGvd7ohAImf639B0oN4FPv9U3uQ6AYjtQcJTWQvTe1BG/E+o/Ug1UvVQbNFUAZDhiZoaJQuhbWyjs22Rkm/QaMBGyMQoOiR1WGSm71wP5AUlIwiczQGRU3lg4ropnJkEeU/j7zbNbxUHTQE+dteUpF3smOP5Fykcbze+c1OYmCu8hNz8YA6kePFRhipj7PGLMbTf4n4JIOjPgFaHNQFOZtZs1gDn9OpmhA07YDrgvkWZ1mRqSYzgWHRDUXD2cvCooI5lJHIsiJTTWYCw6Ibioazl4WFBdWLSGTMp7rJTGCZglVD0bDFCW39qgYti1tFUBxDuvQVmWEyExgW3VA0bHNCm5/UoH3yigUZcYXlaA8y1q9q0FjcKoKMuOIyVmS6yUxgWHRD0bDNCX3dpQaNxS0WpEeVkWVFpprMBIZFNxQNZy8LiwrmUMZSC+ZcHhoB7CYzgWHRDUXDdB2US0iQHlVKaA6xm8wEhkU3FA3TdVAuIUF6VDlNzElTNZkJDItuKBqm66BcSgSpMeUUr7uSOlWTmcCw6IaiYboOyiUkmCedZRQ07YDrgnkX6OjxvGahXEoElYloOjVrUU1mAsOiG4qG6ToolxJBakw5mWdENZOZwLDohqJhug7KJSRIjyolvw4iZDk0po6dLZeKNiqF6ToKHM5DelQ5Xe8YP4dqMhMYFt1QNEzXQbmUCFJjSshydo46fWe5Nq8aioYtTrhzRnEZBU074Lpg7iXyP3Bo1xNUk5nAsOiGomG6DsqlWJDd2uI6Fefx0GVf3WQmMCy6oWiYroNyKRHkbnk+RWsKdAZBNZkJDItuKBqm66BcQoK8Lc+p+C8W+DAIm8wEhkU3FA3TdVAuIUHelueVvKZwjf8mp5rMBIZFNxQN03VQLiWC3C3PK/5/S+0/mqrJTGBYdEPRMF0H5VIsyG6tFzTtgOuCph1wXdC0A64LmnbAdUHTDrguaKDOd9sAvXsnbGu2v/KmAQeKCBZe42wfhHqHHpBVR7B6HARXr4Cx8YBMfYS1C7FxBA88IIuO4KncuLz7ggP68QB6T1jwbOXfW7ByGMxeDaB3/0Lsfj++ArgfzA7YJ4caR4ktEKNYdnq2Iz++qOYvVMteWGqfme33bvPx6AEH9DWwzjXdEgPUqtj9vuab3+yHKZIoRpmLl5SVflNErVUjBNXaW1jTTewwG7DXToIzuMUAiVGbNZ0ZPu0zC4tbOQneAjf8lzcziZpurp4EM9EXM9OzfsR60YNKDkOl3MWlA+JUhO1MNGS6KToY/2INfpqkWHmDoi4HKydRcRnpb8kqqnUhqJK5hHRAYjeRgPjYdDkQrQqOeoeyqTIF/1Si2K5z75hKL5cTUG0egCqZSwiNQe8vCEBRUw1AYhxmA89hVvqP0EpAySwmd4tygJjlX9t8/M2RvqJgHoUU0eXAWAchQMqYojYYRcmSpluKMSV9VcEcyigmtpJ+wVbSB8lKGgFSZiW1wSjqI98K3g2Uqc2ePng7aNcgbTsWw4DCxUsMD33iKFnE0zzpxR5ZXjCPNhfT7O1t9WgeA1JWxm+0T3UlzTtidvqDeL1dXlAt+/ILmnbAdUHTDrguaNoB1wVNO+C6oGkHXBc07YDrgqYdcF3QtAOuC5p2wHVB0w64rv8DWAJ7mlmREMoAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="#cb250-1" tabindex="-1"></a><span class="co">#We find the same most significant SNP with sommer and StageWise</span></span>
<span id="cb250-2"><a href="#cb250-2" tabindex="-1"></a><span class="co">#sommer</span></span>
<span id="cb250-3"><a href="#cb250-3" tabindex="-1"></a>bestSNP <span class="ot">&lt;-</span> <span class="fu">which.max</span>(GWASsommer_additive<span class="sc">$</span>scores)</span>
<span id="cb250-4"><a href="#cb250-4" tabindex="-1"></a><span class="fu">colnames</span>(Mc)[bestSNP]</span></code></pre></div>
<pre><code>## [1] &quot;solcap_snp_c2_22964&quot;</code></pre>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="#cb252-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb252-2"><a href="#cb252-2" tabindex="-1"></a>bestSNP <span class="ot">&lt;-</span> <span class="fu">which.max</span>(gwas.ans<span class="sc">$</span>score)</span>
<span id="cb252-3"><a href="#cb252-3" tabindex="-1"></a><span class="fu">colnames</span>(Mc)[bestSNP]</span></code></pre></div>
<pre><code>## [1] &quot;solcap_snp_c2_22964&quot;</code></pre>
<p>The GWAS peak on chr05 is near the gene <em>CDF1</em>, which is known
to have a large effect on potato maturity <a href="https://doi.org/10.1038/nature11912">(Kloosterman et
al.¬†2013)</a>. The following code extracts the most significant marker
for the large QTL on chr05 and passes it to <code>Stage2</code> as a
fixed effect.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="#cb254-1" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">which.max</span>(gwas.ans<span class="sc">$</span>score)</span>
<span id="cb254-2"><a href="#cb254-2" tabindex="-1"></a>gwas.ans[k,]</span></code></pre></div>
<pre><code>##                   marker chrom position      effect    score
## 5334 solcap_snp_c2_22964 chr05  4422417 -0.01642851 17.05289</code></pre>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="#cb256-1" tabindex="-1"></a>ans2vm<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>ans1vm<span class="sc">$</span>blues,</span>
<span id="cb256-2"><a href="#cb256-2" tabindex="-1"></a>               <span class="at">vcov=</span>ans1vm<span class="sc">$</span>vcov,</span>
<span id="cb256-3"><a href="#cb256-3" tabindex="-1"></a>               <span class="at">geno=</span>geno[[<span class="dv">1</span>]],</span>
<span id="cb256-4"><a href="#cb256-4" tabindex="-1"></a>               <span class="at">fix.eff.marker=</span><span class="st">&quot;solcap_snp_c2_22964&quot;</span>,</span>
<span id="cb256-5"><a href="#cb256-5" tabindex="-1"></a>               <span class="at">non.add=</span><span class="st">&quot;g.resid&quot;</span>)</span>
<span id="cb256-6"><a href="#cb256-6" tabindex="-1"></a></span>
<span id="cb256-7"><a href="#cb256-7" tabindex="-1"></a></span>
<span id="cb256-8"><a href="#cb256-8" tabindex="-1"></a></span>
<span id="cb256-9"><a href="#cb256-9" tabindex="-1"></a><span class="co"># Proportion of variance</span></span>
<span id="cb256-10"><a href="#cb256-10" tabindex="-1"></a><span class="co">#summary(ans2vm.1$vars)</span></span>
<span id="cb256-11"><a href="#cb256-11" tabindex="-1"></a></span>
<span id="cb256-12"><a href="#cb256-12" tabindex="-1"></a><span class="co">#add column into dataframe for fixed effect marker</span></span>
<span id="cb256-13"><a href="#cb256-13" tabindex="-1"></a>fix.eff.marker <span class="ot">&lt;-</span> Mc[,<span class="fu">colnames</span>(Mc)[bestSNP]]</span>
<span id="cb256-14"><a href="#cb256-14" tabindex="-1"></a><span class="co">#make sure names in fix.eff.marker_coefficients match sommerdf_smaller$id</span></span>
<span id="cb256-15"><a href="#cb256-15" tabindex="-1"></a>fix.eff.marker <span class="ot">&lt;-</span> fix.eff.marker[sommerdf_smaller<span class="sc">$</span>id]</span>
<span id="cb256-16"><a href="#cb256-16" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_smaller, fix.eff.marker)</span>
<span id="cb256-17"><a href="#cb256-17" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs))</span></code></pre></div>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="#cb257-1" tabindex="-1"></a><span class="co">#sommer</span></span>
<span id="cb257-2"><a href="#cb257-2" tabindex="-1"></a><span class="co">#without dominance</span></span>
<span id="cb257-3"><a href="#cb257-3" tabindex="-1"></a>ans2vm<span class="fl">.1</span>_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> fix.eff.marker <span class="sc">-</span><span class="dv">1</span>, </span>
<span id="cb257-4"><a href="#cb257-4" tabindex="-1"></a>                      <span class="co">#include best GWAS marker as fixed effects.</span></span>
<span id="cb257-5"><a href="#cb257-5" tabindex="-1"></a>                      <span class="co">#We want to make a regression with it</span></span>
<span id="cb257-6"><a href="#cb257-6" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Hinv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb257-7"><a href="#cb257-7" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb257-8"><a href="#cb257-8" tabindex="-1"></a>                     <span class="fu">vsc</span>(<span class="fu">isc</span>(id)) <span class="sc">+</span> <span class="co">#Genotypic effects not captured by the </span></span>
<span id="cb257-9"><a href="#cb257-9" tabindex="-1"></a>                     <span class="co">#addtivie effect, i.i.d.</span></span>
<span id="cb257-10"><a href="#cb257-10" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb257-11"><a href="#cb257-11" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb257-12"><a href="#cb257-12" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb257-13"><a href="#cb257-13" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb257-14"><a href="#cb257-14" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb257-15"><a href="#cb257-15" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb257-16"><a href="#cb257-16" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb257-17"><a href="#cb257-17" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb257-18"><a href="#cb257-18" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb257-19"><a href="#cb257-19" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb257-20"><a href="#cb257-20" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb257-21"><a href="#cb257-21" tabindex="-1"></a>                           </span>
<span id="cb257-22"><a href="#cb257-22" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb257-23"><a href="#cb257-23" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb257-24"><a href="#cb257-24" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb257-25"><a href="#cb257-25" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb257-26"><a href="#cb257-26" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb257-27"><a href="#cb257-27" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb257-28"><a href="#cb257-28" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb257-29"><a href="#cb257-29" tabindex="-1"></a>                           </span>
<span id="cb257-30"><a href="#cb257-30" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb257-31"><a href="#cb257-31" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb257-32"><a href="#cb257-32" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb257-33"><a href="#cb257-33" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb257-34"><a href="#cb257-34" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb257-35"><a href="#cb257-35" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb257-36"><a href="#cb257-36" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb257-37"><a href="#cb257-37" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units)), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb257-38"><a href="#cb257-38" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb257-39"><a href="#cb257-39" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb257-40"><a href="#cb257-40" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb257-41"><a href="#cb257-41" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb257-42"><a href="#cb257-42" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="#cb258-1" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb258-2"><a href="#cb258-2" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb258-3"><a href="#cb258-3" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>b[<span class="fu">unique</span>(sommerdf_smaller<span class="sc">$</span>Env),]</span>
<span id="cb258-4"><a href="#cb258-4" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb258-5"><a href="#cb258-5" tabindex="-1"></a></span>
<span id="cb258-6"><a href="#cb258-6" tabindex="-1"></a></span>
<span id="cb258-7"><a href="#cb258-7" tabindex="-1"></a></span>
<span id="cb258-8"><a href="#cb258-8" tabindex="-1"></a><span class="co"># #best GWAS marker effects for all observations </span></span>
<span id="cb258-9"><a href="#cb258-9" tabindex="-1"></a><span class="co"># Xfix.eff.marker &lt;- sommerdf_smaller$fix.eff.marker</span></span>
<span id="cb258-10"><a href="#cb258-10" tabindex="-1"></a><span class="co"># fix.eff.markerEffects &lt;- ans2vm.1_open$b[&quot;fix.eff.marker&quot;,]</span></span>
<span id="cb258-11"><a href="#cb258-11" tabindex="-1"></a><span class="co"># fix.eff.marker_variance &lt;- var(Xfix.eff.marker*fix.eff.markerEffects)</span></span>
<span id="cb258-12"><a href="#cb258-12" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb258-13"><a href="#cb258-13" tabindex="-1"></a></span>
<span id="cb258-14"><a href="#cb258-14" tabindex="-1"></a></span>
<span id="cb258-15"><a href="#cb258-15" tabindex="-1"></a></span>
<span id="cb258-16"><a href="#cb258-16" tabindex="-1"></a></span>
<span id="cb258-17"><a href="#cb258-17" tabindex="-1"></a><span class="co">#best GWAS marker effects for each genotype</span></span>
<span id="cb258-18"><a href="#cb258-18" tabindex="-1"></a><span class="co">#extract Xfix.eff.marker from marker matrix, not from the model</span></span>
<span id="cb258-19"><a href="#cb258-19" tabindex="-1"></a><span class="co">#In the model, there are repeated fix.eff.marker coefficients for genotypes </span></span>
<span id="cb258-20"><a href="#cb258-20" tabindex="-1"></a><span class="co">#that had more than one BLUE (genotypes present in more than one environment) </span></span>
<span id="cb258-21"><a href="#cb258-21" tabindex="-1"></a><span class="co">#and there are no fix.eff.marker coefficients for genotypes that were not </span></span>
<span id="cb258-22"><a href="#cb258-22" tabindex="-1"></a><span class="co">#phenotyped and had no BLUEs</span></span>
<span id="cb258-23"><a href="#cb258-23" tabindex="-1"></a><span class="co">#If we take fix.eff.marker from marker matrix, there are no such problems</span></span>
<span id="cb258-24"><a href="#cb258-24" tabindex="-1"></a>Xfix.eff.marker <span class="ot">&lt;-</span> Mc[,<span class="fu">colnames</span>(Mc)[bestSNP]]</span>
<span id="cb258-25"><a href="#cb258-25" tabindex="-1"></a>fix.eff.markerEffects <span class="ot">&lt;-</span> ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>b[<span class="st">&quot;fix.eff.marker&quot;</span>,]</span>
<span id="cb258-26"><a href="#cb258-26" tabindex="-1"></a>fix.eff.marker_effects <span class="ot">&lt;-</span> Xfix.eff.marker<span class="sc">*</span>fix.eff.markerEffects</span>
<span id="cb258-27"><a href="#cb258-27" tabindex="-1"></a><span class="co">#reorder heterosis effects to be in the same order as BLUPs</span></span>
<span id="cb258-28"><a href="#cb258-28" tabindex="-1"></a>fix.eff.marker_effects <span class="ot">&lt;-</span> fix.eff.marker_effects[<span class="fu">match</span>(<span class="fu">rownames</span>(ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span>),</span>
<span id="cb258-29"><a href="#cb258-29" tabindex="-1"></a>                        <span class="fu">names</span>(fix.eff.marker_effects))]</span>
<span id="cb258-30"><a href="#cb258-30" tabindex="-1"></a>fix.eff.marker_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(fix.eff.marker_effects) </span>
<span id="cb258-31"><a href="#cb258-31" tabindex="-1"></a></span>
<span id="cb258-32"><a href="#cb258-32" tabindex="-1"></a></span>
<span id="cb258-33"><a href="#cb258-33" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Hinv)</span><span class="st">`</span> <span class="sc">+</span></span>
<span id="cb258-34"><a href="#cb258-34" tabindex="-1"></a>              fix.eff.marker_effects <span class="sc">+</span></span>
<span id="cb258-35"><a href="#cb258-35" tabindex="-1"></a>              <span class="fu">mean</span>(Env_effects)</span>
<span id="cb258-36"><a href="#cb258-36" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb258-37"><a href="#cb258-37" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>ans1vm<span class="sc">$</span>blues,</span>
<span id="cb258-38"><a href="#cb258-38" tabindex="-1"></a>              <span class="at">vcov=</span>ans1vm<span class="sc">$</span>EEV,</span>
<span id="cb258-39"><a href="#cb258-39" tabindex="-1"></a>              <span class="at">geno=</span>geno[[<span class="dv">1</span>]],</span>
<span id="cb258-40"><a href="#cb258-40" tabindex="-1"></a>              <span class="at">vars=</span>ans2vm<span class="fl">.1</span><span class="sc">$</span>vars)</span>
<span id="cb258-41"><a href="#cb258-41" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">geno=</span>geno[[<span class="dv">1</span>]], <span class="at">what=</span><span class="st">&quot;BV&quot;</span>)</span>
<span id="cb258-42"><a href="#cb258-42" tabindex="-1"></a></span>
<span id="cb258-43"><a href="#cb258-43" tabindex="-1"></a></span>
<span id="cb258-44"><a href="#cb258-44" tabindex="-1"></a><span class="co">#Very similar BLUPs</span></span>
<span id="cb258-45"><a href="#cb258-45" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb258-46"><a href="#cb258-46" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9821844</code></pre>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="#cb260-1" tabindex="-1"></a>ASRemlBLUPs<span class="sc">$</span>value[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 3.977339 5.213230 5.041262 4.354287 5.103266</code></pre>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="#cb262-1" tabindex="-1"></a>SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 3.882425 5.006276 5.040866 4.412567 4.916953</code></pre>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="#cb264-1" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb264-2"><a href="#cb264-2" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(Env_variance, <span class="co">#environmental variance</span></span>
<span id="cb264-3"><a href="#cb264-3" tabindex="-1"></a>                fix.eff.marker_variance,</span>
<span id="cb264-4"><a href="#cb264-4" tabindex="-1"></a>  ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Hinv:isc:isc&quot;</span>]<span class="sc">*</span>meanH, <span class="co">#additive</span></span>
<span id="cb264-5"><a href="#cb264-5" tabindex="-1"></a>  ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:isc:isc&quot;</span>], <span class="co">#non-additive</span></span>
<span id="cb264-6"><a href="#cb264-6" tabindex="-1"></a>  ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;units:isc:isc&quot;</span>], <span class="co">#residual</span></span>
<span id="cb264-7"><a href="#cb264-7" tabindex="-1"></a>  stage1errorvar2) <span class="co">#stage1 estimation error</span></span>
<span id="cb264-8"><a href="#cb264-8" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2vm<span class="fl">.1</span><span class="sc">$</span>vars)<span class="sc">$</span>Variance</span>
<span id="cb264-9"><a href="#cb264-9" tabindex="-1"></a>variances_ans2vm <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb264-10"><a href="#cb264-10" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb264-11"><a href="#cb264-11" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2vm) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2vm<span class="fl">.1</span><span class="sc">$</span>vars))</span>
<span id="cb264-12"><a href="#cb264-12" tabindex="-1"></a><span class="co">#Extremely similar variance components for both models</span></span>
<span id="cb264-13"><a href="#cb264-13" tabindex="-1"></a>variances_ans2vm</span></code></pre></div>
<pre><code>##                 Sommer ASReml
## env          0.4902587  0.485
## fixed.marker 0.3268151  0.313
## additive     0.5945204  0.596
## g.resid      0.4424101  0.445
## g x env      0.2893497  0.297
## Stage1.error 0.8798469  0.884</code></pre>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="#cb266-1" tabindex="-1"></a><span class="co">#almost same aic</span></span>
<span id="cb266-2"><a href="#cb266-2" tabindex="-1"></a>ans2vm<span class="fl">.1</span><span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 1521.565</code></pre>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="#cb268-1" tabindex="-1"></a>ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 1524.398</code></pre>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="#cb270-1" tabindex="-1"></a><span class="co"># Very similar fixed effect for the marker</span></span>
<span id="cb270-2"><a href="#cb270-2" tabindex="-1"></a>ans2vm<span class="fl">.1</span><span class="sc">$</span>params<span class="sc">$</span>marker <span class="co">#stage wise</span></span></code></pre></div>
<pre><code>##                marker   estimate         SE
## 1 solcap_snp_c2_22964 -0.7764807 0.08692611</code></pre>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="#cb272-1" tabindex="-1"></a>ans2vm<span class="fl">.1</span>_open<span class="sc">$</span>b[<span class="st">&quot;fix.eff.marker&quot;</span>,] <span class="co">#sommer</span></span></code></pre></div>
<pre><code>## fix.eff.marker 
##     -0.7688477</code></pre>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="#cb274-1" tabindex="-1"></a><span class="co">#save.image(file=&quot;workspace1.RData&quot;)</span></span>
<span id="cb274-2"><a href="#cb274-2" tabindex="-1"></a><span class="co"># save(ans2_open, ans2a_open, ans2b_open, ans2c_open, ans2d_open, ans2mask, ans2vm.1_open,</span></span>
<span id="cb274-3"><a href="#cb274-3" tabindex="-1"></a><span class="co">#     GWASsommer_additive, model_Sommer, modelGWASvar, file = &quot;Vignette1Models.RData&quot;)</span></span></code></pre></div>
<p>The fixed effect estimate of -0.78 for solcap_snp_c2_22964 implies
that, on average, each additional copy of the alternate allele reduced
vine maturity by 0.78 (on a 1-9 visual scale). According to the
proportion of variance table, the marker accounted for 0.123/(0.123 +
0.235) = 34% of the breeding value.</p>
</div>
</div>
<div id="bibliography" class="section level1">
<h1>Bibliography</h1>
<p>Damesa, T.M., M√∂hring, J., Worku, M. and Piepho, H.-P. (2017), One
Step at a Time: Stage-Wise Analysis of a Series of Experiments. Agronomy
Journal, 109: 845-857.</p>
<p>Endelman, J. B. (2023). Fully efficient, two-stage analysis of
multi-environment trials with directional dominance and multi-trait
genomic selection. Theoretical and Applied Genetics, 136(4), 65.</p>
<p>Piepho, H.P., J. M√∂hring, T. Schulz-Streeck, and J.O. Ogutu. 2012. A
stage-wise approach for analysis of multi-environment trials. Biometrics
54:844‚Äì860. <a href="doi:10.1002/bimj.201100219" class="uri">doi:10.1002/bimj.201100219</a></p>
<p>Robinson, G. K., &amp; Jones, L. P. (1987). Approximations for
prediction error variances. Journal of Dairy Science, 70(8),
1623-1632.</p>
<p>VanRaden, P. M. (2008). Efficient methods to compute genomic
predictions. Journal of dairy science, 91(11), 4414-4423.</p>
<p>Vitezica ZG, Varona L, Legarra A (2013). On the additive and dominant
variance and covariance of individuals within the genomic selection
scope. Genetics, 195(4):1223-30.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
