<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Vignette 2: Single trait analysis at correlated locations</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Vignette 2: Single trait analysis at
correlated locations</h1>



<p>Libraries for open source:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(Matrix) <span class="co">#all StageWise models in the vignettes work for Matrix package </span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#version 1.5-1. Later versions may cause errors in some of the StageWise models</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(lme4)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(matrixcalc)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(sommer)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(pracma)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(StageWise)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;Vignette2Models.RData&quot;</span>)</span></code></pre></div>
<p>If you cite this work, <strong>it is important that you also cite the
StageWise package (Endelman, J. B., 2023)</strong>. For more information
about the fully efficient two-stage modeling strategy, we recommend
reading Damesa et al.Â (2017)</p>
<p>This vignette builds upon the information provided in <a href="https://jendelman.github.io/StageWise/Vignette1.html">Vignette
1</a>, which covers analysis at a single location. This vignette
illustrates how to model the covariance between multiple locations using
a dataset of potato yield trials across 8 locations (TX was excluded)
and 6 years, with a population of 336 genotypes (<a href="https://doi.org/10.2135/cropsci2018.05.0314">Schmitz Carley et
al.Â 2019</a>).</p>
<p>To understand this vignette, familiarity with factor analytics
methodology for exploring location-by-environment interactions is
important. Letâs assume the following model for Stage 2:</p>
<p><span class="math display">\[
BLUEs = \boldsymbol{y} = X\boldsymbol{E} + Z\boldsymbol{g} +
Z_{gl}\boldsymbol{gl} + \boldsymbol{s} + \boldsymbol{\epsilon}
\]</span></p>
<p>Where <span class="math inline">\(BLUEs = \boldsymbol{y}\)</span> is
a vector of stage 1 genotypic BLUEs (BLUEs of different environments are
concatenated), <span class="math inline">\(\boldsymbol{E}\)</span> is a
vector of fixed environmental effects, <span class="math inline">\(\boldsymbol{g} \sim N(\boldsymbol{0},
I\sigma^2_g)\)</span> is a vector of random genotypic effects, <span class="math inline">\(\boldsymbol{gl} \sim N(\boldsymbol{0},
GxL\sigma^2_{gl})\)</span>, <span class="math inline">\(\boldsymbol{s}
\sim N(\boldsymbol{0}, EEV)\)</span> is a vector of random Stage 1
estimation errors, and <span class="math inline">\(\boldsymbol{\epsilon}
\sim N(\boldsymbol{0}, I\sigma^2_{\epsilon})\)</span> is a vector of
residuals that correspond to genotype by environment interaction
unexplained by the other terms. <span class="math inline">\(X\)</span>,
<span class="math inline">\(Z\)</span>, and <span class="math inline">\(Z_{gl}\)</span> are design matrices for their
corresponding effects. <span class="math inline">\(\sigma^2_g\)</span>,
<span class="math inline">\(\sigma^2_{gl}\)</span>, and <span class="math inline">\(\sigma^2_{\epsilon}\)</span> are the genotypic,
genotype by location interaction, and residual variances, respectively.
<span class="math inline">\(I\)</span> is an identity matrix with the
appropriate dimensions. <span class="math inline">\(EEV\)</span> is the
matrix of estimation error variance-covariance of the BLUEs. Finally,
<span class="math inline">\(GxL = L \otimes G\)</span> is the
variance-covariance matrix for the genotype by location interaction.
<span class="math inline">\(G\)</span> is the variance-covariance
structure for the genotypes (the additive relationship matrix if
available, the identity otherwise). <span class="math inline">\(L\)</span> is a variance-covariance structure
across locations that must be estimated by the model.</p>
<p>Various variance-covariance structures for <span class="math inline">\(L\)</span> can be assumed (diagonal, compound
symmetry, unstructured, etc.). Here, we will focus on unstructured,
which is extremely versatile as it independently calculates all
variances and covariances. Therefore, the number of parameters to
estimate scales with the square of the number of locations. This can be
problematic as a large number of parameters to estimate for a large
number of locations can cause convergence issues.</p>
<p>Factor analytics attempts to solve this problem through
dimensionality reduction. At its core, the aim of factor analytics is to
replace the <span class="math inline">\(L\)</span> matrix with a smaller
<span class="math inline">\(FA\)</span> matrix that carries almost as
much information, massively reducing the number of parameters to
estimate. There are a few ways to achieve this, and ASReml and Sommer
use two different approaches. That way, we expect different results from
these packages.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>geno.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;geno2.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>geno <span class="ot">&lt;-</span> <span class="fu">read_geno</span>(<span class="at">filename=</span>geno.file,<span class="at">ploidy=</span><span class="dv">4</span>,<span class="at">map=</span><span class="cn">FALSE</span>,<span class="at">dominance=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Minor allele threshold = 5 genotypes
## Number of markers = 5262
## Number of genotypes = 336</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>pheno.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;vignette_data&quot;</span>, <span class="st">&quot;pheno2.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;StageWise&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(pheno.file)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">head</span>(pheno)</span></code></pre></div>
<pre><code>##           id loc     env Yield.Mg.ha
## 1  A01143-3C  CA CA_2011       90.98
## 2  A01143-3C  CA CA_2011       95.12
## 3 AC00206-2W  CA CA_2011       74.44
## 4 AC01151-5W  CA CA_2011       74.44
## 5 AC01151-5W  CA CA_2011       95.12
## 6 AC03452-2W  CA CA_2011       86.85</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">table</span>(pheno<span class="sc">$</span>loc)</span></code></pre></div>
<pre><code>## 
##   CA   FL   MI   MO   NC   NY   OR   WI 
##  831  841  844  504 1232  869  629  777</code></pre>
<p>The input file contains not only a column named âenvâ for environment
but also one named âlocâ for location. While the location information is
not explicitly used by <code>Stage1</code>, it is automatically detected
and retained in the output for use in <code>Stage2</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>ans1 <span class="ot">&lt;-</span> <span class="fu">Stage1</span>(<span class="at">filename=</span>pheno.file, <span class="at">traits=</span><span class="st">&#39;Yield.Mg.ha&#39;</span>)</span></code></pre></div>
<pre><code>## Online License checked out Fri Jun  7 10:04:55 2024</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">head</span>(ans1<span class="sc">$</span>blues)</span></code></pre></div>
<pre><code>##       env          id    BLUE loc
## 1 CA_2011   A01143-3C  93.050  CA
## 2 CA_2011  AC00206-2W  74.440  CA
## 3 CA_2011  AC01151-5W  84.780  CA
## 4 CA_2011  AC03452-2W  86.850  CA
## 5 CA_2011  AC05153-1W  49.630  CA
## 6 CA_2011 Accumulator 105.455  CA</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>ans1<span class="sc">$</span>fit,<span class="fu">aes</span>(<span class="at">x=</span>loc,<span class="at">y=</span>H2)) <span class="sc">+</span> </span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  <span class="fu">stat_boxplot</span>(<span class="at">outlier.color=</span><span class="st">&quot;red&quot;</span>) <span class="sc">+</span> </span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Location&quot;</span>) <span class="sc">+</span> </span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">paste</span>(<span class="st">&quot;Broad-sense &quot;</span>,H<span class="sc">^</span><span class="dv">2</span>,<span class="st">&quot; (plot basis)&quot;</span>)))</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAEgCAMAAACKBVRjAAABAlBMVEUAAAAAADoAAGYAOpAAZrYzMzM6AAA6ADo6AGY6OmY6kNtNTU1NTW5NTY5NbqtNjshmAABmAGZmOpBmtv9uTU1uTW5uTY5ubo5ubqtujqtujshuq+SOTU2OTW6OTY6Obk2Obm6ObquOjm6Ojo6OjsiOyP+QOgCQOjqQOmaQtpCQ27aQ29uQ2/+rbk2rbm6rbo6rjk2rjqurq46ryKur5Mir5OSr5P+2ZgC2Zjq22/+2///Ijk3Ijm7IyI7I5KvI/+TI///bkDrb/7bb///kq27kq47k/8jk/+Tk///r6+v/AAD/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///9P9qmAAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMWUlEQVR4nO2dC3vbthWG5cRxN6uu09lJszjbmnXWbnayLVu0xWlrLWrn2XInW8L//ysDeBfEAwIgyEPS3/c8zgX6TBzhJW4kCI4ExKoRdwAPXQDALABgFgAwCwCYBQDM8gfw33JR6bBpyQDAawMAZhsAMNsAgNkGAMw2AGC2AQCzDQCYbU0BuL6+biH6AdgaAnB9bUmgS2XBYgMAWk/ayBQAaPUZwCD6gF4D6HTJAkA/bADAbOsagPtX4y+uon8tj8cHlz0GYFmyHQOwPj8V88OIxK/ei3nCAgBaA3D/60uxfK5O/OWLq+h/AOCSaW0AUbHLc79YA55IVf5i92QZc6tfrRrA4osUQKE3EKgBDDVg+ey9WBygCXLLNGAfUKgLANDmKOgkGQWhBvhkGmoeoCrBYjz+PK0AAICZsLMNAJhtAMBs6wWA2SjSPgAEytQJwN3TtORno50zAAiRqQuAuy8/EP8BAO9M0QfkAgBmWy8A3D09kj3BI3PzAwD2mboCmO6K2aMPs10ACJSpIwBZAVaTXXFTXQUaCatZGwcAap2OAcDd030ACJUpuVKNArCa7N/IGcAUTVCYTJ0BiNu90a6YPv4EAEEydQdgrTphPRwAzn0AALBeiojmAJEeVCf8xKhmYkMNKKabvmffALRSfz1tYQHU+6YkAOdLEeb6S0UJABQA50sRAOAVGwXA/VKEHwCWAUlPADheijCvpAQARwDulyIAwCs2CoD9pYgsQK9PWRYqewVDfhgmtvrDUNQAr9gAoJjeJQCyCXK7FAEAXrFRAGQnvJrIgdBRVfkPCoDXtaDGhqFiKoeh1b1wFgcA+MRmAjBzm4h1GIBlyXapCVIzAFn6s2HUAMuS7RQA2QmIaeXCUAEA3RmGdvha0IMA4BcWAJAAbtQ0oHoUCgANAZip8Y+6IAoALACSKVjjK+MAAAC6CSAuepdLEZZl4Rd9pwD4TZjd1gWNUqEGeNvsYisF4CTHstCLJqgNAKrLQi+aoDYAqC4LvWiC2gCguiz0oglqA4DqstCLJqhtKKMggXkAAHjZ6gKwK9kWmqBZNg9o+hmxTgGwzLSVPsBmDjx0AK3UJwqAvRyj179MUFu3AFjZ0oRtADO77YIAgDpaTQDb9wPy7VrX58E27QMAAsD2KCjfvFtcnEZ7hwJAqwDyjVvzjbsBwASgXie81QQVti5+8ZekCVJHMkVTrXaWp/utO695NDMA/feqO+F8w+Ll8WmEI5bj6aOfTUFtLdWAkPM1GoCuYg0It3d0HwEEzdQeQKEPeA0AwTItBVB6SzLfvFuNgh5yExQ001IA5co375b/CvUSHwCwB0CoXvQAQAKY1Vua2C0AduMWy6PZZloPQN2liQAQeibcawBWZWF7NNtMrah3FYA5euq3OgXAzkYBEGqjAsYmyO95mwEBsF+dWC+ssABauYkY1kYBsFe9sAAAADoIoMUXOABAGQDVAyRTsKZfYQIApQCkpu28xAcAKADWqhcWAAAAAJSmAwAAAAAAAICFQgO4UbdhVn+o3jQaAELYtgCsfn8mpkf8ALwuRw8BwN0vPgnxzz8DAGcNkH/8nBuAMc8hAxC3P/uQ/NEIAMtT+wEDEM3uFWFZsv0BQL0eyfJopQAiWWyYGMkx+qEBIF8QZnm0NAp/AI6qKFlHW810i6yqFAGodYRYqAH2X2FD/DXAMfqhAUAfsHFwn0w7OAqKBAAA4FxmQwGAeYDxKwS1lQJwkmNYYQGEmFcDQGnJtmcDAOciA4CiHMMCAC05EQDYf4WgtjSKgQPwubvDDWA2Gh3ZzAUcwwp7PyDsaNXyKwS1kQCmj/8dv9GWCUBQTnoMFuVl+ApBbRSA6HW2Rw08I9biDAsAygQAWjIBQMxUE9TAQ3oAoCVTAOKX+ITftA8AtGQSgK0cwwIALRkASPECiHphi6vRANDUPGBX7dcx850HkHdKAUBLJgDcxbMw32EovVYAALRkGoAagwJA0zYKwGqyr7brmHo2QQBga6MAiNu90a6Yel+MQx9gaSMBbCvfvDveSNoIAKMgS5s9gMLm3ULMxwAQxkYCkE3Q5rKUfONWIZZf/QYAwtgoANt3AgobFq/f/CNugtSleWMRlBSK16dey549f4tH1e+QyTfvFvMT9AGhbBSA1UQHsLF5NwCEslEAtqdgeR8wHyudAEAIGwUgXR5aunl3h4ahXusd+gCgRPnm3QAQzuYAgJBjWGiCtGQSQFPrggBgM5kC0Ni6IADYTCYAYFkKADR3tD4A6Ni6oAcIoFvrgh4iAFs5hgUAWjIAkMK6oKaO1gsANdcFAYCljQJQc10QANjaDADqrAsCAFsbBaDmuqBuA2ilZC1tFIC664IAwNJGArCWY1gAoCUDACl2AHY7BjmGBQBaMgCQAoC2j8ZhAwBa7ADs5BgWAGjJZQA6sGccy9E4bKUAlJI3ale/09wxLADQkgkAjb3QGQC05NoAHOW3oLydo3GKaoIGeVO+DzUguSlf3QUAAIahAOASFgBoyRSArYf0AKARGwVgNdlfTY4wD2jcRgFQRT/dFzfVl4McwwIALdkAYDbUVRF9AKBux8vSt7gg6hhWYABejyj1AoDsBMR0tHNWVf68APSD29l6AcBajmEBgJYMALw2GsCsD88H6Ae3s/UCQD8uxukHt7P1AUBz9wOaHLcAgHNYYYtsQAAaa4I0AQAFoKlOWBMAkABsVS8sACAAbG/YBACN2CgANheiASCAjQJg/0bzemEBAFkDGloZpwkAqBpgrXphAQAA8NoIAFM5BVC35asfkgSAJgBMH39aTUZH0V0ZAGjUVgogmgREl4GKN+Xz3dOXx2Pvzbs1AUAZgGgSEAMo7hua7p6u9s9dPnsPACFs9gDynXMXCsPFKQCEsNkDKOyentQCr93TNYVdQd6n9eiaqgEUdk+P9zFOVO+8QA0oB1DyjFixBty/ysofANqaiBXeoLE8Ps3T64UFANYA8t3TN8ofAFq7FJHtnh6/PwCjoCA2BwCE6oUFAADAawMAZhsAMNsAgNkGAMw2AGC2AQCzDQCYbdwAwtoAoOXodQFAy9HrAoCWo9cFAC1HrwsAWo5eFwC0HP0AbADAbAMAZhsAMNsAgNkGAMw2AGC2AQCzDQCYbfUBQEGEGsBkqw+AkOVSfdgSAQCzDQCYbQDAbMMoiFkAwCwAYBYAMAsAmBUQwGI8jjeVyB/rLpHab2I8PkmeuzT5oufyDy4NzswjxLzw9Nq2TX0SHSeL0eT7/jw67CGVbZJX/E0OyOjUngILZVwcGr5DOABzFcqF+naLb05oW/rEayWAZ7+8EvdfPzcCSD1irp4mf0VkuzxWoanj5DEafWpDDDrbNK/YMSd5zmU4c1UUc9PpFgxAfNqrnT3Wb/7+mgrKAcDz38pS+MYMIPWoZ5npYy6ff3sYfZrHaPaps/+COomyvOLs6PCWL67Wb96pP963AWCRnQky4wuqNXAB8O2p+OFdBYDEE2dOlaw8ggxomdpMocU+eTbJgiNcWV5VNUCSun/9n7fq7GgFQNZkyqq3INvPuOU8tAHw3cv12+8qACSeJD+Cuzr3X18tMxt9uMSnmnnyFMryivsAmqk6818q30vTtw0GYJmeMutzY9dkXwO+/+uPv1tWAEg8caEYakDcDi/J03rTZwwvy0t5Fp8bxhvz0/mpWJzMT1oBkLWv0bek2yB7AJcf/3ZSBSDxVPUBl/J0fGfRByQ+Y3ibfcCcPtUkqo+yi3r50dSjNzAKmp8KYWiDHADI06sSQOyJMqdHQVGRqiFOxSgo8xnDS/OKPOtzulm7/1qei+s3X122AyAZY6/fqrzoqUC28co46gtIxUVfCSA1LEzzgLi3PKicB+Q+4/mR5JWMgo7JMfc6m0+0AwDyEQAwCwCYBQDMAgBmAQCzeg2gcovrn/4lbj+rfi8ppwYNoOuFrwQAzBoGgNUkfuVB+rd6BUL8IoR9BSFJvv3sjzLF9iVFLWkQAFaT3exHbb8cbcE8e/RBFb78ST+63Xv8KX5TYIc0CADpdtdpk/M/9fID+Z8EQPbx3lH3mqVhAFCvm0hLOk6RTdBOBiD5OCp7AAgoCsDd052zYg0AgIaUAZDlXWyCohK/yWtA8WMACKiyTlj9qBK/3ds5U51xsRMGgMBSw8vo3bv6MHQqe4A/ydKfjnY3hqEAAOkCAGYBALMAgFkAwCwAYBYAMAsAmAUAzAIAZv0f9sMy/9iKMp8AAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#one model per environment</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>EEVs <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>Envs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>H2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>residuals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>Locs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno<span class="sc">$</span>env)) {</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> pheno[pheno<span class="sc">$</span>env <span class="sc">==</span> enviro,]</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>  <span class="co">#1) remove entries with missing values for yield</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>  phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>Yield.Mg.ha),]</span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>  </span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>  Envs <span class="ot">&lt;-</span> <span class="fu">c</span>(Envs, <span class="fu">rep</span>(enviro, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))))</span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a>  loc <span class="ot">&lt;-</span> <span class="fu">unique</span>(phenoenv<span class="sc">$</span>loc)</span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a>  Locs <span class="ot">&lt;-</span> <span class="fu">c</span>(Locs, <span class="fu">rep</span>(loc, <span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))))</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>  </span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a>  phenoenv<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">factor</span>(phenoenv<span class="sc">$</span>id)</span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a>  </span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="co"># Remove the intercept in the model!</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="co"># It&#39;s important for correct EEV calculation.</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="co"># We use lm() instead of lmer() because we don&#39;t have any random effects.</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a>  models[[enviro]] <span class="ot">&lt;-</span> <span class="fu">lm</span>(Yield.Mg.ha <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span> id,</span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a>                        <span class="at">data=</span>phenoenv)</span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>  <span class="co">#extract H2</span></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a>  <span class="co">#random effect variances</span></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a>  variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models[[enviro]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a>  </span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a>  <span class="co">#fixed effect variances</span></span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a>  <span class="co">#1) calculate design matrix</span></span>
<span id="cb13-34"><a href="#cb13-34" tabindex="-1"></a>  X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(Yield.Mg.ha<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoenv) <span class="co">#remove intercept.</span></span>
<span id="cb13-35"><a href="#cb13-35" tabindex="-1"></a>  </span>
<span id="cb13-36"><a href="#cb13-36" tabindex="-1"></a>  <span class="co">#calculate fixed effects for each plot</span></span>
<span id="cb13-37"><a href="#cb13-37" tabindex="-1"></a>  <span class="co">#id effects:</span></span>
<span id="cb13-38"><a href="#cb13-38" tabindex="-1"></a>  <span class="co"># y = Xb + ... + epsilon</span></span>
<span id="cb13-39"><a href="#cb13-39" tabindex="-1"></a>  <span class="co"># b (vector of the genotypic fixed effects) can be retrieved from:</span></span>
<span id="cb13-40"><a href="#cb13-40" tabindex="-1"></a>  <span class="co">#&quot;models[[enviro]]@beta[1:length(unique(phenoenv$id))]&quot;</span></span>
<span id="cb13-41"><a href="#cb13-41" tabindex="-1"></a>  <span class="co"># X is the design matrix linking each observation (each plot in the field) with</span></span>
<span id="cb13-42"><a href="#cb13-42" tabindex="-1"></a>  <span class="co"># its corresponding genotypic effect.</span></span>
<span id="cb13-43"><a href="#cb13-43" tabindex="-1"></a>  id_effects <span class="ot">&lt;-</span> X.id<span class="sc">%*%</span>models[[enviro]]<span class="sc">$</span>coefficients[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(phenoenv<span class="sc">$</span>id))]</span>
<span id="cb13-44"><a href="#cb13-44" tabindex="-1"></a></span>
<span id="cb13-45"><a href="#cb13-45" tabindex="-1"></a>  <span class="co">#calculate variance for the fixed effects</span></span>
<span id="cb13-46"><a href="#cb13-46" tabindex="-1"></a>  id_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(id_effects)</span>
<span id="cb13-47"><a href="#cb13-47" tabindex="-1"></a>  </span>
<span id="cb13-48"><a href="#cb13-48" tabindex="-1"></a>  <span class="co">#Plot level heritability</span></span>
<span id="cb13-49"><a href="#cb13-49" tabindex="-1"></a>  H2 <span class="ot">&lt;-</span> <span class="fu">c</span>(H2, id_variance<span class="sc">/</span>(id_variance <span class="sc">+</span></span>
<span id="cb13-50"><a href="#cb13-50" tabindex="-1"></a>                       variance_residual) )</span>
<span id="cb13-51"><a href="#cb13-51" tabindex="-1"></a>  </span>
<span id="cb13-52"><a href="#cb13-52" tabindex="-1"></a>  </span>
<span id="cb13-53"><a href="#cb13-53" tabindex="-1"></a>  <span class="co">#extract residuals</span></span>
<span id="cb13-54"><a href="#cb13-54" tabindex="-1"></a>  resid <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>residuals)</span>
<span id="cb13-55"><a href="#cb13-55" tabindex="-1"></a>  expt <span class="ot">&lt;-</span> <span class="fu">rep</span>(enviro, <span class="fu">length</span>(resid))</span>
<span id="cb13-56"><a href="#cb13-56" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">cbind</span>(expt, resid))</span>
<span id="cb13-57"><a href="#cb13-57" tabindex="-1"></a>  residuals <span class="ot">&lt;-</span> <span class="fu">rbind</span>(residuals, tmp)</span>
<span id="cb13-58"><a href="#cb13-58" tabindex="-1"></a>  </span>
<span id="cb13-59"><a href="#cb13-59" tabindex="-1"></a>  </span>
<span id="cb13-60"><a href="#cb13-60" tabindex="-1"></a></span>
<span id="cb13-61"><a href="#cb13-61" tabindex="-1"></a>  </span>
<span id="cb13-62"><a href="#cb13-62" tabindex="-1"></a>  <span class="co">#calculate EEV (estimation error variance-covariance matrix for id effects)</span></span>
<span id="cb13-63"><a href="#cb13-63" tabindex="-1"></a>  <span class="co">#extract EEV from the model</span></span>
<span id="cb13-64"><a href="#cb13-64" tabindex="-1"></a>  EEV_lm <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>cov.unscaled<span class="sc">*</span>variance_residual, </span>
<span id="cb13-65"><a href="#cb13-65" tabindex="-1"></a>                     <span class="at">nrow =</span> <span class="fu">sqrt</span>(<span class="fu">length</span>(<span class="fu">summary</span>(models[[enviro]])<span class="sc">$</span>cov.unscaled))) </span>
<span id="cb13-66"><a href="#cb13-66" tabindex="-1"></a>  </span>
<span id="cb13-67"><a href="#cb13-67" tabindex="-1"></a>  <span class="fu">rownames</span>(EEV_lm) <span class="ot">&lt;-</span> <span class="fu">names</span>(models[[enviro]]<span class="sc">$</span>coefficients)</span>
<span id="cb13-68"><a href="#cb13-68" tabindex="-1"></a>  <span class="fu">colnames</span>(EEV_lm) <span class="ot">&lt;-</span> <span class="fu">names</span>(models[[enviro]]<span class="sc">$</span>coefficients)</span>
<span id="cb13-69"><a href="#cb13-69" tabindex="-1"></a></span>
<span id="cb13-70"><a href="#cb13-70" tabindex="-1"></a>  </span>
<span id="cb13-71"><a href="#cb13-71" tabindex="-1"></a>  placeholder <span class="ot">&lt;-</span> <span class="fu">names</span>(BLUEs)</span>
<span id="cb13-72"><a href="#cb13-72" tabindex="-1"></a>  BLUEs <span class="ot">&lt;-</span> <span class="fu">c</span>(BLUEs, models[[enviro]]<span class="sc">$</span>coefficients)</span>
<span id="cb13-73"><a href="#cb13-73" tabindex="-1"></a>  <span class="fu">names</span>(BLUEs) <span class="ot">&lt;-</span> <span class="fu">c</span>(placeholder, <span class="fu">names</span>(models[[enviro]]<span class="sc">$</span>coefficients))</span>
<span id="cb13-74"><a href="#cb13-74" tabindex="-1"></a>  </span>
<span id="cb13-75"><a href="#cb13-75" tabindex="-1"></a>  EEVs[[enviro]] <span class="ot">&lt;-</span> EEV_lm</span>
<span id="cb13-76"><a href="#cb13-76" tabindex="-1"></a>}</span>
<span id="cb13-77"><a href="#cb13-77" tabindex="-1"></a><span class="fu">colnames</span>(residuals) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;resid&quot;</span>)</span>
<span id="cb13-78"><a href="#cb13-78" tabindex="-1"></a>residuals<span class="sc">$</span>resid <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(residuals<span class="sc">$</span>resid)</span>
<span id="cb13-79"><a href="#cb13-79" tabindex="-1"></a></span>
<span id="cb13-80"><a href="#cb13-80" tabindex="-1"></a></span>
<span id="cb13-81"><a href="#cb13-81" tabindex="-1"></a></span>
<span id="cb13-82"><a href="#cb13-82" tabindex="-1"></a><span class="co">#exactly the same BLUEs</span></span>
<span id="cb13-83"><a href="#cb13-83" tabindex="-1"></a><span class="fu">cor</span>(BLUEs, <span class="co">#lm</span></span>
<span id="cb13-84"><a href="#cb13-84" tabindex="-1"></a>    ans1<span class="sc">$</span>blues<span class="sc">$</span>BLUE) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>BLUEs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##  idA01143-3C idAC00206-2W idAC01151-5W idAC03452-2W idAC05153-1W 
##        93.05        74.44        84.78        86.85        49.63</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>ans1<span class="sc">$</span>blues<span class="sc">$</span>BLUE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] 93.05 74.44 84.78 86.85 49.63</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">#Almost exactly the same EEV (differences due to rounding errors)</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>differences <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(pheno<span class="sc">$</span>env)) {</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  differences <span class="ot">&lt;-</span> <span class="fu">c</span>(differences, <span class="fu">mean</span>(<span class="fu">abs</span>(EEVs[[enviro]] <span class="sc">-</span> ans1<span class="sc">$</span>vcov[[enviro]])))</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>}</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>differences</span></code></pre></div>
<pre><code>##  [1] 1.182825e-13 1.135094e-14 1.023151e-13 2.678651e-14 1.573423e-14
##  [6] 1.180774e-15 1.083396e-14 1.247924e-15 4.134694e-13 1.098241e-14
## [11] 1.676807e-15 2.353688e-14 2.644988e-11 1.327111e-14 8.166087e-15
## [16] 2.626889e-15 3.900033e-15 1.882794e-13 7.213307e-13 1.401704e-13
## [21] 1.109521e-14 7.708802e-14 4.962385e-15 5.832111e-14 4.407656e-15
## [26] 6.265049e-15 6.994867e-15 4.701868e-15 6.516183e-15 1.444187e-14
## [31] 2.074527e-14 1.241832e-14 1.671286e-15 2.882401e-14 1.756560e-14
## [36] 2.070228e-14 3.972099e-14 6.412185e-14 3.626199e-14 3.475709e-13
## [41] 3.984126e-13 1.610082e-14 2.769563e-14</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>EEVs<span class="sc">$</span>NC_2011[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##              idA01143-3C idAC00206-2W idAC01151-5W idAC03452-2W idAC05153-1W
## idA01143-3C     4.133311     0.000000     0.000000     0.000000     0.000000
## idAC00206-2W    0.000000     8.266623     0.000000     0.000000     0.000000
## idAC01151-5W    0.000000     0.000000     4.133311     0.000000     0.000000
## idAC03452-2W    0.000000     0.000000     0.000000     8.266623     0.000000
## idAC05153-1W    0.000000     0.000000     0.000000     0.000000     8.266623</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>ans1<span class="sc">$</span>vcov<span class="sc">$</span>NC_2011[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## 5 x 5 Matrix of class &quot;dspMatrix&quot;
##                A01143-3C    AC00206-2W    AC01151-5W    AC03452-2W
## A01143-3C   4.133311e+00 -1.332268e-15 -4.440892e-16 -1.332268e-15
## AC00206-2W -1.332268e-15  8.266623e+00 -4.440892e-16 -1.332268e-15
## AC01151-5W -4.440892e-16 -4.440892e-16  4.133311e+00 -1.332268e-15
## AC03452-2W -1.332268e-15 -1.332268e-15 -1.332268e-15  8.266623e+00
## AC05153-1W -8.881784e-16 -8.881784e-16 -8.881784e-16 -8.881784e-16
##               AC05153-1W
## A01143-3C  -8.881784e-16
## AC00206-2W -8.881784e-16
## AC01151-5W -8.881784e-16
## AC03452-2W -8.881784e-16
## AC05153-1W  8.266623e+00</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>variance_residual <span class="ot">&lt;-</span> <span class="fu">sigma</span>(models[[<span class="st">&quot;NC_2011&quot;</span>]])<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="co">#design matrix</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> pheno[pheno<span class="sc">$</span>env <span class="sc">==</span> <span class="st">&quot;NC_2011&quot;</span>,]</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>phenoenv <span class="ot">&lt;-</span> phenoenv[<span class="sc">!</span><span class="fu">is.na</span>(phenoenv<span class="sc">$</span>Yield.Mg.ha),]</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>X.id <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(Yield.Mg.ha<span class="sc">~</span>id<span class="dv">-1</span>, <span class="at">data=</span>phenoenv) <span class="co">#remove intercept.</span></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>EEVNC11 <span class="ot">&lt;-</span> <span class="fu">solve</span>(<span class="fu">t</span>(X.id)<span class="sc">%*%</span>X.id)<span class="sc">*</span>variance_residual</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">abs</span>(EEVs<span class="sc">$</span>NC_2011 <span class="sc">-</span> EEVNC11)) <span class="co">#the only differences are rounding errors</span></span></code></pre></div>
<pre><code>## [1] 2.04281e-14</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co">#merge EEV matrices from each environmet for their use in the second stage</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="co">#It is done using the direct sum operator</span></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>()</span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">unique</span>(pheno<span class="sc">$</span>env))) {</span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>  env1 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno<span class="sc">$</span>env)[i<span class="dv">-1</span>]</span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>  env2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(pheno<span class="sc">$</span>env)[i]</span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(<span class="fu">is.na</span>(placeholder_matrix)) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(EEVs[[env1]],EEVs[[env2]])</span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>    placeholder_matrix <span class="ot">&lt;-</span> <span class="fu">direct.sum</span>(placeholder_matrix,EEVs[[env2]])</span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a>  }</span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a>}</span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a>EEV_full <span class="ot">&lt;-</span> placeholder_matrix</span>
<span id="cb27-14"><a href="#cb27-14" tabindex="-1"></a><span class="fu">dim</span>(EEV_full)</span></code></pre></div>
<pre><code>## [1] 4195 4195</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">length</span>(BLUEs)</span></code></pre></div>
<pre><code>## [1] 4195</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co">#Mostly similar heritability estimations</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="fu">cor</span>(H2, <span class="co">#lm</span></span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>    ans1<span class="sc">$</span>fit<span class="sc">$</span>H2) <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>## [1] 0.9334192</code></pre>
<p>The figure above illustrates the variation in broad-sense
heritability across locations and years.</p>
<p>When the data frame of BLUEs passed to <code>Stage2</code> includes a
column labeled âlocâ, genotype x location effects are incorporated using
a separable covariance structure. The genetic covariance between
locations for the highest-order genetic effect (i.e., additive when
marker data are included) follows a 2nd order factor-analytic (FA2)
model. (For non-additive genetic effects, the correlation is constrained
to be 1.) For large datasets with many locations, such as this one, I
recommend initially analyzing it without marker data, as computation
proceeds more quickly.</p>
<p>Here is a joint analysis of the 8 locations without using the marker
data.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co">#Warning!</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a><span class="co">#This works for version 1.5-1 of the Matrix package, but it may fail </span></span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a><span class="co">#if you are using later versions</span></span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a>ans2a <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>ans1<span class="sc">$</span>blues, <span class="at">vcov=</span>ans1<span class="sc">$</span>vcov, <span class="at">silent =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="co">#Open source</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="co">#1) Prepare data</span></span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>Stage1Output <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">BLUEs =</span> BLUEs,</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>                       <span class="at">id =</span> <span class="fu">names</span>(BLUEs),</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>                       <span class="at">Env =</span> Envs,</span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a>                       <span class="at">Loc =</span> Locs)</span>
<span id="cb34-7"><a href="#cb34-7" tabindex="-1"></a>Stage1Output<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;^id&quot;</span>, <span class="st">&quot;&quot;</span>,Stage1Output<span class="sc">$</span>id) <span class="co">#remove the &quot;id&quot; at the beginning</span></span>
<span id="cb34-8"><a href="#cb34-8" tabindex="-1"></a><span class="co">#of each genotype name</span></span>
<span id="cb34-9"><a href="#cb34-9" tabindex="-1"></a></span>
<span id="cb34-10"><a href="#cb34-10" tabindex="-1"></a><span class="co">#We need to add to Stage1Output a column for stage 1 estimation error </span></span>
<span id="cb34-11"><a href="#cb34-11" tabindex="-1"></a>Stage1Error <span class="ot">&lt;-</span> <span class="fu">paste0</span>(Stage1Output<span class="sc">$</span>id,<span class="st">&quot;:&quot;</span>,Stage1Output<span class="sc">$</span>Env)</span>
<span id="cb34-12"><a href="#cb34-12" tabindex="-1"></a><span class="fu">rownames</span>(EEV_full) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb34-13"><a href="#cb34-13" tabindex="-1"></a><span class="fu">colnames</span>(EEV_full) <span class="ot">&lt;-</span> Stage1Error</span>
<span id="cb34-14"><a href="#cb34-14" tabindex="-1"></a></span>
<span id="cb34-15"><a href="#cb34-15" tabindex="-1"></a><span class="co">#model:</span></span>
<span id="cb34-16"><a href="#cb34-16" tabindex="-1"></a><span class="co">#BLUEs = XE + Zg + s + gE</span></span>
<span id="cb34-17"><a href="#cb34-17" tabindex="-1"></a><span class="co">#We need a column for BLUEs, environment name, genotypic id and stage1error</span></span>
<span id="cb34-18"><a href="#cb34-18" tabindex="-1"></a>sommerdf <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Stage1Output,</span>
<span id="cb34-19"><a href="#cb34-19" tabindex="-1"></a>                  Stage1Error)</span>
<span id="cb34-20"><a href="#cb34-20" tabindex="-1"></a><span class="fu">head</span>(sommerdf)</span></code></pre></div>
<pre><code>##     BLUEs          id     Env Loc         Stage1Error
## 1  93.050   A01143-3C CA_2011  CA   A01143-3C:CA_2011
## 2  74.440  AC00206-2W CA_2011  CA  AC00206-2W:CA_2011
## 3  84.780  AC01151-5W CA_2011  CA  AC01151-5W:CA_2011
## 4  86.850  AC03452-2W CA_2011  CA  AC03452-2W:CA_2011
## 5  49.630  AC05153-1W CA_2011  CA  AC05153-1W:CA_2011
## 6 105.455 Accumulator CA_2011  CA Accumulator:CA_2011</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co"># When you want to use a known variance-covariance matrix in Sommer&#39;s mmec() function,</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="co"># what you actually have to provide is the inverse of the matrix, and it must be of the</span></span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a><span class="co"># &quot;dgCMatrix&quot; class (sparse matrix).</span></span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a>EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(EEV_full),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a><span class="co"># Fix the variance for stage1Error to be 1. We have to scale it by dividing it</span></span>
<span id="cb36-7"><a href="#cb36-7" tabindex="-1"></a><span class="co"># by the variance of the response variable because Sommer works that way.</span></span>
<span id="cb36-8"><a href="#cb36-8" tabindex="-1"></a><span class="co"># We fix the variance to be 1 because we don&#39;t want the model to scale EEV in </span></span>
<span id="cb36-9"><a href="#cb36-9" tabindex="-1"></a><span class="co"># any way, but 1 is not the true stage1Error variance component. More on that later.</span></span>
<span id="cb36-10"><a href="#cb36-10" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf<span class="sc">$</span>BLUEs))</span>
<span id="cb36-11"><a href="#cb36-11" tabindex="-1"></a></span>
<span id="cb36-12"><a href="#cb36-12" tabindex="-1"></a></span>
<span id="cb36-13"><a href="#cb36-13" tabindex="-1"></a><span class="co">#Factor analytics</span></span>
<span id="cb36-14"><a href="#cb36-14" tabindex="-1"></a><span class="co">#1) Extract loadings using sommer functions</span></span>
<span id="cb36-15"><a href="#cb36-15" tabindex="-1"></a>FA2LocLoadings <span class="ot">&lt;-</span> <span class="fu">with</span>(sommerdf, <span class="fu">rrc</span>(Loc, <span class="co">#Factor analytics for location</span></span>
<span id="cb36-16"><a href="#cb36-16" tabindex="-1"></a>                                   id, <span class="co">#Variables characterizing each location</span></span>
<span id="cb36-17"><a href="#cb36-17" tabindex="-1"></a>                                   <span class="co">#are the genotypes</span></span>
<span id="cb36-18"><a href="#cb36-18" tabindex="-1"></a>                                   <span class="co">#Gu = geno@G, #I get inversion issues if I use G</span></span>
<span id="cb36-19"><a href="#cb36-19" tabindex="-1"></a>                                   BLUEs, <span class="co">#response variable</span></span>
<span id="cb36-20"><a href="#cb36-20" tabindex="-1"></a>                                   <span class="at">nPC=</span><span class="dv">2</span>, <span class="co">#FA2 --&gt; 2PCs</span></span>
<span id="cb36-21"><a href="#cb36-21" tabindex="-1"></a>                                   <span class="at">returnGamma =</span> <span class="cn">TRUE</span></span>
<span id="cb36-22"><a href="#cb36-22" tabindex="-1"></a>                  )) </span>
<span id="cb36-23"><a href="#cb36-23" tabindex="-1"></a></span>
<span id="cb36-24"><a href="#cb36-24" tabindex="-1"></a></span>
<span id="cb36-25"><a href="#cb36-25" tabindex="-1"></a><span class="co"># 2) You can directly use the sommer loadings, but with this dataset there were</span></span>
<span id="cb36-26"><a href="#cb36-26" tabindex="-1"></a><span class="co"># matrix inversion issues. To solve them, we can perform a varimax rotation to obtain</span></span>
<span id="cb36-27"><a href="#cb36-27" tabindex="-1"></a><span class="co"># alternative but equivalent loadings that don&#39;t cause inversion issues in the model.</span></span>
<span id="cb36-28"><a href="#cb36-28" tabindex="-1"></a>RotatedLoadings <span class="ot">&lt;-</span> <span class="fu">varimax</span>(FA2LocLoadings<span class="sc">$</span>Gamma)</span>
<span id="cb36-29"><a href="#cb36-29" tabindex="-1"></a><span class="co"># We recommend using the original FA2LocLoadings by default and only using the varimax </span></span>
<span id="cb36-30"><a href="#cb36-30" tabindex="-1"></a><span class="co"># rotation if you encounter inversion issues.</span></span>
<span id="cb36-31"><a href="#cb36-31" tabindex="-1"></a></span>
<span id="cb36-32"><a href="#cb36-32" tabindex="-1"></a><span class="co"># Expand the loading matrix using the locations design matrix.</span></span>
<span id="cb36-33"><a href="#cb36-33" tabindex="-1"></a><span class="co"># The expanded loadings matrix will be the input to the factor analytics term</span></span>
<span id="cb36-34"><a href="#cb36-34" tabindex="-1"></a><span class="co"># in the model.</span></span>
<span id="cb36-35"><a href="#cb36-35" tabindex="-1"></a>Zl <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Loc<span class="dv">-1</span>, <span class="at">data=</span>sommerdf)</span>
<span id="cb36-36"><a href="#cb36-36" tabindex="-1"></a><span class="co">#FA2Loc &lt;- Zl%*%FA2LocLoadings$Gamma #without rotation</span></span>
<span id="cb36-37"><a href="#cb36-37" tabindex="-1"></a>FA2Loc <span class="ot">&lt;-</span> Zl<span class="sc">%*%</span>RotatedLoadings<span class="sc">$</span>loadings <span class="co">#with rotation</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="co">#sommer model. It is quite slow</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="co">#If you are using the full dataset, it will probably take a few hours</span></span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>ans2a_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Fixed effects = Environment. -1 means that</span></span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a>                     <span class="co">#we remove the intercept (you could leave it if you prefer it)</span></span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id)) <span class="sc">+</span> <span class="co">#vsc() is just an auxiliary function</span></span>
<span id="cb37-6"><a href="#cb37-6" tabindex="-1"></a>                       <span class="co">#used every time you want to specify details for a random</span></span>
<span id="cb37-7"><a href="#cb37-7" tabindex="-1"></a>                       <span class="co">#effect.</span></span>
<span id="cb37-8"><a href="#cb37-8" tabindex="-1"></a>                       <span class="co">#isc(id) means that we assume identity variance-covariance</span></span>
<span id="cb37-9"><a href="#cb37-9" tabindex="-1"></a>                       <span class="co">#structure for the &quot;id&quot; effect (genotypic effect)</span></span>
<span id="cb37-10"><a href="#cb37-10" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">usc</span>(FA2Loc),<span class="fu">isc</span>(id)) <span class="sc">+</span> <span class="co">#kronecker factor analytics</span></span>
<span id="cb37-11"><a href="#cb37-11" tabindex="-1"></a>                       <span class="co">#matrix (usc(FA2Loc)) with identity for genotypes (isc(id))</span></span>
<span id="cb37-12"><a href="#cb37-12" tabindex="-1"></a>                       <span class="co">#in usc(FA2Loc) we indicate that the 2 factors</span></span>
<span id="cb37-13"><a href="#cb37-13" tabindex="-1"></a>                       <span class="co">#that characterize the locations follow a unstructured</span></span>
<span id="cb37-14"><a href="#cb37-14" tabindex="-1"></a>                       <span class="co">#variance covariance matrix</span></span>
<span id="cb37-15"><a href="#cb37-15" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb37-16"><a href="#cb37-16" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb37-17"><a href="#cb37-17" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb37-18"><a href="#cb37-18" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb37-19"><a href="#cb37-19" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb37-20"><a href="#cb37-20" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb37-21"><a href="#cb37-21" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb37-22"><a href="#cb37-22" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb37-23"><a href="#cb37-23" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb37-24"><a href="#cb37-24" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb37-25"><a href="#cb37-25" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb37-26"><a href="#cb37-26" tabindex="-1"></a>                           </span>
<span id="cb37-27"><a href="#cb37-27" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb37-28"><a href="#cb37-28" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb37-29"><a href="#cb37-29" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb37-30"><a href="#cb37-30" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb37-31"><a href="#cb37-31" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb37-32"><a href="#cb37-32" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb37-33"><a href="#cb37-33" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb37-34"><a href="#cb37-34" tabindex="-1"></a>                           </span>
<span id="cb37-35"><a href="#cb37-35" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb37-36"><a href="#cb37-36" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb37-37"><a href="#cb37-37" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb37-38"><a href="#cb37-38" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb37-39"><a href="#cb37-39" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb37-40"><a href="#cb37-40" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb37-41"><a href="#cb37-41" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb37-42"><a href="#cb37-42" tabindex="-1"></a>                      <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">dsc</span>(Loc),<span class="fu">isc</span>(units)), <span class="co">#diagonal loc by i.i.d. residuals</span></span>
<span id="cb37-43"><a href="#cb37-43" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb37-44"><a href="#cb37-44" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb37-45"><a href="#cb37-45" tabindex="-1"></a>                     <span class="at">nIters=</span><span class="dv">50</span>, <span class="co">#increased number or iterations</span></span>
<span id="cb37-46"><a href="#cb37-46" tabindex="-1"></a>                     <span class="co">#more EM iterations at the beggining for usc models </span></span>
<span id="cb37-47"><a href="#cb37-47" tabindex="-1"></a>                     <span class="at">emWeight=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">10</span>),sommer<span class="sc">::</span><span class="fu">logspace</span>(<span class="dv">10</span>,<span class="dv">1</span>,.<span class="dv">05</span>),<span class="fu">rep</span>(.<span class="dv">05</span>,<span class="dv">80</span>)), </span>
<span id="cb37-48"><a href="#cb37-48" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf,</span>
<span id="cb37-49"><a href="#cb37-49" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb37-50"><a href="#cb37-50" tabindex="-1"></a></span>
<span id="cb37-51"><a href="#cb37-51" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="co">#very different aic</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>ans2a<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 22921.91</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a>ans2a_open<span class="sc">$</span>AIC</span></code></pre></div>
<pre><code>## [1] 130105</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>## $var
##              Variance   PVE
## genotype         53.0 0.360
## g x loc          14.7 0.100
## g x env          26.9 0.183
## Stage1.error     52.6 0.358
## 
## $cor
##       CA    FL    MI    MO    NC    NY    OR    WI
## CA 1.000 0.636 0.691 0.736 0.837 0.686 0.870 0.753
## FL 0.636 1.000 0.670 0.716 0.696 0.702 0.734 0.720
## MI 0.691 0.670 1.000 0.935 0.860 0.930 0.912 0.934
## MO 0.736 0.716 0.935 1.000 0.919 0.996 0.974 1.000
## NC 0.837 0.696 0.860 0.919 1.000 0.899 0.949 0.924
## NY 0.686 0.702 0.930 0.996 0.899 1.000 0.954 0.994
## OR 0.870 0.734 0.912 0.974 0.949 0.954 1.000 0.979
## WI 0.753 0.720 0.934 1.000 0.924 0.994 0.979 1.000</code></pre>
<p>In the single-location analysis, running the <code>summary</code>
command on the âvarsâ output displays the partitioning of variance. In
this case, a genotype x location effect is also present. However,
because genotype x year effects are often not repeatable and tend to be
small, they are not included in the <code>Stage2</code> model.</p>
<p>The FA2 model should provide enough complexity for a set of
correlated locations used within a single breeding program. However, it
may be inadequate when analyzing disparate locations. The factor
loadings returned by <code>Stage2</code> can be used with the
<code>uniplot</code> function to visualize the modelâs sufficiency and
correlation structure (<a href="https://doi.org/10.1139/G10-080">Cullis
et al.Â 2010</a>).</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="co"># Different uniplot</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a><span class="co"># Please note that the uniplot() function belongs to the StageWise package, but </span></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a><span class="co"># it does not rely on ASReml and can be used without any license.</span></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="fu">uniplot</span>(ans2a<span class="sc">$</span>loadings)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAw1BMVEUAAAAAADUAADoAAF4AAGYAAJAAANsAAP8ANYQAOpAAXqgAZrYzMzM1AAA1ADU6AAA6ADo6AGY6Ojo6OpA6kNteAABeADVeqOtmAABmADpmAGZmOpBmZmZmkJBmtrZmtttmtv+EyeuQOgCQOjqQOmaQkGaQtpCQ27aQ2/+oXgCo6+u2ZgC2Zjq225C2/9u2///JhDXJ6+vbkDrbtmbb/7bb/9vb///rqF7ryYTr66jr6+v/AAD/tmb/25D//7b//9v///+rMZbOAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALmUlEQVR4nO2dbX/bNBDAxVqWdWHrxrKtZQPSwqAQICywssVO6u//qbBkO/GTHk4+WedI92LJb+ezTv/IsnQ6qSyLohTm2wHqwnw7QF2YbweoC/PtAHVhvh2gLsy3A9SFKbXfBCxmgGrft4rrVLqJGkZAGmUEpFFGQBplBKRRRkAaZQSkUUZAGmUEpFFGQBqlIaCtD0na4sMJii2o4tFRthVoJSqUtAB16t9vWLsqHEC9aFSGhUEYgORsdHdVW54EoKKKQxoCtOnpdYQAYXUl/YimDqjvRQW9a+MNZ2WoUXoD5OJt3UY0YUCQqkBKbN53soBgDwOsxPq9pwmo952DOZw5FjBFQJI3MvJ4ryxlgoBkozr0AXFia9hWjgpIPurFnzHohp8UAVlOCmxLTJJJAbKebg34SWznaT4AITV4qKGcEC1A4qf0AUjeiEgBwnvpWhhazPTHBmQwY3cIyGKmPyog3HGtneGQsbtrQMgzI1vDLiEigExDGq4BdQnRAIQefLA3bBMiAShRKY11SIaWESiXgBzE94YYmgfTxgHUeXX4BmTeI44CCPbeGAeQ6TvVAhB4zd9LooFe4G45akF9g1cCLajmmN9HzNky3nBDk3G9c0DwuPN4gExmhq4BOVwpxzBMtPFqt4CsIjBjAip+QW+A7GJ44wLSxTddAnKcrYNlaJ55hQzINjAfAekMVaW7MFQuCbkDpOJDC5Cxq6iAjH8WkM6RoWFjxwQ0Qs4ppqHZ+yRgQGYjEjxAQ5NVPRia5FKgAQKN30E6h4YGg34sQPrF0whI4xFFQAbzaiRAY20tQTfURmZwAEGjUCBdBOTVUBf8RAFUFrJhjD26K78uVKWCdKcC6OH24n6bk1mK7+/P14pSVR7BlMMNNesLwwHlzUZ8ri7uuW6Tf2Tp+Z+XS0WpKo9gSgTDnhAoJiDGCkL7q+VRt7r4dDtXlKryCKbEMBwH0O7F3UG3v1pkm84zRhaQMjcGD9DL9UGX5nB2nWcsUEBVH1RvQas576fbzxhdQKrkocFr89Vq95fXb4vPNzdfXotmdfa7xeK5H1Es2Q9tQUf2m/Itdr4uup/OM0a4BSlSmQYCqt34MA4qH67OMxY6oLzrKUbSVdPZlKPqfpekHsGUWIbSbMFhgGwzI0G6CIiEoSyBMQIqxQkgh5uXxzeUZHhGQJU4AGS/AYKkYX8KbAR0lKRPGQEdBRuQ4wMUPBj25QhHQDXBBQTNFY+ANB5NAFDf8pU1oEE7ZakaRkA6XXcJ3RbQKKfcjG8YAWl0aIAGHmdA1jAC0uk6eU4RUFMiII0OCdBoR5GNb9hORo2AWhIBaXQRkEaHAmjE0/7GN2zt17RKXiB6agCOtCoXW1Bb7FpQ884RUFsiIBAgu53fUwHU2jM+AND+SiQApSJdqkgip1RPa0NEQCJpnAPiyVMc0lvVTQIE9P2r+wIQTyLPZfPk3sqlkwW0XC0EIJFErvNoMoCae28tAFV9dI6FJ0fngEQKsM6jIAFlm4UA9HKt9yhMQPvru9NrQY0DEgYCyjbzWh+0f3OXySVMQA8/fcjfYpvyLXZ23MDycCvS7Xna9GKYu9MGlKWsfxx04BI6oGwlRtKr9kg6AtKUGgFpSo2ANKWWnfQiUEDVTl55qeG2oJJPL6EIKAsCUD0kCAYkyERALTkCKtGA+qCytw4KkKxUpVJRQCiAlNowAMkeroMyeEDKIyzDBSRfu25KzsfBcvnoUqsjbgsi1hDoPWLE6hkBRUCeDJEApSLmnB3jzwti9fQMSMRa91cX99W0Ynd5EmvzGRKgciWDw6nmXaunteuK9foNm4uL5rsXN7buThPQw+2y+JKer3sBibSYh9v3nNP+ajElQCgh18Nqav6lBJQ20l/EKTn761+v8+vSR3fhAtpfLQ+LhPVSBbX04hP/yB/HgAFdFy1od3kM+cxms6KTWi2yTd4J5R/BAer2QWnRgmaCTqVY8vMD+RJjcICqkyRrb7HV+QEOl7xp8byYMo0oPECtcVCO5uvLpw2r1YI/XuVHgIAOI+nZY/ZYNJzWIcmbOU/RKz9CBMSl8VC1St19Kzry4iMcQIeY4awJp1uqePyyMjcmGEBFVJWz6Qmu0qqnreGwHMUcj2g4veFnUvX0B0iAEaAgHoUCqLa2WrYlQ49CA1R11g1EpOppazh0KwJvOI2ra82IUj2tDQfv1eCEWvqKEaV6WhuibGbpdM8FI0r1tDYcDkgQ6iLK/49SPW0NMfasiv65h1DWGVobuaTRTQ9Q+R6b8SCZiHmUc4q0++avXbB7rtjIMXlAjaV9ASj/nG0/P/vqJv/y+dmTj9vtb/zPIHx5/eRjzqi6tHbB5+c342QeDBSEgwWOs4zd5XsR0OBLF+VfRSiCZ1Uzql0wkRaEcnbHYZ66u/zh3T3f8XNxL0Kw/MbFLuiyO6pdEBSgKiS0u/zx53W2e/UPX7Uo/zLLIZrPm1H9gjABLTfLLF2kNUCHHb7tCwIFlM6z1bIB6LoGon7BJADhnGE2q9af8/rv3/373Trt9kGdC0IFlP3xYZ6l/C2Wg9nWEsjbFwQLiC9lcEB8GWhbjRkrqV0QFqDyRrz+vPNJjyPpZcOwdsEkAOEcVTo73gng0STmYniA4nnSDQkFENKJ5LPGzUw9ioA0HgUHKP5djZpEQBEQFyxAnfsZeUQfUNKnjICOEgGpdUmv0hxQda7k2V+35V/inXcIUainrWHSr4QAKs6VPFvzjGixF+OUACUSJQRQca7k2VqEfsROgzYh//W0NsQAVJwrebYWTxv/k+kRUP2q6lxJfpBb+uiXYu2iRch/PW0NE5kSBEicK3kmdjmVGdER0PGq6lxJcRRgFZk/FUDyepivzX95/Xa7/fvpf3wFflv8213JnqrIqwFrQXwRudGC5J1bVwi3IMXLBggoS1kLkGyApXEJoJsWoGzVBiQZomtcAujcG6oGvEPmYt3bTxOQck6JAKh/Fqx0CaSLgDwbqgN/GIB6I3HKUkE6x4aaxYfgAenW91AA9S2XKEsF6U4BUM+SrbJUkM6poTbHIHBAksOQ8AF182qUpYJ0pwGok/yoLBWkc2goO03LBaB2Cr+yVJDOnaH0tDEngFrbiJSlgnQnA6i511NZKkjnzFB+XF0ExHWK4/wcAWqcWKAsFaRzZGh43CEqIOUZi8QAmbqKC8j0ZwHpnBgaN3ZkQIYPNkjnwtC8uwwTEGBEgg3I7N0J0p0YoCwxGL+DdPiGkEE/PiCTGSBIh22YgKaNLgBZbOQYERAwMuMEEHyfwniAoMFPN4DAafijAQKHz20OFhiWDeBV4G45akHQFNiRWpDFIrAzQN3XvX9ANnkW7gCBsqvGANT8xSgAkue16QzNleaGlslwTgGZZ1e5B2SbkOsWUKNZ+wTUnf8QAWSa+uEY0ICdW84BHX88f4CG7B51D8ho9uMSUH94gRIgg/mzO0Cy6AspQPoIjDNAJmuDSuVIgHQxPFeA7OKbPgDlTd12xcMekKpIcoBypcJfB4DyXwSjzY4JCKFHMNdh9XrjAhr8TjHV4b03RwYk64twASWII6/RAWW9zQgRUO0XmCogYDANFj8wDYrRBtRBhAQIcFfqgFqdEQagbu82bUBZvUqDAfX2/ZMHlB0eioGAJKOHUwAkfvtByY2JfBJzGoC4KCqpNFTbnRAgMQ2R1lbybIrLXYUBDkIHkJCkD1P39YfYu+uUrtbmB0milHF9IdaC6BlGQBplBKRRRkAaZQSkUUZAGmUEpFFGQBplBKRRRkAaZQSkURoCCliMAEWJgHTCfDtAXZhvB6gL8+0AdWG+HaAuzLcD1IX5doC6MN8OUBfm2wHqwnw7QF2YbweoC/PtAHVhvh2gLsy3A9SF+XaAujDfDlAX5tsB6sJ8O0BdmG8HqAvz7QB1Yb4doC7MtwPUhfl2gLow3w5QF+bbAerCfDtAXZhvB6gL8+0AdfkfqpMUW23lUsEAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a>Gamma <span class="ot">&lt;-</span> RotatedLoadings<span class="sc">$</span>loadings</span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a><span class="fu">uniplot</span>(Gamma, <span class="at">nudge =</span> <span class="fl">0.3</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAyVBMVEUAAAAAADoAAF4AAGYAAP8ANYQAOpAAXqgAZrYzMzM1AAA1ADU1hMk6AAA6ADo6AGY6Ojo6OpA6kNteAABeADVehIReqOtmAABmADpmAGZmOpBmZmZmkJBmtrZmtttmtv+ENQCEyeuQOgCQOjqQOmaQkGaQtpCQ27aQ2/+oXgCo6+u2ZgC2Zjq225C2/9u2///JhDXJ6+vbAADbkDrbtmbb/7bb/9vb///rqF7ryYTr66jr68nr6+v/AAD/tmb/25D//7b//9v////KruVzAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALRklEQVR4nO2dDZujthWFvRniOM26babtzjTb1qmTtLNx2tmN28xuCv7g//+oIvFhEBJHiCuBzT3Pk7HXB8nwWoB0pUsWKatTi7F3YOpiQEAMCIgBATEgIAYExICAGBAQAwJiQEAMCKgb0Jczlh2g2vvYvJmbNe0aGRCwGBCwGBCwGBCwGBCwbhXQebuQ2py393bFdssP2d/9YpX9PW9Xx98+yY9vF1DBxRrQ4e5HsfXb5Uuanh7uGZCq4/pNRubxn48ZmMOrJwak6rx9nZFZ/iK23y9fegKKgyrRqV8V//v2tfIG6v3nH+J3r+P3X8Ti5dNXf5efTqoFlSy0xVTP6iJ9n1q3oPTw2fN5u0mPv3s+PWQvEzvFFDDmYhUm6lMsPf3xKYOTXaA34mVCgAxsOot1FMmLOQBK393vxT1+J1+mAkgcqGONnVxdAL1f7cS2e/kyCUDFEQ6o0cDIDdCn30soR/kyPqDLsQ2rUceoAai4WuMaf30QncSshHgZGZDpVuVWYwvRtY/FlAPyWmPUt0alwAiA/PzepjbZF5C6fXBA2ivGoBorXaque8oRgxpVPqEBaW86dFeMsnpnQC0+QQGZ+i2UPStNx6F50J01tvkEBGTu+lJ3PZ0BafiEA9QxMqAevKh988Zxd9So4xMKEBg5EVvKuWwJKBqxH9Q53vISHzASMhaLxusotq4KdsX6WMf13XMqYslysHAQg4uHi9sAdHoQQ/b0cPfvrSxzXK+KbcYChAakNIBeyXHmWgDaifD76WHxUtk1QhmgxSYVgJ6PazFC20mmkenLfAMqLwfeAb2VkQoxJ7FfvgjvvF1X51kT0F++fpGA0n3233G9KTcYA5C2b4uL9baO679+85Kev/th+SLCptLLEJRf3wS0EfEeAUicbbtV5Y8AyDaCPNg6rv/2fdYavv5PPh0hPfGmTSguQqoCUHp49Q8R1Yg6vswrILuQBg2gzX6THu4PNUAi9F72L5qA0v19Dui8FYGiSFdjKZ+AErPVVczFygAdstNl0wD0+HTZi4qQBJRZEpDAFEXaGkv5A9TsrPkHdPrmv39+PjSvQfmOZPd8BVC6X5WAIkONpbwBSsxWVzFHS9yLfvphlR7EXezuOb+LFWFWGXK9nEcSUHY5L1sQ+DJfgHoExagA7bPryaHWD1oW/aB8mUcTUNaXlGjuqAD1m/eN454TxYP16Tdv5GTxx88/ZP/6KIi8Kb1FQaj4569/kM67z36M4+ij+NspPy2oV9TQy1isrkYT6htL8wHIbg65T409LdVbRJerUM2yCld7AGSYykPFCC3Vk3wSFVAr/BMIUO+wqndABYmsHdWtdngsDCBDZGxEQCWf/ESLm5/CGqkB9V2mgWt0sLRX4uJmFjc/hTUSA3IJzI8BSBt+niegC4oaIC2fEIDMofmxADVQZHySWP0Q1EgKyG1qxyegNoraqMOqRkpACU2vZbjV4Qk+pmP1Dch1LV1oQAvDse4z59VT8bZacDVDQFEUaU6y81bkauzljMd5+7Ya5ZMBApNf4QDJSbH8MBfN1XdC5dA+UjHt8rmQ/VLOePxLznUIUQEasBqTFtBO/PZytWERMTtWxypVEJLva5hElKiqcbf8Zbsq3hMBgpNfoQDJFpDGAk4ZUtyt6hvEsv3UD1dCqiLZqcz1kVNmUjcGSASkpXe4ezYAkoOy1kVIzASVNR6q+cSUCpBmYZeiQICKxbuxeFMAOiw2rWJRohKqtyBB9FyeYySAhi149gIou6ZUSYftYlGqEiqvQafHp9ODLFacYzcL6DFvQfkChVaxNqFiRj+79uyLRR85WApANhPwo12DDroWJIYgSfM2X/WDipOrPMcIAFlNwAe7i8mfv34X29VnduqA1JHaLu9Jl01nn/eqbw2Qrh+k3MWkdIQ8jcUGJ13460nnl599oy/dAGSxfPH2ANkVE2RUQn4AWS7hmBqgsgk11lbptp8zIJXQEEDGqevQk/B0EnP1SfnGrIEtiCLxa5wWdGlCXleYXTEgqQYhD4DsV0lNGlAa+VoGfPWAGstg6QHR5FaGBHR6+EK85CvvRJR+8VQ4MhWhltFAkvV8jYDKPIQ0z/7eV2PZSAKqMhooALXmCa8B0J/KPIQ8RyP9uVzKKNKhahkNBICosnPDAnpT5CHkazkbRxHVMxrmC6jIQygBNIIRSkaDkDugfgsRpwOoyEMQnHLV4n1KRoPQDAHleQhVC6pHjJWMBqEZAsrzEMprUIarAajMaBgMiO4JAcEB5XkI+RSjjNFWs1a1jIY5A8rzEGr9oAagIqNh1oCKQP6uXO/SACTIEADqu1p8IoAMVsfMOQMSYkDAogdE+RiXCQDqWN7EgKQYELCoAfXPWGFATvs1pBhtjcanwzKgXAwIWLSAHJLCbh1Qczr2euebzTIdE7egQqZEAQZUiAEBixIQ8RMjpwHIlKzEgEoxIGAxIGAxIGDRAXLK/WZArvvlXIy8Rn3SNgOqxICAxYCAxYCAxYCAxYCAxYCAxYCAxYCARQWI/H+QMRVA+iNjQBcxIGAxIGAxIGAxIGAxIGANAWQxR3v90h4Zt6CL+BQDFgMCFgMCFgMCFgMCFgMCFgPqtjhgBiwGBCwGBCwGBCwGBCwGBCwGBCwGBCwGBCwGBCwG1G3xMmBgMSBgMSBgMSBgMSBgEQLihDrjVsU3MCDTVsU3MKDWVhaT2FcuTgvvtvjRFMBiQMBiQMAiBsRPoDJsVVXDgPRbVdUwIP1WVTU3BogfNAksBgQsBgQsD4Cu9HHJeisxWwxIiAF1W4nZGgDoOh/6z4AciiVmKx0CSEOIATWquRFAidkSmj2gxGxJDQDUJsSAlGpUQlcIKDFbuQYBUgkxILWaqweUmK1CwwAphBhQq5orB2Sx+wMBWfwE0LoJQL1nbK9CNjs/tAXh2wC0xmtBNveYOQOy6sYNBgS7otBiQMAaC5DdSGk4IDQchtbNAwIBFWiNBMgyGDFbQLYBUQpA3UFdaI0CKAkKqHPeBFpjALKf95wnoB5rL2gAdc3ddpTC3s0A6lhg01UKen5q7LOCkAqQeZFfZynkeamx1ypmMkDGperdpYB3S4BM6TKgVLfno8Z+yUqEgAxZsahUaEA9U25nB6jv448oATk+e2lGgNwyxoMC6t3KaQE5ZWuGBNT/RkIMyCVTKiAgh64INSCHPJdggBKXziw5oP6L8EMBchsvzgeQY8SBHlDvNeazA2QfrbOusael81yjnkOnnt3ndMPKeY98tKBU04jGbUG13ZnCKZbvk8WX96vR3lK9IdPj3gDZLx3yDmjQEh1/gKz3yzegYYu8PAKybdmeAQ1cR+kTkOW6Bq+A+twtwgOq7d44gDQ9smkBuuziKIAoctp8A7IYI3oDRJJX6x8QjDJ4AqQb8DjUGAAQilP5AUT1jKMggLJfMzAgQ/NxqDEMoM7JBHpAZjzTBdTxm1IDSlyneMcFJHac9LlMBssp8NxhBQSU0j40xnDLpK4xLCD91YHucCy6pRMHRNL7N1lWA5vJAyIYP2qtxhXuugHJgxmeA1O31Ov/tQMSGhAmVq32WXsLgGo/+8DLPtlDRKYGKC2PbkjPynvHYVxAacdBohq7yt0SoFQOQ3p2skWBQMPfSQASf5JEh0ktllw2mxugXEmSNEjF+o971DjU8jM3P1xJQ8G//qKJtaDp1ciAgMWAgMWAgMWAgMWAgMWAgMWAgMWAgMWAgGUJaMayAsRiQEgMCIgBATEgIAYExICAGBAQAwJiQEAMCIgBATEgIAYExICAGBAQAwJiQEAMCIgBATEgIAYExICAGBAQAwJiQEAMCIgBATEgIAYExICAGBAQAwJiQED/B//SijVmSHjAAAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="co"># The squared radius for each location represents the proportion of variance explained (PVE) </span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a><span class="co"># by the latent factors. With the exception of FL, the FA2 model appears to provide</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a><span class="co"># a good representation of the covariance structure. The numeric PVE values can be </span></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a><span class="co"># obtained as follows:</span></span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a><span class="fu">apply</span>(ans2a<span class="sc">$</span>loadings,<span class="dv">1</span>,norm,<span class="at">type=</span><span class="st">&quot;f&quot;</span>)<span class="sc">^</span><span class="dv">2</span> <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##        CA        FL        MI        MO        NC        NY        OR        WI 
## 1.0000000 0.5383870 0.8735862 1.0000000 0.9006442 0.9963027 1.0000000 1.0000000</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="fu">apply</span>(Gamma,<span class="dv">1</span>,norm,<span class="at">type=</span><span class="st">&quot;f&quot;</span>)<span class="sc">^</span><span class="dv">2</span> <span class="co">#sommer</span></span></code></pre></div>
<pre><code>##        MI        NY        OR        WI        CA        FL        MO        NC 
## 1.0000000 1.0000000 0.9872895 0.9942892 0.9775128 0.9801256 0.9920671 0.9909906</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="co">#different PVE in sommer and StageWise</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a><span class="co"># The cosine of the angle between locations in the uniplot equals the correlation </span></span>
<span id="cb50-5"><a href="#cb50-5" tabindex="-1"></a><span class="co"># due to the latent factors (recall cos(0) = 1). We can use this concept to construct the </span></span>
<span id="cb50-6"><a href="#cb50-6" tabindex="-1"></a><span class="co"># correlation matrix between environments in the sommer output.</span></span>
<span id="cb50-7"><a href="#cb50-7" tabindex="-1"></a><span class="co"># Calculate correlations as the cosine of the angle between loadings:</span></span>
<span id="cb50-8"><a href="#cb50-8" tabindex="-1"></a>Locs_cor <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">nrow</span>(Gamma), <span class="at">ncol =</span> <span class="fu">nrow</span>(Gamma))</span>
<span id="cb50-9"><a href="#cb50-9" tabindex="-1"></a><span class="fu">rownames</span>(Locs_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>cor)</span>
<span id="cb50-10"><a href="#cb50-10" tabindex="-1"></a><span class="fu">colnames</span>(Locs_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>cor)</span>
<span id="cb50-11"><a href="#cb50-11" tabindex="-1"></a><span class="cf">for</span> (loc1 <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb50-12"><a href="#cb50-12" tabindex="-1"></a>  <span class="cf">for</span> (loc2 <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb50-13"><a href="#cb50-13" tabindex="-1"></a>    <span class="co">#find slopes of lines that go from (0,0) to each location in the uniplot</span></span>
<span id="cb50-14"><a href="#cb50-14" tabindex="-1"></a>    m1 <span class="ot">&lt;-</span> (Gamma[loc1,<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">0</span>)<span class="sc">/</span>(Gamma[loc1,<span class="dv">1</span>] <span class="sc">-</span> <span class="dv">0</span>)</span>
<span id="cb50-15"><a href="#cb50-15" tabindex="-1"></a>    m2 <span class="ot">&lt;-</span> (Gamma[loc2,<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">0</span>)<span class="sc">/</span>(Gamma[loc2,<span class="dv">1</span>] <span class="sc">-</span> <span class="dv">0</span>)</span>
<span id="cb50-16"><a href="#cb50-16" tabindex="-1"></a>    <span class="co">#angle from slopes</span></span>
<span id="cb50-17"><a href="#cb50-17" tabindex="-1"></a>    angle <span class="ot">&lt;-</span> <span class="fu">atan</span>((m1<span class="sc">-</span>m2)<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span>m1<span class="sc">*</span>m2))</span>
<span id="cb50-18"><a href="#cb50-18" tabindex="-1"></a>    <span class="co">#correlation from angle</span></span>
<span id="cb50-19"><a href="#cb50-19" tabindex="-1"></a>    correlation <span class="ot">&lt;-</span> <span class="fu">cos</span>(angle)</span>
<span id="cb50-20"><a href="#cb50-20" tabindex="-1"></a>    Locs_cor[loc1,loc2] <span class="ot">&lt;-</span> correlation</span>
<span id="cb50-21"><a href="#cb50-21" tabindex="-1"></a>    Locs_cor[loc2,loc1] <span class="ot">&lt;-</span> correlation</span>
<span id="cb50-22"><a href="#cb50-22" tabindex="-1"></a>  }</span>
<span id="cb50-23"><a href="#cb50-23" tabindex="-1"></a>  </span>
<span id="cb50-24"><a href="#cb50-24" tabindex="-1"></a>}</span>
<span id="cb50-25"><a href="#cb50-25" tabindex="-1"></a></span>
<span id="cb50-26"><a href="#cb50-26" tabindex="-1"></a><span class="co"># The correlation matrices between locations are different.</span></span>
<span id="cb50-27"><a href="#cb50-27" tabindex="-1"></a><span class="co"># This difference arises because the correlation depends solely on the loadings of the factor</span></span>
<span id="cb50-28"><a href="#cb50-28" tabindex="-1"></a><span class="co"># analytics. ASReml updates the loadings during the REML while estimating</span></span>
<span id="cb50-29"><a href="#cb50-29" tabindex="-1"></a><span class="co"># the variances, whereas sommer doesn&#39;t (Sommer retains the initial loadings).</span></span>
<span id="cb50-30"><a href="#cb50-30" tabindex="-1"></a><span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>cor <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##       CA    FL    MI    MO    NC    NY    OR    WI
## CA 1.000 0.636 0.691 0.736 0.837 0.686 0.870 0.753
## FL 0.636 1.000 0.670 0.716 0.696 0.702 0.734 0.720
## MI 0.691 0.670 1.000 0.935 0.860 0.930 0.912 0.934
## MO 0.736 0.716 0.935 1.000 0.919 0.996 0.974 1.000
## NC 0.837 0.696 0.860 0.919 1.000 0.899 0.949 0.924
## NY 0.686 0.702 0.930 0.996 0.899 1.000 0.954 0.994
## OR 0.870 0.734 0.912 0.974 0.949 0.954 1.000 0.979
## WI 0.753 0.720 0.934 1.000 0.924 0.994 0.979 1.000</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a>Locs_cor <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>##           CA        FL        MI        MO        NC        NY        OR
## CA 1.0000000 0.9966790 0.9995382 0.9846991 0.9995779 0.9975216 0.9879114
## FL 0.9966790 1.0000000 0.9937442 0.9956193 0.9938926 0.9999384 0.9972540
## MI 0.9995382 0.9937442 1.0000000 0.9789490 0.9999991 0.9949228 0.9827446
## MO 0.9846991 0.9956193 0.9789490 1.0000000 0.9792209 0.9945200 0.9998097
## NC 0.9995779 0.9938926 0.9999991 0.9792209 1.0000000 0.9950565 0.9829909
## NY 0.9975216 0.9999384 0.9949228 0.9945200 0.9950565 1.0000000 0.9963703
## OR 0.9879114 0.9972540 0.9827446 0.9998097 0.9829909 0.9963703 1.0000000
## WI 0.9999919 0.9969987 0.9994078 0.9853927 0.9994529 0.9977968 0.9885275
##           WI
## CA 0.9999919
## FL 0.9969987
## MI 0.9994078
## MO 0.9853927
## NC 0.9994529
## NY 0.9977968
## OR 0.9885275
## WI 1.0000000</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="co"># Using vcov between PCs, loadings, and the correlation between locations,</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="co"># rebuild the vcov between locations and calculate the variance for the GxL term.</span></span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="co"># Variance of a weighted sum of two distributions:</span></span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a><span class="co"># (Link to explanation: </span></span>
<span id="cb54-5"><a href="#cb54-5" tabindex="-1"></a><span class="co"># https://stats.stackexchange.com/questions/5392/variance-of-two-weighted-random-variables)</span></span>
<span id="cb54-6"><a href="#cb54-6" tabindex="-1"></a><span class="co"># Weights = loadings</span></span>
<span id="cb54-7"><a href="#cb54-7" tabindex="-1"></a></span>
<span id="cb54-8"><a href="#cb54-8" tabindex="-1"></a><span class="co"># In factor analytics, the original variable = latent variables (factors) + error (psi).</span></span>
<span id="cb54-9"><a href="#cb54-9" tabindex="-1"></a><span class="co"># In sommer, we assume that psi = 0.</span></span>
<span id="cb54-10"><a href="#cb54-10" tabindex="-1"></a><span class="co"># Therefore, for variance calculation, we can extract the variance of the locations</span></span>
<span id="cb54-11"><a href="#cb54-11" tabindex="-1"></a><span class="co"># from the variance of the latent factors and the </span></span>
<span id="cb54-12"><a href="#cb54-12" tabindex="-1"></a><span class="co"># loadings that link the latent factors to the original variables (locations).</span></span>
<span id="cb54-13"><a href="#cb54-13" tabindex="-1"></a></span>
<span id="cb54-14"><a href="#cb54-14" tabindex="-1"></a><span class="co"># loc_{loc1} = loading_{loc1,pc1}*PC1 + loading_{loc1,pc2}*PC2</span></span>
<span id="cb54-15"><a href="#cb54-15" tabindex="-1"></a><span class="co"># Var(loc_{loc1}) = Var(loading_{loc1,pc1}*PC1 + loading_{loc1,pc2}*PC2)</span></span>
<span id="cb54-16"><a href="#cb54-16" tabindex="-1"></a><span class="co"># Var(loc_{loc1}) = loading_{loc1,pc1}^2*Var(PC1) + </span></span>
<span id="cb54-17"><a href="#cb54-17" tabindex="-1"></a><span class="co">#                   loading_{loc1,pc2}^2*Var(PC2) + </span></span>
<span id="cb54-18"><a href="#cb54-18" tabindex="-1"></a><span class="co">#                   loading_{loc1,pc1}*loading_{loc1,pc2}*cov(PC1,PC2)</span></span>
<span id="cb54-19"><a href="#cb54-19" tabindex="-1"></a></span>
<span id="cb54-20"><a href="#cb54-20" tabindex="-1"></a></span>
<span id="cb54-21"><a href="#cb54-21" tabindex="-1"></a>PC1var <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:PC1:PC1&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb54-22"><a href="#cb54-22" tabindex="-1"></a>PC2var <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:PC2:PC2&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb54-23"><a href="#cb54-23" tabindex="-1"></a>PCscov <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:PC1:PC2&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb54-24"><a href="#cb54-24" tabindex="-1"></a>Gamma</span></code></pre></div>
<pre><code>## 
## Loadings:
##    PC1   PC2  
## MI 0.787 0.617
## NY 0.721 0.693
## OR 0.655 0.747
## WI 0.763 0.642
## CA 0.759 0.633
## FL 0.706 0.694
## MO 0.642 0.762
## NC 0.783 0.615
## 
##                  PC1   PC2
## SS loadings    4.249 3.674
## Proportion Var 0.531 0.459
## Cumulative Var 0.531 0.990</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a>locvars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">nrow</span>(Gamma)))</span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a><span class="fu">names</span>(locvars) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(Gamma)</span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a><span class="cf">for</span> (loc <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>  w1 <span class="ot">&lt;-</span> Gamma[loc,<span class="st">&quot;PC1&quot;</span>]</span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a>  w2 <span class="ot">&lt;-</span> Gamma[loc,<span class="st">&quot;PC2&quot;</span>]</span>
<span id="cb56-6"><a href="#cb56-6" tabindex="-1"></a>  locvars[loc] <span class="ot">&lt;-</span> w1<span class="sc">^</span><span class="dv">2</span><span class="sc">*</span>PC1var <span class="sc">+</span> w2<span class="sc">^</span><span class="dv">2</span><span class="sc">*</span>PC2var <span class="sc">+</span> <span class="dv">2</span><span class="sc">*</span>w1<span class="sc">*</span>w2<span class="sc">*</span>PCscov</span>
<span id="cb56-7"><a href="#cb56-7" tabindex="-1"></a>}</span>
<span id="cb56-8"><a href="#cb56-8" tabindex="-1"></a></span>
<span id="cb56-9"><a href="#cb56-9" tabindex="-1"></a>Locs_vcov <span class="ot">&lt;-</span> Locs_cor</span>
<span id="cb56-10"><a href="#cb56-10" tabindex="-1"></a></span>
<span id="cb56-11"><a href="#cb56-11" tabindex="-1"></a><span class="cf">for</span> (loc1 <span class="cf">in</span> <span class="fu">rownames</span>(Locs_vcov)) {</span>
<span id="cb56-12"><a href="#cb56-12" tabindex="-1"></a>  <span class="cf">for</span> (loc2 <span class="cf">in</span> <span class="fu">colnames</span>(Locs_vcov)) {</span>
<span id="cb56-13"><a href="#cb56-13" tabindex="-1"></a>    <span class="co">#diagonal</span></span>
<span id="cb56-14"><a href="#cb56-14" tabindex="-1"></a>    <span class="cf">if</span> (loc1 <span class="sc">==</span> loc2) {</span>
<span id="cb56-15"><a href="#cb56-15" tabindex="-1"></a>      Locs_vcov[loc1,loc2] <span class="ot">&lt;-</span> locvars[loc1]</span>
<span id="cb56-16"><a href="#cb56-16" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb56-17"><a href="#cb56-17" tabindex="-1"></a>      <span class="co">#cov(x,y) = cor(x,y)*sd(x)*sd(y)</span></span>
<span id="cb56-18"><a href="#cb56-18" tabindex="-1"></a>      Locs_vcov[loc1,loc2] <span class="ot">&lt;-</span> Locs_vcov[loc1,loc2]<span class="sc">*</span><span class="fu">sqrt</span>(locvars[loc1])<span class="sc">*</span><span class="fu">sqrt</span>(locvars[loc2])</span>
<span id="cb56-19"><a href="#cb56-19" tabindex="-1"></a>    }</span>
<span id="cb56-20"><a href="#cb56-20" tabindex="-1"></a>    </span>
<span id="cb56-21"><a href="#cb56-21" tabindex="-1"></a>  }</span>
<span id="cb56-22"><a href="#cb56-22" tabindex="-1"></a>  </span>
<span id="cb56-23"><a href="#cb56-23" tabindex="-1"></a>}</span>
<span id="cb56-24"><a href="#cb56-24" tabindex="-1"></a></span>
<span id="cb56-25"><a href="#cb56-25" tabindex="-1"></a><span class="co">#weights = frequencies for each observed genotype-location combination</span></span>
<span id="cb56-26"><a href="#cb56-26" tabindex="-1"></a>gL <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sommerdf<span class="sc">$</span>id, <span class="st">&quot;:&quot;</span>, sommerdf<span class="sc">$</span>Loc)</span>
<span id="cb56-27"><a href="#cb56-27" tabindex="-1"></a>gL.weights <span class="ot">&lt;-</span> <span class="fu">table</span>(gL)</span>
<span id="cb56-28"><a href="#cb56-28" tabindex="-1"></a>gL.weights <span class="ot">&lt;-</span> gL.weights<span class="sc">/</span><span class="fu">sum</span>(gL.weights)</span>
<span id="cb56-29"><a href="#cb56-29" tabindex="-1"></a><span class="co">#vcov for all possible combinations of genotypes and locations</span></span>
<span id="cb56-30"><a href="#cb56-30" tabindex="-1"></a>Imat <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">length</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>id)))</span>
<span id="cb56-31"><a href="#cb56-31" tabindex="-1"></a><span class="fu">rownames</span>(Imat) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sommerdf<span class="sc">$</span>id)</span>
<span id="cb56-32"><a href="#cb56-32" tabindex="-1"></a><span class="fu">colnames</span>(Imat) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sommerdf<span class="sc">$</span>id)</span>
<span id="cb56-33"><a href="#cb56-33" tabindex="-1"></a>gLvcov <span class="ot">&lt;-</span> <span class="fu">kronecker</span>(Imat,Locs_vcov,<span class="at">make.dimnames =</span> T)</span>
<span id="cb56-34"><a href="#cb56-34" tabindex="-1"></a><span class="co">#leave only observed genotype-location combinations</span></span>
<span id="cb56-35"><a href="#cb56-35" tabindex="-1"></a>gLvcov <span class="ot">&lt;-</span> gLvcov[<span class="fu">names</span>(gL.weights),<span class="fu">names</span>(gL.weights)]</span>
<span id="cb56-36"><a href="#cb56-36" tabindex="-1"></a></span>
<span id="cb56-37"><a href="#cb56-37" tabindex="-1"></a>GxLvar <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sum</span>(<span class="fu">diag</span>(gLvcov)<span class="sc">*</span>gL.weights) <span class="sc">-</span> </span>
<span id="cb56-38"><a href="#cb56-38" tabindex="-1"></a>  <span class="fu">matrix</span>(gL.weights,<span class="at">nrow=</span><span class="dv">1</span>)<span class="sc">%*%</span>gLvcov<span class="sc">%*%</span><span class="fu">matrix</span>(gL.weights,<span class="at">ncol=</span><span class="dv">1</span>))</span>
<span id="cb56-39"><a href="#cb56-39" tabindex="-1"></a></span>
<span id="cb56-40"><a href="#cb56-40" tabindex="-1"></a></span>
<span id="cb56-41"><a href="#cb56-41" tabindex="-1"></a></span>
<span id="cb56-42"><a href="#cb56-42" tabindex="-1"></a><span class="co">#residual variance per location</span></span>
<span id="cb56-43"><a href="#cb56-43" tabindex="-1"></a>res_vars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp[(<span class="fu">nrow</span>(<span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp)<span class="sc">-</span></span>
<span id="cb56-44"><a href="#cb56-44" tabindex="-1"></a>                                            <span class="fu">length</span>(<span class="fu">unique</span>(sommerdf<span class="sc">$</span>Loc))<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span></span>
<span id="cb56-45"><a href="#cb56-45" tabindex="-1"></a>                                            <span class="fu">nrow</span>(<span class="fu">summary</span>(ans2a_open)<span class="sc">$</span>varcomp)</span>
<span id="cb56-46"><a href="#cb56-46" tabindex="-1"></a>                                           ,<span class="dv">1</span>]</span>
<span id="cb56-47"><a href="#cb56-47" tabindex="-1"></a><span class="co">#Weighted average</span></span>
<span id="cb56-48"><a href="#cb56-48" tabindex="-1"></a>Loc_weights <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">table</span>(sommerdf<span class="sc">$</span>Loc))</span>
<span id="cb56-49"><a href="#cb56-49" tabindex="-1"></a>Loc_weights <span class="ot">&lt;-</span> Loc_weights<span class="sc">/</span><span class="fu">sum</span>(Loc_weights)</span>
<span id="cb56-50"><a href="#cb56-50" tabindex="-1"></a>Residual_var <span class="ot">&lt;-</span> <span class="fu">sum</span>(res_vars<span class="sc">*</span>Loc_weights)</span>
<span id="cb56-51"><a href="#cb56-51" tabindex="-1"></a></span>
<span id="cb56-52"><a href="#cb56-52" tabindex="-1"></a></span>
<span id="cb56-53"><a href="#cb56-53" tabindex="-1"></a><span class="co">#extract stage1error variance component from EEV</span></span>
<span id="cb56-54"><a href="#cb56-54" tabindex="-1"></a>Stage1Errorvar <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(EEV_full)) <span class="sc">-</span> <span class="fu">mean</span>(EEV_full)</span>
<span id="cb56-55"><a href="#cb56-55" tabindex="-1"></a></span>
<span id="cb56-56"><a href="#cb56-56" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb56-57"><a href="#cb56-57" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="co">#Env_variance, #environmental variance</span></span>
<span id="cb56-58"><a href="#cb56-58" tabindex="-1"></a>  ans2a_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2a_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:isc:isc&quot;</span>], <span class="co">#genetic</span></span>
<span id="cb56-59"><a href="#cb56-59" tabindex="-1"></a>  GxLvar, <span class="co">#genotype by location interaction</span></span>
<span id="cb56-60"><a href="#cb56-60" tabindex="-1"></a>  Residual_var, <span class="co">#residual</span></span>
<span id="cb56-61"><a href="#cb56-61" tabindex="-1"></a>  Stage1Errorvar) <span class="co">#stage1 estimation error</span></span>
<span id="cb56-62"><a href="#cb56-62" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>var</span>
<span id="cb56-63"><a href="#cb56-63" tabindex="-1"></a>variances_ans2a <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb56-64"><a href="#cb56-64" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb56-65"><a href="#cb56-65" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2a) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2a<span class="sc">$</span>vars)<span class="sc">$</span>var)</span>
<span id="cb56-66"><a href="#cb56-66" tabindex="-1"></a>                                      </span>
<span id="cb56-67"><a href="#cb56-67" tabindex="-1"></a><span class="co"># Very different variance components for genotype, genotype by location, and residual </span></span>
<span id="cb56-68"><a href="#cb56-68" tabindex="-1"></a><span class="co"># (genotype by environment) The difference is mostly due to the fact that ASReml updates </span></span>
<span id="cb56-69"><a href="#cb56-69" tabindex="-1"></a><span class="co"># FA loadings while estimating variances, while sommer doesn&#39;t.</span></span>
<span id="cb56-70"><a href="#cb56-70" tabindex="-1"></a>variances_ans2a</span></code></pre></div>
<pre><code>##                 Sommer ASReml.Variance ASReml.PVE
## genotype      2.928802            53.0      0.360
## g x loc      32.689134            14.7      0.100
## g x env      48.126774            26.9      0.183
## Stage1.error 52.613111            52.6      0.358</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="co">#BLUPs</span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>Gamma<span class="ot">=</span>RotatedLoadings<span class="sc">$</span>loadings <span class="co">#loadings after rotation</span></span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a>score.mat<span class="ot">&lt;-</span>ans2a_open<span class="sc">$</span>uList[[<span class="dv">2</span>]];<span class="co">#extractfactorscores </span></span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a>BLUPs_GxE<span class="ot">&lt;-</span>score.mat<span class="sc">%*%</span><span class="fu">t</span>(Gamma)<span class="co">#BLUPsforallenvironments</span></span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a><span class="fu">head</span>(BLUPs_GxE)</span></code></pre></div>
<pre><code>##                    MI         NY         OR         WI         CA         FL
## A00188-3C   0.9434937  1.0359506  1.0988931  0.9733942  0.9614247  1.0350640
## A01143-3C   8.4357163  9.2702047  9.8394443  8.7057272  8.5983760  9.2630503
## A03440-2C   2.3068115  2.5228995  2.6685580  2.3765195  2.3476772  2.5197490
## A07117-2C  -5.7161832 -6.2552211 -6.6191151 -5.8901376 -5.8185156 -6.2477674
## AC00206-2W  0.8518717  0.9324869  0.9869516  0.8778924  0.8672067  0.9314041
## AC01151-5W  4.0718489  4.4521194  4.7082727  4.1944995  4.1436377  4.4464445
##                   MO         NC
## A00188-3C   1.116401  0.9405185
## A01143-3C   9.997496  8.4092195
## A03440-2C   2.709432  2.2994057
## A07117-2C  -6.721092 -5.6978792
## AC00206-2W  1.002204  0.8491476
## AC01151-5W  4.780196  4.0587613</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a>BLUPs_G <span class="ot">&lt;-</span> ans2a_open<span class="sc">$</span>uList[[<span class="dv">1</span>]]</span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a><span class="co">#Weights for each location:</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>Locs.index <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">WI=</span><span class="dv">1</span>, <span class="at">NC=</span><span class="dv">1</span>, <span class="at">OR=</span><span class="dv">1</span>, <span class="at">NY=</span><span class="dv">1</span>, <span class="at">MO=</span><span class="dv">1</span>, <span class="at">MI=</span><span class="dv">1</span>, <span class="at">CA =</span> <span class="dv">1</span>, <span class="at">FL =</span> <span class="dv">1</span>)  </span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a><span class="co">#scale the weights to unit variance:</span></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a>index.scaled <span class="ot">&lt;-</span> Locs.index<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">diag</span>(Locs_vcov))[<span class="fu">match</span>(<span class="fu">names</span>(Locs.index),</span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a>                                                       <span class="fu">rownames</span>(Locs_vcov))]</span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>index.scaled <span class="ot">&lt;-</span> index.scaled<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">sum</span>(index.scaled<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a></span>
<span id="cb60-10"><a href="#cb60-10" tabindex="-1"></a>weighted_BLUPs_GxE <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">nrow</span>(BLUPs_GxE)))</span>
<span id="cb60-11"><a href="#cb60-11" tabindex="-1"></a><span class="cf">for</span> (loc <span class="cf">in</span> <span class="fu">names</span>(index.scaled)) {</span>
<span id="cb60-12"><a href="#cb60-12" tabindex="-1"></a>  weighted_BLUPs_GxE <span class="ot">&lt;-</span> weighted_BLUPs_GxE <span class="sc">+</span> index.scaled[loc]<span class="sc">*</span>BLUPs_GxE[,loc]</span>
<span id="cb60-13"><a href="#cb60-13" tabindex="-1"></a>}</span>
<span id="cb60-14"><a href="#cb60-14" tabindex="-1"></a>weighted_BLUPs_GxE <span class="ot">&lt;-</span> weighted_BLUPs_GxE<span class="sc">/</span><span class="fu">sum</span>(index.scaled)</span>
<span id="cb60-15"><a href="#cb60-15" tabindex="-1"></a></span>
<span id="cb60-16"><a href="#cb60-16" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> BLUPs_G <span class="sc">+</span> weighted_BLUPs_GxE <span class="sc">+</span> <span class="fu">mean</span>(ans2a_open<span class="sc">$</span>b)</span>
<span id="cb60-17"><a href="#cb60-17" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb60-18"><a href="#cb60-18" tabindex="-1"></a></span>
<span id="cb60-19"><a href="#cb60-19" tabindex="-1"></a>prep0 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>ans1<span class="sc">$</span>blues,</span>
<span id="cb60-20"><a href="#cb60-20" tabindex="-1"></a>              <span class="at">vcov=</span>ans1<span class="sc">$</span>vcov,</span>
<span id="cb60-21"><a href="#cb60-21" tabindex="-1"></a>              <span class="at">vars=</span>ans2a<span class="sc">$</span>vars)</span>
<span id="cb60-22"><a href="#cb60-22" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(prep0, <span class="at">what=</span><span class="st">&quot;GV&quot;</span>, <span class="at">index.coeff =</span> Locs.index)</span>
<span id="cb60-23"><a href="#cb60-23" tabindex="-1"></a></span>
<span id="cb60-24"><a href="#cb60-24" tabindex="-1"></a><span class="co"># Mostly similar BLUPs</span></span>
<span id="cb60-25"><a href="#cb60-25" tabindex="-1"></a><span class="co"># The main difference is due to the fact that ASReml updates FA loadings while</span></span>
<span id="cb60-26"><a href="#cb60-26" tabindex="-1"></a><span class="co"># estimating variances, while sommer doesn&#39;t.</span></span>
<span id="cb60-27"><a href="#cb60-27" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb60-28"><a href="#cb60-28" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.9880297</code></pre>
<p>Focusing on the 6 highly correlated locations in StageWise
resultsâWI, MI, NC, OR, NY, MOâwe now add the marker data to
<code>Stage2</code> to estimate the additive correlation and predict
breeding values.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a>locs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;WI&quot;</span>,<span class="st">&quot;MI&quot;</span>,<span class="st">&quot;OR&quot;</span>,<span class="st">&quot;NY&quot;</span>,<span class="st">&quot;NC&quot;</span>,<span class="st">&quot;MO&quot;</span>)</span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a>blues <span class="ot">&lt;-</span> ans1<span class="sc">$</span>blues[ans1<span class="sc">$</span>blues<span class="sc">$</span>loc <span class="sc">%in%</span> locs,]</span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">names</span>(ans1<span class="sc">$</span>vcov),<span class="at">split=</span><span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">1</span>)</span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a>EEV <span class="ot">&lt;-</span> ans1<span class="sc">$</span>vcov[tmp <span class="sc">%in%</span> locs]</span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a></span>
<span id="cb62-6"><a href="#cb62-6" tabindex="-1"></a></span>
<span id="cb62-7"><a href="#cb62-7" tabindex="-1"></a><span class="co">#open source</span></span>
<span id="cb62-8"><a href="#cb62-8" tabindex="-1"></a>locs_index <span class="ot">&lt;-</span> <span class="fu">which</span>(sommerdf<span class="sc">$</span>Loc <span class="sc">%in%</span> locs)</span>
<span id="cb62-9"><a href="#cb62-9" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> sommerdf[locs_index,]</span>
<span id="cb62-10"><a href="#cb62-10" tabindex="-1"></a>EEV_smaller <span class="ot">&lt;-</span> EEV_full[locs_index,locs_index]</span>
<span id="cb62-11"><a href="#cb62-11" tabindex="-1"></a></span>
<span id="cb62-12"><a href="#cb62-12" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb62-13"><a href="#cb62-13" tabindex="-1"></a>ans2b <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>blues,<span class="at">vcov=</span>EEV,<span class="at">geno=</span>geno,<span class="at">non.add=</span><span class="st">&quot;g.resid&quot;</span>)</span>
<span id="cb62-14"><a href="#cb62-14" tabindex="-1"></a>ans2c <span class="ot">&lt;-</span> <span class="fu">Stage2</span>(<span class="at">data=</span>blues,<span class="at">vcov=</span>EEV,<span class="at">geno=</span>geno,<span class="at">non.add=</span><span class="st">&quot;dom&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">non.add=</span><span class="fu">c</span>(<span class="st">&quot;g.resid&quot;</span>,<span class="st">&quot;dom&quot;</span>), <span class="at">AIC=</span><span class="fu">c</span>(ans2b<span class="sc">$</span>aic,ans2c<span class="sc">$</span>aic))</span></code></pre></div>
<pre><code>##   non.add      AIC
## 1 g.resid 16715.49
## 2     dom 16685.18</code></pre>
<p>The dominance model is selected over the genetic residual model
because of its lower AIC.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="co">#Open source dominance model:</span></span>
<span id="cb65-2"><a href="#cb65-2" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> geno<span class="sc">@</span>Fg</span>
<span id="cb65-3"><a href="#cb65-3" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" tabindex="-1"></a><span class="co">#include Fg into sommerdf:</span></span>
<span id="cb65-5"><a href="#cb65-5" tabindex="-1"></a><span class="co">#1) reorder</span></span>
<span id="cb65-6"><a href="#cb65-6" tabindex="-1"></a>Fg <span class="ot">&lt;-</span> Fg[<span class="fu">match</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))]</span>
<span id="cb65-7"><a href="#cb65-7" tabindex="-1"></a><span class="fu">identical</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="fu">names</span>(Fg))</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a>sommerdf_smaller <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_smaller, Fg)</span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" tabindex="-1"></a><span class="co"># When you want to use a known variance-covariance matrix in Sommer&#39;s `mmec()` function,</span></span>
<span id="cb67-5"><a href="#cb67-5" tabindex="-1"></a><span class="co"># what you actually have to provide is the inverse of the matrix, and it must be of the</span></span>
<span id="cb67-6"><a href="#cb67-6" tabindex="-1"></a><span class="co"># &quot;dgCMatrix&quot; class (sparse matrix).</span></span>
<span id="cb67-7"><a href="#cb67-7" tabindex="-1"></a></span>
<span id="cb67-8"><a href="#cb67-8" tabindex="-1"></a>EEVInv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(EEV_smaller),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb67-9"><a href="#cb67-9" tabindex="-1"></a>Ginv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(geno<span class="sc">@</span>G)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb67-10"><a href="#cb67-10" tabindex="-1"></a>Dinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(geno<span class="sc">@</span>D)),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb67-11"><a href="#cb67-11" tabindex="-1"></a></span>
<span id="cb67-12"><a href="#cb67-12" tabindex="-1"></a><span class="co">#needed later to scale variances</span></span>
<span id="cb67-13"><a href="#cb67-13" tabindex="-1"></a>meanG <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno<span class="sc">@</span>G))<span class="sc">-</span><span class="fu">mean</span>(geno<span class="sc">@</span>G)</span>
<span id="cb67-14"><a href="#cb67-14" tabindex="-1"></a>meanD <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(geno<span class="sc">@</span>D))<span class="sc">-</span><span class="fu">mean</span>(geno<span class="sc">@</span>D)</span>
<span id="cb67-15"><a href="#cb67-15" tabindex="-1"></a></span>
<span id="cb67-16"><a href="#cb67-16" tabindex="-1"></a><span class="co"># Fix the variance for stage1Error to be 1. We have to scale it by dividing it</span></span>
<span id="cb67-17"><a href="#cb67-17" tabindex="-1"></a><span class="co"># by the variance of the response variable because Sommer works that way.</span></span>
<span id="cb67-18"><a href="#cb67-18" tabindex="-1"></a><span class="co"># We fix the variance to be 1 because we don&#39;t want the model to scale EEV in </span></span>
<span id="cb67-19"><a href="#cb67-19" tabindex="-1"></a><span class="co"># any way. However, 1 is not the true stage1Error variance component. More on that later.</span></span>
<span id="cb67-20"><a href="#cb67-20" tabindex="-1"></a></span>
<span id="cb67-21"><a href="#cb67-21" tabindex="-1"></a>Variance_scaled <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs))</span>
<span id="cb67-22"><a href="#cb67-22" tabindex="-1"></a></span>
<span id="cb67-23"><a href="#cb67-23" tabindex="-1"></a></span>
<span id="cb67-24"><a href="#cb67-24" tabindex="-1"></a><span class="co">#Factor analytics</span></span>
<span id="cb67-25"><a href="#cb67-25" tabindex="-1"></a><span class="co">#1) Extract loadings using sommer functions</span></span>
<span id="cb67-26"><a href="#cb67-26" tabindex="-1"></a>FA2LocLoadings <span class="ot">&lt;-</span> <span class="fu">with</span>(sommerdf_smaller, <span class="fu">rrc</span>(Loc, <span class="co">#Factor analytics for location</span></span>
<span id="cb67-27"><a href="#cb67-27" tabindex="-1"></a>                                   id, <span class="co">#Variables characterizing each location</span></span>
<span id="cb67-28"><a href="#cb67-28" tabindex="-1"></a>                                   <span class="co">#are the genotypes</span></span>
<span id="cb67-29"><a href="#cb67-29" tabindex="-1"></a>                                   BLUEs, <span class="co">#response variable</span></span>
<span id="cb67-30"><a href="#cb67-30" tabindex="-1"></a>                                   <span class="at">nPC=</span><span class="dv">2</span>, <span class="co">#FA2 --&gt; 2PCs</span></span>
<span id="cb67-31"><a href="#cb67-31" tabindex="-1"></a>                                   <span class="at">returnGamma =</span> <span class="cn">TRUE</span></span>
<span id="cb67-32"><a href="#cb67-32" tabindex="-1"></a>                  )) </span>
<span id="cb67-33"><a href="#cb67-33" tabindex="-1"></a></span>
<span id="cb67-34"><a href="#cb67-34" tabindex="-1"></a><span class="co"># 2) Rotate loadings (optional)</span></span>
<span id="cb67-35"><a href="#cb67-35" tabindex="-1"></a><span class="co"># You can directly use the FA2LocLoadings, but with this dataset, there were</span></span>
<span id="cb67-36"><a href="#cb67-36" tabindex="-1"></a><span class="co"># matrix inversion issues. To solve them, we can perform a varimax rotation to obtain</span></span>
<span id="cb67-37"><a href="#cb67-37" tabindex="-1"></a><span class="co"># alternative but equivalent loadings that don&#39;t cause inversion issues in the model.</span></span>
<span id="cb67-38"><a href="#cb67-38" tabindex="-1"></a>RotatedLoadings <span class="ot">&lt;-</span> <span class="fu">varimax</span>(FA2LocLoadings<span class="sc">$</span>Gamma)</span>
<span id="cb67-39"><a href="#cb67-39" tabindex="-1"></a><span class="co"># I recommend using the original FA2LocLoadings by default and only using the varimax </span></span>
<span id="cb67-40"><a href="#cb67-40" tabindex="-1"></a><span class="co"># rotation if you encounter inversion issues.</span></span>
<span id="cb67-41"><a href="#cb67-41" tabindex="-1"></a></span>
<span id="cb67-42"><a href="#cb67-42" tabindex="-1"></a></span>
<span id="cb67-43"><a href="#cb67-43" tabindex="-1"></a><span class="co"># 3) Expand loadings</span></span>
<span id="cb67-44"><a href="#cb67-44" tabindex="-1"></a><span class="co"># Expand the loadings matrix using the locations design matrix.</span></span>
<span id="cb67-45"><a href="#cb67-45" tabindex="-1"></a><span class="co"># The expanded loadings matrix will be the input to the factor analytics term</span></span>
<span id="cb67-46"><a href="#cb67-46" tabindex="-1"></a><span class="co"># in the model.</span></span>
<span id="cb67-47"><a href="#cb67-47" tabindex="-1"></a>Zl <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Loc<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb67-48"><a href="#cb67-48" tabindex="-1"></a><span class="co">#FA2Loc &lt;- Zl%*%FA2LocLoadings$Gamma #without rotation</span></span>
<span id="cb67-49"><a href="#cb67-49" tabindex="-1"></a>FA2Loc <span class="ot">&lt;-</span> Zl<span class="sc">%*%</span>RotatedLoadings<span class="sc">$</span>loadings <span class="co">#with rotation</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a><span class="co">#with dominance</span></span>
<span id="cb68-2"><a href="#cb68-2" tabindex="-1"></a>ans2c_open <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb68-3"><a href="#cb68-3" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb68-4"><a href="#cb68-4" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Ginv) <span class="sc">+</span> <span class="co">#addtitive genotypic effects</span></span>
<span id="cb68-5"><a href="#cb68-5" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id), <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic effects</span></span>
<span id="cb68-6"><a href="#cb68-6" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">usc</span>(FA2Loc),<span class="fu">isc</span>(id), <span class="at">Gu =</span> Ginv) <span class="sc">+</span> <span class="co">#kronecker unstructured</span></span>
<span id="cb68-7"><a href="#cb68-7" tabindex="-1"></a>                     <span class="co">#factor analytics for location with additive effects</span></span>
<span id="cb68-8"><a href="#cb68-8" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co">#stage 1 estimation error effect.</span></span>
<span id="cb68-9"><a href="#cb68-9" tabindex="-1"></a>                               <span class="co">#as it&#39;s inside isc() function we are assuming </span></span>
<span id="cb68-10"><a href="#cb68-10" tabindex="-1"></a>                               <span class="co">#identity variance-covariance structure. That is</span></span>
<span id="cb68-11"><a href="#cb68-11" tabindex="-1"></a>                               <span class="co">#undesired and it will be overriden later. Ideally,</span></span>
<span id="cb68-12"><a href="#cb68-12" tabindex="-1"></a>                               <span class="co">#we would skip isc() function altogether, but </span></span>
<span id="cb68-13"><a href="#cb68-13" tabindex="-1"></a>                               <span class="co">#Sommer syntax needs it even if we will overwrite</span></span>
<span id="cb68-14"><a href="#cb68-14" tabindex="-1"></a>                               <span class="co">#it later.</span></span>
<span id="cb68-15"><a href="#cb68-15" tabindex="-1"></a>                               <span class="at">theta =</span> Variance_scaled, <span class="co">#initial value for the</span></span>
<span id="cb68-16"><a href="#cb68-16" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb68-17"><a href="#cb68-17" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb68-18"><a href="#cb68-18" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb68-19"><a href="#cb68-19" tabindex="-1"></a>                           </span>
<span id="cb68-20"><a href="#cb68-20" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb68-21"><a href="#cb68-21" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb68-22"><a href="#cb68-22" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb68-23"><a href="#cb68-23" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb68-24"><a href="#cb68-24" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb68-25"><a href="#cb68-25" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb68-26"><a href="#cb68-26" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb68-27"><a href="#cb68-27" tabindex="-1"></a>                           </span>
<span id="cb68-28"><a href="#cb68-28" tabindex="-1"></a>                                <span class="co">#As we are using constraint 3, we disable</span></span>
<span id="cb68-29"><a href="#cb68-29" tabindex="-1"></a>                                <span class="co">#variance estimation, the model will use the</span></span>
<span id="cb68-30"><a href="#cb68-30" tabindex="-1"></a>                                <span class="co">#value we provided in &quot;theta&quot; argument</span></span>
<span id="cb68-31"><a href="#cb68-31" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb68-32"><a href="#cb68-32" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb68-33"><a href="#cb68-33" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb68-34"><a href="#cb68-34" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb68-35"><a href="#cb68-35" tabindex="-1"></a>                      <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">dsc</span>(Loc),<span class="fu">isc</span>(units)), <span class="co">#diagonal loc by i.i.d. residuals</span></span>
<span id="cb68-36"><a href="#cb68-36" tabindex="-1"></a>                     <span class="at">nIters=</span><span class="dv">50</span>, <span class="co">#increased number or iterations</span></span>
<span id="cb68-37"><a href="#cb68-37" tabindex="-1"></a>                     <span class="co">#more EM iterations at the beggining for usc models </span></span>
<span id="cb68-38"><a href="#cb68-38" tabindex="-1"></a>                     <span class="at">emWeight=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">10</span>),sommer<span class="sc">::</span><span class="fu">logspace</span>(<span class="dv">10</span>,<span class="dv">1</span>,.<span class="dv">05</span>),<span class="fu">rep</span>(.<span class="dv">05</span>,<span class="dv">80</span>)), </span>
<span id="cb68-39"><a href="#cb68-39" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_smaller,</span>
<span id="cb68-40"><a href="#cb68-40" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb68-41"><a href="#cb68-41" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)</span></code></pre></div>
<pre><code>## $var
##              Variance   PVE
## additive         35.3 0.211
## add x loc         8.0 0.048
## dominance        35.9 0.215
## heterosis         4.0 0.024
## g x env          24.3 0.145
## Stage1.error     59.9 0.357
## 
## $cor
##       MI    MO    NC    NY    OR    WI
## MI 1.000 0.892 0.708 0.883 0.898 0.908
## MO 0.892 1.000 0.897 0.930 0.981 0.941
## NC 0.708 0.897 1.000 0.719 0.831 0.697
## NY 0.883 0.930 0.719 1.000 0.941 0.961
## OR 0.898 0.981 0.831 0.941 1.000 0.959
## WI 0.908 0.941 0.697 0.961 0.959 1.000</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="fu">uniplot</span>(ans2c<span class="sc">$</span>loadings)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAzFBMVEUAAAAAADoAAF4AAGYAAP8ANYQAOpAAXqgAZrYzMzM1AAA1ADU1AF41hMk6AAA6ADo6AGY6Ojo6OpA6kNteAABeXl5ehIReqOtmAABmADpmAGZmOpBmZmZmkJBmtrZmtttmtv+ENQCEyeuQOgCQOjqQOmaQkGaQtpCQ27aQ2/+oXgCo6+u2AAC2ZgC2Zjq2/9u2///JhDXJ6+vbAADbkDrbtmbb/7bb/9vb///rqF7ryYTr66jr68nr6+v/AAD/tmb/25D//7b//9v////3gKYcAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAKi0lEQVR4nO2dfX/bNBDHXdZ14WEZLBtN2YBsgUGBQDcMdA9xkur9vycs22lsy9Lp4STL9t0fSz87/2LpG0WWdCclYWRKS/ouQOxGgAAjQIARIMAIEGAECDA1oM8nbHqAan9vFdepfAMVEiDASYAAJwECnAQIcBIgwEmAACcBApwECHASIMCpCWjbh2Vt66MQMbagIw/B2Xag3VHhjAuQUP9uYe2q6QDqRKMSloJpAJKzgd5VrRwFoLKKLg3BtOnBvogAYXUl3YiGDqjrQWX6ro0nnJUQcPYGyMfTuo1owIBMqmJyx+b7DhaQ2ZfB7I719x4moM5nDuZw5nSDIQKSPJGRx3vVXQYISDaqQx8QZ7bCtjMoIPmoF3/GAA0/YwRkOSmwvWOWDQqQ9XTL4SOxnaf1AQipwZsK5YTiAlR8lH0AkjeiqADhPXQthBYz/dCANGbsHgFZzPSDAsId19oJXcbuvgEhz4xshSKhSADpLmn4BiQSigMQ+uKDvbBNKApAmcqp7UMSWq5A+QTkYX3PRai/mBYGkPDo6BuQfo8YBJDZcyMMIN1nqgUg45h/L4kGsJkXy1ML6hq8RtCCagXr9yvmLYznLtQZ13sHZL7uHA6QzszQNyCPkXIMYQauV/sFZLUCExJQ+Qn2BshuDS8sIGh90ycgz9k6WEL9zCtkQLYL8wQIEqru7kOoDAn5A6TiExcg7aKiAtL+WIx8noSajR0TUICcU0yh3vNkwoD0RiR4gFyTVXsQ6uRSoAEyGr8b+TwKNQb9WIDg4CkBAkoUIyCNeTUSoFBbS9CF4MoMDiDTVSgjHwHqVQgtfqIAQtxREFxIgCAfEF/AAJQpfEqhns+3sGMJFBeQRTTOyOddSIAgnyo3xh1Q0D3IfoQECPIp8iycY/ORBuHNTFEJ1xaknyMVcwtSpDI5AvK0cTC4kABBPuk3wQ2QbWakkY8ARSGUJTASoMq8APK4eTm8UDKeI0BH8wDIfgNElMLuOSUBOlnnsg0BOhk2IM8HKPQg7FoZJUA1wwVkmitOgIASDQBQV/jKGpDTTtlYhQQI8okhdFtAQU65CS90BLS5uM3FaTLL3+luPds/uTYo0RQA7R7cMPZp/fLiNmOH5YIAta/az1eMfbz69eo6Y7uz6xECEvOcjADdrReMvb/4b73IWHpxS4CEq3Iq7LcFS2cZ2ywYARKuyjuhu1er/Kt2c1iuRglISEY1A3S4ut5/dZN30Kv91zcESLwq/2Klj/KXPxZp/qgnQMJVee+zeZy//D3LuyACJF7F9t88eZ2/Sf5yPVJA7f2ahskLH589fLfdZp9e8ZcPX74OkVkQ2FqJDFZzsYDnRYYX2rWg5jsToLYRICNA7qc3Ri3MCJDa5wwoSaQXx1RPa6EroCSRE4qpntZCAgT5MgKk9jkCGn0f5AzI8vyS4QBqHJDgDuiwnPGX3YO/1g9+Z3zhemZepHEDSlasWGzcz/lSSBEbMi3SyAA1eqHD8vunt2VIKP3spox8oBZ3eICaz7HDcsVXzzigw7MZ20i+YJMGxBeoi6Di+7NfhEU05+KOABBLF6yKuiYLqyKNC1C7D1rxaEcBaFv8a1GkkQEqFRUlDoilMwJ0r64Ex+9ZAejupzcE6KiuBA1AbJeMCFB9xcsJ0Dn/owTENgSoJeYN6Pzc4K5TA1RaE1Fk9YwBUBNRZPWMA1AdUWT1jAXQCVFk9QwISB67ruz83F+0PLzV6ojTgrjxVhRZQ4jnK1Za3opUb0KA2g99gyKNCZAiqJHfVYFoIoBqi0E7/iefZ9yti/9NFvyuR0T7eTHtOCyPCdWTA1TMvnj9ywRzjuSycJSI9vOz6+JlqoDSMnrB4VSA2OZRdRFHtJ+/5P+bvpwYoGMfdLeuohc8ubwNKEeUA/rhxS1fKBoeIJQVxfsU1/yPCtAuuaxdt5//+PMN2z/9Z+qADsvVsZNe1e+6n6/SFdstdpMHdFW2oP18wVqAdjO2WU0VkNgH7YQWdHjx73c3UwXE0nJ1tfYU2xTJC0fjMeg/38zYZAF1jYMe1a7jgNJkMWFA9ZF0+ZhP63FVDoiTmTAgs7tOBpB6smpZpLiEToBUOZyR1dNW6JajSIC6jABRH1Qz1zRg2swiGgEiQCcjQIDPHRBtyRRMAFQLWbDapCyieloLsQDdhyzq0/qI6mktdDtY4Bjb//DF88f5H2+fP3y3fcsPGdhuP716HDC9wJ/hHCxQC1nUlxYjagi2QsezOypxJoQsWLFKHU89rYVogFohC1aEN+Kpp7UQD1AzZMGK8EY89bQW4gE6hizG1Qc5nmF2L85OIYt6eCOaeloLMQFVIYtxjYMwAVUhCzaqkTQWIDrstmkEiM6TLo0AqX2YJ5LTmfZ1I0D0uxrc6IdHAB8BAnzIgOjHj8SrmmICJFzVFI8MEP4PsNFP+AlXtcRjApRJnE6ABEL919NaSIAAnx9AbUL919NWmMmcBKg0AqT2yethFZuXR7KHavJqOLYgeecmWsQtSPGwcQYkG2ABRTLwDR2QZIgOFMnA51+oGvASIGBOiQCoexasLJKRbzyADstiQ1150FuaJMnZteSuqhKZORGE6oU/DEDHW5wOTmZ3ax6TTpOVWqjl8ywEgg84gMr9G4flt8eDk6tDpdP62dJxAoLieyiAqh1Ah+VldXBydbai9K5GvlEBqg5OFn95rFuo5fMqBHMMkAFVBydzTqq7Gvl8CiV7BrABsWMfdFkdnDyYFhQKUHkrDqg8OPnYB+W4TIpr4XQTyjad+ADEshJQeXBy9fhK62ffxgdIZ1MOHiCWFYDKg5OHMQ4KDIhlBaAyp5Nt4h9Ja+17QwU0rLmY3gkSmIAaJxYo72rk8yRUnQfpC5DyEMrIAOkWFReQ7sdi5PMi1G7syIA0v9hGPh9C/e5ymoAyfSE2IMs94xMCxDKN8buRD1+YKXxtIT4gnRmgkQ9bmLX3zCmFPgBZbOQICEjcbaAUegFkvk8hHKCOXHGl0A8g4zT8YIC6Mn2VQtfkBfNsgF7NvFieWpBpCmygFmQRBPYGSHzc9w/IJs/CHyCj7KoQgJqfWAyA5HltkFDfqS+0TIbzCkg/u8o/INuEXL+AGs26T0Di/CcSQLq5MZ4BOezc8g7o9OH1B8hl96h/QFqzH5+AupcXYgKkMX/2B0i2+hIVIHgFxhsg5xMNAwGC1vB8AXI/EzMUoLyp20Y87AGpbhkdoNypKK8HQPkngtFmQwJC6BH0fVi9XlhAzs8UXR/eczMwIFlfhAsoQxx5BQfEOpsRIqDaJzBUQIaLaWbrB7qLYnEDEhAhATJ419gBtTojDEBi7zZsQKxeJWdAnX3/4AGx+y+FIyDJ6GEMgIrP3im5MZNPYsYBiJuikkqhWjciQMU0RFpbyXezuNzXMsC9xQOosKwLk/j4Q+zdIaev2LyTZUoLW5bIWlB8QgIEOAkQ4CRAgJMAAU4CBDgJEOAkQICTAAFOAgQ4CRDg1AQ0YdMCREaAICNAgBEgwAgQYAQIMAIEGAECjAABRoAAI0CAESDACBBgBAgwAgQYAQKMAAFGgAAjQIARIMAIEGAECDACBBgBAowAAUaAACNAgBEgwAgQYP8DAdntZJUEMYMAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a><span class="co"># The squared radius for each location represents the proportion of variance explained (PVE)</span></span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a><span class="co"># by the latent factors. With the exception of FL, the FA2 model appears to provide</span></span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a><span class="co"># a good representation of the covariance structure. The numeric PVE values can be </span></span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a><span class="co"># obtained as follows:</span></span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a><span class="fu">apply</span>(ans2c<span class="sc">$</span>loadings,<span class="dv">1</span>,norm,<span class="at">type=</span><span class="st">&quot;f&quot;</span>)<span class="sc">^</span><span class="dv">2</span> <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##        MI        MO        NC        NY        OR        WI 
## 0.8385931 1.0000000 1.0000000 0.9323721 0.9747025 0.9947429</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a>Gamma <span class="ot">&lt;-</span> RotatedLoadings<span class="sc">$</span>loadings</span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a><span class="fu">apply</span>(Gamma,<span class="dv">1</span>,norm,<span class="at">type=</span><span class="st">&quot;f&quot;</span>)<span class="sc">^</span><span class="dv">2</span> <span class="co">#sommer</span></span></code></pre></div>
<pre><code>##        MI        NY        OR        WI        MO        NC 
## 1.0000000 1.0000000 0.9845461 0.9924632 0.9836232 0.9916222</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="co">#different PVE in sommer and StageWise</span></span>
<span id="cb76-2"><a href="#cb76-2" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" tabindex="-1"></a></span>
<span id="cb76-4"><a href="#cb76-4" tabindex="-1"></a><span class="co"># The cosine of the angle between locations equals the correlation due to the latent</span></span>
<span id="cb76-5"><a href="#cb76-5" tabindex="-1"></a><span class="co"># factors (recall cos(0) = 1). We can use this concept to construct the correlation matrix</span></span>
<span id="cb76-6"><a href="#cb76-6" tabindex="-1"></a><span class="co"># between environments in the sommer output.</span></span>
<span id="cb76-7"><a href="#cb76-7" tabindex="-1"></a><span class="co"># Calculate correlations as the cosine of the angle between loadings:</span></span>
<span id="cb76-8"><a href="#cb76-8" tabindex="-1"></a>Locs_cor <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">nrow</span>(Gamma), <span class="at">ncol =</span> <span class="fu">nrow</span>(Gamma))</span>
<span id="cb76-9"><a href="#cb76-9" tabindex="-1"></a><span class="fu">rownames</span>(Locs_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>cor)</span>
<span id="cb76-10"><a href="#cb76-10" tabindex="-1"></a><span class="fu">colnames</span>(Locs_cor) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>cor)</span>
<span id="cb76-11"><a href="#cb76-11" tabindex="-1"></a><span class="cf">for</span> (loc1 <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb76-12"><a href="#cb76-12" tabindex="-1"></a>  <span class="cf">for</span> (loc2 <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb76-13"><a href="#cb76-13" tabindex="-1"></a>    <span class="co">#find slopes of lines that go from (0,0) to each location in the uniplot</span></span>
<span id="cb76-14"><a href="#cb76-14" tabindex="-1"></a>    m1 <span class="ot">&lt;-</span> (Gamma[loc1,<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">0</span>)<span class="sc">/</span>(Gamma[loc1,<span class="dv">1</span>] <span class="sc">-</span> <span class="dv">0</span>)</span>
<span id="cb76-15"><a href="#cb76-15" tabindex="-1"></a>    m2 <span class="ot">&lt;-</span> (Gamma[loc2,<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">0</span>)<span class="sc">/</span>(Gamma[loc2,<span class="dv">1</span>] <span class="sc">-</span> <span class="dv">0</span>)</span>
<span id="cb76-16"><a href="#cb76-16" tabindex="-1"></a>    <span class="co">#angle from slopes</span></span>
<span id="cb76-17"><a href="#cb76-17" tabindex="-1"></a>    angle <span class="ot">&lt;-</span> <span class="fu">atan</span>((m1<span class="sc">-</span>m2)<span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span>m1<span class="sc">*</span>m2))</span>
<span id="cb76-18"><a href="#cb76-18" tabindex="-1"></a>    <span class="co">#correlation from angle</span></span>
<span id="cb76-19"><a href="#cb76-19" tabindex="-1"></a>    correlation <span class="ot">&lt;-</span> <span class="fu">cos</span>(angle)</span>
<span id="cb76-20"><a href="#cb76-20" tabindex="-1"></a>    Locs_cor[loc1,loc2] <span class="ot">&lt;-</span> correlation</span>
<span id="cb76-21"><a href="#cb76-21" tabindex="-1"></a>    Locs_cor[loc2,loc1] <span class="ot">&lt;-</span> correlation</span>
<span id="cb76-22"><a href="#cb76-22" tabindex="-1"></a>  }</span>
<span id="cb76-23"><a href="#cb76-23" tabindex="-1"></a>  </span>
<span id="cb76-24"><a href="#cb76-24" tabindex="-1"></a>}</span>
<span id="cb76-25"><a href="#cb76-25" tabindex="-1"></a></span>
<span id="cb76-26"><a href="#cb76-26" tabindex="-1"></a><span class="co"># The correlation matrices between locations are different.</span></span>
<span id="cb76-27"><a href="#cb76-27" tabindex="-1"></a><span class="co"># This difference arises because the correlation depends solely on the loadings of the factor</span></span>
<span id="cb76-28"><a href="#cb76-28" tabindex="-1"></a><span class="co"># analytics. ASReml updates the loadings during the REML while estimating</span></span>
<span id="cb76-29"><a href="#cb76-29" tabindex="-1"></a><span class="co"># the variances, whereas sommer doesn&#39;t (Sommer retains the initial loadings).</span></span>
<span id="cb76-30"><a href="#cb76-30" tabindex="-1"></a></span>
<span id="cb76-31"><a href="#cb76-31" tabindex="-1"></a><span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>cor <span class="co">#StageWise</span></span></code></pre></div>
<pre><code>##       MI    MO    NC    NY    OR    WI
## MI 1.000 0.892 0.708 0.883 0.898 0.908
## MO 0.892 1.000 0.897 0.930 0.981 0.941
## NC 0.708 0.897 1.000 0.719 0.831 0.697
## NY 0.883 0.930 0.719 1.000 0.941 0.961
## OR 0.898 0.981 0.831 0.941 1.000 0.959
## WI 0.908 0.941 0.697 0.961 0.959 1.000</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a>Locs_cor <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>##           MI        MO        NC        NY        OR        WI
## MI 1.0000000 0.9693045 0.9993596 0.9922168 0.9789307 0.9990856
## MO 0.9693045 1.0000000 0.9774816 0.9923757 0.9990855 0.9789300
## NC 0.9993596 0.9774816 1.0000000 0.9960372 0.9856105 0.9999757
## NY 0.9922168 0.9923757 0.9960372 1.0000000 0.9967380 0.9966334
## OR 0.9789307 0.9990855 0.9856105 0.9967380 1.0000000 0.9867657
## WI 0.9990856 0.9789300 0.9999757 0.9966334 0.9867657 1.0000000</code></pre>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a><span class="co"># Using vcov between PCs, loadings, and correlation between locations, </span></span>
<span id="cb80-2"><a href="#cb80-2" tabindex="-1"></a><span class="co"># rebuild the vcov between locations and calculate the variance for the GxL term.</span></span>
<span id="cb80-3"><a href="#cb80-3" tabindex="-1"></a><span class="co"># Variance of the weighted sum of two distributions:</span></span>
<span id="cb80-4"><a href="#cb80-4" tabindex="-1"></a><span class="co"># (Link to explanation: </span></span>
<span id="cb80-5"><a href="#cb80-5" tabindex="-1"></a><span class="co"># https://stats.stackexchange.com/questions/5392/variance-of-two-weighted-random-variables)</span></span>
<span id="cb80-6"><a href="#cb80-6" tabindex="-1"></a><span class="co"># Weights = loadings</span></span>
<span id="cb80-7"><a href="#cb80-7" tabindex="-1"></a></span>
<span id="cb80-8"><a href="#cb80-8" tabindex="-1"></a>PC1var <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:Ginv:PC1:PC1&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb80-9"><a href="#cb80-9" tabindex="-1"></a>PC2var <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:Ginv:PC2:PC2&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb80-10"><a href="#cb80-10" tabindex="-1"></a>PCscov <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp[<span class="st">&quot;FA2Loc:id:Ginv:PC1:PC2&quot;</span>,<span class="dv">1</span>]</span>
<span id="cb80-11"><a href="#cb80-11" tabindex="-1"></a>Gamma</span></code></pre></div>
<pre><code>## 
## Loadings:
##    PC1   PC2  
## MI 0.801 0.599
## NY 0.720 0.694
## OR 0.657 0.744
## WI 0.772 0.630
## MO 0.624 0.771
## NC 0.776 0.624
## 
##                  PC1   PC2
## SS loadings    3.178 2.774
## Proportion Var 0.530 0.462
## Cumulative Var 0.530 0.992</code></pre>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a>locvars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">nrow</span>(Gamma)))</span>
<span id="cb82-2"><a href="#cb82-2" tabindex="-1"></a><span class="fu">names</span>(locvars) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(Gamma)</span>
<span id="cb82-3"><a href="#cb82-3" tabindex="-1"></a><span class="cf">for</span> (loc <span class="cf">in</span> <span class="fu">rownames</span>(Gamma)) {</span>
<span id="cb82-4"><a href="#cb82-4" tabindex="-1"></a>  w1 <span class="ot">&lt;-</span> Gamma[loc,<span class="st">&quot;PC1&quot;</span>]</span>
<span id="cb82-5"><a href="#cb82-5" tabindex="-1"></a>  w2 <span class="ot">&lt;-</span> Gamma[loc,<span class="st">&quot;PC2&quot;</span>]</span>
<span id="cb82-6"><a href="#cb82-6" tabindex="-1"></a>  locvars[loc] <span class="ot">&lt;-</span> w1<span class="sc">^</span><span class="dv">2</span><span class="sc">*</span>PC1var <span class="sc">+</span> w2<span class="sc">^</span><span class="dv">2</span><span class="sc">*</span>PC2var <span class="sc">+</span> <span class="dv">2</span><span class="sc">*</span>w1<span class="sc">*</span>w2<span class="sc">*</span>PCscov</span>
<span id="cb82-7"><a href="#cb82-7" tabindex="-1"></a>}</span>
<span id="cb82-8"><a href="#cb82-8" tabindex="-1"></a></span>
<span id="cb82-9"><a href="#cb82-9" tabindex="-1"></a>Locs_vcov <span class="ot">&lt;-</span> Locs_cor</span>
<span id="cb82-10"><a href="#cb82-10" tabindex="-1"></a></span>
<span id="cb82-11"><a href="#cb82-11" tabindex="-1"></a><span class="cf">for</span> (loc1 <span class="cf">in</span> <span class="fu">rownames</span>(Locs_vcov)) {</span>
<span id="cb82-12"><a href="#cb82-12" tabindex="-1"></a>  <span class="cf">for</span> (loc2 <span class="cf">in</span> <span class="fu">colnames</span>(Locs_vcov)) {</span>
<span id="cb82-13"><a href="#cb82-13" tabindex="-1"></a>    <span class="co">#diagonal</span></span>
<span id="cb82-14"><a href="#cb82-14" tabindex="-1"></a>    <span class="cf">if</span> (loc1 <span class="sc">==</span> loc2) {</span>
<span id="cb82-15"><a href="#cb82-15" tabindex="-1"></a>      Locs_vcov[loc1,loc2] <span class="ot">&lt;-</span> locvars[loc1]</span>
<span id="cb82-16"><a href="#cb82-16" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb82-17"><a href="#cb82-17" tabindex="-1"></a>      <span class="co">#cov(x,y) = cor(x,y)*sd(x)*sd(y)</span></span>
<span id="cb82-18"><a href="#cb82-18" tabindex="-1"></a>      Locs_vcov[loc1,loc2] <span class="ot">&lt;-</span> Locs_vcov[loc1,loc2]<span class="sc">*</span><span class="fu">sqrt</span>(locvars[loc1])<span class="sc">*</span><span class="fu">sqrt</span>(locvars[loc2])</span>
<span id="cb82-19"><a href="#cb82-19" tabindex="-1"></a>    }</span>
<span id="cb82-20"><a href="#cb82-20" tabindex="-1"></a>    </span>
<span id="cb82-21"><a href="#cb82-21" tabindex="-1"></a>  }</span>
<span id="cb82-22"><a href="#cb82-22" tabindex="-1"></a>  </span>
<span id="cb82-23"><a href="#cb82-23" tabindex="-1"></a>}</span>
<span id="cb82-24"><a href="#cb82-24" tabindex="-1"></a></span>
<span id="cb82-25"><a href="#cb82-25" tabindex="-1"></a><span class="co">#weights = frequencies for each observed genotype-location combination</span></span>
<span id="cb82-26"><a href="#cb82-26" tabindex="-1"></a>gL <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sommerdf_smaller<span class="sc">$</span>id, <span class="st">&quot;:&quot;</span>, sommerdf_smaller<span class="sc">$</span>Loc)</span>
<span id="cb82-27"><a href="#cb82-27" tabindex="-1"></a>gL.weights <span class="ot">&lt;-</span> <span class="fu">table</span>(gL)</span>
<span id="cb82-28"><a href="#cb82-28" tabindex="-1"></a>gL.weights <span class="ot">&lt;-</span> gL.weights<span class="sc">/</span><span class="fu">sum</span>(gL.weights)</span>
<span id="cb82-29"><a href="#cb82-29" tabindex="-1"></a><span class="co">#vcov for all possible combinations of genotypes and locations</span></span>
<span id="cb82-30"><a href="#cb82-30" tabindex="-1"></a>gLvcov <span class="ot">&lt;-</span> <span class="fu">kronecker</span>(geno<span class="sc">@</span>G,Locs_vcov,<span class="at">make.dimnames =</span> T)</span>
<span id="cb82-31"><a href="#cb82-31" tabindex="-1"></a><span class="co">#leave only observed genotype-location combinations</span></span>
<span id="cb82-32"><a href="#cb82-32" tabindex="-1"></a>gLvcov <span class="ot">&lt;-</span> gLvcov[<span class="fu">names</span>(gL.weights),<span class="fu">names</span>(gL.weights)]</span>
<span id="cb82-33"><a href="#cb82-33" tabindex="-1"></a></span>
<span id="cb82-34"><a href="#cb82-34" tabindex="-1"></a>GxLvar <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">sum</span>(<span class="fu">diag</span>(gLvcov)<span class="sc">*</span>gL.weights) <span class="sc">-</span> </span>
<span id="cb82-35"><a href="#cb82-35" tabindex="-1"></a>  <span class="fu">matrix</span>(gL.weights,<span class="at">nrow=</span><span class="dv">1</span>)<span class="sc">%*%</span>gLvcov<span class="sc">%*%</span><span class="fu">matrix</span>(gL.weights,<span class="at">ncol=</span><span class="dv">1</span>))</span>
<span id="cb82-36"><a href="#cb82-36" tabindex="-1"></a></span>
<span id="cb82-37"><a href="#cb82-37" tabindex="-1"></a></span>
<span id="cb82-38"><a href="#cb82-38" tabindex="-1"></a>res_vars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp[(<span class="fu">nrow</span>(<span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp)<span class="sc">-</span></span>
<span id="cb82-39"><a href="#cb82-39" tabindex="-1"></a>                                            <span class="fu">length</span>(<span class="fu">unique</span>(sommerdf_smaller<span class="sc">$</span>Loc))<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span></span>
<span id="cb82-40"><a href="#cb82-40" tabindex="-1"></a>                                            <span class="fu">nrow</span>(<span class="fu">summary</span>(ans2c_open)<span class="sc">$</span>varcomp)</span>
<span id="cb82-41"><a href="#cb82-41" tabindex="-1"></a>                                           ,<span class="dv">1</span>]</span>
<span id="cb82-42"><a href="#cb82-42" tabindex="-1"></a>Loc_weights <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">table</span>(sommerdf_smaller<span class="sc">$</span>Loc))</span>
<span id="cb82-43"><a href="#cb82-43" tabindex="-1"></a>Loc_weights <span class="ot">&lt;-</span> Loc_weights<span class="sc">/</span><span class="fu">sum</span>(Loc_weights)</span>
<span id="cb82-44"><a href="#cb82-44" tabindex="-1"></a>Residual_var <span class="ot">&lt;-</span> <span class="fu">sum</span>(res_vars<span class="sc">*</span>Loc_weights)</span>
<span id="cb82-45"><a href="#cb82-45" tabindex="-1"></a></span>
<span id="cb82-46"><a href="#cb82-46" tabindex="-1"></a></span>
<span id="cb82-47"><a href="#cb82-47" tabindex="-1"></a></span>
<span id="cb82-48"><a href="#cb82-48" tabindex="-1"></a><span class="co">#calculate fixed effect variance in Sommer as variance of X%*%beta</span></span>
<span id="cb82-49"><a href="#cb82-49" tabindex="-1"></a>Xenv <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(BLUEs<span class="sc">~</span>Env<span class="dv">-1</span>, <span class="at">data=</span>sommerdf_smaller)</span>
<span id="cb82-50"><a href="#cb82-50" tabindex="-1"></a>Env_effects <span class="ot">&lt;-</span> ans2c_open<span class="sc">$</span>b[<span class="fu">which</span>(<span class="fu">rownames</span>(ans2c_open<span class="sc">$</span>b) <span class="sc">!=</span> <span class="st">&quot;Fg&quot;</span>)]</span>
<span id="cb82-51"><a href="#cb82-51" tabindex="-1"></a>Env_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(Xenv<span class="sc">%*%</span>Env_effects)</span>
<span id="cb82-52"><a href="#cb82-52" tabindex="-1"></a></span>
<span id="cb82-53"><a href="#cb82-53" tabindex="-1"></a><span class="co">#extract stage1error variance component from vcov</span></span>
<span id="cb82-54"><a href="#cb82-54" tabindex="-1"></a>Stage1Errorvar <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">diag</span>(EEV_smaller)) <span class="sc">-</span> <span class="fu">mean</span>(EEV_smaller)</span>
<span id="cb82-55"><a href="#cb82-55" tabindex="-1"></a></span>
<span id="cb82-56"><a href="#cb82-56" tabindex="-1"></a></span>
<span id="cb82-57"><a href="#cb82-57" tabindex="-1"></a></span>
<span id="cb82-58"><a href="#cb82-58" tabindex="-1"></a></span>
<span id="cb82-59"><a href="#cb82-59" tabindex="-1"></a><span class="co"># #heterosis effects for all observations </span></span>
<span id="cb82-60"><a href="#cb82-60" tabindex="-1"></a><span class="co"># XFg &lt;- sommerdf_smaller$Fg</span></span>
<span id="cb82-61"><a href="#cb82-61" tabindex="-1"></a>FgEffects <span class="ot">&lt;-</span> ans2c_open<span class="sc">$</span>b[<span class="st">&quot;Fg&quot;</span>,]</span>
<span id="cb82-62"><a href="#cb82-62" tabindex="-1"></a><span class="co"># Fg_variance &lt;- var(XFg*FgEffects)</span></span>
<span id="cb82-63"><a href="#cb82-63" tabindex="-1"></a></span>
<span id="cb82-64"><a href="#cb82-64" tabindex="-1"></a><span class="co"># Heterosis effects for each genotype</span></span>
<span id="cb82-65"><a href="#cb82-65" tabindex="-1"></a><span class="co"># Extract XFg from geno, not from the model</span></span>
<span id="cb82-66"><a href="#cb82-66" tabindex="-1"></a><span class="co"># In the model, there are repeated Fg coefficients for genotypes that had more</span></span>
<span id="cb82-67"><a href="#cb82-67" tabindex="-1"></a><span class="co"># than one BLUE (genotypes present in more than one environment), and there are </span></span>
<span id="cb82-68"><a href="#cb82-68" tabindex="-1"></a><span class="co"># no Fg coefficients for genotypes that were not phenotyped and had no BLUEs.</span></span>
<span id="cb82-69"><a href="#cb82-69" tabindex="-1"></a><span class="co"># If we take Fg from geno, there are no such problems.</span></span>
<span id="cb82-70"><a href="#cb82-70" tabindex="-1"></a></span>
<span id="cb82-71"><a href="#cb82-71" tabindex="-1"></a>XFg <span class="ot">&lt;-</span> geno<span class="sc">@</span>Fg</span>
<span id="cb82-72"><a href="#cb82-72" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> XFg<span class="sc">*</span>FgEffects</span>
<span id="cb82-73"><a href="#cb82-73" tabindex="-1"></a><span class="co">#reorder heterosis effects to be in the same order as BLUPs</span></span>
<span id="cb82-74"><a href="#cb82-74" tabindex="-1"></a>heterosis_effects <span class="ot">&lt;-</span> heterosis_effects[<span class="fu">match</span>(<span class="fu">rownames</span>(ans2c_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Ginv)</span><span class="st">`</span>),</span>
<span id="cb82-75"><a href="#cb82-75" tabindex="-1"></a>                        <span class="fu">names</span>(heterosis_effects))]</span>
<span id="cb82-76"><a href="#cb82-76" tabindex="-1"></a>Fg_variance <span class="ot">&lt;-</span> <span class="fu">var</span>(heterosis_effects)</span>
<span id="cb82-77"><a href="#cb82-77" tabindex="-1"></a></span>
<span id="cb82-78"><a href="#cb82-78" tabindex="-1"></a></span>
<span id="cb82-79"><a href="#cb82-79" tabindex="-1"></a></span>
<span id="cb82-80"><a href="#cb82-80" tabindex="-1"></a></span>
<span id="cb82-81"><a href="#cb82-81" tabindex="-1"></a></span>
<span id="cb82-82"><a href="#cb82-82" tabindex="-1"></a></span>
<span id="cb82-83"><a href="#cb82-83" tabindex="-1"></a><span class="co">#Put Everything together in a dataframe</span></span>
<span id="cb82-84"><a href="#cb82-84" tabindex="-1"></a>SommerVars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="co">#Env_variance, #environmental variance</span></span>
<span id="cb82-85"><a href="#cb82-85" tabindex="-1"></a>  ans2c_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2c_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Ginv:isc:isc&quot;</span>]<span class="sc">*</span>meanG, <span class="co">#additive</span></span>
<span id="cb82-86"><a href="#cb82-86" tabindex="-1"></a>  GxLvar, <span class="co">#additive by location interaction</span></span>
<span id="cb82-87"><a href="#cb82-87" tabindex="-1"></a>  ans2c_open<span class="sc">$</span>sigma[<span class="fu">names</span>(ans2c_open<span class="sc">$</span>sigma) <span class="sc">==</span> <span class="st">&quot;id:Dinv:isc:isc&quot;</span>]<span class="sc">*</span>meanD, <span class="co">#dominance</span></span>
<span id="cb82-88"><a href="#cb82-88" tabindex="-1"></a>  Fg_variance, <span class="co">#heterosis</span></span>
<span id="cb82-89"><a href="#cb82-89" tabindex="-1"></a>  Residual_var, <span class="co">#residual</span></span>
<span id="cb82-90"><a href="#cb82-90" tabindex="-1"></a>  Stage1Errorvar) <span class="co">#stage1 estimation error</span></span>
<span id="cb82-91"><a href="#cb82-91" tabindex="-1"></a>ASRemlVars <span class="ot">&lt;-</span> <span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>var</span>
<span id="cb82-92"><a href="#cb82-92" tabindex="-1"></a>variances_ans2c <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sommer =</span> <span class="fu">c</span>(SommerVars),</span>
<span id="cb82-93"><a href="#cb82-93" tabindex="-1"></a>                        <span class="at">ASReml =</span> <span class="fu">c</span>(ASRemlVars))</span>
<span id="cb82-94"><a href="#cb82-94" tabindex="-1"></a><span class="fu">rownames</span>(variances_ans2c) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">summary</span>(ans2c<span class="sc">$</span>vars)<span class="sc">$</span>var)</span>
<span id="cb82-95"><a href="#cb82-95" tabindex="-1"></a>                                      </span>
<span id="cb82-96"><a href="#cb82-96" tabindex="-1"></a><span class="co">#Very different variance components </span></span>
<span id="cb82-97"><a href="#cb82-97" tabindex="-1"></a><span class="co">#The difference is due to ASReml updating the FA loadings while estimating </span></span>
<span id="cb82-98"><a href="#cb82-98" tabindex="-1"></a><span class="co">#variances, which sommer doesn&#39;t do</span></span>
<span id="cb82-99"><a href="#cb82-99" tabindex="-1"></a>variances_ans2c</span></code></pre></div>
<pre><code>##                 Sommer ASReml.Variance ASReml.PVE
## additive      1.334965            35.3      0.211
## add x loc    31.408256             8.0      0.048
## dominance    18.553387            35.9      0.215
## heterosis     2.232232             4.0      0.024
## g x env      42.708615            24.3      0.145
## Stage1.error 59.873645            59.9      0.357</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a><span class="co">#BLUPs</span></span>
<span id="cb84-2"><a href="#cb84-2" tabindex="-1"></a>Gamma<span class="ot">=</span>RotatedLoadings<span class="sc">$</span>loadings <span class="co">#loadings after rotation</span></span>
<span id="cb84-3"><a href="#cb84-3" tabindex="-1"></a>score.mat<span class="ot">&lt;-</span>ans2a_open<span class="sc">$</span>uList[[<span class="dv">2</span>]];<span class="co">#extractfactorscores </span></span>
<span id="cb84-4"><a href="#cb84-4" tabindex="-1"></a>BLUPs_GxE<span class="ot">&lt;-</span>score.mat<span class="sc">%*%</span><span class="fu">t</span>(Gamma)<span class="co">#BLUPsforallenvironments</span></span>
<span id="cb84-5"><a href="#cb84-5" tabindex="-1"></a><span class="fu">head</span>(BLUPs_GxE)</span></code></pre></div>
<pre><code>##                    MI         NY         OR         WI        MO         NC
## A00188-3C   0.9208769  1.0367201  1.0946311  0.9586373  1.126152  0.9516248
## A01143-3C   8.2316992  9.2771533  9.8010492  8.5726367 10.086105  8.5093963
## A03440-2C   2.2538062  2.5246933  2.6585055  2.3419016  2.731469  2.3254449
## A07117-2C  -5.5840149 -6.2596975 -6.5940733 -5.8038306 -6.776348 -5.7628039
## AC00206-2W  0.8321097  0.9331565  0.9832092  0.8649886  1.010490  0.8588549
## AC01151-5W  3.9785528  4.4552756  4.6905712  4.1335636  4.818886  4.1045949</code></pre>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a><span class="co">#Weights for each location:</span></span>
<span id="cb86-2"><a href="#cb86-2" tabindex="-1"></a>Locs.index <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">WI=</span><span class="dv">1</span>, <span class="at">NC=</span><span class="dv">1</span>, <span class="at">OR=</span><span class="dv">1</span>, <span class="at">NY=</span><span class="dv">1</span>, <span class="at">MO=</span><span class="dv">1</span>, <span class="at">MI=</span><span class="dv">1</span>)  </span>
<span id="cb86-3"><a href="#cb86-3" tabindex="-1"></a><span class="co">#scale the weights to unit variance:</span></span>
<span id="cb86-4"><a href="#cb86-4" tabindex="-1"></a>index.scaled <span class="ot">&lt;-</span> Locs.index<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">diag</span>(Locs_vcov))[<span class="fu">match</span>(<span class="fu">names</span>(Locs.index),</span>
<span id="cb86-5"><a href="#cb86-5" tabindex="-1"></a>                                                       <span class="fu">rownames</span>(Locs_vcov))]</span>
<span id="cb86-6"><a href="#cb86-6" tabindex="-1"></a>index.scaled <span class="ot">&lt;-</span> index.scaled<span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">sum</span>(index.scaled<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb86-7"><a href="#cb86-7" tabindex="-1"></a></span>
<span id="cb86-8"><a href="#cb86-8" tabindex="-1"></a>weighted_BLUPs_GxE <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">nrow</span>(BLUPs_GxE)))</span>
<span id="cb86-9"><a href="#cb86-9" tabindex="-1"></a><span class="cf">for</span> (loc <span class="cf">in</span> <span class="fu">names</span>(index.scaled)) {</span>
<span id="cb86-10"><a href="#cb86-10" tabindex="-1"></a>  weighted_BLUPs_GxE <span class="ot">&lt;-</span> weighted_BLUPs_GxE <span class="sc">+</span> index.scaled[loc]<span class="sc">*</span>BLUPs_GxE[,loc]</span>
<span id="cb86-11"><a href="#cb86-11" tabindex="-1"></a>}</span>
<span id="cb86-12"><a href="#cb86-12" tabindex="-1"></a>weighted_BLUPs_GxE <span class="ot">&lt;-</span> weighted_BLUPs_GxE<span class="sc">/</span><span class="fu">sum</span>(index.scaled)</span>
<span id="cb86-13"><a href="#cb86-13" tabindex="-1"></a></span>
<span id="cb86-14"><a href="#cb86-14" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2c_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Ginv)</span><span class="st">`</span> <span class="sc">+</span></span>
<span id="cb86-15"><a href="#cb86-15" tabindex="-1"></a>              ans2c_open<span class="sc">$</span>uList<span class="sc">$</span><span class="st">`</span><span class="at">vsc(isc(id), Gu = Dinv)</span><span class="st">`</span> <span class="sc">+</span> </span>
<span id="cb86-16"><a href="#cb86-16" tabindex="-1"></a>              weighted_BLUPs_GxE<span class="sc">+</span></span>
<span id="cb86-17"><a href="#cb86-17" tabindex="-1"></a>              heterosis_effects <span class="sc">+</span></span>
<span id="cb86-18"><a href="#cb86-18" tabindex="-1"></a>              <span class="fu">mean</span>(Env_effects)</span>
<span id="cb86-19"><a href="#cb86-19" tabindex="-1"></a></span>
<span id="cb86-20"><a href="#cb86-20" tabindex="-1"></a></span>
<span id="cb86-21"><a href="#cb86-21" tabindex="-1"></a><span class="co">#SommerBLUPs</span></span>
<span id="cb86-22"><a href="#cb86-22" tabindex="-1"></a></span>
<span id="cb86-23"><a href="#cb86-23" tabindex="-1"></a>prep1 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>blues,<span class="at">vcov=</span>EEV,<span class="at">geno=</span>geno,<span class="at">vars=</span>ans2c<span class="sc">$</span>vars)</span>
<span id="cb86-24"><a href="#cb86-24" tabindex="-1"></a> </span>
<span id="cb86-25"><a href="#cb86-25" tabindex="-1"></a></span>
<span id="cb86-26"><a href="#cb86-26" tabindex="-1"></a>ASRemlBLUPs <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep1,<span class="at">geno=</span>geno,<span class="at">index.coeff=</span>index.scaled,<span class="at">what=</span><span class="st">&quot;GV&quot;</span>)</span>
<span id="cb86-27"><a href="#cb86-27" tabindex="-1"></a><span class="co">#Mostly similar correlation</span></span>
<span id="cb86-28"><a href="#cb86-28" tabindex="-1"></a><span class="co">#The difference is due to ASReml updating the FA loadings while estimating </span></span>
<span id="cb86-29"><a href="#cb86-29" tabindex="-1"></a><span class="co">#variances, which sommer doesn&#39;t do</span></span>
<span id="cb86-30"><a href="#cb86-30" tabindex="-1"></a><span class="fu">cor</span>(ASRemlBLUPs<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb86-31"><a href="#cb86-31" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(ASRemlBLUPs<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.946667</code></pre>
<p>The correlation matrix and <code>uniplot</code>, which are based on
the additive values, indicate that the yield in NC differs somewhat from
the other five sites.</p>
<p>In comparison to the single-location analysis in Vignette 1, an
additional argument is required for the <code>blup</code> function to
specify the index coefficients. These coefficients are interpreted as
the relative weights for each location after standardization to unit
variance. The following code compares the reliability of breeding value
(BV) predictions between WI and NC.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="co">#prep1 &lt;- blup_prep(data=blues,vcov=vcov,geno=geno,vars=ans2c$vars)</span></span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a></span>
<span id="cb88-3"><a href="#cb88-3" tabindex="-1"></a>WI.index <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">WI=</span><span class="dv">1</span>, <span class="at">NC=</span><span class="dv">0</span>, <span class="at">OR=</span><span class="dv">0</span>, <span class="at">NY=</span><span class="dv">0</span>, <span class="at">MO=</span><span class="dv">0</span>, <span class="at">MI=</span><span class="dv">0</span>)  </span>
<span id="cb88-4"><a href="#cb88-4" tabindex="-1"></a>NC.index <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">WI=</span><span class="dv">0</span>, <span class="at">NC=</span><span class="dv">1</span>, <span class="at">OR=</span><span class="dv">0</span>, <span class="at">NY=</span><span class="dv">0</span>, <span class="at">MO=</span><span class="dv">0</span>, <span class="at">MI=</span><span class="dv">0</span>)  </span>
<span id="cb88-5"><a href="#cb88-5" tabindex="-1"></a></span>
<span id="cb88-6"><a href="#cb88-6" tabindex="-1"></a>WI.pred <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep1,<span class="at">geno=</span>geno,<span class="at">index.coeff=</span>WI.index,<span class="at">what=</span><span class="st">&quot;BV&quot;</span>)</span>
<span id="cb88-7"><a href="#cb88-7" tabindex="-1"></a>NC.pred <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep1,<span class="at">geno=</span>geno,<span class="at">index.coeff=</span>NC.index,<span class="at">what=</span><span class="st">&quot;BV&quot;</span>)</span>
<span id="cb88-8"><a href="#cb88-8" tabindex="-1"></a></span>
<span id="cb88-9"><a href="#cb88-9" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">merge</span>(NC.pred,WI.pred,<span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb88-10"><a href="#cb88-10" tabindex="-1"></a><span class="fu">colnames</span>(pred) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;BV.NC&quot;</span>,<span class="st">&quot;r2.NC&quot;</span>,<span class="st">&quot;BV.WI&quot;</span>,<span class="st">&quot;r2.WI&quot;</span>)</span>
<span id="cb88-11"><a href="#cb88-11" tabindex="-1"></a></span>
<span id="cb88-12"><a href="#cb88-12" tabindex="-1"></a><span class="fu">ggplot</span>(pred,<span class="fu">aes</span>(<span class="at">x=</span>r2.NC,<span class="at">y=</span>r2.WI)) <span class="sc">+</span></span>
<span id="cb88-13"><a href="#cb88-13" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb88-14"><a href="#cb88-14" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="at">ratio=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb88-15"><a href="#cb88-15" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb88-16"><a href="#cb88-16" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb88-17"><a href="#cb88-17" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>),<span class="at">y=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>)),</span>
<span id="cb88-18"><a href="#cb88-18" tabindex="-1"></a>            <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y),<span class="at">linetype=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb88-19"><a href="#cb88-19" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb88-20"><a href="#cb88-20" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;NC&quot;</span>) <span class="sc">+</span></span>
<span id="cb88-21"><a href="#cb88-21" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;WI&quot;</span>) <span class="sc">+</span></span>
<span id="cb88-22"><a href="#cb88-22" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;BV Reliability&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABEVBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYzMzM6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmOgBmOmZmkJBmkNtmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQOjqQZgCQZmaQkGaQkLaQttuQ2/+rbk2rbm6rbo6rjk2r5OSr5P+2ZgC2Zjq2kDq2tpC2ttu22/+2///Ijk3I///bkDrbtmbbtpDb29vb/7bb/9vb///kq27k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///90+w9dAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAL2klEQVR4nO2dDXvktBGAlcB1WWj5alJoev0gob2jLdl+kaP0Utok7LWQkE3o5jb+/z+kkuy1JHvk8dqSLckzz8OdGcuS/N5oJI9nZZaRNAobuwOhCwFChAAhQoAQIUCIECBECBAiBAiRXoCWTMoHF9nD0ZNrqdos9i/1c+zdC/2KzeLJtfhPaR6OZpVz8vz9Ly/79MyduADEOJMlO5aq8n6359i+fqetAS33kwAk72LzTw7nfp7f56oAtb1DfvJAu8KEI0QBMs6nBIjD4QzO8+Pz7Y1t7/DhSAD64Wk+2JQFvfoxY3u/loBWc/bmhWFB3y246R3ksPMKRhMXFvT13pkwHXEfyh4KQP87F/r7uRhsvJgaQvn4O+ZXvDE3z5WA8spWovbxxIkPEmhyuyhHmOGDNou9T7LsP3wUKgj7F5kclw9H7KfXm6/ZDPBB0hzPq4NyWHEC6ENxC0vxL70s76Y4tyfOyTEojUpz0v/911/mbFbMf2LyqwNaSQsbdYQ58UGvmBgK4k42iwPj3Kv5B4LFqgC5d6amcTnoJKAcnjqnAIkxNvIIcwKo8Mz8z/u3z8xzK5YPvCogPrDe/9u/vz1qBiSqHHmEuQR0Pz9WI2x7rvDRW8+0BZBrch80y2xDjJP9zcgjzNEQk1w2ix/9+aB67uEod9Jf8Kleg8DRXIu5X9oSH5p1Jy1H1mbxk5FHmBsnXUxdK6bdzbJ85JiVY+xAg5BrBKD359u5TgeUz41LNvaC0QWgdwss5QOZPFfcGUfB8d3zheIbn2gLRbFyfPMLPja5D+ILxfcuKo8aD0/Z7DorF+jjSeBP82pdNZaEDeiHpyPPYWED4v6bjTyHhQ2Iu68Px+5D0ICCEAKECAFChAAhQoAQIUCI9AH0VoLiFlB5dFs/6VLlu36lIkAEqJ+KABGgfioC5AnQ688OP/5eHq1PDj/6plbb1AE9vjjNbn4ujl7//mV2U7CKAxBjLJOBYfkndmFHQK//8E22/q0wnPWz7+X/VWoLFBCrix9AEgu3Hd2CxLLzNmwB+DDkko6A7j7eAtK8UfgWBAHybUHrT19mdx9FMsRAPp59kGZLkQLKKemlnMxiz4tZLB4LgsZXOaHp5TIHgArPI4zo7vDwZy9rtYUHCOZTB8TMCyeykrbQAS3IrCt9QCUEmA/gg7JJAbKzYc0XFjJlQFtG6kKAWSqA+M0CxZrpbAkpTwVUnwgg0NPa+agzel1aBZMHVJ5Dm5wsoExNX1MAVPFBxa0jgFo1mQogQ1XzNM2E5GrJVleKgOxcYEK39ZtODRDTwzp1GCig1B818lu2AmphQfbqHQMaJHBaE3nL+nEuYIQ117avOzULMm3kFjCnHZtMApDmg3QQsPup1sSmAEipNBDan/oIq1zlKWhvkfEBoU/v1SU41uS0AHVockKAujWZGqBm8+nQZGKAbGxKPvqFgKpefUqAmk2nduHkYtKId65eCMVXAVUKgBgS/TEITdAHbQHYnA8BygFAeG618x2bTAYQaD+3W3SVC21+G1AlAMj+chmK5cvFwA5N9s5yvTkUclqtbehZDHbMMCB7fBVQ9c5yFXI3epYrRCezAdqpyd4ZZlmmEsxGDZjV+BioOjfZO0eRS2FK3rNcGRwqBd2Ppu7XaP8sV82A/FqQYQ7aE0OD+6kbEJTi4NuCSg8UDCClrVaFxFcBlQMfdPUcqG0gQBoUCFBtrhvMB6ks1+zxy3KEeXbSrLYgBoeXejBz0Yv+Wa6aCxp6FrPAyYpzTpqMeSVtsx6nTcYLCB5e9pfKUwu5gmiqhPRH2o5NRgoIMp3bJkDt4quAKk5A4NBqBNRZFSUg2PUUfttxkzEC0rFontlPk5ED0icu65W9jCo+QNaZ3XblDvFVQBUdICsf25XjxIMs4h+QAcactMgHZRX7gd5XOG8yLkDQyPLcZFxZropOy/L9m4zSgtpeuXN8FVDFBSgDnY/tylEDZrAMM82D77si8UGd+7GbigAhKgKEqVAfxFw1GSkgdMHtrMk0AXWNrwKqNAE5VBEgAtRPlR4gCto3F6OgffPDar/4KqCKDlCxiA7eB6ndEx9fDLqHWSSAtCzXq9PBsly1OHTogFSGWbmPq1mbj96WkXpLsRYhosEAaTtxPvuqGGJDZLk2xVodxFcB6Z3luj45lbgquP1ZkK2Yi/gqoHJgQcNtVaoFOSLyQZ9HtJdrB1X/LNerwYaYr8o8roNklutnA+9pD0zzlP6iSe6qzcWjxyZTAOQuvgqoUgDktcn4ADnLoW+nigmQ/2QgQBURIONtYQnNXf2wKnJAjuOrgKoVIPEF4UKav2U+NCDX8VVA1RJQy2+8T9YHnePGU6ltgo8a93Mc0qQBSVmO6IMquvDSXzYL9GPvQ/5ezGv9HQBxN/TkGj6lZDhAQC/HBLRkbO8M0Fcl9O+LdZCW0/xxCzrZhJ00dz+tVkKTBZTJWX4Gn9FkoN+Lea4/6wQoE54Ic0TDhDs812+qdlsHbRYDroPKJwuzlM/4KqAKdyWtnk3D90FiiTj4s1hMgEZ5mo8JUGvx7YOCDXe0Fd/v5l1W1lIVEyD/8VVAFRMgX5URoD4qApQIoGqvggek0oDFbrfev/U8UHwVUDlJAwZqc9vboeKrgKp3Cp6+jSL5oFJU6iYfa8V20t7TgEeR/mnAn75UVkQWtBUj+Vf5IR+AwGSp4AEZG257BQQnSwUPSKUBi8H2+Hdv07wl/hw8IC0NmK+Dhvw51DD1x7eS9lkZAeqjChlQw/4JBEjvCAECVY3xZwKEqAgQoiJAiGrqgND9EyYOiKGlpg2oRfx52oBaqAgQoooW0NgBUvcSmgW13D8hWgvq3I+ycQLUpGr9Y9SpAmqtIkCIigAhqikC2mn/hAkCaoivEqCsOb5KgHZXESBERYAQ1bQAddg/YVKAuuxPOyVAeHw1SkAqyzXP6KzWNnUfpGW5ikRgAlQVPcNs/bs/EqCqaDmKj1/+40XXLFc/+9M6ld5ZrtnN884+qPv+tBFZED/qCqjHZn/BA1I+SPxU4/DwebW2qfsgbbNbmuZBUVmuBGgHaQ+o5/4JyQPquz9t6oB2i69OEFBvFQFCVAQIUaUMyMn+CQkDcrM/bbqAusRXJwXIkYoAISoChKjSBORw/4RoATUFLyOIrwIynAU5/f5XtBbUuR87qggQoiJAiCo1QM73p00MkPv9adMC1De+mjwgDyoChKgIEKJKB5Cn/WmTAeRrf9pUALmJryYMqH0/dlQRIEQVPiCV5Xp3OPCnjIeovzcgleUq93nbZgrpgLzuTxs8IHMfxfJIA+R3f9rgAZk7ceYWZGS5xhlfBaR/lmu2PqF9FGtiWlB5RIC2MtxerrAqeEDmXq5kQXWhvVw7CQFChAAhQoAQIUCIvJWgOAWkoRq81GBNEiCkFAFCSrkBlLAQIEQIECIECBEChEg/QPDbDnsp/Rev9lL6V/Aaij2+UIEFa6n8Z9qWNlVd6xNbi70AWd52WEtl+s4f9lJXNoa1Yncl+IYmW5QSIa4buFQvQJa3HQ2ltJ0/rKWMr+DZi4mjVh3TPvxlKyVjynB9vQBBbzsaS5U7fzSVUl/Bayy2fvaVdYgZHbP1S2/SkwVZ3nbYS+k7f1hLGV/Bayh2cipvEeuY1YD0UrqbNMWhBdl6Ytn5o7Euqx8yKmvRpN0DGXXxf5M72Es79EG2m7Ls/NGpLt0HfW4HpFd2BbdnlDLeRZjScxYD33ZYS2UN07zlK3gNlV3Zh5hWqsHlq1K+LMjytsNeCl8HYXVpxfgRslqSpawuSC91Z22SVtKIECBECBAiBAgRAoQIAUIkYEAPRzPx12r/kv+5ZIztnY3Qi6ABseMsB7RZPLkWkI6H70XQgH713nUO6Fzw4YTyvwaVoAEdnx9IQPxotF6EDej+nUsB6P7tMbxPLmEDypYHEtA7l6P1InBAD784IwuyifQ8y5nmgziuwXsROqDNn/7KZ7Fi+lruDz/UQgeUrRitgyxSDKxzaTbntJIOVAgQIgQIEQKECAFChAAhQoAQIUCIECBECBAi/we8ICZ5K6r3XAAAAABJRU5ErkJggg==" /><!-- --></p>
<p>The figure above illustrates that breeding values were predicted with
higher reliability in WI compared to NC. This outcome aligns with the
high correlation between WI and the other four sites (OR, MO, MI, NY)
when contrasted with NC.</p>
<p>You can use the <code>mask</code> argument for <code>blup_prep</code>
to selectively exclude individuals at one or more locations. This allows
you to assess the accuracy of predictions in new environments. While the
previous analysis assessed the reliability of GEBVs in WI when WI
phenotypes were available, the next analysis excludes WI phenotypes:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" tabindex="-1"></a><span class="co">#StageWise</span></span>
<span id="cb89-2"><a href="#cb89-2" tabindex="-1"></a>WI.env <span class="ot">&lt;-</span> <span class="fu">unique</span>(blues<span class="sc">$</span>env[blues<span class="sc">$</span>loc<span class="sc">==</span><span class="st">&quot;WI&quot;</span>])</span>
<span id="cb89-3"><a href="#cb89-3" tabindex="-1"></a></span>
<span id="cb89-4"><a href="#cb89-4" tabindex="-1"></a>prep2 <span class="ot">&lt;-</span> <span class="fu">blup_prep</span>(<span class="at">data=</span>blues,<span class="at">vcov=</span>EEV,<span class="at">geno=</span>geno,<span class="at">vars=</span>ans2c<span class="sc">$</span>vars,</span>
<span id="cb89-5"><a href="#cb89-5" tabindex="-1"></a>                   <span class="at">mask=</span><span class="fu">data.frame</span>(<span class="at">env=</span>WI.env))</span>
<span id="cb89-6"><a href="#cb89-6" tabindex="-1"></a></span>
<span id="cb89-7"><a href="#cb89-7" tabindex="-1"></a>pred2 <span class="ot">&lt;-</span> <span class="fu">blup</span>(<span class="at">data=</span>prep2,<span class="at">geno=</span>geno,<span class="at">index.coeff=</span>WI.index,<span class="at">what=</span><span class="st">&quot;BV&quot;</span>)</span>
<span id="cb89-8"><a href="#cb89-8" tabindex="-1"></a></span>
<span id="cb89-9"><a href="#cb89-9" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(pred2[,<span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;r2&quot;</span>)],WI.pred[,<span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;r2&quot;</span>)],<span class="at">by=</span><span class="st">&quot;id&quot;</span>)</span>
<span id="cb89-10"><a href="#cb89-10" tabindex="-1"></a><span class="fu">colnames</span>(plot.data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;without.WI.pheno&quot;</span>,<span class="st">&quot;with.WI.pheno&quot;</span>)</span>
<span id="cb89-11"><a href="#cb89-11" tabindex="-1"></a></span>
<span id="cb89-12"><a href="#cb89-12" tabindex="-1"></a><span class="fu">ggplot</span>(plot.data,<span class="fu">aes</span>(<span class="at">x=</span>without.WI.pheno,<span class="at">y=</span>with.WI.pheno)) <span class="sc">+</span></span>
<span id="cb89-13"><a href="#cb89-13" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">coord_fixed</span>(<span class="at">ratio=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb89-14"><a href="#cb89-14" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb89-15"><a href="#cb89-15" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>) <span class="sc">+</span> </span>
<span id="cb89-16"><a href="#cb89-16" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>),<span class="at">y=</span><span class="fu">c</span>(<span class="fl">0.3</span>,<span class="fl">0.8</span>)),</span>
<span id="cb89-17"><a href="#cb89-17" tabindex="-1"></a>            <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y),</span>
<span id="cb89-18"><a href="#cb89-18" tabindex="-1"></a>            <span class="at">linetype=</span><span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb89-19"><a href="#cb89-19" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb89-20"><a href="#cb89-20" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Without WI phenotypes&quot;</span>) <span class="sc">+</span> </span>
<span id="cb89-21"><a href="#cb89-21" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;With WI phenotypes&quot;</span>) <span class="sc">+</span> </span>
<span id="cb89-22"><a href="#cb89-22" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;WI BV Reliability&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABHVBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYzMzM6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmOgBmOmZmZmZmkJBmkNtmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2ObquOyP+QOgCQZgCQZmaQkGaQkLaQtpCQttuQ27aQ29uQ2/+rbk2rbm6rbo6rjk2r5OSr5P+2ZgC2Zjq2kDq2tpC2ttu22/+2/9u2///Ijk3I///bkDrbtmbbtpDb29vb/7bb/9vb///kq27k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///+TBl7xAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAO20lEQVR4nO2dDXvbthGA4bSeqmTt0s5eOyVbl8rdkm5rtK9uTtdF7ebaZbrGbmRlVmT8/58xfJEASIAASQAESdzzJJKPAEi+OhxA8HgEMEmtgL4PIHZJgAySABkkATJIAmSQBMggCZBBEiCDdAG0PjjFHxk4vEQf+9XhyxX5RnRE3j+DuwXT7Vd3zsVt4L0zsTFU/XJf1MeyW8xK28j27a/OOxxzY+kCaAOW+GMNCCh0PvsyIICYZLQUP998G7gjnqk1oOzOYADRM9gtPl4cQYJLAETOYv81grOd0/NkOPNteOOR0JgMhzdf2T4gQPSYNwf/WOEzQQdeBoTgIAZr+n2dn1h+hjvC9fVD2tm4Bb34KQAHvyGANnPw9plkQS9XyPSOKGzagGfp5KTJqWaHl/hT7iLMgr7BnW9DLIXbAwP0vzXWb+e4s6FivAvR/rdENd6ay9sKQLSxDfWBfqUToO18iQ5+RnoP/jmrPgijYYaW9zDJB+1XB59A+D3qhRzCnTNI+uVuAX5+uf8GzBQ+iJjjutwpfUgnQBgKhoQ/MYAqoA/w3xn+pbPibNi2A7yN9MGie7L6//33n+dgxsY/PPhVAeU/iX/pNg9Cv+GG/ZZ5N6MbWDd6Aagbx8ZVnA3Z9mL+Pi66YSAPTvkwTjodAUTh8W0cEO5jQXpYR0CIDrFzdNDrGawCYp4Z/b+9W5wN3bYBtOOVAaGOdf+v//lhUQ8INxmkh3UEtJ0/ImfBPnWAUDfkPSzfxnx07plyAFRDfdAM6roYIvsoSA/rCGi/uk8sg30quhibZP/kT/xs2LbdgjrpL9BQL0BAaC7x2E9sCXXNqpM+oHt8N0gP63otloHcHKhDLTtpNnRtgHA2uXVl2EFt8tGOQ6AaDOj+PB/rREB0bMzkibg36QhoA2bCZwXQewxLcUFGtrEzQygQvi2aKL71iTBRxDPHt79AfRP5IDRR/NlZ6VJj9xDMLmExQfctg72a5/MqvzJUQK8fBhnDhgoI+W8QZAwbKiDkvj4ItKthAgooCZBBEiCDJEAGSYAMkgAZpAugd0YobgEV315VN7pU+W6fqxKgBKibKgFKgLqpEiBPgN58dvzRj+Tbzcnxh99VWps6oNtnT+HVL/G3N79/Dq8YqwECAoVoSrUE9OYP38Gb32LDuXn8I/mr1NogAAFJ1KVaAiJYkO2IFoSnna8GJKAk6lItAV1/lAMSvNGwLKjMx5cF3Xz6HF5/OMAuVsfHpQ8SbGksgICTUewJG8UGa0FlQlwPHQBingcb0fXx8S+eV1qLH5DOgoBccaIz6YqHdu2DNDIEQJiFic+UASnYKM44AdL751LFiAEVR+4OkIoOmkIXmxUV4wXEf1tngNR8ilJAVXHygEwVpwNIYz6mivECcuuD1Hi0P8AgANWpuk6bRT6k1NTnQTV4SKnySU8MUK394FKVc54WIA0fxaxQ0ZZjQGEXTe1Ex8eu9vgtyGQ/UxnFgO6a0wQITANQySiEUkY+UwYE1Gs/8vrzhAGZ3Y95l2MBpPJBSjhNdzkaQFaX7gkQF13vEkoBZcWSaqyAtN5HuMC3amukgPTemV+/2rU1TkA1w1fyQbB+ePcGaLdY7ham54qjBtRul9aA1jP8cHJW/2BxFIDq7IeV0t1nVqhsASEDIlld6k0oBkA6PjbrqwpVA0Ako0sOqIgruzrG8rTcWnyAeCkvlxr71RHOrbJmXYxHuWK5jijKtYZNq11a+6AtTuBTpFfhEWYQ8gCz/gFV0JT5hBrmeYwiEmZK7aJcge3ap1VTJXHQaPcoV8GAWlhQyy6g0im6VrWUx4vVDACe80e0oMIDxQZIUcqwvqpQ2c+DDn+gIz0R0QddPFG0Fh6QyjeXS5nWVxWqRsP8shjmeZQrvP2y6GFtnHQ7J6qwDdXY5WBcaAlIiHIVXFBvo5huZA8ICGa4i5ny8PUESMcnKCCaOsuQgSYWQLwU/8P3xaqV9ANIhwfCV/zvtkY1fEBa84EiILv1VYWq0TwImLKE9QCojo9oQW13ae+kaT7Q2Jw056LgI/mglrtsMMzjj9jWg8p83O9yLICUfFipMMsdJNNuZF2s1v/wUp12aW9BxVHUGFFYQEY+tFS3XQ53mBe40P/97HKwgCTDAcq1jcBO2iYXcS+A/O7S3oLWABiTpYaLcs3pGEp131GTLmbOlxrOgpSzwkqh4D4II6qbCQUHJEKqFHPit5sAyvCbCWrfXRHYB8GYfBBOnG6cTIeeSUcEiCTkN8qkAcV6LaYuZvGonJ1q4IB0qtCL9sU7H8AA4oNg+/VVhaqhBRkkFkAOVYO9FgulargmHedtH5+qoa9JV1XAqpS1anSjWKf1VYVqDICAEAHdbX1VoRpBF8un03E5aZ498fZZzznMYgEkiRDlevG0tyhXCiZKQDzCrMjjKrcWAlBOJvdB6rsaHVVNwoCFez5CJs7HX7EuFjyXa2nN1Vm0rCT2geTSVRiPcr05eUpwlXCHtCAqLtZXFaqW12KiBfWXqlRele4X0H4lARJ80OcDzeVqqbK/Ny9NEYUo14veulgQVdN784oo1896yWkvda9WzxLaqYa63CE5aGfrqwrVGAC5W19VqFo+q6GRnizI5y5bPquhkb58kMddtn0UQS1BAHmMdVGohgdIjpYCumKuVMN7VkMCFM9MOp5nNURArtdXFaoBDvNx+iA7mfClRmk9SCMTBmSaAZVb8w9IE1bvWjXYe/Ne1lcVqpbrQRoJeKnht/0WPsgwRWSteVkW7lWsANk9qAEn7aTtJAEySBhA/u4SKlQDfGbV4/qqQhV98AIoZxf1ub6qUMUe/qJYOXTavlGVADmbB/XzzOpgAPX2zKrkg2oe7ekbkJ14HsW8r68qVEMC5H99VaEaEiBfjSVAXVQJ0EgAhbmNqlC1BcTDgHG2W+/veg60vqpQtVwPksOAFa25PdpQ66sKlSWg8uSQh+CJaRSn64PW5Rk0D91EfY2lkw4eBhxE7H0QuTGWQxLCgD99zq1oshZUSFYJJCdy4TEjuTJ/QpSAcGqB/GJeSrjtFZD6PnyEgNZS8hceBow72+0/vQ3zmvXn2ABlAJTSB/EwYDQP6vdxKL8qy2He5rYqnC4g7H5sbqxOFxBkb0UwiVtANfkTIgQEsScy5TFzCqgurUucgHBXC3dXo3b9OUZA0kxaI1P1QXiKaIQjtzYtQHbpp+B0AVmLK0DG/AkTB2R+jGfagCzWn6cNyEKVABlUEQKyi7Tve4HUvcQWaW+ZPyE+CwoUaW+bnzY+QGEi7a2fc4oPkGWkfevjaKiKDFCKtBclDfOwDSD/Ua6N8idMEFDN+uoAAPnPBly3vjoAQBE+cRhKlZy0m2F+WU6wpJbpAvK65Noif0JkgKoBVBppBahNftrYAGHxddunVZqkGAERyeQkb1RwRGe5tWn5ICpiAJUQ5YoDgROgcgCVGGF287s/Th5QJYBKiFG8/fJfz9pGufrJT+tUWgZQ8ShXePWktQ9qn582NguqBFBJyW7bAurw/q/YAMFyABX3QfhRjePjJ+XWJuaDqAgBVEKy2zTMC8IDqHiUawLExO1MumP+hMgAuQ+g6pqfNjJAzq/mm62vxg/IWibtg2wkATJIAmQQG0BO8ieMGJCb/LTjBdRmfXVSgBypEiCDKgEyqMYJyGH+hMECqlu8HMD6qkLCWZDT938N1oJaH0dDVQJkUCVABtXYADnPTzsyQO7z044LUNf11dED8qBKgAyqBMigGg8gT/lpRwPIV37asQBys746YkD2x9FQlQAZVPED4lGu18e9vMo4lKolIB7lSvK85ZFCIiCv+WmjByTnUSy+CYD85qeNHpCciZNakBTlOsz1VYW0BCREucKbk8nnUayKbEHFtwQol3C5XNWq6AHJuVyTBVUl5XJtJQmQQRIggyRABkmADPLOCMUpIAFV8FLBdpkAGUolQIZSbgCNWBIggyRABkmADJIAGaQbIPXdDn0p8YlXfSnxLXg1xW6f8YUFbSn6mLZmn7ytmxPdHjsB0tzt0JaCYuYPfakLHcNKsesCfM0uLUrhJa4rdalOgDR3O2pKCZk/tKWkt+Dpi+FvVgcmvPhLV4qsKavb6wRIdbejtlSR+aOuFH8LXm2xm8dfabuYdGC64xJ36cmCNHc79KXEzB/aUtJb8GqKnTwlp2g6MK0BiaVENymLQwvSHYkm80dtW1o/JDVmsUu9B5LaQr/JtdpLO/RBupPSZP5o1Zbogz7XAxIbu1DvTyol3YuQpeMoprzboS0Fa4Z5zVvwahq70HcxoVSNy+elfFmQ5m6HvpR5HmRqSyiGvhlmS6SU1gWJpa61u0wzaYMkQAZJgAySABkkATJIAmQQf4DWJFNlRpLF7Vez7d1T+PoM4g+loG2aWoJoa5dacSj+AJGE+PvVr/EZ7xY46yk+Pd0p5vpSrYaAjACbiz9A2/kSneODvz9Ax7whCfVsAJVqjRnQfoWsZnP4En9kh5fbu3+bA4BM4i8AgCXNlDZjp8S2KWuRMya1kFFJtZGt0YpFazgz3dEa985sllfJi9K0dZYvaRbEo5NG5wfXR+hYIf5Ap0osaI602Z1z/P4p/I8B4qZSqYWV2zmtIdYm/BSt4U66Xy3zKkVRsqN5Y0IeAW3wES7h9t45zg5bAFoSIMTXoP8qgCq1sDKvJdVG21WtkWr3zitF79llriuLR0C7B6f4qNjB5YAYkA22/hyMCKhSCyvJB1ax0uzNH0fK1pDxZbOiSl4UZ4C2eWlaWXzOg9ZHuKOwD1tAlVpYVwHEkl2rW7v37WrJAfG82LuF3VvQJfEJKJutj4qPMiA8rqm6WKUW1tH+cu9cqg2hurX96mNse6zKRkyNbfdCFUl8Atq+S06PfqBjL5wK+i/3n3iGtF8dnPJDL9ciutxJF7WRXeSzAKk13AqZZnInTYsSX9RiGuAT0G5BzJscIjm2NZjlp5QP82T0ffSAbNPUgmyY55MCWvvgFFZaW1M2S14lLwo3oJxo3UqGcS3W7JcnA5arOeMYAWX5dY0LGR8gMnmcGqAeJQEySAJkkATIIAmQQRIggyRABvk/VZFJB8D33XIAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="co">#Sommer</span></span>
<span id="cb90-2"><a href="#cb90-2" tabindex="-1"></a><span class="co">#Mask the data (replace BLUEs with NAs)</span></span>
<span id="cb90-3"><a href="#cb90-3" tabindex="-1"></a>sommerdf_mask <span class="ot">&lt;-</span> sommerdf_smaller</span>
<span id="cb90-4"><a href="#cb90-4" tabindex="-1"></a>sommerdf_mask<span class="sc">$</span>BLUEs[<span class="fu">which</span>(sommerdf_mask<span class="sc">$</span>Env <span class="sc">%in%</span> WI.env)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb90-5"><a href="#cb90-5" tabindex="-1"></a><span class="co">#sommerdf_mask$BLUEs[sample(1:nrow(sommerdf_mask),30)] &lt;- NA</span></span>
<span id="cb90-6"><a href="#cb90-6" tabindex="-1"></a></span>
<span id="cb90-7"><a href="#cb90-7" tabindex="-1"></a><span class="co">#diagonal variance matrix for locations in residuals:</span></span>
<span id="cb90-8"><a href="#cb90-8" tabindex="-1"></a>Locdiag <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb90-9"><a href="#cb90-9" tabindex="-1"></a><span class="cf">for</span> (enviro <span class="cf">in</span> <span class="fu">unique</span>(sommerdf_mask<span class="sc">$</span>Loc)) {</span>
<span id="cb90-10"><a href="#cb90-10" tabindex="-1"></a>  Locdiag <span class="ot">&lt;-</span> <span class="fu">c</span>(Locdiag, ans2c_open<span class="sc">$</span>sigma[[<span class="fu">paste0</span>(<span class="st">&quot;Loc:units:&quot;</span>, enviro, <span class="st">&quot;:&quot;</span>, enviro)]])</span>
<span id="cb90-11"><a href="#cb90-11" tabindex="-1"></a>}</span>
<span id="cb90-12"><a href="#cb90-12" tabindex="-1"></a>Locdiag <span class="ot">&lt;-</span> <span class="fu">diag</span>(Locdiag)</span>
<span id="cb90-13"><a href="#cb90-13" tabindex="-1"></a><span class="fu">rownames</span>(Locdiag) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sommerdf_mask<span class="sc">$</span>Loc)</span>
<span id="cb90-14"><a href="#cb90-14" tabindex="-1"></a><span class="fu">colnames</span>(Locdiag) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sommerdf_mask<span class="sc">$</span>Loc)</span>
<span id="cb90-15"><a href="#cb90-15" tabindex="-1"></a><span class="co">#LocdiagInv &lt;-as(solve(Locdiag),Class=&quot;dgCMatrix&quot;)</span></span>
<span id="cb90-16"><a href="#cb90-16" tabindex="-1"></a></span>
<span id="cb90-17"><a href="#cb90-17" tabindex="-1"></a><span class="co">#Residual vcov</span></span>
<span id="cb90-18"><a href="#cb90-18" tabindex="-1"></a>Rmat <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">diag</span>(Locdiag[sommerdf_mask<span class="sc">$</span>Loc,sommerdf_mask<span class="sc">$</span>Loc]))</span>
<span id="cb90-19"><a href="#cb90-19" tabindex="-1"></a>Rinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(Rmat),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb90-20"><a href="#cb90-20" tabindex="-1"></a><span class="fu">rownames</span>(Rinv) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;u&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Rinv)) <span class="co">#sommer naming convention for residuals</span></span>
<span id="cb90-21"><a href="#cb90-21" tabindex="-1"></a><span class="fu">colnames</span>(Rinv) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;u&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Rinv)) <span class="co">#sommer naming convention for residuals</span></span>
<span id="cb90-22"><a href="#cb90-22" tabindex="-1"></a></span>
<span id="cb90-23"><a href="#cb90-23" tabindex="-1"></a><span class="co">#We already know the estimated vcov between locs from previous models</span></span>
<span id="cb90-24"><a href="#cb90-24" tabindex="-1"></a>Locs_vcov</span></code></pre></div>
<pre><code>##          MI       MO       NC       NY       OR       WI
## MI 38.07112 29.92236 37.00346 34.52890 31.51140 36.83195
## MO 29.92236 25.03084 29.34736 28.00220 26.07704 29.26263
## NC 37.00346 29.34736 36.01186 33.71139 30.85645 35.85389
## NY 34.52890 28.00220 33.71139 31.80949 29.32765 33.58443
## OR 31.51140 26.07704 30.85645 29.32765 27.21672 30.75786
## WI 36.83195 29.26263 35.85389 33.58443 30.75786 35.69835</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a>LxG <span class="ot">&lt;-</span> <span class="fu">kronecker</span>(Locs_vcov,geno<span class="sc">@</span>G, <span class="at">make.dimnames =</span> T)</span>
<span id="cb92-2"><a href="#cb92-2" tabindex="-1"></a>LxGinv <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">solve</span>(<span class="fu">as.matrix</span>(LxG)<span class="sc">+</span><span class="fl">1e-6</span><span class="sc">*</span><span class="fu">diag</span>(<span class="fu">nrow</span>(LxG))),<span class="at">Class=</span><span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb92-3"><a href="#cb92-3" tabindex="-1"></a></span>
<span id="cb92-4"><a href="#cb92-4" tabindex="-1"></a>LxGcol <span class="ot">&lt;-</span> <span class="fu">paste0</span>(sommerdf_mask<span class="sc">$</span>Loc, <span class="st">&quot;:&quot;</span>, sommerdf_mask<span class="sc">$</span>id)</span>
<span id="cb92-5"><a href="#cb92-5" tabindex="-1"></a>sommerdf_mask <span class="ot">&lt;-</span> <span class="fu">cbind</span>(sommerdf_mask, LxGcol)</span></code></pre></div>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="co"># To calculate BLUPs with known variances, simply fix all variance terms </span></span>
<span id="cb93-2"><a href="#cb93-2" tabindex="-1"></a><span class="co"># and perform a single REML iteration:</span></span>
<span id="cb93-3"><a href="#cb93-3" tabindex="-1"></a>ans2mask <span class="ot">&lt;-</span> <span class="fu">mmec</span>(BLUEs <span class="sc">~</span> Env <span class="sc">+</span> Fg <span class="sc">-</span><span class="dv">1</span>, <span class="co">#Include inbreeding coefficients in</span></span>
<span id="cb93-4"><a href="#cb93-4" tabindex="-1"></a>                   <span class="co">#fixed effects. We want to make a regression with them</span></span>
<span id="cb93-5"><a href="#cb93-5" tabindex="-1"></a>                     <span class="at">random =</span> <span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(id,</span>
<span id="cb93-6"><a href="#cb93-6" tabindex="-1"></a>                                   <span class="at">theta =</span> <span class="fu">matrix</span>(ans2c_open<span class="sc">$</span>sigma[[<span class="st">&quot;id:Ginv:isc:isc&quot;</span>]]<span class="sc">/</span></span>
<span id="cb93-7"><a href="#cb93-7" tabindex="-1"></a>                                         <span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb93-8"><a href="#cb93-8" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)), </span>
<span id="cb93-9"><a href="#cb93-9" tabindex="-1"></a>                                  <span class="at">Gu =</span> Ginv) <span class="sc">+</span> <span class="co">#addtitive genotypic</span></span>
<span id="cb93-10"><a href="#cb93-10" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb93-11"><a href="#cb93-11" tabindex="-1"></a>                               <span class="fu">vsc</span>(<span class="fu">isc</span>(id,</span>
<span id="cb93-12"><a href="#cb93-12" tabindex="-1"></a>                                   <span class="at">theta =</span> <span class="fu">matrix</span>(ans2c_open<span class="sc">$</span>sigma[[<span class="st">&quot;id:Dinv:isc:isc&quot;</span>]]<span class="sc">/</span></span>
<span id="cb93-13"><a href="#cb93-13" tabindex="-1"></a>                                         <span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb93-14"><a href="#cb93-14" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)),</span>
<span id="cb93-15"><a href="#cb93-15" tabindex="-1"></a>                                  <span class="at">Gu =</span> Dinv) <span class="sc">+</span> <span class="co">#dominance genotypic</span></span>
<span id="cb93-16"><a href="#cb93-16" tabindex="-1"></a>                   </span>
<span id="cb93-17"><a href="#cb93-17" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(LxGcol,</span>
<span id="cb93-18"><a href="#cb93-18" tabindex="-1"></a>                                   <span class="at">theta =</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb93-19"><a href="#cb93-19" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)</span>
<span id="cb93-20"><a href="#cb93-20" tabindex="-1"></a>                               ), <span class="at">Gu =</span> LxGinv) <span class="sc">+</span> <span class="co">#kronecker unstructured</span></span>
<span id="cb93-21"><a href="#cb93-21" tabindex="-1"></a>                                <span class="co">#effects</span></span>
<span id="cb93-22"><a href="#cb93-22" tabindex="-1"></a>                       <span class="fu">vsc</span>(<span class="fu">isc</span>(Stage1Error, <span class="co"># Stage 1 estimation error effect.</span></span>
<span id="cb93-23"><a href="#cb93-23" tabindex="-1"></a>                      <span class="co"># Since it&#39;s inside the `isc()` function, we are assuming an identity </span></span>
<span id="cb93-24"><a href="#cb93-24" tabindex="-1"></a>                      <span class="co"># variance-covariance structure.This is undesired and will be overridden later.</span></span>
<span id="cb93-25"><a href="#cb93-25" tabindex="-1"></a>                      <span class="co"># Ideally, we would skip the `isc()` function altogether, but Sommer syntax </span></span>
<span id="cb93-26"><a href="#cb93-26" tabindex="-1"></a>                      <span class="co">#requires it, even if we will overwrite it later.</span></span>
<span id="cb93-27"><a href="#cb93-27" tabindex="-1"></a>                               <span class="at">theta =</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)), <span class="co">#initial value for the</span></span>
<span id="cb93-28"><a href="#cb93-28" tabindex="-1"></a>                               <span class="co">#variance.</span></span>
<span id="cb93-29"><a href="#cb93-29" tabindex="-1"></a>                               <span class="at">thetaC =</span> <span class="fu">matrix</span>(<span class="dv">3</span>)), <span class="co">#In thetaC we can add </span></span>
<span id="cb93-30"><a href="#cb93-30" tabindex="-1"></a>                                <span class="co">#constraints for the variance estimation:</span></span>
<span id="cb93-31"><a href="#cb93-31" tabindex="-1"></a>                           </span>
<span id="cb93-32"><a href="#cb93-32" tabindex="-1"></a>                                <span class="co">#0: not to be estimated</span></span>
<span id="cb93-33"><a href="#cb93-33" tabindex="-1"></a>                                <span class="co">#1: estimated and constrained to be positive </span></span>
<span id="cb93-34"><a href="#cb93-34" tabindex="-1"></a>                                <span class="co">#(this is the default)</span></span>
<span id="cb93-35"><a href="#cb93-35" tabindex="-1"></a>                                <span class="co">#2: estimated and unconstrained</span></span>
<span id="cb93-36"><a href="#cb93-36" tabindex="-1"></a>                                <span class="co">#(can be negative or positive)</span></span>
<span id="cb93-37"><a href="#cb93-37" tabindex="-1"></a>                                <span class="co">#3: not to be estimated but fixed (value has to </span></span>
<span id="cb93-38"><a href="#cb93-38" tabindex="-1"></a>                                <span class="co">#be provided in the theta argument)</span></span>
<span id="cb93-39"><a href="#cb93-39" tabindex="-1"></a>                           </span>
<span id="cb93-40"><a href="#cb93-40" tabindex="-1"></a>                                <span class="co"># As we are using constraint 3, we disable </span></span>
<span id="cb93-41"><a href="#cb93-41" tabindex="-1"></a>                                <span class="co">#variance estimation.</span></span>
<span id="cb93-42"><a href="#cb93-42" tabindex="-1"></a>                                <span class="co"># The model will use the value we provided </span></span>
<span id="cb93-43"><a href="#cb93-43" tabindex="-1"></a>                                <span class="co">#in the &quot;theta&quot; argument.</span></span>
<span id="cb93-44"><a href="#cb93-44" tabindex="-1"></a></span>
<span id="cb93-45"><a href="#cb93-45" tabindex="-1"></a>                           <span class="at">Gu=</span>EEVInv), <span class="co">#inverse of known variance-covariance </span></span>
<span id="cb93-46"><a href="#cb93-46" tabindex="-1"></a>                     <span class="co">#matrix. This overwrites the identity variance-covariance </span></span>
<span id="cb93-47"><a href="#cb93-47" tabindex="-1"></a>                     <span class="co">#structure that we had assumed when we used isc() function</span></span>
<span id="cb93-48"><a href="#cb93-48" tabindex="-1"></a>                     <span class="co">#for the Stage1Error effect.</span></span>
<span id="cb93-49"><a href="#cb93-49" tabindex="-1"></a>                     <span class="at">rcov=</span><span class="sc">~</span><span class="fu">vsc</span>(<span class="fu">isc</span>(units,</span>
<span id="cb93-50"><a href="#cb93-50" tabindex="-1"></a>                                   <span class="at">theta =</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">/</span><span class="fu">var</span>(sommerdf_smaller<span class="sc">$</span>BLUEs)),</span>
<span id="cb93-51"><a href="#cb93-51" tabindex="-1"></a>                                  <span class="at">thetaC =</span>  <span class="fu">matrix</span>(<span class="dv">3</span>)), </span>
<span id="cb93-52"><a href="#cb93-52" tabindex="-1"></a>                               <span class="at">Gu =</span> Rinv), <span class="co">#residuals. isc(units) means that we </span></span>
<span id="cb93-53"><a href="#cb93-53" tabindex="-1"></a>                     <span class="co">#assume identity variance-covariance structure for the </span></span>
<span id="cb93-54"><a href="#cb93-54" tabindex="-1"></a>                     <span class="co">#&quot;units&quot; effect (residuals are always called &quot;units&quot;)</span></span>
<span id="cb93-55"><a href="#cb93-55" tabindex="-1"></a>                     <span class="at">data=</span> sommerdf_mask,</span>
<span id="cb93-56"><a href="#cb93-56" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">TRUE</span>,</span>
<span id="cb93-57"><a href="#cb93-57" tabindex="-1"></a>                     <span class="at">nIters=</span><span class="dv">1</span> <span class="co">#as we already know the variances, a single </span></span>
<span id="cb93-58"><a href="#cb93-58" tabindex="-1"></a>                    <span class="co">#REML iteration is enough</span></span>
<span id="cb93-59"><a href="#cb93-59" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="co">#rename BLUP names to be shorter</span></span>
<span id="cb94-2"><a href="#cb94-2" tabindex="-1"></a><span class="fu">names</span>(ans2mask<span class="sc">$</span>uList) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;additive&quot;</span>, <span class="st">&quot;dominance&quot;</span>, <span class="st">&quot;GxL&quot;</span>, <span class="st">&quot;Stage1Error&quot;</span>)</span>
<span id="cb94-3"><a href="#cb94-3" tabindex="-1"></a></span>
<span id="cb94-4"><a href="#cb94-4" tabindex="-1"></a></span>
<span id="cb94-5"><a href="#cb94-5" tabindex="-1"></a></span>
<span id="cb94-6"><a href="#cb94-6" tabindex="-1"></a>WI.GxE.BLUPs <span class="ot">&lt;-</span> <span class="fu">c</span>(ans2mask<span class="sc">$</span>uList<span class="sc">$</span>GxL[<span class="fu">paste0</span>(<span class="st">&quot;WI:&quot;</span>,<span class="fu">unique</span>(sommerdf_mask<span class="sc">$</span>id)),])</span>
<span id="cb94-7"><a href="#cb94-7" tabindex="-1"></a></span>
<span id="cb94-8"><a href="#cb94-8" tabindex="-1"></a>SommerBLUPs <span class="ot">&lt;-</span> ans2mask<span class="sc">$</span>uList<span class="sc">$</span>additive<span class="sc">+</span> WI.GxE.BLUPs</span>
<span id="cb94-9"><a href="#cb94-9" tabindex="-1"></a>   </span>
<span id="cb94-10"><a href="#cb94-10" tabindex="-1"></a><span class="co"># Very poor correlation.</span></span>
<span id="cb94-11"><a href="#cb94-11" tabindex="-1"></a><span class="co"># We don&#39;t recommend using sommer for this application.</span></span>
<span id="cb94-12"><a href="#cb94-12" tabindex="-1"></a><span class="co"># The GxE BLUPs in sommer are not reliable for locations for which we don&#39;t have</span></span>
<span id="cb94-13"><a href="#cb94-13" tabindex="-1"></a><span class="co"># any data.</span></span>
<span id="cb94-14"><a href="#cb94-14" tabindex="-1"></a><span class="fu">cor</span>(pred2<span class="sc">$</span>value, <span class="co">#StageWise</span></span>
<span id="cb94-15"><a href="#cb94-15" tabindex="-1"></a>    SommerBLUPs[<span class="fu">match</span>(pred2<span class="sc">$</span>id,<span class="fu">rownames</span>(SommerBLUPs))]) <span class="co">#Sommer</span></span></code></pre></div>
<pre><code>## [1] 0.0957398</code></pre>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="co">#save.image(&quot;workspace2.RData&quot;)</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a><span class="co">#save(ans2a_open, ans2c_open, ans2mask,ans2a, file = &quot;Vignette2Models.RData&quot;)</span></span></code></pre></div>
<p>As expected, the reliability using StageWise was higher with WI
phenotypes. In contrast, predictions with sommer have been extremely
poor. The likely reason for this is that the WI GxL BLUPs rely heavily
on the variance-covariance matrix between locations, which is estimated
using the factor analytics loadings. ASReml (StageWise) updates these
loadings as the variances are estimated in the model, while sommer
retains the initial values for the loadings. As a result, the loadings
and therefore the covariance between locations is better estimated by
ASReml than by sommer.</p>
<div id="bibliography" class="section level1">
<h1>Bibliography</h1>
<p>Damesa, T.M., MÃ¶hring, J., Worku, M. and Piepho, H.-P. (2017), One
Step at a Time: Stage-Wise Analysis of a Series of Experiments. Agronomy
Journal, 109: 845-857.</p>
<p>Endelman, J. B. (2023). Fully efficient, two-stage analysis of
multi-environment trials with directional dominance and multi-trait
genomic selection. Theoretical and Applied Genetics, 136(4), 65.</p>
<p>Giovanny C (2016). âGenome assisted prediction of quantitative traits
using the R package sommer.â PLoS ONE, 11, 1-15.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
